{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ccce551",
   "metadata": {},
   "source": [
    "### Clean and slim the repertoire files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42109600",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3632e7c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files with CMV and HLA annotation: 223\n",
      "Number of files with CMV +: 98\n",
      "Number of files with CMV -: 125\n"
     ]
    }
   ],
   "source": [
    "# Set HLA of interest - change dynamically\n",
    "hla = 'A*01'\n",
    "hla_name = hla.replace('*','')\n",
    "\n",
    "# Generate file names\n",
    "\n",
    "os.chdir(\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/Repertoires\")\n",
    "\n",
    "metadata = pd.read_csv(\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/Repertoires/Cohort01_whole_metadata.tsv\", sep=\"\\t\") \n",
    "\n",
    "\n",
    "metadata_annotat = metadata[metadata['sample_tags'].str.contains(f'HLA-{hla}',case=False, regex = False)]\n",
    "\n",
    "metadata_annotat = metadata_annotat[metadata_annotat['sample_tags'].str.contains(r'\\bcytomegalovirus|CMV\\b', case=False, na=False)]\n",
    "metadata_annotat = metadata_annotat.reset_index(drop=True)\n",
    "\n",
    "#Extract pos and neg patient IDs\n",
    "pos_names = metadata_annotat[metadata_annotat['sample_tags'].str.contains(r'\\bcytomegalovirus \\+|CMV \\+\\b', case=False, na=False)]\n",
    "neg_names = metadata_annotat[metadata_annotat['sample_tags'].str.contains(r'\\bcytomegalovirus -|CMV -\\b', case=False, na=False)]\n",
    "\n",
    "#Turn into file names\n",
    "file_names = [name + '.tsv' for name in metadata_annotat['sample_name'].tolist()]\n",
    "pos_names = [name + '.tsv' for name in pos_names['sample_name'].tolist()]\n",
    "neg_names = [name + '.tsv' for name in neg_names['sample_name'].tolist()]\n",
    "\n",
    "\n",
    "metadata_annotat.loc[10,'sample_tags']\n",
    "\n",
    "print(f'Number of files with CMV and HLA annotation: {len(file_names)}')\n",
    "print(f'Number of files with CMV +: {len(pos_names)}')\n",
    "print(f'Number of files with CMV -: {len(neg_names)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "17b33b34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "📄 Found the following TSV files for processing:\n",
      " Checking file: P00404.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00404.tsv\n",
      " Checking file: P00416.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00416.tsv\n",
      " Checking file: P00037.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00037.tsv\n",
      " Checking file: P00346.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00346.tsv\n",
      " Checking file: P00278.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00278.tsv\n",
      " Checking file: P00260.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00260.tsv\n",
      " Checking file: P00210.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00210.tsv\n",
      " Checking file: P00208.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00208.tsv\n",
      " Checking file: P00134.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00134.tsv\n",
      " Checking file: P00088.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00088.tsv\n",
      " Checking file: P00150.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00150.tsv\n",
      " Checking file: P00378.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00378.tsv\n",
      " Checking file: P00377.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00377.tsv\n",
      " Checking file: P00559.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00559.tsv\n",
      " Checking file: P00459.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00459.tsv\n",
      " Checking file: P00367.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00367.tsv\n",
      " Checking file: P00360.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00360.tsv\n",
      " Checking file: P00328.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00328.tsv\n",
      " Checking file: P00213.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00213.tsv\n",
      " Checking file: P00157.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00157.tsv\n",
      " Checking file: P00148.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00148.tsv\n",
      " Checking file: P00071.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00071.tsv\n",
      " Checking file: P00018.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00018.tsv\n",
      " Checking file: P00542.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00542.tsv\n",
      " Checking file: P00575.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00575.tsv\n",
      " Checking file: P00178.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00178.tsv\n",
      " Checking file: P00120.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00120.tsv\n",
      " Checking file: P00365.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00365.tsv\n",
      " Checking file: P00303.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00303.tsv\n",
      " Checking file: P00139.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00139.tsv\n",
      " Checking file: P00190.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00190.tsv\n",
      " Checking file: P00199.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00199.tsv\n",
      " Checking file: P00161.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00161.tsv\n",
      " Checking file: P00309.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00309.tsv\n",
      " Checking file: P00533.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00533.tsv\n",
      " Checking file: P00301.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00301.tsv\n",
      " Checking file: P00361.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00361.tsv\n",
      " Checking file: P00540.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00540.tsv\n",
      " Checking file: P00002.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00002.tsv\n",
      " Checking file: P00243.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00243.tsv\n",
      " Checking file: P00546.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00546.tsv\n",
      " Checking file: P00387.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00387.tsv\n",
      " Checking file: P00424.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00424.tsv\n",
      " Checking file: P00545.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00545.tsv\n",
      " Checking file: P00571.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00571.tsv\n",
      " Checking file: P00581.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00581.tsv\n",
      " Checking file: P00272.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00272.tsv\n",
      " Checking file: P00373.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00373.tsv\n",
      " Checking file: P00446.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00446.tsv\n",
      " Checking file: P00368.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00368.tsv\n",
      " Checking file: P00457.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00457.tsv\n",
      " Checking file: P00554.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00554.tsv\n",
      " Checking file: P00557.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00557.tsv\n",
      " Checking file: P00582.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00582.tsv\n",
      " Checking file: P00544.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00544.tsv\n",
      " Checking file: P00576.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00576.tsv\n",
      " Checking file: P00556.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00556.tsv\n",
      " Checking file: P00016.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00016.tsv\n",
      " Checking file: P00467.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00467.tsv\n",
      " Checking file: P00104.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00104.tsv\n",
      " Checking file: P00155.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00155.tsv\n",
      " Checking file: P00485.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00485.tsv\n",
      " Checking file: P00512.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00512.tsv\n",
      " Checking file: P00081.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00081.tsv\n",
      " Checking file: P00488.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00488.tsv\n",
      " Checking file: P00611.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00611.tsv\n",
      " Checking file: P00603.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00603.tsv\n",
      " Checking file: P00494.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00494.tsv\n",
      " Checking file: P00522.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00522.tsv\n",
      " Checking file: P00466.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00466.tsv\n",
      " Checking file: P00600.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00600.tsv\n",
      " Checking file: P00521.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00521.tsv\n",
      " Checking file: P00515.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00515.tsv\n",
      " Checking file: P00517.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00517.tsv\n",
      " Checking file: P00470.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00470.tsv\n",
      " Checking file: P00595.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00595.tsv\n",
      " Checking file: P00069.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00069.tsv\n",
      " Checking file: P00218.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00218.tsv\n",
      " Checking file: P00236.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00236.tsv\n",
      " Checking file: P00080.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00080.tsv\n",
      " Checking file: P00239.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00239.tsv\n",
      " Checking file: P00329.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00329.tsv\n",
      " Checking file: P00169.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00169.tsv\n",
      " Checking file: P00108.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00108.tsv\n",
      " Checking file: P00105.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00105.tsv\n",
      " Checking file: P00192.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00192.tsv\n",
      " Checking file: P00252.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00252.tsv\n",
      " Checking file: P00175.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00175.tsv\n",
      " Checking file: P00230.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00230.tsv\n",
      " Checking file: P00056.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00056.tsv\n",
      " Checking file: P00257.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00257.tsv\n",
      " Checking file: P00115.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00115.tsv\n",
      " Checking file: P00265.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00265.tsv\n",
      " Checking file: P00166.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00166.tsv\n",
      " Checking file: P00552.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00552.tsv\n",
      " Checking file: P00497.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00497.tsv\n",
      " Checking file: P00375.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00375.tsv\n",
      " Checking file: P00128.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00128.tsv\n",
      " Checking file: P00147.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00147.tsv\n",
      " Checking file: P00129.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00129.tsv\n",
      " Checking file: P00454.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00454.tsv\n",
      " Checking file: P00124.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00124.tsv\n",
      " Checking file: P00211.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00211.tsv\n",
      " Checking file: P00118.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00118.tsv\n",
      " Checking file: P00137.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00137.tsv\n",
      " Checking file: P00505.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00505.tsv\n",
      " Checking file: P00354.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00354.tsv\n",
      " Checking file: P00117.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00117.tsv\n",
      " Checking file: P00359.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00359.tsv\n",
      " Checking file: P00349.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00349.tsv\n",
      " Checking file: P00389.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00389.tsv\n",
      " Checking file: P00511.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00511.tsv\n",
      " Checking file: P00599.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00599.tsv\n",
      " Checking file: P00273.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00273.tsv\n",
      " Checking file: P00275.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00275.tsv\n",
      " Checking file: P00478.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00478.tsv\n",
      " Checking file: P00434.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00434.tsv\n",
      " Checking file: P00615.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00615.tsv\n",
      " Checking file: P00547.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00547.tsv\n",
      " Checking file: P00605.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00605.tsv\n",
      " Checking file: P00338.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00338.tsv\n",
      " Checking file: P00455.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00455.tsv\n",
      " Checking file: P00657.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00657.tsv\n",
      " Checking file: P00180.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00180.tsv\n",
      " Checking file: P00420.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00420.tsv\n",
      " Checking file: P00648.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00648.tsv\n",
      " Checking file: P00383.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00383.tsv\n",
      " Checking file: P00580.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00580.tsv\n",
      " Checking file: P00376.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00376.tsv\n",
      " Checking file: P00355.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00355.tsv\n",
      " Checking file: P00524.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00524.tsv\n",
      " Checking file: P00356.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00356.tsv\n",
      " Checking file: P00617.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00617.tsv\n",
      " Checking file: P00025.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00025.tsv\n",
      " Checking file: P00465.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00465.tsv\n",
      " Checking file: P00039.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00039.tsv\n",
      " Checking file: P00238.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00238.tsv\n",
      " Checking file: P00036.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00036.tsv\n",
      " Checking file: P00102.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00102.tsv\n",
      " Checking file: P00103.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00103.tsv\n",
      " Checking file: P00028.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00028.tsv\n",
      " Checking file: P00141.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00141.tsv\n",
      " Checking file: P00114.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00114.tsv\n",
      " Checking file: P00233.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00233.tsv\n",
      " Checking file: P00013.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00013.tsv\n",
      " Checking file: P00184.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00184.tsv\n",
      " Checking file: P00235.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00235.tsv\n",
      " Checking file: P00164.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00164.tsv\n",
      " Checking file: P00072.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00072.tsv\n",
      " Checking file: P00188.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00188.tsv\n",
      " Checking file: P00113.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00113.tsv\n",
      " Checking file: P00196.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00196.tsv\n",
      " Checking file: P00078.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00078.tsv\n",
      " Checking file: P00176.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00176.tsv\n",
      " Checking file: P00264.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00264.tsv\n",
      " Checking file: P00382.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00382.tsv\n",
      " Checking file: P00263.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00263.tsv\n",
      " Checking file: P00298.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00298.tsv\n",
      " Checking file: P00306.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00306.tsv\n",
      " Checking file: P00333.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00333.tsv\n",
      " Checking file: P00168.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00168.tsv\n",
      " Checking file: P00348.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00348.tsv\n",
      " Checking file: P00062.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00062.tsv\n",
      " Checking file: P00181.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00181.tsv\n",
      " Checking file: P00302.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00302.tsv\n",
      " Checking file: P00345.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00345.tsv\n",
      " Checking file: P00154.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00154.tsv\n",
      " Checking file: P00174.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00174.tsv\n",
      " Checking file: P00047.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00047.tsv\n",
      " Checking file: P00327.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00327.tsv\n",
      " Checking file: P00050.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00050.tsv\n",
      " Checking file: P00067.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00067.tsv\n",
      " Checking file: P00283.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00283.tsv\n",
      " Checking file: P00480.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00480.tsv\n",
      " Checking file: P00482.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00482.tsv\n",
      " Checking file: P00418.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00418.tsv\n",
      " Checking file: P00250.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00250.tsv\n",
      " Checking file: P00415.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00415.tsv\n",
      " Checking file: P00240.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00240.tsv\n",
      " Checking file: P00395.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00395.tsv\n",
      " Checking file: P00215.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00215.tsv\n",
      " Checking file: P00668.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00668.tsv\n",
      " Checking file: P00634.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00634.tsv\n",
      " Checking file: P00228.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00228.tsv\n",
      " Checking file: P00652.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00652.tsv\n",
      " Checking file: P00656.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00656.tsv\n",
      " Checking file: P00665.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00665.tsv\n",
      " Checking file: P00286.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00286.tsv\n",
      " Checking file: P00628.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00628.tsv\n",
      " Checking file: P00491.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00491.tsv\n",
      " Checking file: P00639.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00639.tsv\n",
      " Checking file: P00538.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00538.tsv\n",
      " Checking file: P00399.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00399.tsv\n",
      " Checking file: P00662.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00662.tsv\n",
      " Checking file: P00584.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00584.tsv\n",
      " Checking file: P00651.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00651.tsv\n",
      " Checking file: P00659.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00659.tsv\n",
      " Checking file: P00284.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00284.tsv\n",
      " Checking file: P00406.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00406.tsv\n",
      " Checking file: P00507.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00507.tsv\n",
      " Checking file: P00653.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00653.tsv\n",
      " Checking file: P00342.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00342.tsv\n",
      " Checking file: P00642.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00642.tsv\n",
      " Checking file: P00616.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00616.tsv\n",
      " Checking file: P00566.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00566.tsv\n",
      " Checking file: P00531.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00531.tsv\n",
      " Checking file: P00049.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00049.tsv\n",
      " Checking file: P00227.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00227.tsv\n",
      " Checking file: P00461.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00461.tsv\n",
      " Checking file: P00529.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00529.tsv\n",
      " Checking file: P00481.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00481.tsv\n",
      " Checking file: P00411.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00411.tsv\n",
      " Checking file: P00520.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00520.tsv\n",
      " Checking file: P00447.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00447.tsv\n",
      " Checking file: P00469.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00469.tsv\n",
      " Checking file: P00513.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00513.tsv\n",
      " Checking file: P00206.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00206.tsv\n",
      " Checking file: P00369.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00369.tsv\n",
      " Checking file: P00558.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00558.tsv\n",
      " Checking file: P00384.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00384.tsv\n",
      " Checking file: P00197.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00197.tsv\n",
      " Checking file: P00237.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00237.tsv\n",
      " Checking file: P00289.tsv\n",
      "  ✅ Accepted: /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01/P00289.tsv\n",
      "Processing the following TSV files\n",
      "  P00404.tsv\n",
      "  P00416.tsv\n",
      "  P00037.tsv\n",
      "  P00346.tsv\n",
      "  P00278.tsv\n",
      "  P00260.tsv\n",
      "  P00210.tsv\n",
      "  P00208.tsv\n",
      "  P00134.tsv\n",
      "  P00088.tsv\n",
      "  P00150.tsv\n",
      "  P00378.tsv\n",
      "  P00377.tsv\n",
      "  P00559.tsv\n",
      "  P00459.tsv\n",
      "  P00367.tsv\n",
      "  P00360.tsv\n",
      "  P00328.tsv\n",
      "  P00213.tsv\n",
      "  P00157.tsv\n",
      "  P00148.tsv\n",
      "  P00071.tsv\n",
      "  P00018.tsv\n",
      "  P00542.tsv\n",
      "  P00575.tsv\n",
      "  P00178.tsv\n",
      "  P00120.tsv\n",
      "  P00365.tsv\n",
      "  P00303.tsv\n",
      "  P00139.tsv\n",
      "  P00190.tsv\n",
      "  P00199.tsv\n",
      "  P00161.tsv\n",
      "  P00309.tsv\n",
      "  P00533.tsv\n",
      "  P00301.tsv\n",
      "  P00361.tsv\n",
      "  P00540.tsv\n",
      "  P00002.tsv\n",
      "  P00243.tsv\n",
      "  P00546.tsv\n",
      "  P00387.tsv\n",
      "  P00424.tsv\n",
      "  P00545.tsv\n",
      "  P00571.tsv\n",
      "  P00581.tsv\n",
      "  P00272.tsv\n",
      "  P00373.tsv\n",
      "  P00446.tsv\n",
      "  P00368.tsv\n",
      "  P00457.tsv\n",
      "  P00554.tsv\n",
      "  P00557.tsv\n",
      "  P00582.tsv\n",
      "  P00544.tsv\n",
      "  P00576.tsv\n",
      "  P00556.tsv\n",
      "  P00016.tsv\n",
      "  P00467.tsv\n",
      "  P00104.tsv\n",
      "  P00155.tsv\n",
      "  P00485.tsv\n",
      "  P00512.tsv\n",
      "  P00081.tsv\n",
      "  P00488.tsv\n",
      "  P00611.tsv\n",
      "  P00603.tsv\n",
      "  P00494.tsv\n",
      "  P00522.tsv\n",
      "  P00466.tsv\n",
      "  P00600.tsv\n",
      "  P00521.tsv\n",
      "  P00515.tsv\n",
      "  P00517.tsv\n",
      "  P00470.tsv\n",
      "  P00595.tsv\n",
      "  P00069.tsv\n",
      "  P00218.tsv\n",
      "  P00236.tsv\n",
      "  P00080.tsv\n",
      "  P00239.tsv\n",
      "  P00329.tsv\n",
      "  P00169.tsv\n",
      "  P00108.tsv\n",
      "  P00105.tsv\n",
      "  P00192.tsv\n",
      "  P00252.tsv\n",
      "  P00175.tsv\n",
      "  P00230.tsv\n",
      "  P00056.tsv\n",
      "  P00257.tsv\n",
      "  P00115.tsv\n",
      "  P00265.tsv\n",
      "  P00166.tsv\n",
      "  P00552.tsv\n",
      "  P00497.tsv\n",
      "  P00375.tsv\n",
      "  P00128.tsv\n",
      "  P00147.tsv\n",
      "  P00129.tsv\n",
      "  P00454.tsv\n",
      "  P00124.tsv\n",
      "  P00211.tsv\n",
      "  P00118.tsv\n",
      "  P00137.tsv\n",
      "  P00505.tsv\n",
      "  P00354.tsv\n",
      "  P00117.tsv\n",
      "  P00359.tsv\n",
      "  P00349.tsv\n",
      "  P00389.tsv\n",
      "  P00511.tsv\n",
      "  P00599.tsv\n",
      "  P00273.tsv\n",
      "  P00275.tsv\n",
      "  P00478.tsv\n",
      "  P00434.tsv\n",
      "  P00615.tsv\n",
      "  P00547.tsv\n",
      "  P00605.tsv\n",
      "  P00338.tsv\n",
      "  P00455.tsv\n",
      "  P00657.tsv\n",
      "  P00180.tsv\n",
      "  P00420.tsv\n",
      "  P00648.tsv\n",
      "  P00383.tsv\n",
      "  P00580.tsv\n",
      "  P00376.tsv\n",
      "  P00355.tsv\n",
      "  P00524.tsv\n",
      "  P00356.tsv\n",
      "  P00617.tsv\n",
      "  P00025.tsv\n",
      "  P00465.tsv\n",
      "  P00039.tsv\n",
      "  P00238.tsv\n",
      "  P00036.tsv\n",
      "  P00102.tsv\n",
      "  P00103.tsv\n",
      "  P00028.tsv\n",
      "  P00141.tsv\n",
      "  P00114.tsv\n",
      "  P00233.tsv\n",
      "  P00013.tsv\n",
      "  P00184.tsv\n",
      "  P00235.tsv\n",
      "  P00164.tsv\n",
      "  P00072.tsv\n",
      "  P00188.tsv\n",
      "  P00113.tsv\n",
      "  P00196.tsv\n",
      "  P00078.tsv\n",
      "  P00176.tsv\n",
      "  P00264.tsv\n",
      "  P00382.tsv\n",
      "  P00263.tsv\n",
      "  P00298.tsv\n",
      "  P00306.tsv\n",
      "  P00333.tsv\n",
      "  P00168.tsv\n",
      "  P00348.tsv\n",
      "  P00062.tsv\n",
      "  P00181.tsv\n",
      "  P00302.tsv\n",
      "  P00345.tsv\n",
      "  P00154.tsv\n",
      "  P00174.tsv\n",
      "  P00047.tsv\n",
      "  P00327.tsv\n",
      "  P00050.tsv\n",
      "  P00067.tsv\n",
      "  P00283.tsv\n",
      "  P00480.tsv\n",
      "  P00482.tsv\n",
      "  P00418.tsv\n",
      "  P00250.tsv\n",
      "  P00415.tsv\n",
      "  P00240.tsv\n",
      "  P00395.tsv\n",
      "  P00215.tsv\n",
      "  P00668.tsv\n",
      "  P00634.tsv\n",
      "  P00228.tsv\n",
      "  P00652.tsv\n",
      "  P00656.tsv\n",
      "  P00665.tsv\n",
      "  P00286.tsv\n",
      "  P00628.tsv\n",
      "  P00491.tsv\n",
      "  P00639.tsv\n",
      "  P00538.tsv\n",
      "  P00399.tsv\n",
      "  P00662.tsv\n",
      "  P00584.tsv\n",
      "  P00651.tsv\n",
      "  P00659.tsv\n",
      "  P00284.tsv\n",
      "  P00406.tsv\n",
      "  P00507.tsv\n",
      "  P00653.tsv\n",
      "  P00342.tsv\n",
      "  P00642.tsv\n",
      "  P00616.tsv\n",
      "  P00566.tsv\n",
      "  P00531.tsv\n",
      "  P00049.tsv\n",
      "  P00227.tsv\n",
      "  P00461.tsv\n",
      "  P00529.tsv\n",
      "  P00481.tsv\n",
      "  P00411.tsv\n",
      "  P00520.tsv\n",
      "  P00447.tsv\n",
      "  P00469.tsv\n",
      "  P00513.tsv\n",
      "  P00206.tsv\n",
      "  P00369.tsv\n",
      "  P00558.tsv\n",
      "  P00384.tsv\n",
      "  P00197.tsv\n",
      "  P00237.tsv\n",
      "  P00289.tsv\n",
      "\n",
      "223 files found in /Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01.\n",
      "\n",
      "[1/223] Processing P00404.tsv...\n",
      "Read 345364 rows and 13 columns from P00404.tsv\n",
      "📝 Processed P00404.tsv → P00404.slim.tsv\n",
      "\n",
      "[2/223] Processing P00416.tsv...\n",
      "Read 260017 rows and 13 columns from P00416.tsv\n",
      "📝 Processed P00416.tsv → P00416.slim.tsv\n",
      "\n",
      "[3/223] Processing P00037.tsv...\n",
      "Read 330252 rows and 13 columns from P00037.tsv\n",
      "📝 Processed P00037.tsv → P00037.slim.tsv\n",
      "\n",
      "[4/223] Processing P00346.tsv...\n",
      "Read 123519 rows and 13 columns from P00346.tsv\n",
      "📝 Processed P00346.tsv → P00346.slim.tsv\n",
      "\n",
      "[5/223] Processing P00278.tsv...\n",
      "Read 240732 rows and 13 columns from P00278.tsv\n",
      "📝 Processed P00278.tsv → P00278.slim.tsv\n",
      "\n",
      "[6/223] Processing P00260.tsv...\n",
      "Read 137887 rows and 13 columns from P00260.tsv\n",
      "📝 Processed P00260.tsv → P00260.slim.tsv\n",
      "\n",
      "[7/223] Processing P00210.tsv...\n",
      "Read 139764 rows and 13 columns from P00210.tsv\n",
      "📝 Processed P00210.tsv → P00210.slim.tsv\n",
      "\n",
      "[8/223] Processing P00208.tsv...\n",
      "Read 179361 rows and 13 columns from P00208.tsv\n",
      "📝 Processed P00208.tsv → P00208.slim.tsv\n",
      "\n",
      "[9/223] Processing P00134.tsv...\n",
      "Read 281910 rows and 13 columns from P00134.tsv\n",
      "📝 Processed P00134.tsv → P00134.slim.tsv\n",
      "\n",
      "[10/223] Processing P00088.tsv...\n",
      "Read 253566 rows and 13 columns from P00088.tsv\n",
      "📝 Processed P00088.tsv → P00088.slim.tsv\n",
      "\n",
      "[11/223] Processing P00150.tsv...\n",
      "Read 160458 rows and 13 columns from P00150.tsv\n",
      "📝 Processed P00150.tsv → P00150.slim.tsv\n",
      "\n",
      "[12/223] Processing P00378.tsv...\n",
      "Read 133383 rows and 13 columns from P00378.tsv\n",
      "📝 Processed P00378.tsv → P00378.slim.tsv\n",
      "\n",
      "[13/223] Processing P00377.tsv...\n",
      "Read 220378 rows and 13 columns from P00377.tsv\n",
      "📝 Processed P00377.tsv → P00377.slim.tsv\n",
      "\n",
      "[14/223] Processing P00559.tsv...\n",
      "Read 175397 rows and 13 columns from P00559.tsv\n",
      "📝 Processed P00559.tsv → P00559.slim.tsv\n",
      "\n",
      "[15/223] Processing P00459.tsv...\n",
      "Read 42434 rows and 13 columns from P00459.tsv\n",
      "📝 Processed P00459.tsv → P00459.slim.tsv\n",
      "\n",
      "[16/223] Processing P00367.tsv...\n",
      "Read 71609 rows and 13 columns from P00367.tsv\n",
      "📝 Processed P00367.tsv → P00367.slim.tsv\n",
      "\n",
      "[17/223] Processing P00360.tsv...\n",
      "Read 108578 rows and 13 columns from P00360.tsv\n",
      "📝 Processed P00360.tsv → P00360.slim.tsv\n",
      "\n",
      "[18/223] Processing P00328.tsv...\n",
      "Read 121598 rows and 13 columns from P00328.tsv\n",
      "📝 Processed P00328.tsv → P00328.slim.tsv\n",
      "\n",
      "[19/223] Processing P00213.tsv...\n",
      "Read 89179 rows and 13 columns from P00213.tsv\n",
      "📝 Processed P00213.tsv → P00213.slim.tsv\n",
      "\n",
      "[20/223] Processing P00157.tsv...\n",
      "Read 122475 rows and 13 columns from P00157.tsv\n",
      "📝 Processed P00157.tsv → P00157.slim.tsv\n",
      "\n",
      "[21/223] Processing P00148.tsv...\n",
      "Read 99027 rows and 13 columns from P00148.tsv\n",
      "📝 Processed P00148.tsv → P00148.slim.tsv\n",
      "\n",
      "[22/223] Processing P00071.tsv...\n",
      "Read 49874 rows and 13 columns from P00071.tsv\n",
      "📝 Processed P00071.tsv → P00071.slim.tsv\n",
      "\n",
      "[23/223] Processing P00018.tsv...\n",
      "Read 164437 rows and 13 columns from P00018.tsv\n",
      "📝 Processed P00018.tsv → P00018.slim.tsv\n",
      "\n",
      "[24/223] Processing P00542.tsv...\n",
      "Read 60835 rows and 13 columns from P00542.tsv\n",
      "📝 Processed P00542.tsv → P00542.slim.tsv\n",
      "\n",
      "[25/223] Processing P00575.tsv...\n",
      "Read 17673 rows and 13 columns from P00575.tsv\n",
      "📝 Processed P00575.tsv → P00575.slim.tsv\n",
      "\n",
      "[26/223] Processing P00178.tsv...\n",
      "Read 131246 rows and 13 columns from P00178.tsv\n",
      "📝 Processed P00178.tsv → P00178.slim.tsv\n",
      "\n",
      "[27/223] Processing P00120.tsv...\n",
      "Read 48163 rows and 13 columns from P00120.tsv\n",
      "📝 Processed P00120.tsv → P00120.slim.tsv\n",
      "\n",
      "[28/223] Processing P00365.tsv...\n",
      "Read 135388 rows and 13 columns from P00365.tsv\n",
      "📝 Processed P00365.tsv → P00365.slim.tsv\n",
      "\n",
      "[29/223] Processing P00303.tsv...\n",
      "Read 74993 rows and 13 columns from P00303.tsv\n",
      "📝 Processed P00303.tsv → P00303.slim.tsv\n",
      "\n",
      "[30/223] Processing P00139.tsv...\n",
      "Read 96490 rows and 13 columns from P00139.tsv\n",
      "📝 Processed P00139.tsv → P00139.slim.tsv\n",
      "\n",
      "[31/223] Processing P00190.tsv...\n",
      "Read 112191 rows and 13 columns from P00190.tsv\n",
      "📝 Processed P00190.tsv → P00190.slim.tsv\n",
      "\n",
      "[32/223] Processing P00199.tsv...\n",
      "Read 194696 rows and 13 columns from P00199.tsv\n",
      "📝 Processed P00199.tsv → P00199.slim.tsv\n",
      "\n",
      "[33/223] Processing P00161.tsv...\n",
      "Read 231049 rows and 13 columns from P00161.tsv\n",
      "📝 Processed P00161.tsv → P00161.slim.tsv\n",
      "\n",
      "[34/223] Processing P00309.tsv...\n",
      "Read 173545 rows and 13 columns from P00309.tsv\n",
      "📝 Processed P00309.tsv → P00309.slim.tsv\n",
      "\n",
      "[35/223] Processing P00533.tsv...\n",
      "Read 258616 rows and 13 columns from P00533.tsv\n",
      "📝 Processed P00533.tsv → P00533.slim.tsv\n",
      "\n",
      "[36/223] Processing P00301.tsv...\n",
      "Read 256585 rows and 13 columns from P00301.tsv\n",
      "📝 Processed P00301.tsv → P00301.slim.tsv\n",
      "\n",
      "[37/223] Processing P00361.tsv...\n",
      "Read 193482 rows and 13 columns from P00361.tsv\n",
      "📝 Processed P00361.tsv → P00361.slim.tsv\n",
      "\n",
      "[38/223] Processing P00540.tsv...\n",
      "Read 217290 rows and 13 columns from P00540.tsv\n",
      "📝 Processed P00540.tsv → P00540.slim.tsv\n",
      "\n",
      "[39/223] Processing P00002.tsv...\n",
      "Read 115919 rows and 13 columns from P00002.tsv\n",
      "📝 Processed P00002.tsv → P00002.slim.tsv\n",
      "\n",
      "[40/223] Processing P00243.tsv...\n",
      "Read 209471 rows and 13 columns from P00243.tsv\n",
      "📝 Processed P00243.tsv → P00243.slim.tsv\n",
      "\n",
      "[41/223] Processing P00546.tsv...\n",
      "Read 252635 rows and 13 columns from P00546.tsv\n",
      "📝 Processed P00546.tsv → P00546.slim.tsv\n",
      "\n",
      "[42/223] Processing P00387.tsv...\n",
      "Read 126631 rows and 13 columns from P00387.tsv\n",
      "📝 Processed P00387.tsv → P00387.slim.tsv\n",
      "\n",
      "[43/223] Processing P00424.tsv...\n",
      "Read 205900 rows and 13 columns from P00424.tsv\n",
      "📝 Processed P00424.tsv → P00424.slim.tsv\n",
      "\n",
      "[44/223] Processing P00545.tsv...\n",
      "Read 322748 rows and 13 columns from P00545.tsv\n",
      "📝 Processed P00545.tsv → P00545.slim.tsv\n",
      "\n",
      "[45/223] Processing P00571.tsv...\n",
      "Read 269862 rows and 13 columns from P00571.tsv\n",
      "📝 Processed P00571.tsv → P00571.slim.tsv\n",
      "\n",
      "[46/223] Processing P00581.tsv...\n",
      "Read 360025 rows and 13 columns from P00581.tsv\n",
      "📝 Processed P00581.tsv → P00581.slim.tsv\n",
      "\n",
      "[47/223] Processing P00272.tsv...\n",
      "Read 330239 rows and 13 columns from P00272.tsv\n",
      "📝 Processed P00272.tsv → P00272.slim.tsv\n",
      "\n",
      "[48/223] Processing P00373.tsv...\n",
      "Read 219036 rows and 13 columns from P00373.tsv\n",
      "📝 Processed P00373.tsv → P00373.slim.tsv\n",
      "\n",
      "[49/223] Processing P00446.tsv...\n",
      "Read 64230 rows and 13 columns from P00446.tsv\n",
      "📝 Processed P00446.tsv → P00446.slim.tsv\n",
      "\n",
      "[50/223] Processing P00368.tsv...\n",
      "Read 265991 rows and 13 columns from P00368.tsv\n",
      "📝 Processed P00368.tsv → P00368.slim.tsv\n",
      "\n",
      "[51/223] Processing P00457.tsv...\n",
      "Read 218890 rows and 13 columns from P00457.tsv\n",
      "📝 Processed P00457.tsv → P00457.slim.tsv\n",
      "\n",
      "[52/223] Processing P00554.tsv...\n",
      "Read 401403 rows and 13 columns from P00554.tsv\n",
      "📝 Processed P00554.tsv → P00554.slim.tsv\n",
      "\n",
      "[53/223] Processing P00557.tsv...\n",
      "Read 161435 rows and 13 columns from P00557.tsv\n",
      "📝 Processed P00557.tsv → P00557.slim.tsv\n",
      "\n",
      "[54/223] Processing P00582.tsv...\n",
      "Read 271889 rows and 13 columns from P00582.tsv\n",
      "📝 Processed P00582.tsv → P00582.slim.tsv\n",
      "\n",
      "[55/223] Processing P00544.tsv...\n",
      "Read 220382 rows and 13 columns from P00544.tsv\n",
      "📝 Processed P00544.tsv → P00544.slim.tsv\n",
      "\n",
      "[56/223] Processing P00576.tsv...\n",
      "Read 110792 rows and 13 columns from P00576.tsv\n",
      "📝 Processed P00576.tsv → P00576.slim.tsv\n",
      "\n",
      "[57/223] Processing P00556.tsv...\n",
      "Read 337257 rows and 13 columns from P00556.tsv\n",
      "📝 Processed P00556.tsv → P00556.slim.tsv\n",
      "\n",
      "[58/223] Processing P00016.tsv...\n",
      "Read 297977 rows and 13 columns from P00016.tsv\n",
      "📝 Processed P00016.tsv → P00016.slim.tsv\n",
      "\n",
      "[59/223] Processing P00467.tsv...\n",
      "Read 284760 rows and 13 columns from P00467.tsv\n",
      "📝 Processed P00467.tsv → P00467.slim.tsv\n",
      "\n",
      "[60/223] Processing P00104.tsv...\n",
      "Read 404802 rows and 13 columns from P00104.tsv\n",
      "📝 Processed P00104.tsv → P00104.slim.tsv\n",
      "\n",
      "[61/223] Processing P00155.tsv...\n",
      "Read 337792 rows and 13 columns from P00155.tsv\n",
      "📝 Processed P00155.tsv → P00155.slim.tsv\n",
      "\n",
      "[62/223] Processing P00485.tsv...\n",
      "Read 235103 rows and 13 columns from P00485.tsv\n",
      "📝 Processed P00485.tsv → P00485.slim.tsv\n",
      "\n",
      "[63/223] Processing P00512.tsv...\n",
      "Read 236415 rows and 13 columns from P00512.tsv\n",
      "📝 Processed P00512.tsv → P00512.slim.tsv\n",
      "\n",
      "[64/223] Processing P00081.tsv...\n",
      "Read 254651 rows and 13 columns from P00081.tsv\n",
      "📝 Processed P00081.tsv → P00081.slim.tsv\n",
      "\n",
      "[65/223] Processing P00488.tsv...\n",
      "Read 74875 rows and 13 columns from P00488.tsv\n",
      "📝 Processed P00488.tsv → P00488.slim.tsv\n",
      "\n",
      "[66/223] Processing P00611.tsv...\n",
      "Read 308687 rows and 13 columns from P00611.tsv\n",
      "📝 Processed P00611.tsv → P00611.slim.tsv\n",
      "\n",
      "[67/223] Processing P00603.tsv...\n",
      "Read 77954 rows and 13 columns from P00603.tsv\n",
      "📝 Processed P00603.tsv → P00603.slim.tsv\n",
      "\n",
      "[68/223] Processing P00494.tsv...\n",
      "Read 171000 rows and 13 columns from P00494.tsv\n",
      "📝 Processed P00494.tsv → P00494.slim.tsv\n",
      "\n",
      "[69/223] Processing P00522.tsv...\n",
      "Read 274167 rows and 13 columns from P00522.tsv\n",
      "📝 Processed P00522.tsv → P00522.slim.tsv\n",
      "\n",
      "[70/223] Processing P00466.tsv...\n",
      "Read 217587 rows and 13 columns from P00466.tsv\n",
      "📝 Processed P00466.tsv → P00466.slim.tsv\n",
      "\n",
      "[71/223] Processing P00600.tsv...\n",
      "Read 187435 rows and 13 columns from P00600.tsv\n",
      "📝 Processed P00600.tsv → P00600.slim.tsv\n",
      "\n",
      "[72/223] Processing P00521.tsv...\n",
      "Read 159266 rows and 13 columns from P00521.tsv\n",
      "📝 Processed P00521.tsv → P00521.slim.tsv\n",
      "\n",
      "[73/223] Processing P00515.tsv...\n",
      "Read 68798 rows and 13 columns from P00515.tsv\n",
      "📝 Processed P00515.tsv → P00515.slim.tsv\n",
      "\n",
      "[74/223] Processing P00517.tsv...\n",
      "Read 259136 rows and 13 columns from P00517.tsv\n",
      "📝 Processed P00517.tsv → P00517.slim.tsv\n",
      "\n",
      "[75/223] Processing P00470.tsv...\n",
      "Read 171227 rows and 13 columns from P00470.tsv\n",
      "📝 Processed P00470.tsv → P00470.slim.tsv\n",
      "\n",
      "[76/223] Processing P00595.tsv...\n",
      "Read 300027 rows and 13 columns from P00595.tsv\n",
      "📝 Processed P00595.tsv → P00595.slim.tsv\n",
      "\n",
      "[77/223] Processing P00069.tsv...\n",
      "Read 254507 rows and 13 columns from P00069.tsv\n",
      "📝 Processed P00069.tsv → P00069.slim.tsv\n",
      "\n",
      "[78/223] Processing P00218.tsv...\n",
      "Read 226260 rows and 13 columns from P00218.tsv\n",
      "📝 Processed P00218.tsv → P00218.slim.tsv\n",
      "\n",
      "[79/223] Processing P00236.tsv...\n",
      "Read 387364 rows and 13 columns from P00236.tsv\n",
      "📝 Processed P00236.tsv → P00236.slim.tsv\n",
      "\n",
      "[80/223] Processing P00080.tsv...\n",
      "Read 358816 rows and 13 columns from P00080.tsv\n",
      "📝 Processed P00080.tsv → P00080.slim.tsv\n",
      "\n",
      "[81/223] Processing P00239.tsv...\n",
      "Read 319860 rows and 13 columns from P00239.tsv\n",
      "📝 Processed P00239.tsv → P00239.slim.tsv\n",
      "\n",
      "[82/223] Processing P00329.tsv...\n",
      "Read 269182 rows and 13 columns from P00329.tsv\n",
      "📝 Processed P00329.tsv → P00329.slim.tsv\n",
      "\n",
      "[83/223] Processing P00169.tsv...\n",
      "Read 356464 rows and 13 columns from P00169.tsv\n",
      "📝 Processed P00169.tsv → P00169.slim.tsv\n",
      "\n",
      "[84/223] Processing P00108.tsv...\n",
      "Read 272019 rows and 13 columns from P00108.tsv\n",
      "📝 Processed P00108.tsv → P00108.slim.tsv\n",
      "\n",
      "[85/223] Processing P00105.tsv...\n",
      "Read 52989 rows and 13 columns from P00105.tsv\n",
      "📝 Processed P00105.tsv → P00105.slim.tsv\n",
      "\n",
      "[86/223] Processing P00192.tsv...\n",
      "Read 89098 rows and 13 columns from P00192.tsv\n",
      "📝 Processed P00192.tsv → P00192.slim.tsv\n",
      "\n",
      "[87/223] Processing P00252.tsv...\n",
      "Read 370704 rows and 13 columns from P00252.tsv\n",
      "📝 Processed P00252.tsv → P00252.slim.tsv\n",
      "\n",
      "[88/223] Processing P00175.tsv...\n",
      "Read 318576 rows and 13 columns from P00175.tsv\n",
      "📝 Processed P00175.tsv → P00175.slim.tsv\n",
      "\n",
      "[89/223] Processing P00230.tsv...\n",
      "Read 411496 rows and 13 columns from P00230.tsv\n",
      "📝 Processed P00230.tsv → P00230.slim.tsv\n",
      "\n",
      "[90/223] Processing P00056.tsv...\n",
      "Read 191541 rows and 13 columns from P00056.tsv\n",
      "📝 Processed P00056.tsv → P00056.slim.tsv\n",
      "\n",
      "[91/223] Processing P00257.tsv...\n",
      "Read 292185 rows and 13 columns from P00257.tsv\n",
      "📝 Processed P00257.tsv → P00257.slim.tsv\n",
      "\n",
      "[92/223] Processing P00115.tsv...\n",
      "Read 314312 rows and 13 columns from P00115.tsv\n",
      "📝 Processed P00115.tsv → P00115.slim.tsv\n",
      "\n",
      "[93/223] Processing P00265.tsv...\n",
      "Read 356571 rows and 13 columns from P00265.tsv\n",
      "📝 Processed P00265.tsv → P00265.slim.tsv\n",
      "\n",
      "[94/223] Processing P00166.tsv...\n",
      "Read 180670 rows and 13 columns from P00166.tsv\n",
      "📝 Processed P00166.tsv → P00166.slim.tsv\n",
      "\n",
      "[95/223] Processing P00552.tsv...\n",
      "Read 282001 rows and 13 columns from P00552.tsv\n",
      "📝 Processed P00552.tsv → P00552.slim.tsv\n",
      "\n",
      "[96/223] Processing P00497.tsv...\n",
      "Read 470362 rows and 13 columns from P00497.tsv\n",
      "📝 Processed P00497.tsv → P00497.slim.tsv\n",
      "\n",
      "[97/223] Processing P00375.tsv...\n",
      "Read 445329 rows and 13 columns from P00375.tsv\n",
      "📝 Processed P00375.tsv → P00375.slim.tsv\n",
      "\n",
      "[98/223] Processing P00128.tsv...\n",
      "Read 386072 rows and 13 columns from P00128.tsv\n",
      "📝 Processed P00128.tsv → P00128.slim.tsv\n",
      "\n",
      "[99/223] Processing P00147.tsv...\n",
      "Read 151016 rows and 13 columns from P00147.tsv\n",
      "📝 Processed P00147.tsv → P00147.slim.tsv\n",
      "\n",
      "[100/223] Processing P00129.tsv...\n",
      "Read 248596 rows and 13 columns from P00129.tsv\n",
      "📝 Processed P00129.tsv → P00129.slim.tsv\n",
      "\n",
      "[101/223] Processing P00454.tsv...\n",
      "Read 473359 rows and 13 columns from P00454.tsv\n",
      "📝 Processed P00454.tsv → P00454.slim.tsv\n",
      "\n",
      "[102/223] Processing P00124.tsv...\n",
      "Read 382105 rows and 13 columns from P00124.tsv\n",
      "📝 Processed P00124.tsv → P00124.slim.tsv\n",
      "\n",
      "[103/223] Processing P00211.tsv...\n",
      "Read 328357 rows and 13 columns from P00211.tsv\n",
      "📝 Processed P00211.tsv → P00211.slim.tsv\n",
      "\n",
      "[104/223] Processing P00118.tsv...\n",
      "Read 222097 rows and 13 columns from P00118.tsv\n",
      "📝 Processed P00118.tsv → P00118.slim.tsv\n",
      "\n",
      "[105/223] Processing P00137.tsv...\n",
      "Read 150008 rows and 13 columns from P00137.tsv\n",
      "📝 Processed P00137.tsv → P00137.slim.tsv\n",
      "\n",
      "[106/223] Processing P00505.tsv...\n",
      "Read 302180 rows and 13 columns from P00505.tsv\n",
      "📝 Processed P00505.tsv → P00505.slim.tsv\n",
      "\n",
      "[107/223] Processing P00354.tsv...\n",
      "Read 351298 rows and 13 columns from P00354.tsv\n",
      "📝 Processed P00354.tsv → P00354.slim.tsv\n",
      "\n",
      "[108/223] Processing P00117.tsv...\n",
      "Read 75725 rows and 13 columns from P00117.tsv\n",
      "📝 Processed P00117.tsv → P00117.slim.tsv\n",
      "\n",
      "[109/223] Processing P00359.tsv...\n",
      "Read 137090 rows and 13 columns from P00359.tsv\n",
      "📝 Processed P00359.tsv → P00359.slim.tsv\n",
      "\n",
      "[110/223] Processing P00349.tsv...\n",
      "Read 419633 rows and 13 columns from P00349.tsv\n",
      "📝 Processed P00349.tsv → P00349.slim.tsv\n",
      "\n",
      "[111/223] Processing P00389.tsv...\n",
      "Read 419433 rows and 13 columns from P00389.tsv\n",
      "📝 Processed P00389.tsv → P00389.slim.tsv\n",
      "\n",
      "[112/223] Processing P00511.tsv...\n",
      "Read 341879 rows and 13 columns from P00511.tsv\n",
      "📝 Processed P00511.tsv → P00511.slim.tsv\n",
      "\n",
      "[113/223] Processing P00599.tsv...\n",
      "Read 336436 rows and 13 columns from P00599.tsv\n",
      "📝 Processed P00599.tsv → P00599.slim.tsv\n",
      "\n",
      "[114/223] Processing P00273.tsv...\n",
      "Read 249311 rows and 13 columns from P00273.tsv\n",
      "📝 Processed P00273.tsv → P00273.slim.tsv\n",
      "\n",
      "[115/223] Processing P00275.tsv...\n",
      "Read 190175 rows and 13 columns from P00275.tsv\n",
      "📝 Processed P00275.tsv → P00275.slim.tsv\n",
      "\n",
      "[116/223] Processing P00478.tsv...\n",
      "Read 173173 rows and 13 columns from P00478.tsv\n",
      "📝 Processed P00478.tsv → P00478.slim.tsv\n",
      "\n",
      "[117/223] Processing P00434.tsv...\n",
      "Read 274544 rows and 13 columns from P00434.tsv\n",
      "📝 Processed P00434.tsv → P00434.slim.tsv\n",
      "\n",
      "[118/223] Processing P00615.tsv...\n",
      "Read 278533 rows and 13 columns from P00615.tsv\n",
      "📝 Processed P00615.tsv → P00615.slim.tsv\n",
      "\n",
      "[119/223] Processing P00547.tsv...\n",
      "Read 221386 rows and 13 columns from P00547.tsv\n",
      "📝 Processed P00547.tsv → P00547.slim.tsv\n",
      "\n",
      "[120/223] Processing P00605.tsv...\n",
      "Read 236771 rows and 13 columns from P00605.tsv\n",
      "📝 Processed P00605.tsv → P00605.slim.tsv\n",
      "\n",
      "[121/223] Processing P00338.tsv...\n",
      "Read 167818 rows and 13 columns from P00338.tsv\n",
      "📝 Processed P00338.tsv → P00338.slim.tsv\n",
      "\n",
      "[122/223] Processing P00455.tsv...\n",
      "Read 224997 rows and 13 columns from P00455.tsv\n",
      "📝 Processed P00455.tsv → P00455.slim.tsv\n",
      "\n",
      "[123/223] Processing P00657.tsv...\n",
      "Read 251413 rows and 13 columns from P00657.tsv\n",
      "📝 Processed P00657.tsv → P00657.slim.tsv\n",
      "\n",
      "[124/223] Processing P00180.tsv...\n",
      "Read 213933 rows and 13 columns from P00180.tsv\n",
      "📝 Processed P00180.tsv → P00180.slim.tsv\n",
      "\n",
      "[125/223] Processing P00420.tsv...\n",
      "Read 209249 rows and 13 columns from P00420.tsv\n",
      "📝 Processed P00420.tsv → P00420.slim.tsv\n",
      "\n",
      "[126/223] Processing P00648.tsv...\n",
      "Read 297058 rows and 13 columns from P00648.tsv\n",
      "📝 Processed P00648.tsv → P00648.slim.tsv\n",
      "\n",
      "[127/223] Processing P00383.tsv...\n",
      "Read 224639 rows and 13 columns from P00383.tsv\n",
      "📝 Processed P00383.tsv → P00383.slim.tsv\n",
      "\n",
      "[128/223] Processing P00580.tsv...\n",
      "Read 187208 rows and 13 columns from P00580.tsv\n",
      "📝 Processed P00580.tsv → P00580.slim.tsv\n",
      "\n",
      "[129/223] Processing P00376.tsv...\n",
      "Read 392580 rows and 13 columns from P00376.tsv\n",
      "📝 Processed P00376.tsv → P00376.slim.tsv\n",
      "\n",
      "[130/223] Processing P00355.tsv...\n",
      "Read 217409 rows and 13 columns from P00355.tsv\n",
      "📝 Processed P00355.tsv → P00355.slim.tsv\n",
      "\n",
      "[131/223] Processing P00524.tsv...\n",
      "Read 227163 rows and 13 columns from P00524.tsv\n",
      "📝 Processed P00524.tsv → P00524.slim.tsv\n",
      "\n",
      "[132/223] Processing P00356.tsv...\n",
      "Read 285537 rows and 13 columns from P00356.tsv\n",
      "📝 Processed P00356.tsv → P00356.slim.tsv\n",
      "\n",
      "[133/223] Processing P00617.tsv...\n",
      "Read 341828 rows and 13 columns from P00617.tsv\n",
      "📝 Processed P00617.tsv → P00617.slim.tsv\n",
      "\n",
      "[134/223] Processing P00025.tsv...\n",
      "Read 266994 rows and 13 columns from P00025.tsv\n",
      "📝 Processed P00025.tsv → P00025.slim.tsv\n",
      "\n",
      "[135/223] Processing P00465.tsv...\n",
      "Read 156297 rows and 13 columns from P00465.tsv\n",
      "📝 Processed P00465.tsv → P00465.slim.tsv\n",
      "\n",
      "[136/223] Processing P00039.tsv...\n",
      "Read 309878 rows and 13 columns from P00039.tsv\n",
      "📝 Processed P00039.tsv → P00039.slim.tsv\n",
      "\n",
      "[137/223] Processing P00238.tsv...\n",
      "Read 374602 rows and 13 columns from P00238.tsv\n",
      "📝 Processed P00238.tsv → P00238.slim.tsv\n",
      "\n",
      "[138/223] Processing P00036.tsv...\n",
      "Read 311298 rows and 13 columns from P00036.tsv\n",
      "📝 Processed P00036.tsv → P00036.slim.tsv\n",
      "\n",
      "[139/223] Processing P00102.tsv...\n",
      "Read 287833 rows and 13 columns from P00102.tsv\n",
      "📝 Processed P00102.tsv → P00102.slim.tsv\n",
      "\n",
      "[140/223] Processing P00103.tsv...\n",
      "Read 283037 rows and 13 columns from P00103.tsv\n",
      "📝 Processed P00103.tsv → P00103.slim.tsv\n",
      "\n",
      "[141/223] Processing P00028.tsv...\n",
      "Read 367425 rows and 13 columns from P00028.tsv\n",
      "📝 Processed P00028.tsv → P00028.slim.tsv\n",
      "\n",
      "[142/223] Processing P00141.tsv...\n",
      "Read 313900 rows and 13 columns from P00141.tsv\n",
      "📝 Processed P00141.tsv → P00141.slim.tsv\n",
      "\n",
      "[143/223] Processing P00114.tsv...\n",
      "Read 163156 rows and 13 columns from P00114.tsv\n",
      "📝 Processed P00114.tsv → P00114.slim.tsv\n",
      "\n",
      "[144/223] Processing P00233.tsv...\n",
      "Read 464856 rows and 13 columns from P00233.tsv\n",
      "📝 Processed P00233.tsv → P00233.slim.tsv\n",
      "\n",
      "[145/223] Processing P00013.tsv...\n",
      "Read 238171 rows and 13 columns from P00013.tsv\n",
      "📝 Processed P00013.tsv → P00013.slim.tsv\n",
      "\n",
      "[146/223] Processing P00184.tsv...\n",
      "Read 317177 rows and 13 columns from P00184.tsv\n",
      "📝 Processed P00184.tsv → P00184.slim.tsv\n",
      "\n",
      "[147/223] Processing P00235.tsv...\n",
      "Read 300445 rows and 13 columns from P00235.tsv\n",
      "📝 Processed P00235.tsv → P00235.slim.tsv\n",
      "\n",
      "[148/223] Processing P00164.tsv...\n",
      "Read 140592 rows and 13 columns from P00164.tsv\n",
      "📝 Processed P00164.tsv → P00164.slim.tsv\n",
      "\n",
      "[149/223] Processing P00072.tsv...\n",
      "Read 180087 rows and 13 columns from P00072.tsv\n",
      "📝 Processed P00072.tsv → P00072.slim.tsv\n",
      "\n",
      "[150/223] Processing P00188.tsv...\n",
      "Read 392899 rows and 13 columns from P00188.tsv\n",
      "📝 Processed P00188.tsv → P00188.slim.tsv\n",
      "\n",
      "[151/223] Processing P00113.tsv...\n",
      "Read 255770 rows and 13 columns from P00113.tsv\n",
      "📝 Processed P00113.tsv → P00113.slim.tsv\n",
      "\n",
      "[152/223] Processing P00196.tsv...\n",
      "Read 192489 rows and 13 columns from P00196.tsv\n",
      "📝 Processed P00196.tsv → P00196.slim.tsv\n",
      "\n",
      "[153/223] Processing P00078.tsv...\n",
      "Read 217016 rows and 13 columns from P00078.tsv\n",
      "📝 Processed P00078.tsv → P00078.slim.tsv\n",
      "\n",
      "[154/223] Processing P00176.tsv...\n",
      "Read 328088 rows and 13 columns from P00176.tsv\n",
      "📝 Processed P00176.tsv → P00176.slim.tsv\n",
      "\n",
      "[155/223] Processing P00264.tsv...\n",
      "Read 149375 rows and 13 columns from P00264.tsv\n",
      "📝 Processed P00264.tsv → P00264.slim.tsv\n",
      "\n",
      "[156/223] Processing P00382.tsv...\n",
      "Read 280706 rows and 13 columns from P00382.tsv\n",
      "📝 Processed P00382.tsv → P00382.slim.tsv\n",
      "\n",
      "[157/223] Processing P00263.tsv...\n",
      "Read 367285 rows and 13 columns from P00263.tsv\n",
      "📝 Processed P00263.tsv → P00263.slim.tsv\n",
      "\n",
      "[158/223] Processing P00298.tsv...\n",
      "Read 235998 rows and 13 columns from P00298.tsv\n",
      "📝 Processed P00298.tsv → P00298.slim.tsv\n",
      "\n",
      "[159/223] Processing P00306.tsv...\n",
      "Read 193978 rows and 13 columns from P00306.tsv\n",
      "📝 Processed P00306.tsv → P00306.slim.tsv\n",
      "\n",
      "[160/223] Processing P00333.tsv...\n",
      "Read 366052 rows and 13 columns from P00333.tsv\n",
      "📝 Processed P00333.tsv → P00333.slim.tsv\n",
      "\n",
      "[161/223] Processing P00168.tsv...\n",
      "Read 208964 rows and 13 columns from P00168.tsv\n",
      "📝 Processed P00168.tsv → P00168.slim.tsv\n",
      "\n",
      "[162/223] Processing P00348.tsv...\n",
      "Read 269490 rows and 13 columns from P00348.tsv\n",
      "📝 Processed P00348.tsv → P00348.slim.tsv\n",
      "\n",
      "[163/223] Processing P00062.tsv...\n",
      "Read 249686 rows and 13 columns from P00062.tsv\n",
      "📝 Processed P00062.tsv → P00062.slim.tsv\n",
      "\n",
      "[164/223] Processing P00181.tsv...\n",
      "Read 283424 rows and 13 columns from P00181.tsv\n",
      "📝 Processed P00181.tsv → P00181.slim.tsv\n",
      "\n",
      "[165/223] Processing P00302.tsv...\n",
      "Read 298530 rows and 13 columns from P00302.tsv\n",
      "📝 Processed P00302.tsv → P00302.slim.tsv\n",
      "\n",
      "[166/223] Processing P00345.tsv...\n",
      "Read 239806 rows and 13 columns from P00345.tsv\n",
      "📝 Processed P00345.tsv → P00345.slim.tsv\n",
      "\n",
      "[167/223] Processing P00154.tsv...\n",
      "Read 221593 rows and 13 columns from P00154.tsv\n",
      "📝 Processed P00154.tsv → P00154.slim.tsv\n",
      "\n",
      "[168/223] Processing P00174.tsv...\n",
      "Read 303116 rows and 13 columns from P00174.tsv\n",
      "📝 Processed P00174.tsv → P00174.slim.tsv\n",
      "\n",
      "[169/223] Processing P00047.tsv...\n",
      "Read 252653 rows and 13 columns from P00047.tsv\n",
      "📝 Processed P00047.tsv → P00047.slim.tsv\n",
      "\n",
      "[170/223] Processing P00327.tsv...\n",
      "Read 122079 rows and 13 columns from P00327.tsv\n",
      "📝 Processed P00327.tsv → P00327.slim.tsv\n",
      "\n",
      "[171/223] Processing P00050.tsv...\n",
      "Read 313976 rows and 13 columns from P00050.tsv\n",
      "📝 Processed P00050.tsv → P00050.slim.tsv\n",
      "\n",
      "[172/223] Processing P00067.tsv...\n",
      "Read 320028 rows and 13 columns from P00067.tsv\n",
      "📝 Processed P00067.tsv → P00067.slim.tsv\n",
      "\n",
      "[173/223] Processing P00283.tsv...\n",
      "Read 388638 rows and 13 columns from P00283.tsv\n",
      "📝 Processed P00283.tsv → P00283.slim.tsv\n",
      "\n",
      "[174/223] Processing P00480.tsv...\n",
      "Read 125326 rows and 13 columns from P00480.tsv\n",
      "📝 Processed P00480.tsv → P00480.slim.tsv\n",
      "\n",
      "[175/223] Processing P00482.tsv...\n",
      "Read 212200 rows and 13 columns from P00482.tsv\n",
      "📝 Processed P00482.tsv → P00482.slim.tsv\n",
      "\n",
      "[176/223] Processing P00418.tsv...\n",
      "Read 232322 rows and 13 columns from P00418.tsv\n",
      "📝 Processed P00418.tsv → P00418.slim.tsv\n",
      "\n",
      "[177/223] Processing P00250.tsv...\n",
      "Read 222991 rows and 13 columns from P00250.tsv\n",
      "📝 Processed P00250.tsv → P00250.slim.tsv\n",
      "\n",
      "[178/223] Processing P00415.tsv...\n",
      "Read 238992 rows and 13 columns from P00415.tsv\n",
      "📝 Processed P00415.tsv → P00415.slim.tsv\n",
      "\n",
      "[179/223] Processing P00240.tsv...\n",
      "Read 260718 rows and 13 columns from P00240.tsv\n",
      "📝 Processed P00240.tsv → P00240.slim.tsv\n",
      "\n",
      "[180/223] Processing P00395.tsv...\n",
      "Read 259319 rows and 13 columns from P00395.tsv\n",
      "📝 Processed P00395.tsv → P00395.slim.tsv\n",
      "\n",
      "[181/223] Processing P00215.tsv...\n",
      "Read 334490 rows and 13 columns from P00215.tsv\n",
      "📝 Processed P00215.tsv → P00215.slim.tsv\n",
      "\n",
      "[182/223] Processing P00668.tsv...\n",
      "Read 341001 rows and 13 columns from P00668.tsv\n",
      "📝 Processed P00668.tsv → P00668.slim.tsv\n",
      "\n",
      "[183/223] Processing P00634.tsv...\n",
      "Read 460895 rows and 13 columns from P00634.tsv\n",
      "📝 Processed P00634.tsv → P00634.slim.tsv\n",
      "\n",
      "[184/223] Processing P00228.tsv...\n",
      "Read 291141 rows and 13 columns from P00228.tsv\n",
      "📝 Processed P00228.tsv → P00228.slim.tsv\n",
      "\n",
      "[185/223] Processing P00652.tsv...\n",
      "Read 338037 rows and 13 columns from P00652.tsv\n",
      "📝 Processed P00652.tsv → P00652.slim.tsv\n",
      "\n",
      "[186/223] Processing P00656.tsv...\n",
      "Read 255775 rows and 13 columns from P00656.tsv\n",
      "📝 Processed P00656.tsv → P00656.slim.tsv\n",
      "\n",
      "[187/223] Processing P00665.tsv...\n",
      "Read 67186 rows and 13 columns from P00665.tsv\n",
      "📝 Processed P00665.tsv → P00665.slim.tsv\n",
      "\n",
      "[188/223] Processing P00286.tsv...\n",
      "Read 453794 rows and 13 columns from P00286.tsv\n",
      "📝 Processed P00286.tsv → P00286.slim.tsv\n",
      "\n",
      "[189/223] Processing P00628.tsv...\n",
      "Read 310208 rows and 13 columns from P00628.tsv\n",
      "📝 Processed P00628.tsv → P00628.slim.tsv\n",
      "\n",
      "[190/223] Processing P00491.tsv...\n",
      "Read 211843 rows and 13 columns from P00491.tsv\n",
      "📝 Processed P00491.tsv → P00491.slim.tsv\n",
      "\n",
      "[191/223] Processing P00639.tsv...\n",
      "Read 340570 rows and 13 columns from P00639.tsv\n",
      "📝 Processed P00639.tsv → P00639.slim.tsv\n",
      "\n",
      "[192/223] Processing P00538.tsv...\n",
      "Read 253850 rows and 13 columns from P00538.tsv\n",
      "📝 Processed P00538.tsv → P00538.slim.tsv\n",
      "\n",
      "[193/223] Processing P00399.tsv...\n",
      "Read 421237 rows and 13 columns from P00399.tsv\n",
      "📝 Processed P00399.tsv → P00399.slim.tsv\n",
      "\n",
      "[194/223] Processing P00662.tsv...\n",
      "Read 259959 rows and 13 columns from P00662.tsv\n",
      "📝 Processed P00662.tsv → P00662.slim.tsv\n",
      "\n",
      "[195/223] Processing P00584.tsv...\n",
      "Read 248192 rows and 13 columns from P00584.tsv\n",
      "📝 Processed P00584.tsv → P00584.slim.tsv\n",
      "\n",
      "[196/223] Processing P00651.tsv...\n",
      "Read 266080 rows and 13 columns from P00651.tsv\n",
      "📝 Processed P00651.tsv → P00651.slim.tsv\n",
      "\n",
      "[197/223] Processing P00659.tsv...\n",
      "Read 232236 rows and 13 columns from P00659.tsv\n",
      "📝 Processed P00659.tsv → P00659.slim.tsv\n",
      "\n",
      "[198/223] Processing P00284.tsv...\n",
      "Read 297607 rows and 13 columns from P00284.tsv\n",
      "📝 Processed P00284.tsv → P00284.slim.tsv\n",
      "\n",
      "[199/223] Processing P00406.tsv...\n",
      "Read 398714 rows and 13 columns from P00406.tsv\n",
      "📝 Processed P00406.tsv → P00406.slim.tsv\n",
      "\n",
      "[200/223] Processing P00507.tsv...\n",
      "Read 85837 rows and 13 columns from P00507.tsv\n",
      "📝 Processed P00507.tsv → P00507.slim.tsv\n",
      "\n",
      "[201/223] Processing P00653.tsv...\n",
      "Read 280770 rows and 13 columns from P00653.tsv\n",
      "📝 Processed P00653.tsv → P00653.slim.tsv\n",
      "\n",
      "[202/223] Processing P00342.tsv...\n",
      "Read 313233 rows and 13 columns from P00342.tsv\n",
      "📝 Processed P00342.tsv → P00342.slim.tsv\n",
      "\n",
      "[203/223] Processing P00642.tsv...\n",
      "Read 231936 rows and 13 columns from P00642.tsv\n",
      "📝 Processed P00642.tsv → P00642.slim.tsv\n",
      "\n",
      "[204/223] Processing P00616.tsv...\n",
      "Read 299354 rows and 13 columns from P00616.tsv\n",
      "📝 Processed P00616.tsv → P00616.slim.tsv\n",
      "\n",
      "[205/223] Processing P00566.tsv...\n",
      "Read 232844 rows and 13 columns from P00566.tsv\n",
      "📝 Processed P00566.tsv → P00566.slim.tsv\n",
      "\n",
      "[206/223] Processing P00531.tsv...\n",
      "Read 266591 rows and 13 columns from P00531.tsv\n",
      "📝 Processed P00531.tsv → P00531.slim.tsv\n",
      "\n",
      "[207/223] Processing P00049.tsv...\n",
      "Read 102017 rows and 13 columns from P00049.tsv\n",
      "📝 Processed P00049.tsv → P00049.slim.tsv\n",
      "\n",
      "[208/223] Processing P00227.tsv...\n",
      "Read 242389 rows and 13 columns from P00227.tsv\n",
      "📝 Processed P00227.tsv → P00227.slim.tsv\n",
      "\n",
      "[209/223] Processing P00461.tsv...\n",
      "Read 157436 rows and 13 columns from P00461.tsv\n",
      "📝 Processed P00461.tsv → P00461.slim.tsv\n",
      "\n",
      "[210/223] Processing P00529.tsv...\n",
      "Read 105882 rows and 13 columns from P00529.tsv\n",
      "📝 Processed P00529.tsv → P00529.slim.tsv\n",
      "\n",
      "[211/223] Processing P00481.tsv...\n",
      "Read 161169 rows and 13 columns from P00481.tsv\n",
      "📝 Processed P00481.tsv → P00481.slim.tsv\n",
      "\n",
      "[212/223] Processing P00411.tsv...\n",
      "Read 292766 rows and 13 columns from P00411.tsv\n",
      "📝 Processed P00411.tsv → P00411.slim.tsv\n",
      "\n",
      "[213/223] Processing P00520.tsv...\n",
      "Read 363702 rows and 13 columns from P00520.tsv\n",
      "📝 Processed P00520.tsv → P00520.slim.tsv\n",
      "\n",
      "[214/223] Processing P00447.tsv...\n",
      "Read 236598 rows and 13 columns from P00447.tsv\n",
      "📝 Processed P00447.tsv → P00447.slim.tsv\n",
      "\n",
      "[215/223] Processing P00469.tsv...\n",
      "Read 149119 rows and 13 columns from P00469.tsv\n",
      "📝 Processed P00469.tsv → P00469.slim.tsv\n",
      "\n",
      "[216/223] Processing P00513.tsv...\n",
      "Read 306691 rows and 13 columns from P00513.tsv\n",
      "📝 Processed P00513.tsv → P00513.slim.tsv\n",
      "\n",
      "[217/223] Processing P00206.tsv...\n",
      "Read 111659 rows and 13 columns from P00206.tsv\n",
      "📝 Processed P00206.tsv → P00206.slim.tsv\n",
      "\n",
      "[218/223] Processing P00369.tsv...\n",
      "Read 234627 rows and 13 columns from P00369.tsv\n",
      "📝 Processed P00369.tsv → P00369.slim.tsv\n",
      "\n",
      "[219/223] Processing P00558.tsv...\n",
      "Read 310173 rows and 13 columns from P00558.tsv\n",
      "📝 Processed P00558.tsv → P00558.slim.tsv\n",
      "\n",
      "[220/223] Processing P00384.tsv...\n",
      "Read 429849 rows and 13 columns from P00384.tsv\n",
      "📝 Processed P00384.tsv → P00384.slim.tsv\n",
      "\n",
      "[221/223] Processing P00197.tsv...\n",
      "Read 307882 rows and 13 columns from P00197.tsv\n",
      "📝 Processed P00197.tsv → P00197.slim.tsv\n",
      "\n",
      "[222/223] Processing P00237.tsv...\n",
      "Read 115853 rows and 13 columns from P00237.tsv\n",
      "📝 Processed P00237.tsv → P00237.slim.tsv\n",
      "\n",
      "[223/223] Processing P00289.tsv...\n",
      "Read 82691 rows and 13 columns from P00289.tsv\n",
      "📝 Processed P00289.tsv → P00289.slim.tsv\n",
      "\n",
      " Done. Slim files and combined metadata written to: /Volumes/IshaVerbat/Isha/TCR/Filtered_HLA-A01\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "# === USER SETTINGS ===\n",
    "INPUT_DIR    = Path(\"/Volumes/IshaVerbat/Isha/TCR/All_Emerson_Cohort01\")      # your raw .tsv files\n",
    "OUTPUT_DIR   = Path(\"/Volumes/IshaVerbat/Isha/TCR/Filtered_HLA-A01\")    # where slim TSVs & metadata CSV will go\n",
    "CHAIN        = \"beta\"                          # always \"beta\"\n",
    "ID_COLUMN    = \"sample_name\"                   # the column that uniquely IDs each file/patient\n",
    "DATA_COLUMNS = [\n",
    "    \"rearrangement\",\n",
    "    \"amino_acid\",           # CDR3 sequence\n",
    "    \"v_gene\",               # V gene\n",
    "    \"d_gene\",               # D gene\n",
    "    \"j_gene\",               # J gene\n",
    "    \"frequency\",            # clone freq\n",
    "    \"productive_frequency\", # productive freq\n",
    "]\n",
    "METADATA_COLUMNS = [\n",
    "    \"total_templates\",\n",
    "    \"productive_templates\",\n",
    "    \"total_reads\",\n",
    "    \"total_productive_reads\",\n",
    "    \"fraction_productive\"\n",
    "    \n",
    "    # add more if you need them\n",
    "]\n",
    "# =====================\n",
    "\n",
    "print(\"📄 Found the following TSV files for processing:\")\n",
    "\n",
    "tsv_files = []\n",
    "\n",
    "for fname in file_names:\n",
    "    print(f\" Checking file: {fname}\")\n",
    "    \n",
    "    if not fname.endswith(\".tsv\"):\n",
    "        print(f\" Skipped (not .tsv): {fname}\")\n",
    "        continue\n",
    "    if fname.startswith(\".\") or fname.startswith(\"._\"):\n",
    "        print(f\" Skipped (hidden or Apple metadata): {fname}\")\n",
    "        continue\n",
    "\n",
    "    full_path = INPUT_DIR / fname\n",
    "    print(f\"  ✅ Accepted: {full_path}\")\n",
    "    tsv_files.append(full_path)\n",
    "\n",
    "\n",
    "print(\"Processing the following TSV files\")\n",
    "for f in tsv_files:\n",
    "    print(f\"  {f.name}\")\n",
    "print(f\"\\n{len(tsv_files)} files found in {INPUT_DIR}.\")\n",
    "\n",
    "OUTPUT_DIR.mkdir(parents=True, exist_ok=True)\n",
    "meta_csv = OUTPUT_DIR / \"tcrdist_combined_metadata.csv\"\n",
    "\n",
    "# Load any existing metadata so we can append new rows\n",
    "if meta_csv.exists():\n",
    "    metadata_df = pd.read_csv(meta_csv, index_col=ID_COLUMN)\n",
    "    completed_ids = set(metadata_df.index)\n",
    "else:\n",
    "    metadata_df = pd.DataFrame(\n",
    "        columns=METADATA_COLUMNS,\n",
    "    )\n",
    "    metadata_df.index.name = ID_COLUMN\n",
    "    completed_ids = set()\n",
    "\n",
    "# How to rename columns for tcrdist3 β-chain\n",
    "rename_map = {\n",
    "    \"amino_acid\":           \"cdr3_b_aa\",\n",
    "    \"v_gene\":               \"v_b_gene\",\n",
    "    \"j_gene\":               \"j_b_gene\",\n",
    "    # d_gene, frequency, productive_frequency stay unchanged\n",
    "}\n",
    "\n",
    "# Process files one by one\n",
    "for idx, in_path in enumerate(tsv_files, start=1):\n",
    "    print(f\"\\n[{idx}/{len(tsv_files)}] Processing {in_path.name}...\")\n",
    "    \n",
    "    # Read just the sample ID first to check if we’ve already processed this\n",
    "    try:\n",
    "        sample_id = pd.read_csv(in_path, sep=\"\\t\", usecols=[ID_COLUMN], nrows=1)[ID_COLUMN][0]\n",
    "    except Exception as e:\n",
    "        print(f\"Could not read sample ID from {in_path.name}: {e}\")\n",
    "        continue\n",
    "\n",
    "    if sample_id in completed_ids:\n",
    "        print(f\"Skipping {in_path.name} (already processed)\")\n",
    "        continue\n",
    "\n",
    "    try:\n",
    "        # Load only relevant columns\n",
    "        usecols = [ID_COLUMN] + DATA_COLUMNS + METADATA_COLUMNS\n",
    "        df = pd.read_csv(in_path, sep=\"\\t\", usecols=usecols, dtype=str, low_memory=False)\n",
    "        print(f'Read {df.shape[0]} rows and {df.shape[1]} columns from {in_path.name}')\n",
    "\n",
    "        # Extract metadata\n",
    "        meta = {ID_COLUMN: sample_id}\n",
    "        for col in METADATA_COLUMNS:\n",
    "            meta[col] = df[col].iat[0] if df[col].nunique(dropna=False) == 1 else df[col].iat[0]\n",
    "            df.drop(columns=col, inplace=True)\n",
    "\n",
    "        # Remove sample ID from table and store metadata\n",
    "        df.drop(columns=[ID_COLUMN], inplace=True)\n",
    "        df.rename(columns=rename_map, inplace=True)\n",
    "        df[\"frequency\"] = df[\"frequency\"].astype(float)\n",
    "        df[\"productive_frequency\"] = df[\"productive_frequency\"].astype(float)\n",
    "\n",
    "        # Save slimmed TSV\n",
    "        out_tsv = OUTPUT_DIR / in_path.name.replace(\".tsv\", \".slim.tsv\")\n",
    "        df.to_csv(out_tsv, sep=\"\\t\", index=False)\n",
    "\n",
    "        # Append metadata\n",
    "        metadata_df.loc[sample_id] = meta\n",
    "        metadata_df.to_csv(meta_csv)\n",
    "        print(f\"📝 Processed {in_path.name} → {out_tsv.name}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {in_path.name}: {e}\")\n",
    "        continue\n",
    "\n",
    "print(f\"\\n Done. Slim files and combined metadata written to: {OUTPUT_DIR}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b57751",
   "metadata": {},
   "source": [
    "### Load in highconf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee0f614e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clonotype_id</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "      <th>v_call</th>\n",
       "      <th>d_call</th>\n",
       "      <th>j_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC02_clonotype9</td>\n",
       "      <td>CASSPYTGTGYGYTF</td>\n",
       "      <td>110.0</td>\n",
       "      <td>TRBV6-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC02_clonotype9</td>\n",
       "      <td>CASSPYTGTGYGYTF</td>\n",
       "      <td>85.0</td>\n",
       "      <td>TRBV6-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC02_clonotype1</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>74.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC02_clonotype1</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>71.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AC02_clonotype5</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>44.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4121</th>\n",
       "      <td>AC02_clonotype27</td>\n",
       "      <td>CASSPYTGTGYGYTF</td>\n",
       "      <td>184.0</td>\n",
       "      <td>TRBV6-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4122</th>\n",
       "      <td>AC02_clonotype1</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>31.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4123</th>\n",
       "      <td>AC02_clonotype8</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>52.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4124</th>\n",
       "      <td>AC02_clonotype8</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>48.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4125</th>\n",
       "      <td>AC02_clonotype5</td>\n",
       "      <td>CASSSVNEQFF</td>\n",
       "      <td>55.0</td>\n",
       "      <td>TRBV12-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4126 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          clonotype_id       amino_acid  count    v_call d_call   j_call\n",
       "0      AC02_clonotype9  CASSPYTGTGYGYTF  110.0   TRBV6-5    NaN  TRBJ1-2\n",
       "1      AC02_clonotype9  CASSPYTGTGYGYTF   85.0   TRBV6-5    NaN  TRBJ1-2\n",
       "2      AC02_clonotype1      CASSSVNEQFF   74.0  TRBV12-4    NaN  TRBJ2-1\n",
       "3      AC02_clonotype1      CASSSVNEQFF   71.0  TRBV12-4    NaN  TRBJ2-1\n",
       "4      AC02_clonotype5      CASSSVNEQFF   44.0  TRBV12-4    NaN  TRBJ2-1\n",
       "...                ...              ...    ...       ...    ...      ...\n",
       "4121  AC02_clonotype27  CASSPYTGTGYGYTF  184.0   TRBV6-5    NaN  TRBJ1-2\n",
       "4122   AC02_clonotype1      CASSSVNEQFF   31.0  TRBV12-4    NaN  TRBJ2-1\n",
       "4123   AC02_clonotype8      CASSSVNEQFF   52.0  TRBV12-4    NaN  TRBJ2-1\n",
       "4124   AC02_clonotype8      CASSSVNEQFF   48.0  TRBV12-4    NaN  TRBJ2-1\n",
       "4125   AC02_clonotype5      CASSSVNEQFF   55.0  TRBV12-4    NaN  TRBJ2-1\n",
       "\n",
       "[4126 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load high conf data\n",
    "\n",
    "highconf_file_name = 'VDJ_INFO_AC02_high_conf_CMV_TCRs.csv'\n",
    "\n",
    "highconf = pd.read_csv(f'/Users/ishaharris/Projects/TCR/TCR-Isha/data/High_Confidence_CMV_TCR/{highconf_file_name}')\n",
    "\n",
    "highconf\n",
    "\n",
    "# ------------------ COMMENTED OUT BECAUSE VDJ INFO FILES HAVE AMINO_ACID == ONLY TRB --------------------\n",
    "# Add a column with just the TRB sequence \n",
    "\n",
    "# highconf_trb = highconf.loc[:, 'cdr3s_aa']\n",
    "# highconf_trb=highconf_trb.apply(lambda x: re.search(r'TRB:([^;]+)', x).group(1) if 'TRB:' in x else None)\n",
    "# len(highconf_trb)\n",
    "\n",
    "# highconf['TRB'] = highconf_trb\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f2db2e14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rearrangement</th>\n",
       "      <th>cdr3_b_aa</th>\n",
       "      <th>frequency</th>\n",
       "      <th>productive_frequency</th>\n",
       "      <th>v_b_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_b_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCAGCC...</td>\n",
       "      <td>CASSPRGARSLETQYF</td>\n",
       "      <td>8.532311e-07</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TCRBV05-05</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBJ02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCA...</td>\n",
       "      <td>CASSPGTGAWQPQHF</td>\n",
       "      <td>1.023877e-06</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TCRBV05-05</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>TCRBJ01-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTGAATGTGAACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCT...</td>\n",
       "      <td>CASSRDGYGYTF</td>\n",
       "      <td>1.023877e-06</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TCRBV05-05</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCAGCTGGGGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.023877e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBV05-05</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBJ01-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATGTGAACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTG...</td>\n",
       "      <td>CASSLYTNTEAFF</td>\n",
       "      <td>1.194524e-06</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TCRBV05-05</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187249</th>\n",
       "      <td>CTGAAGATCCAGCCCTCAGAACCCAGGGACTCAGCTGTGTACTTCT...</td>\n",
       "      <td>CASSSSAREQYF</td>\n",
       "      <td>1.692811e-04</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBJ02-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187250</th>\n",
       "      <td>GTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGGG...</td>\n",
       "      <td>CRAR*QSSGNTIYF</td>\n",
       "      <td>2.214988e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>TCRBJ01-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187251</th>\n",
       "      <td>CAGCCCTCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCA...</td>\n",
       "      <td>CASSLGGTLNTEAFF</td>\n",
       "      <td>2.517032e-04</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187252</th>\n",
       "      <td>TAGAGTCTGCCATCCCCAACCAGACAGCTCTTTACTTCTGTGCCAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.288353e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187253</th>\n",
       "      <td>GTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTG...</td>\n",
       "      <td>CSARGLAVQETQYF</td>\n",
       "      <td>1.772502e-03</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>unresolved</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBJ02-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>187254 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            rearrangement         cdr3_b_aa  \\\n",
       "0       GCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCAGCC...  CASSPRGARSLETQYF   \n",
       "1       AACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCA...   CASSPGTGAWQPQHF   \n",
       "2       CTGAATGTGAACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCT...      CASSRDGYGYTF   \n",
       "3       GTTGCTGGGGGACTCGGCCCTGTATCTCTGTGCCAGCAGCTGGGGG...               NaN   \n",
       "4       AATGTGAACGCCTTGTTGCTGGGGGACTCGGCCCTGTATCTCTGTG...     CASSLYTNTEAFF   \n",
       "...                                                   ...               ...   \n",
       "187249  CTGAAGATCCAGCCCTCAGAACCCAGGGACTCAGCTGTGTACTTCT...      CASSSSAREQYF   \n",
       "187250  GTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGGG...    CRAR*QSSGNTIYF   \n",
       "187251  CAGCCCTCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCA...   CASSLGGTLNTEAFF   \n",
       "187252  TAGAGTCTGCCATCCCCAACCAGACAGCTCTTTACTTCTGTGCCAC...               NaN   \n",
       "187253  GTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTG...    CSARGLAVQETQYF   \n",
       "\n",
       "           frequency  productive_frequency    v_b_gene      d_gene    j_b_gene  \n",
       "0       8.532311e-07              0.000001  TCRBV05-05  unresolved  TCRBJ02-05  \n",
       "1       1.023877e-06              0.000001  TCRBV05-05  TCRBD01-01  TCRBJ01-05  \n",
       "2       1.023877e-06              0.000001  TCRBV05-05  unresolved  TCRBJ01-02  \n",
       "3       1.023877e-06                   NaN  TCRBV05-05  TCRBD02-01  TCRBJ01-05  \n",
       "4       1.194524e-06              0.000001  TCRBV05-05  TCRBD01-01  TCRBJ01-01  \n",
       "...              ...                   ...         ...         ...         ...  \n",
       "187249  1.692811e-04              0.000202  unresolved  TCRBD02-01  TCRBJ02-07  \n",
       "187250  2.214988e-04                   NaN  unresolved  TCRBD01-01  TCRBJ01-03  \n",
       "187251  2.517032e-04              0.000300  unresolved  unresolved  TCRBJ01-01  \n",
       "187252  3.288353e-04                   NaN  unresolved  TCRBD01-01  TCRBJ01-02  \n",
       "187253  1.772502e-03              0.002112  unresolved  TCRBD02-01  TCRBJ02-05  \n",
       "\n",
       "[187254 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load a sample repertoire\n",
    "rep = pd.read_csv('/Volumes/IshaVerbat/Isha/TCR/filtered_for_tcrdist/P00021.slim.tsv', sep=\"\\t\")\n",
    "\n",
    "# count \n",
    "rep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "332b8649",
   "metadata": {},
   "source": [
    "### Try to fix v gene naming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70e32601",
   "metadata": {},
   "source": [
    "For v gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d53f383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No hyphen numbers:['1', '1', '10', '14', '15', '16', '16', '16', '16', '17', '19', '19', '19', '2', '20', '20', '21', '23', '24', '24', '24', '24', '26', '26', '29', '29', '3', '3', '30', '31', '31', '4', '4', '5', '5', '5', '5', '5', '8', '9', '1', '13', '13', '14', '14', '15', '15', '15', '16', '16', '16', '17', '18', '19', '19', '19', '2', '2', '2', '26', '27', '28', '30', '30', '30', '30', '9', '9', '9']\n",
      "Hyphen numbers:['12', '12', '12', '12', '13', '13', '13', '13', '13', '13', '13', '13', '10', '10', '10', '10', '10', '10', '10', '10', '11', '11', '11', '11', '11', '11', '11', '12', '12', '12', '12', '12', '12', '20', '20', '20', '20', '20', '20', '20', '20', '20', '20', '21', '21', '23', '23', '23', '24', '24', '25', '26', '26', '29', '29', '29', '29', '29', '3', '3', '3', '3', '3', '4', '4', '4', '4', '4', '4', '4', '4', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7']\n",
      "1 is in the list\n"
     ]
    }
   ],
   "source": [
    "db_dir = '/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/db'\n",
    "db_file = 'alphabeta_gammadelta_db.tsv'\n",
    "\n",
    "# Load the database\n",
    "db = pd.read_csv(f'{db_dir}/{db_file}', sep=\"\\t\")\n",
    "\n",
    "# find rows starting with TRB27\n",
    "trb_v = db[db['id'].str.startswith('TRBV')]\n",
    "# turn into a list of ids\n",
    "trb_v_list = trb_v['id'].tolist()\n",
    "\n",
    "# Get a list of all id that don't have a hyphen in it\n",
    "trb_v_no_hyphen = trb_v[~trb_v['id'].str.contains('-')]\n",
    "trb_v_no_hyphen_list = trb_v_no_hyphen['id'].tolist()\n",
    "v_no_hyphen_numbers = [re.search(r'TRBV(\\d+)', item).group(1) for item in trb_v_no_hyphen_list]\n",
    "print(f'No hyphen numbers:{v_no_hyphen_numbers}')\n",
    "\n",
    "\n",
    "# Get a list of all id that have a hyphen in it\n",
    "trb_v_hyphen = trb_v[trb_v['id'].str.contains('-')]\n",
    "trb_v_hyphen_list = trb_v_hyphen['id'].tolist()\n",
    "trb_v_hyphen_list\n",
    "v_hyphen_numbers = [re.search(r'TRBV(\\d+)', item).group(1) for item in trb_v_hyphen_list]\n",
    "print(f'Hyphen numbers:{v_hyphen_numbers}')\n",
    "\n",
    "\n",
    "trb_v_no_hyphen_list\n",
    "trb_v_hyphen_list\n",
    "\n",
    "# ccheck if a number is in the list\n",
    "number = '1'\n",
    "if number in v_no_hyphen_numbers:\n",
    "    print(f'{number} is in the list')\n",
    "\n",
    "\n",
    "# Save to pickes\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "directory = '/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/pickles'\n",
    "\n",
    "# Make sure the directory exists\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "\n",
    "\n",
    "\n",
    "# save trb_v_no_hyphen_list to picke\n",
    "with open(f'{directory}/trb_v_no_hyphen_list.pkl', 'wb') as f:\n",
    "    pickle.dump(trb_v_no_hyphen_list, f)\n",
    "\n",
    "with open(f'{directory}/trb_v_hyphen_list.pkl', 'wb') as f:\n",
    "    pickle.dump(trb_v_hyphen_list, f)\n",
    "with open(f'{directory}/v_no_hyphen_numbers.pkl', 'wb') as f:\n",
    "    pickle.dump(v_no_hyphen_numbers, f)\n",
    "with open(f'{directory}/v_hyphen_numbers.pkl', 'wb') as f:\n",
    "    pickle.dump(v_hyphen_numbers, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bdaac900",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>organism</th>\n",
       "      <th>chain</th>\n",
       "      <th>region</th>\n",
       "      <th>nucseq</th>\n",
       "      <th>frame</th>\n",
       "      <th>aligned_protseq</th>\n",
       "      <th>cdr_columns</th>\n",
       "      <th>cdrs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>TRBV7-1*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtccctgagacacaaggtagcaaagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSLRHKVAKKGKDVALRYDPISGH.......NALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......NA;FQG....KDA;S.EGSI;CASSS.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....NSA;T.GGSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>681</th>\n",
       "      <td>TRBV7-2*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQRL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....NSA;T.GESV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>TRBV7-2*03</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQRL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....NSA;T.GESV;CTSSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>TRBV7-2*04</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggagctggagtttcccagtcccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....NSA;T.GGSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>TRBV7-3*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQTPSNKVTEKGKYVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....TGA;P.EGSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>TRBV7-3*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQTPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....TGA;P.EGSV;RASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>TRBV7-3*03</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQTPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....TGA;P.EGSV;RASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>TRBV7-3*04</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQTPSNKVTEKGKYVELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....TGA;P.EGSV;CASS..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>TRBV7-3*05</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>tgggagctcaggtgtgatccaatttcaggtcatactgccctttact...</td>\n",
       "      <td>1</td>\n",
       "      <td>..................WELRCDPISGH.......TALYWYRQSL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......TA;FQG....TGA;P.EGSV;CASS..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>TRBV7-4*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtccccaaggtacaaagtcgcaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVAKRGRDVALRCDSISGH.......VTLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VT;SQS....DAQ;P.ERSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>TRBV7-6*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVTKRGQDVALRCDPISGH.......VSLYWYRQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VS;FNY....EAQ;P.EGSI;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>TRBV7-6*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVTKRGQDVALRCDPISGH.......VSLYWYRQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VS;FNY....EAQ;P.EGSI;CASS..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>TRBV7-7*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVTKRGQDVTLRCDPISSH.......ATLYWYQQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SSH.......AT;FNY....EAQ;P.EGSI;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>TRBV7-7*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVTKRGQDVTLRCDPISSH.......VTLYWYQQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SSH.......VT;FNY....EAQ;P.EGSI;CASS..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>TRBV7-8*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VS;FQN....EAQ;P.EGSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>TRBV7-8*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VS;FQN....EAQ;P.EGSV;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>TRBV7-8*03</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SGH.......VS;FQN....EAQ;P.EGSV;CASSR.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>697</th>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>gatactggagtctcccagaaccccagacacaagatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSF;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>TRBV7-9*02</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>gatactggagtctcccagaaccccagacacaacatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>DTGVSQNPRHNITKRGQNVTFRCDPISEH.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSF;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>TRBV7-9*03</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>gatactggagtctcccaggaccccagacacaagatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>DTGVSQDPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSF;CASSL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>TRBV7-9*04</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>atatctggagtctcccacaaccccagacacaagatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>ISGVSHNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQNP...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSF;CASS..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>TRBV7-9*05</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>gatactggagtctcccagaaccccagacacaagatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSL;CASTK.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>TRBV7-9*06</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>gatactggagtctcccagaaccccagacacaagatcacaaagaggg...</td>\n",
       "      <td>1</td>\n",
       "      <td>DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>SEH.......NR;FQN....EAQ;P.KGSL;CASTL.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>TRBV7-9*07</td>\n",
       "      <td>human</td>\n",
       "      <td>B</td>\n",
       "      <td>V</td>\n",
       "      <td>cacaaccgcctttattggtaccgacagaccctggggcagggcccag...</td>\n",
       "      <td>1</td>\n",
       "      <td>............................H.......NRLYWYRQTL...</td>\n",
       "      <td>27-38;56-65;81-86;104-109</td>\n",
       "      <td>..H.......NR;FQN....EAQ;P.KGSF;CASS..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id organism chain region  \\\n",
       "679  TRBV7-1*01    human     B      V   \n",
       "680  TRBV7-2*01    human     B      V   \n",
       "681  TRBV7-2*02    human     B      V   \n",
       "682  TRBV7-2*03    human     B      V   \n",
       "683  TRBV7-2*04    human     B      V   \n",
       "684  TRBV7-3*01    human     B      V   \n",
       "685  TRBV7-3*02    human     B      V   \n",
       "686  TRBV7-3*03    human     B      V   \n",
       "687  TRBV7-3*04    human     B      V   \n",
       "688  TRBV7-3*05    human     B      V   \n",
       "689  TRBV7-4*01    human     B      V   \n",
       "690  TRBV7-6*01    human     B      V   \n",
       "691  TRBV7-6*02    human     B      V   \n",
       "692  TRBV7-7*01    human     B      V   \n",
       "693  TRBV7-7*02    human     B      V   \n",
       "694  TRBV7-8*01    human     B      V   \n",
       "695  TRBV7-8*02    human     B      V   \n",
       "696  TRBV7-8*03    human     B      V   \n",
       "697  TRBV7-9*01    human     B      V   \n",
       "698  TRBV7-9*02    human     B      V   \n",
       "699  TRBV7-9*03    human     B      V   \n",
       "700  TRBV7-9*04    human     B      V   \n",
       "701  TRBV7-9*05    human     B      V   \n",
       "702  TRBV7-9*06    human     B      V   \n",
       "703  TRBV7-9*07    human     B      V   \n",
       "\n",
       "                                                nucseq  frame  \\\n",
       "679  ggtgctggagtctcccagtccctgagacacaaggtagcaaagaagg...      1   \n",
       "680  ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...      1   \n",
       "681  ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...      1   \n",
       "682  ggagctggagtctcccagtcccccagtaacaaggtcacagagaagg...      1   \n",
       "683  ggagctggagtttcccagtcccccagtaacaaggtcacagagaagg...      1   \n",
       "684  ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...      1   \n",
       "685  ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...      1   \n",
       "686  ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...      1   \n",
       "687  ggtgctggagtctcccagacccccagtaacaaggtcacagagaagg...      1   \n",
       "688  tgggagctcaggtgtgatccaatttcaggtcatactgccctttact...      1   \n",
       "689  ggtgctggagtctcccagtccccaaggtacaaagtcgcaaagaggg...      1   \n",
       "690  ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...      1   \n",
       "691  ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...      1   \n",
       "692  ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...      1   \n",
       "693  ggtgctggagtctcccagtctcccaggtacaaagtcacaaagaggg...      1   \n",
       "694  ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...      1   \n",
       "695  ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...      1   \n",
       "696  ggtgctggagtctcccagtcccctaggtacaaagtcgcaaagagag...      1   \n",
       "697  gatactggagtctcccagaaccccagacacaagatcacaaagaggg...      1   \n",
       "698  gatactggagtctcccagaaccccagacacaacatcacaaagaggg...      1   \n",
       "699  gatactggagtctcccaggaccccagacacaagatcacaaagaggg...      1   \n",
       "700  atatctggagtctcccacaaccccagacacaagatcacaaagaggg...      1   \n",
       "701  gatactggagtctcccagaaccccagacacaagatcacaaagaggg...      1   \n",
       "702  gatactggagtctcccagaaccccagacacaagatcacaaagaggg...      1   \n",
       "703  cacaaccgcctttattggtaccgacagaccctggggcagggcccag...      1   \n",
       "\n",
       "                                       aligned_protseq  \\\n",
       "679  GAGVSQSLRHKVAKKGKDVALRYDPISGH.......NALYWYRQSL...   \n",
       "680  GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...   \n",
       "681  GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQRL...   \n",
       "682  GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQRL...   \n",
       "683  GAGVSQSPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...   \n",
       "684  GAGVSQTPSNKVTEKGKYVELRCDPISGH.......TALYWYRQSL...   \n",
       "685  GAGVSQTPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...   \n",
       "686  GAGVSQTPSNKVTEKGKDVELRCDPISGH.......TALYWYRQSL...   \n",
       "687  GAGVSQTPSNKVTEKGKYVELRCDPISGH.......TALYWYRQSL...   \n",
       "688  ..................WELRCDPISGH.......TALYWYRQSL...   \n",
       "689  GAGVSQSPRYKVAKRGRDVALRCDSISGH.......VTLYWYRQTL...   \n",
       "690  GAGVSQSPRYKVTKRGQDVALRCDPISGH.......VSLYWYRQAL...   \n",
       "691  GAGVSQSPRYKVTKRGQDVALRCDPISGH.......VSLYWYRQAL...   \n",
       "692  GAGVSQSPRYKVTKRGQDVTLRCDPISSH.......ATLYWYQQAL...   \n",
       "693  GAGVSQSPRYKVTKRGQDVTLRCDPISSH.......VTLYWYQQAL...   \n",
       "694  GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...   \n",
       "695  GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...   \n",
       "696  GAGVSQSPRYKVAKRGQDVALRCDPISGH.......VSLFWYQQAL...   \n",
       "697  DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...   \n",
       "698  DTGVSQNPRHNITKRGQNVTFRCDPISEH.......NRLYWYRQTL...   \n",
       "699  DTGVSQDPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...   \n",
       "700  ISGVSHNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQNP...   \n",
       "701  DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...   \n",
       "702  DTGVSQNPRHKITKRGQNVTFRCDPISEH.......NRLYWYRQTL...   \n",
       "703  ............................H.......NRLYWYRQTL...   \n",
       "\n",
       "                   cdr_columns                                   cdrs  \n",
       "679  27-38;56-65;81-86;104-109  SGH.......NA;FQG....KDA;S.EGSI;CASSS.  \n",
       "680  27-38;56-65;81-86;104-109  SGH.......TA;FQG....NSA;T.GGSV;CASSL.  \n",
       "681  27-38;56-65;81-86;104-109  SGH.......TA;FQG....NSA;T.GESV;CASSL.  \n",
       "682  27-38;56-65;81-86;104-109  SGH.......TA;FQG....NSA;T.GESV;CTSSL.  \n",
       "683  27-38;56-65;81-86;104-109  SGH.......TA;FQG....NSA;T.GGSV;CASSL.  \n",
       "684  27-38;56-65;81-86;104-109  SGH.......TA;FQG....TGA;P.EGSV;CASSL.  \n",
       "685  27-38;56-65;81-86;104-109  SGH.......TA;FQG....TGA;P.EGSV;RASSL.  \n",
       "686  27-38;56-65;81-86;104-109  SGH.......TA;FQG....TGA;P.EGSV;RASSL.  \n",
       "687  27-38;56-65;81-86;104-109  SGH.......TA;FQG....TGA;P.EGSV;CASS..  \n",
       "688  27-38;56-65;81-86;104-109  SGH.......TA;FQG....TGA;P.EGSV;CASS..  \n",
       "689  27-38;56-65;81-86;104-109  SGH.......VT;SQS....DAQ;P.ERSV;CASSL.  \n",
       "690  27-38;56-65;81-86;104-109  SGH.......VS;FNY....EAQ;P.EGSI;CASSL.  \n",
       "691  27-38;56-65;81-86;104-109  SGH.......VS;FNY....EAQ;P.EGSI;CASS..  \n",
       "692  27-38;56-65;81-86;104-109  SSH.......AT;FNY....EAQ;P.EGSI;CASSL.  \n",
       "693  27-38;56-65;81-86;104-109  SSH.......VT;FNY....EAQ;P.EGSI;CASS..  \n",
       "694  27-38;56-65;81-86;104-109  SGH.......VS;FQN....EAQ;P.EGSV;CASSL.  \n",
       "695  27-38;56-65;81-86;104-109  SGH.......VS;FQN....EAQ;P.EGSV;CASSL.  \n",
       "696  27-38;56-65;81-86;104-109  SGH.......VS;FQN....EAQ;P.EGSV;CASSR.  \n",
       "697  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSF;CASSL.  \n",
       "698  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSF;CASSL.  \n",
       "699  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSF;CASSL.  \n",
       "700  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSF;CASS..  \n",
       "701  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSL;CASTK.  \n",
       "702  27-38;56-65;81-86;104-109  SEH.......NR;FQN....EAQ;P.KGSL;CASTL.  \n",
       "703  27-38;56-65;81-86;104-109  ..H.......NR;FQN....EAQ;P.KGSF;CASS..  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# db id cols starging with TRBV7\n",
    "trb_v7 = db[db['id'].str.startswith('TRBV7')]\n",
    "trb_v7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "576ecc89",
   "metadata": {},
   "source": [
    "For J gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f06fb959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No hyphen numbers: []\n",
      "Hyphen numbers: ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '2', '2', '2', '2', '2', '2', '2', '2', '1', '1', '1', '1', '1', '1', '1', '2', '2', '2', '2', '2', '2', '2', '2', '2']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "db_dir = '/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/db'\n",
    "db_file = 'alphabeta_gammadelta_db.tsv'\n",
    "\n",
    "# Load the database\n",
    "db = pd.read_csv(f'{db_dir}/{db_file}', sep=\"\\t\")\n",
    "\n",
    "# Find rows starting with TRBJ\n",
    "trb_j = db[db['id'].str.startswith('TRBJ')]\n",
    "\n",
    "# Turn into a list of ids\n",
    "trb_j_list = trb_j['id'].tolist()\n",
    "\n",
    "# Get a list of all ids that don't have a hyphen in them\n",
    "trb_j_no_hyphen = trb_j[~trb_j['id'].str.contains('-')]\n",
    "trb_j_no_hyphen_list = trb_j_no_hyphen['id'].tolist()\n",
    "j_no_hyphen_numbers = [re.search(r'TRBJ(\\d+)', item).group(1) for item in trb_j_no_hyphen_list]\n",
    "print(f'No hyphen numbers: {j_no_hyphen_numbers}')\n",
    "\n",
    "# Get a list of all ids that do have a hyphen in them\n",
    "trb_j_hyphen = trb_j[trb_j['id'].str.contains('-')]\n",
    "trb_j_hyphen_list = trb_j_hyphen['id'].tolist()\n",
    "j_hyphen_numbers = [re.search(r'TRBJ(\\d+)', item).group(1) for item in trb_j_hyphen_list]\n",
    "print(f'Hyphen numbers: {j_hyphen_numbers}')\n",
    "\n",
    "# Example: check if a number is in the no-hyphen list\n",
    "number = '2'\n",
    "if number in j_no_hyphen_numbers:\n",
    "    print(f'{number} is in the no-hyphen list')\n",
    "\n",
    "\n",
    "# Save to pickle\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "directory = '/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/pickles'\n",
    "\n",
    "# Make sure the directory exists\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "\n",
    "\n",
    "\n",
    "with open(f'{directory}/trb_j_no_hyphen_list.pkl', 'wb') as f:\n",
    "    pickle.dump(trb_v_no_hyphen_list, f)\n",
    "with open(f'{directory}/trb_j_hyphen_list.pkl', 'wb') as f:\n",
    "    pickle.dump(trb_v_hyphen_list, f)\n",
    "with open(f'{directory}/j_no_hyphen_numbers.pkl', 'wb') as f:\n",
    "    pickle.dump(v_no_hyphen_numbers, f)\n",
    "with open(f'{directory}/j_hyphen_numbers.pkl', 'wb') as f:\n",
    "    pickle.dump(v_hyphen_numbers, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cb45f520",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TRBV1*01',\n",
       " 'TRBV1*02',\n",
       " 'TRBV10*01',\n",
       " 'TRBV12-1*01',\n",
       " 'TRBV12-1*02',\n",
       " 'TRBV12-2*01',\n",
       " 'TRBV12-2*02',\n",
       " 'TRBV13-1*01',\n",
       " 'TRBV13-1*02',\n",
       " 'TRBV13-2*01',\n",
       " 'TRBV13-2*02',\n",
       " 'TRBV13-2*03',\n",
       " 'TRBV13-2*04',\n",
       " 'TRBV13-2*05',\n",
       " 'TRBV13-3*01',\n",
       " 'TRBV14*01',\n",
       " 'TRBV15*01',\n",
       " 'TRBV16*01',\n",
       " 'TRBV16*02',\n",
       " 'TRBV16*03',\n",
       " 'TRBV16*04',\n",
       " 'TRBV17*01',\n",
       " 'TRBV19*01',\n",
       " 'TRBV19*02',\n",
       " 'TRBV19*03',\n",
       " 'TRBV2*01',\n",
       " 'TRBV20*01',\n",
       " 'TRBV20*02',\n",
       " 'TRBV21*01',\n",
       " 'TRBV23*01',\n",
       " 'TRBV24*01',\n",
       " 'TRBV24*02',\n",
       " 'TRBV24*03',\n",
       " 'TRBV24*04',\n",
       " 'TRBV26*01',\n",
       " 'TRBV26*02',\n",
       " 'TRBV29*01',\n",
       " 'TRBV29*02',\n",
       " 'TRBV3*01',\n",
       " 'TRBV3*02',\n",
       " 'TRBV30*01',\n",
       " 'TRBV31*01',\n",
       " 'TRBV31*02',\n",
       " 'TRBV4*01',\n",
       " 'TRBV4*02',\n",
       " 'TRBV5*01',\n",
       " 'TRBV5*02',\n",
       " 'TRBV5*03',\n",
       " 'TRBV5*04',\n",
       " 'TRBV5*05',\n",
       " 'TRBV8*01',\n",
       " 'TRBV9*01',\n",
       " 'TRBV1*01',\n",
       " 'TRBV10-1*01',\n",
       " 'TRBV10-1*02',\n",
       " 'TRBV10-2*01',\n",
       " 'TRBV10-2*02',\n",
       " 'TRBV10-3*01',\n",
       " 'TRBV10-3*02',\n",
       " 'TRBV10-3*03',\n",
       " 'TRBV10-3*04',\n",
       " 'TRBV11-1*01',\n",
       " 'TRBV11-2*01',\n",
       " 'TRBV11-2*02',\n",
       " 'TRBV11-2*03',\n",
       " 'TRBV11-3*01',\n",
       " 'TRBV11-3*02',\n",
       " 'TRBV11-3*03',\n",
       " 'TRBV12-1*01',\n",
       " 'TRBV12-2*01',\n",
       " 'TRBV12-3*01',\n",
       " 'TRBV12-4*01',\n",
       " 'TRBV12-4*02',\n",
       " 'TRBV12-5*01',\n",
       " 'TRBV13*01',\n",
       " 'TRBV13*02',\n",
       " 'TRBV14*01',\n",
       " 'TRBV14*02',\n",
       " 'TRBV15*01',\n",
       " 'TRBV15*02',\n",
       " 'TRBV15*03',\n",
       " 'TRBV16*01',\n",
       " 'TRBV16*02',\n",
       " 'TRBV16*03',\n",
       " 'TRBV17*01',\n",
       " 'TRBV18*01',\n",
       " 'TRBV19*01',\n",
       " 'TRBV19*02',\n",
       " 'TRBV19*03',\n",
       " 'TRBV2*01',\n",
       " 'TRBV2*02',\n",
       " 'TRBV2*03',\n",
       " 'TRBV20-1*01',\n",
       " 'TRBV20-1*02',\n",
       " 'TRBV20-1*03',\n",
       " 'TRBV20-1*04',\n",
       " 'TRBV20-1*05',\n",
       " 'TRBV20-1*06',\n",
       " 'TRBV20-1*07',\n",
       " 'TRBV20/OR9-2*01',\n",
       " 'TRBV20/OR9-2*02',\n",
       " 'TRBV20/OR9-2*03',\n",
       " 'TRBV21-1*01',\n",
       " 'TRBV21/OR9-2*01',\n",
       " 'TRBV23-1*01',\n",
       " 'TRBV23/OR9-2*01',\n",
       " 'TRBV23/OR9-2*02',\n",
       " 'TRBV24-1*01',\n",
       " 'TRBV24/OR9-2*01',\n",
       " 'TRBV25-1*01',\n",
       " 'TRBV26*01',\n",
       " 'TRBV26/OR9-2*01',\n",
       " 'TRBV26/OR9-2*02',\n",
       " 'TRBV27*01',\n",
       " 'TRBV28*01',\n",
       " 'TRBV29-1*01',\n",
       " 'TRBV29-1*02',\n",
       " 'TRBV29-1*03',\n",
       " 'TRBV29/OR9-2*01',\n",
       " 'TRBV29/OR9-2*02',\n",
       " 'TRBV3-1*01',\n",
       " 'TRBV3-1*02',\n",
       " 'TRBV3-2*01',\n",
       " 'TRBV3-2*02',\n",
       " 'TRBV3-2*03',\n",
       " 'TRBV30*01',\n",
       " 'TRBV30*02',\n",
       " 'TRBV30*04',\n",
       " 'TRBV30*05',\n",
       " 'TRBV4-1*01',\n",
       " 'TRBV4-1*02',\n",
       " 'TRBV4-2*01',\n",
       " 'TRBV4-2*02',\n",
       " 'TRBV4-3*01',\n",
       " 'TRBV4-3*02',\n",
       " 'TRBV4-3*03',\n",
       " 'TRBV4-3*04',\n",
       " 'TRBV5-1*01',\n",
       " 'TRBV5-1*02',\n",
       " 'TRBV5-3*01',\n",
       " 'TRBV5-3*02',\n",
       " 'TRBV5-4*01',\n",
       " 'TRBV5-4*02',\n",
       " 'TRBV5-4*03',\n",
       " 'TRBV5-4*04',\n",
       " 'TRBV5-5*01',\n",
       " 'TRBV5-5*02',\n",
       " 'TRBV5-5*03',\n",
       " 'TRBV5-6*01',\n",
       " 'TRBV5-7*01',\n",
       " 'TRBV5-8*01',\n",
       " 'TRBV5-8*02',\n",
       " 'TRBV6-1*01',\n",
       " 'TRBV6-2*01',\n",
       " 'TRBV6-3*01',\n",
       " 'TRBV6-4*01',\n",
       " 'TRBV6-4*02',\n",
       " 'TRBV6-5*01',\n",
       " 'TRBV6-6*01',\n",
       " 'TRBV6-6*02',\n",
       " 'TRBV6-6*03',\n",
       " 'TRBV6-6*04',\n",
       " 'TRBV6-6*05',\n",
       " 'TRBV6-7*01',\n",
       " 'TRBV6-8*01',\n",
       " 'TRBV6-9*01',\n",
       " 'TRBV7-1*01',\n",
       " 'TRBV7-2*01',\n",
       " 'TRBV7-2*02',\n",
       " 'TRBV7-2*03',\n",
       " 'TRBV7-2*04',\n",
       " 'TRBV7-3*01',\n",
       " 'TRBV7-3*02',\n",
       " 'TRBV7-3*03',\n",
       " 'TRBV7-3*04',\n",
       " 'TRBV7-3*05',\n",
       " 'TRBV7-4*01',\n",
       " 'TRBV7-6*01',\n",
       " 'TRBV7-6*02',\n",
       " 'TRBV7-7*01',\n",
       " 'TRBV7-7*02',\n",
       " 'TRBV7-8*01',\n",
       " 'TRBV7-8*02',\n",
       " 'TRBV7-8*03',\n",
       " 'TRBV7-9*01',\n",
       " 'TRBV7-9*02',\n",
       " 'TRBV7-9*03',\n",
       " 'TRBV7-9*04',\n",
       " 'TRBV7-9*05',\n",
       " 'TRBV7-9*06',\n",
       " 'TRBV7-9*07',\n",
       " 'TRBV9*01',\n",
       " 'TRBV9*02',\n",
       " 'TRBV9*03']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trb_v_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66e531fe",
   "metadata": {},
   "source": [
    "### Write TCRdist calculator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c7bae02",
   "metadata": {},
   "source": [
    "##### Ignore this below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03f58312",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " - Loaded libraries.\n",
      " - Set parameters.\n",
      "Defined functions.\n",
      "Loading high‑confidence CMV search set …\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"Column(s) ['productive_frequency'] do not exist\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[31], line 150\u001b[0m\n\u001b[1;32m    148\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLoading high‑confidence CMV search set …\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    149\u001b[0m high_raw  \u001b[38;5;241m=\u001b[39m highconf\n\u001b[0;32m--> 150\u001b[0m high_df   \u001b[38;5;241m=\u001b[39m prepare_df(high_raw)\n\u001b[1;32m    151\u001b[0m \u001b[38;5;28mprint\u001b[39m(high_df\u001b[38;5;241m.\u001b[39mhead())\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m → \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(high_df)\u001b[38;5;132;01m:\u001b[39;00m\u001b[38;5;124m,\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m unique TRB clonotypes ready for search.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[31], line 139\u001b[0m, in \u001b[0;36mprepare_df\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m    134\u001b[0m df \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mdrop_duplicates()\u001b[38;5;241m.\u001b[39mreset_index(drop\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    136\u001b[0m \u001b[38;5;66;03m# if the same CDR3/v_gene/j_gene appears multiple times, sum its frequencies\u001b[39;00m\n\u001b[1;32m    137\u001b[0m df \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    138\u001b[0m     df\u001b[38;5;241m.\u001b[39mgroupby([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcdr3_b_aa\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mv_b_gene\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mj_b_gene\u001b[39m\u001b[38;5;124m\"\u001b[39m], as_index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m--> 139\u001b[0m     \u001b[38;5;241m.\u001b[39magg({\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mproductive_frequency\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msum\u001b[39m\u001b[38;5;124m\"\u001b[39m})\n\u001b[1;32m    140\u001b[0m )\n\u001b[1;32m    141\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m df\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/groupby/generic.py:1432\u001b[0m, in \u001b[0;36mDataFrameGroupBy.aggregate\u001b[0;34m(self, func, engine, engine_kwargs, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1429\u001b[0m     kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mengine_kwargs\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m engine_kwargs\n\u001b[1;32m   1431\u001b[0m op \u001b[38;5;241m=\u001b[39m GroupByApply(\u001b[38;5;28mself\u001b[39m, func, args\u001b[38;5;241m=\u001b[39margs, kwargs\u001b[38;5;241m=\u001b[39mkwargs)\n\u001b[0;32m-> 1432\u001b[0m result \u001b[38;5;241m=\u001b[39m op\u001b[38;5;241m.\u001b[39magg()\n\u001b[1;32m   1433\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_dict_like(func) \u001b[38;5;129;01mand\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1434\u001b[0m     \u001b[38;5;66;03m# GH #52849\u001b[39;00m\n\u001b[1;32m   1435\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mas_index \u001b[38;5;129;01mand\u001b[39;00m is_list_like(func):\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/apply.py:190\u001b[0m, in \u001b[0;36mApply.agg\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    187\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_str()\n\u001b[1;32m    189\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_dict_like(func):\n\u001b[0;32m--> 190\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39magg_dict_like()\n\u001b[1;32m    191\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_list_like(func):\n\u001b[1;32m    192\u001b[0m     \u001b[38;5;66;03m# we require a list, but not a 'str'\u001b[39;00m\n\u001b[1;32m    193\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39magg_list_like()\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/apply.py:423\u001b[0m, in \u001b[0;36mApply.agg_dict_like\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    415\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21magg_dict_like\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m    416\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    417\u001b[0m \u001b[38;5;124;03m    Compute aggregation in the case of a dict-like argument.\u001b[39;00m\n\u001b[1;32m    418\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    421\u001b[0m \u001b[38;5;124;03m    Result of aggregation.\u001b[39;00m\n\u001b[1;32m    422\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 423\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39magg_or_apply_dict_like(op_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124magg\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/apply.py:1608\u001b[0m, in \u001b[0;36mGroupByApply.agg_or_apply_dict_like\u001b[0;34m(self, op_name)\u001b[0m\n\u001b[1;32m   1603\u001b[0m     kwargs\u001b[38;5;241m.\u001b[39mupdate({\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mengine\u001b[39m\u001b[38;5;124m\"\u001b[39m: engine, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mengine_kwargs\u001b[39m\u001b[38;5;124m\"\u001b[39m: engine_kwargs})\n\u001b[1;32m   1605\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m com\u001b[38;5;241m.\u001b[39mtemp_setattr(\n\u001b[1;32m   1606\u001b[0m     obj, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas_index\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m, condition\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mhasattr\u001b[39m(obj, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas_index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1607\u001b[0m ):\n\u001b[0;32m-> 1608\u001b[0m     result_index, result_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcompute_dict_like(\n\u001b[1;32m   1609\u001b[0m         op_name, selected_obj, selection, kwargs\n\u001b[1;32m   1610\u001b[0m     )\n\u001b[1;32m   1611\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwrap_results_dict_like(selected_obj, result_index, result_data)\n\u001b[1;32m   1612\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/apply.py:462\u001b[0m, in \u001b[0;36mApply.compute_dict_like\u001b[0;34m(self, op_name, selected_obj, selection, kwargs)\u001b[0m\n\u001b[1;32m    460\u001b[0m is_groupby \u001b[38;5;241m=\u001b[39m \u001b[38;5;28misinstance\u001b[39m(obj, (DataFrameGroupBy, SeriesGroupBy))\n\u001b[1;32m    461\u001b[0m func \u001b[38;5;241m=\u001b[39m cast(AggFuncTypeDict, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfunc)\n\u001b[0;32m--> 462\u001b[0m func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnormalize_dictlike_arg(op_name, selected_obj, func)\n\u001b[1;32m    464\u001b[0m is_non_unique_col \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    465\u001b[0m     selected_obj\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m\n\u001b[1;32m    466\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m selected_obj\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnunique() \u001b[38;5;241m<\u001b[39m \u001b[38;5;28mlen\u001b[39m(selected_obj\u001b[38;5;241m.\u001b[39mcolumns)\n\u001b[1;32m    467\u001b[0m )\n\u001b[1;32m    469\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m selected_obj\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    470\u001b[0m     \u001b[38;5;66;03m# key only used for output\u001b[39;00m\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pandas/core/apply.py:663\u001b[0m, in \u001b[0;36mApply.normalize_dictlike_arg\u001b[0;34m(self, how, obj, func)\u001b[0m\n\u001b[1;32m    661\u001b[0m     cols \u001b[38;5;241m=\u001b[39m Index(\u001b[38;5;28mlist\u001b[39m(func\u001b[38;5;241m.\u001b[39mkeys()))\u001b[38;5;241m.\u001b[39mdifference(obj\u001b[38;5;241m.\u001b[39mcolumns, sort\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    662\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(cols) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 663\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mColumn(s) \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlist\u001b[39m(cols)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m do not exist\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    665\u001b[0m aggregator_types \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28mtuple\u001b[39m, \u001b[38;5;28mdict\u001b[39m)\n\u001b[1;32m    667\u001b[0m \u001b[38;5;66;03m# if we have a dict of any non-scalars\u001b[39;00m\n\u001b[1;32m    668\u001b[0m \u001b[38;5;66;03m# eg. {'A' : ['mean']}, normalize all to\u001b[39;00m\n\u001b[1;32m    669\u001b[0m \u001b[38;5;66;03m# be list-likes\u001b[39;00m\n\u001b[1;32m    670\u001b[0m \u001b[38;5;66;03m# Cannot use func.values() because arg may be a Series\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"Column(s) ['productive_frequency'] do not exist\""
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "\"\"\"\n",
    "Compute TCRdist between a panel of high‑confidence CMV‑associated TRB clonotypes and each patient's bulk TCRB repertoire.\n",
    "The script mirrors the logic of the previous Levenshtein implementation but leverages **tcrdist3** for a biochemically‑weighted distance.\n",
    "\n",
    "Dependencies\n",
    "------------\n",
    "    pip install tcrdist3 pandas numpy\n",
    "\n",
    "Input\n",
    "-----\n",
    "* A TSV (or CSV) file containing the CMV‑associated search set (≈200 rows).\n",
    "* A directory of TSV/CSV files – one per patient – already filtered to unique productive TRB clonotypes (≈3×10⁵ rows each).\n",
    "\n",
    "Both inputs must expose **cdr3_b_aa**, **v_b_gene**, **j_b_gene** columns (see `prepare_df`).\n",
    "\n",
    "Output\n",
    "------\n",
    "`tcrdist_output.csv` – One row per patient with a histogram of TCRdist values (0‥MAX_DIST).\n",
    "\n",
    "Notes\n",
    "-----\n",
    "* The tcrdist scale is *not* the same as edit distance. A single amino‑acid substitution in CDR3 is ≈ 12 distance units when TRBV and TRBJ genes match. A radius of 36 therefore captures most functional \"one‑to‑two‑sub\" neighbours (Dash et al., 2017).\n",
    "* The computation is executed as a sparse **rectangular** distance matrix; only distances ≤ `RADIUS` are kept in memory.\n",
    "* Adapt `MAX_DIST`, `RADIUS`, and path variables as needed.\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import re\n",
    "import multiprocessing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tcrdist.repertoire import TCRrep\n",
    "from tcrdist.breadth import get_safe_chunk\n",
    "\n",
    "excluded_genes = []\n",
    "\n",
    "print(' - Loaded libraries.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# User‑configurable parameters\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "HLA = 'A02'\n",
    "MAX_DIST   = 50        # highest TCRdist bucket to tally (inclusive)\n",
    "RADIUS     = MAX_DIST  # distances > RADIUS are discarded by tcrdist3\n",
    "HIGHCONF   = highconf   # search‑set file\n",
    "PATIENT_DIR = f\"/Volumes/IshaVerbat/Isha/TCR/Filtered_HLA-{HLA}\"  # directory with patient files\n",
    "OUTFILE    = \"tcrdist_output_freqs.csv\"\n",
    "N_CPUS     = min(multiprocessing.cpu_count(), 8)\n",
    "\n",
    "# set wd\n",
    "os.chdir(\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist\")\n",
    "\n",
    "MAX_FILES = 4\n",
    "\n",
    "print(' - Set parameters.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# Helper functions\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "GENE_PAT = re.compile(r\"[*!].*$\")\n",
    "\n",
    "\n",
    "def _normalize_gene(s: str) -> str:\n",
    "    \"\"\"Strip prefixes / allele info to IMGT-like gene names understood by tcrdist3,\n",
    "    remove leading zeros from gene numbers (e.g. TRBV05 → TRBV5),\n",
    "    and from any numeric subfields (e.g. TRBV5-05 → TRBV5-5).\"\"\"\n",
    "    if pd.isna(s):\n",
    "        return np.nan\n",
    "    # Standardize prefix\n",
    "    s = str(s).replace(\"TCR\", \"TR\")  # e.g. TCRBV05-05 → TRBV05-05\n",
    "\n",
    "    # Remove leading zeros from primary gene number (e.g. TRBV05 → TRBV5)\n",
    "    s = re.sub(\n",
    "        r\"^([A-Z]+)(0+)(\\d+)\",\n",
    "        lambda m: f\"{m.group(1)}{int(m.group(2) + m.group(3))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Remove leading zeros from any subfield after hyphens (e.g. -05 → -5)\n",
    "    s = re.sub(\n",
    "        r\"-(0+)(\\d+)\",\n",
    "        lambda m: f\"-{int(m.group(1) + m.group(2))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Normalize allele information to *01\n",
    "    if \"*\" in s:\n",
    "        s = re.sub(r\"[*!].*$\", \"*01\", s)\n",
    "    else:\n",
    "        s = s + \"*01\"\n",
    "\n",
    "    # Fix adaptive errors in V gene names - they added in -01 for families with only one gene. \n",
    "    # Get the number from the string\n",
    "\n",
    "    if not re.match(r'^TRB[JV]\\d+(-\\d+)?\\*\\d+$', s):\n",
    "        excluded_genes.append(s)\n",
    "        print(f\"Excluded malformed gene: {s}\")\n",
    "        return np.nan\n",
    "\n",
    "    number = re.search(r'TRB[JV](\\d+)', s).group(1)\n",
    "\n",
    "    if number in no_hyphen_numbers and s not in trb_v_list:\n",
    "        print(f'{s}: no match')\n",
    "        s = re.sub(r'-\\d+(?=\\*)', '', s)\n",
    "        print(f'Changed to {s}')\n",
    "\n",
    "    return s\n",
    "\n",
    "\n",
    "def prepare_df(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Bring an arbitrary clonotype table to tcrdist3‑ready shape.\"\"\"\n",
    "    # Remove d_gene column\n",
    "    if \"d_gene\" in df.columns:\n",
    "        df = df.drop(columns=[\"d_gene\"])\n",
    "    \n",
    "    rename_map = {\n",
    "        \"amino_acid\": \"cdr3_b_aa\",\n",
    "        \"cdr3s_aa\": \"cdr3_b_aa\",\n",
    "        \"v_call\": \"v_b_gene\",\n",
    "        \"j_call\": \"j_b_gene\",\n",
    "    }\n",
    "\n",
    "    df = df.rename(columns=rename_map)\n",
    "\n",
    "    has_freq = \"productive_frequency\" in df.columns\n",
    "\n",
    "    base_cols = [\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"]\n",
    "    use_cols = base_cols + (['productive_frequency'] if has_freq else [])\n",
    "\n",
    "    df[\"v_b_gene\"] = df[\"v_b_gene\"].apply(_normalize_gene)\n",
    "    df[\"j_b_gene\"] = df[\"j_b_gene\"].apply(_normalize_gene)\n",
    "  \n",
    "\n",
    "    df = df.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "    # if the same CDR3/v_gene/j_gene appears multiple times, sum its frequencies\n",
    "    df = (\n",
    "        df.groupby([\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"], as_index=False)\n",
    "        .agg({\"productive_frequency\": \"sum\"})\n",
    "    )\n",
    "    return df\n",
    "\n",
    "print('Defined functions.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# Load the CMV‑associated search set once\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "print(\"Loading high‑confidence CMV search set …\")\n",
    "high_raw  = highconf\n",
    "high_df   = prepare_df(high_raw)\n",
    "print(high_df.head())\n",
    "\n",
    "print(f\" → {len(high_df):,} unique TRB clonotypes ready for search.\")\n",
    "\n",
    "tr_search = TCRrep(cell_df=high_df, organism=\"human\", chains=[\"beta\"], compute_distances=False)\n",
    "\n",
    "tr_search.cpus = N_CPUS\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# Prepare output buffer and file names\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "results_buffer = []\n",
    "outfile = OUTFILE\n",
    "\n",
    "file_names = sorted(fn for fn in os.listdir(PATIENT_DIR) if fn.endswith(('.tsv', '.csv')) and fn.startswith('P'))\n",
    "\n",
    "# cols = [\"patient_id\"] + [f\"dist_{d}\" for d in range(MAX_DIST + 1)]\n",
    "# results = pd.DataFrame(columns=cols)\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# Main loop over patient repertoires\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "file_names = sorted(fn for fn in os.listdir(PATIENT_DIR) if fn.endswith(('.tsv', '.csv')) and fn.startswith('P'))\n",
    "\n",
    "for idx, fname in enumerate(file_names, 1):\n",
    "    if idx > MAX_FILES:\n",
    "        print(f\"Reached maximum number of files to process ({MAX_FILES}).\")\n",
    "        break\n",
    "    patient_id = os.path.splitext(fname)[0]\n",
    "    print(f\"Processing ({idx}/{len(file_names)}): {patient_id}\")\n",
    "\n",
    "    # Load & clean bulk repertoire\n",
    "    rep_raw = pd.read_csv(os.path.join(PATIENT_DIR, fname), sep=\"\\t\", low_memory=False)\n",
    "    rep_df  = prepare_df(rep_raw)\n",
    "\n",
    "    if rep_df.empty:\n",
    "        print(f\" ! No usable sequences in {patient_id} – skipping.\")\n",
    "        continue\n",
    "\n",
    "    # extract and sum productive frequencies\n",
    "    rep_freq = rep_raw.rename(columns = {\n",
    "        \"amino_acid\": \"cdr3_b_aa\",\n",
    "        'cdr3s_aa': \"cdr3_b_aa\",\n",
    "        \"v_call\": \"v_b_gene\",\n",
    "        \"j_call\": \"j_b_gene\",\n",
    "    })\n",
    "\n",
    "    # keep rows with required fields\n",
    "    rep_freq = rep_freq[[\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\", \"productive_frequency\"]].dropna()\n",
    "\n",
    "    # normalise gene names\n",
    "    rep_freq[\"v_b_gene\"] = rep_freq[\"v_b_gene\"].apply(_normalize_gene)\n",
    "    rep_freq[\"j_b_gene\"] = rep_freq[\"j_b_gene\"].apply(_normalize_gene)\n",
    "\n",
    "    # sum productive frequencies for each unique clono\n",
    "    rep_freq = (\n",
    "        rep_freq.groupby([\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"], as_index=False)\n",
    "        .agg({\"productive_frequency\": \"sum\"})\n",
    "    )\n",
    "\n",
    "    # merge back into cleaned core\n",
    "    rep_df = pd.merge(rep_core, rep_freq, on=[\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"], how=\"left\").fillna({'productive_frequency': 0})\n",
    "\n",
    "\n",
    "    tr_bulk = TCRrep(cell_df=rep_df, organism=\"human\", chains=[\"beta\"], compute_distances=False)\n",
    "\n",
    "\n",
    "\n",
    "    # Compute sparse rectangular distances (rows=search, cols=bulk)\n",
    "    chunk_size = get_safe_chunk(tr_search.clone_df.shape[0], tr_bulk.clone_df.shape[0])\n",
    "    tr_search.compute_sparse_rect_distances(\n",
    "        df=tr_search.clone_df,\n",
    "        df2=tr_bulk.clone_df,\n",
    "        radius=RADIUS,\n",
    "        chunk_size=chunk_size,\n",
    "    )\n",
    "\n",
    "    # Histogram distances ≤ RADIUS\n",
    "    coo = tr_search.rw_beta.tocoo()\n",
    "    dists = coo.data.astype(int)\n",
    "    cols = coo.col\n",
    "    freqs = tr_bulk.clone_df['productive_frequency'].values\n",
    "\n",
    "    bins = np.zeros(MAX_DIST + 1, dtype=float)\n",
    "    np.add.at(bins, dists, freqs[cols])\n",
    "\n",
    "    # Append to results\n",
    "    row = {\"patient_id\": patient_id, **{f\"dist_{d}\": bins[d] for d in range(MAX_DIST + 1)}}\n",
    "    results = pd.concat([results, pd.DataFrame([row])], ignore_index=True)\n",
    "\n",
    "    # Persist after each patient\n",
    "    results.to_csv(OUTFILE, index=False)\n",
    "    print(f\"   ↳ saved ➜ {OUTFILE}\")\n",
    "\n",
    "    # Free distance matrix to keep RAM stable before next iteration\n",
    "    del tr_search.rw_beta\n",
    "\n",
    "print(\"✔︎ All patient repertoires processed.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60916e3",
   "metadata": {},
   "source": [
    "### Tester area"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "4377690e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV6-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV12-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV12-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV28 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV29-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV19 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV27 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV6-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV4-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV20-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV24-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV30 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV3-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV6-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV15 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV4-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV12-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV14 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV6-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV6-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV25-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-8 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV4-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV10-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV11-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV11-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV6-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV12-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV12-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV28 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV29-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV19 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV27 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV6-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV4-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV20-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV24-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV30 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV3-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV6-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV15 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV4-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV12-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV14 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV6-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV6-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV25-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-8 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV4-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV10-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV11-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV11-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV6-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV12-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV12-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV28 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV29-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV19 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-9 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV27 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV6-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV4-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV20-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV24-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV30 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV3-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV6-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV15 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV4-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV12-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV14 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV6-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV6-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5-5 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV25-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-8 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5-4 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV4-1 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV10-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-6 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV11-3 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV11-2 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.216 of 216 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "zero-size array to reduction operation maximum which has no identity",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[91], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m tr_search \u001b[38;5;241m=\u001b[39m TCRrep(cell_df\u001b[38;5;241m=\u001b[39mhigh_df, organism\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhuman\u001b[39m\u001b[38;5;124m\"\u001b[39m, chains\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m\"\u001b[39m], compute_distances\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m----> 2\u001b[0m tr_search\u001b[38;5;241m.\u001b[39mcompute_distances()\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclone_df shape:\u001b[39m\u001b[38;5;124m\"\u001b[39m, tr_search\u001b[38;5;241m.\u001b[39mclone_df\u001b[38;5;241m.\u001b[39mshape)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:325\u001b[0m, in \u001b[0;36mTCRrep.compute_distances\u001b[0;34m(self, df)\u001b[0m\n\u001b[1;32m    323\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_distance_attributes(d \u001b[38;5;241m=\u001b[39m pw_alpha, chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124malpha\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    324\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchains:\n\u001b[0;32m--> 325\u001b[0m     pw_beta \u001b[38;5;241m=\u001b[39m _pws(\n\u001b[1;32m    326\u001b[0m         df \u001b[38;5;241m=\u001b[39m df, \n\u001b[1;32m    327\u001b[0m         metrics \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmetrics_b, \n\u001b[1;32m    328\u001b[0m         weights \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mweights_b, \n\u001b[1;32m    329\u001b[0m         kargs   \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkargs_b, \n\u001b[1;32m    330\u001b[0m         cpu     \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcpus, \n\u001b[1;32m    331\u001b[0m         store   \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstore_all_cdr)\n\u001b[1;32m    332\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_distance_attributes(d \u001b[38;5;241m=\u001b[39m pw_beta, chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    333\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgamma\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchains:\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/rep_funcs.py:115\u001b[0m, in \u001b[0;36m_pws\u001b[0;34m(df, metrics, weights, kargs, df2, cpu, uniquify, store)\u001b[0m\n\u001b[1;32m    113\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m metric_keys:\n\u001b[1;32m    114\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m df2 \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 115\u001b[0m         pw_mat \u001b[38;5;241m=\u001b[39m _pw(seqs1 \u001b[38;5;241m=\u001b[39m df[k]\u001b[38;5;241m.\u001b[39mvalues, metric \u001b[38;5;241m=\u001b[39m metrics[k], ncpus \u001b[38;5;241m=\u001b[39m cpu, uniqify\u001b[38;5;241m=\u001b[39m uniquify, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkargs[k])\n\u001b[1;32m    116\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    117\u001b[0m         pw_mat \u001b[38;5;241m=\u001b[39m _pw(seqs1 \u001b[38;5;241m=\u001b[39m df[k]\u001b[38;5;241m.\u001b[39mvalues, seqs2 \u001b[38;5;241m=\u001b[39m df2[k]\u001b[38;5;241m.\u001b[39mvalues, metric \u001b[38;5;241m=\u001b[39m metrics[k], ncpus \u001b[38;5;241m=\u001b[39m cpu, uniqify\u001b[38;5;241m=\u001b[39m uniquify, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkargs[k])\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/rep_funcs.py:134\u001b[0m, in \u001b[0;36m_pw\u001b[0;34m(metric, seqs1, seqs2, ncpus, uniqify, use_numba, **kwargs)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_pw\u001b[39m(metric, seqs1, seqs2\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, ncpus\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, uniqify\u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, use_numba \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    130\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    131\u001b[0m \u001b[38;5;124;03m    This is a wrapper for accessing pwseqdist version > 0.2.\u001b[39;00m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;124;03m    No matter what, it returns squareform results\u001b[39;00m\n\u001b[1;32m    133\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 134\u001b[0m     pw_mat \u001b[38;5;241m=\u001b[39m pw\u001b[38;5;241m.\u001b[39mapply_pairwise_rect(metric \u001b[38;5;241m=\u001b[39m metric, \n\u001b[1;32m    135\u001b[0m                                     seqs1  \u001b[38;5;241m=\u001b[39m seqs1, \n\u001b[1;32m    136\u001b[0m                                     seqs2  \u001b[38;5;241m=\u001b[39m seqs2, \n\u001b[1;32m    137\u001b[0m                                     ncpus  \u001b[38;5;241m=\u001b[39m ncpus, \n\u001b[1;32m    138\u001b[0m                                     uniqify \u001b[38;5;241m=\u001b[39m uniqify, \n\u001b[1;32m    139\u001b[0m                                     use_numba \u001b[38;5;241m=\u001b[39m use_numba,\n\u001b[1;32m    140\u001b[0m                                     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    141\u001b[0m     \u001b[38;5;66;03m# if len(pw_mat.shape) == 1:\u001b[39;00m\n\u001b[1;32m    142\u001b[0m     \u001b[38;5;66;03m#     from scipy.spatial.distance import squareform\u001b[39;00m\n\u001b[1;32m    143\u001b[0m     \u001b[38;5;66;03m#     pw_mat = squareform(pw_mat)\u001b[39;00m\n\u001b[1;32m    145\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m pw_mat\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pwseqdist/pairwise.py:140\u001b[0m, in \u001b[0;36mapply_pairwise_rect\u001b[0;34m(metric, seqs1, seqs2, ncpus, use_numba, uniqify, reexpand, alphabet, *args, **kwargs)\u001b[0m\n\u001b[1;32m    137\u001b[0m         urect \u001b[38;5;241m=\u001b[39m compute_many(pw_indices, metric, useqs, dtype, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    139\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 140\u001b[0m     seqs_mat, seqs_L \u001b[38;5;241m=\u001b[39m seqs2mat(useqs, alphabet\u001b[38;5;241m=\u001b[39malphabet)\n\u001b[1;32m    142\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ncpus \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(useqs) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m10\u001b[39m:\n\u001b[1;32m    143\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"Now a list of the chunked [chunksz x 2] arrays\"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pwseqdist/matrices.py:240\u001b[0m, in \u001b[0;36mseqs2mat\u001b[0;34m(seqs, alphabet, max_len)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Convert a collection of AA sequences into a\u001b[39;00m\n\u001b[1;32m    214\u001b[0m \u001b[38;5;124;03mnumpy matrix of integers for fast comparison.\u001b[39;00m\n\u001b[1;32m    215\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    237\u001b[0m \u001b[38;5;124;03mwherase mat in dict_from_matrix() is a parasail substitution Matrix object\u001b[39;00m\n\u001b[1;32m    238\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    239\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m max_len \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 240\u001b[0m     max_len \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmax([\u001b[38;5;28mlen\u001b[39m(s) \u001b[38;5;28;01mfor\u001b[39;00m s \u001b[38;5;129;01min\u001b[39;00m seqs])\n\u001b[1;32m    241\u001b[0m mat \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;241m*\u001b[39m np\u001b[38;5;241m.\u001b[39mones((\u001b[38;5;28mlen\u001b[39m(seqs), max_len), dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mint8)\n\u001b[1;32m    242\u001b[0m L \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mlen\u001b[39m(seqs), dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mint8)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/numpy/_core/fromnumeric.py:2899\u001b[0m, in \u001b[0;36mmax\u001b[0;34m(a, axis, out, keepdims, initial, where)\u001b[0m\n\u001b[1;32m   2781\u001b[0m \u001b[38;5;129m@array_function_dispatch\u001b[39m(_max_dispatcher)\n\u001b[1;32m   2782\u001b[0m \u001b[38;5;129m@set_module\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnumpy\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   2783\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmax\u001b[39m(a, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, out\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, keepdims\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue, initial\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue,\n\u001b[1;32m   2784\u001b[0m          where\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue):\n\u001b[1;32m   2785\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   2786\u001b[0m \u001b[38;5;124;03m    Return the maximum of an array or maximum along an axis.\u001b[39;00m\n\u001b[1;32m   2787\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   2897\u001b[0m \u001b[38;5;124;03m    5\u001b[39;00m\n\u001b[1;32m   2898\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 2899\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _wrapreduction(a, np\u001b[38;5;241m.\u001b[39mmaximum, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m'\u001b[39m, axis, \u001b[38;5;28;01mNone\u001b[39;00m, out,\n\u001b[1;32m   2900\u001b[0m                           keepdims\u001b[38;5;241m=\u001b[39mkeepdims, initial\u001b[38;5;241m=\u001b[39minitial, where\u001b[38;5;241m=\u001b[39mwhere)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/numpy/_core/fromnumeric.py:86\u001b[0m, in \u001b[0;36m_wrapreduction\u001b[0;34m(obj, ufunc, method, axis, dtype, out, **kwargs)\u001b[0m\n\u001b[1;32m     83\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     84\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m reduction(axis\u001b[38;5;241m=\u001b[39maxis, out\u001b[38;5;241m=\u001b[39mout, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpasskwargs)\n\u001b[0;32m---> 86\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ufunc\u001b[38;5;241m.\u001b[39mreduce(obj, axis, dtype, out, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpasskwargs)\n",
      "\u001b[0;31mValueError\u001b[0m: zero-size array to reduction operation maximum which has no identity"
     ]
    }
   ],
   "source": [
    "tr_search = TCRrep(cell_df=high_df, organism=\"human\", chains=[\"beta\"], compute_distances=False)\n",
    "tr_search.compute_distances()\n",
    "print(\"clone_df shape:\", tr_search.clone_df.shape)  # should be ≈ 200\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "98372d24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "bins  = np.zeros(5, dtype=int)\n",
    "freqs = np.array([0.1, 0.3, 0.6])\n",
    "# pretend we saw one of each at distance 0,1,2\n",
    "np.add.at(bins, [0,1,2], freqs)\n",
    "print(bins)   # → [0 0 0 0 0]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2fb1be4",
   "metadata": {},
   "source": [
    "##### Original raw counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b80f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#!/usr/bin/env python\n",
    "\"\"\"\n",
    "Compute TCRdist between a panel of high‑confidence CMV‑associated TRB clonotypes and each patient's bulk TCRB repertoire.\n",
    "The script mirrors the logic of the previous Levenshtein implementation but leverages **tcrdist3** for a biochemically‑weighted distance.\n",
    "\n",
    "## Dependencies\n",
    "\n",
    "```\n",
    "pip install tcrdist3 pandas numpy\n",
    "```\n",
    "\n",
    "## Input\n",
    "\n",
    "* A TSV (or CSV) file containing the CMV‑associated search set (≈200 rows).\n",
    "* A directory of TSV/CSV files – one per patient – already filtered to unique productive TRB clonotypes (≈3×10⁵ rows each).\n",
    "\n",
    "Both inputs must expose **cdr3\\_b\\_aa**, **v\\_b\\_gene**, **j\\_b\\_gene** columns (see `prepare_df`).\n",
    "\n",
    "## Output\n",
    "\n",
    "`tcrdist_output.csv` – One row per patient with a histogram of TCRdist values (0‥MAX\\_DIST).\n",
    "\n",
    "## Notes\n",
    "\n",
    "* The tcrdist scale is *not* the same as edit distance. A single amino‑acid substitution in CDR3 is ≈ 12 distance units when TRBV and TRBJ genes match. A radius of 36 therefore captures most functional \"one‑to‑two‑sub\" neighbours (Dash et al., 2017).\n",
    "* The computation is executed as a sparse **rectangular** distance matrix; only distances ≤ `RADIUS` are kept in memory.\n",
    "* Adapt `MAX_DIST`, `RADIUS`, and path variables as needed.\n",
    "  \"\"\"\n",
    "\n",
    "import os\n",
    "import re\n",
    "import multiprocessing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tcrdist.repertoire import TCRrep\n",
    "from tcrdist.breadth import get_safe_chunk\n",
    "\n",
    "excluded_genes = []\n",
    "\n",
    "print(' - Loaded libraries.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "# User‑configurable parameters\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "MAX_DIST   = 50        # highest TCRdist bucket to tally (inclusive)\n",
    "RADIUS     = MAX_DIST  #distances > RADIUS are discarded by tcrdist3\n",
    "HIGHCONF   = highconf   # search‑set file\n",
    "PATIENT_DIR = \"/Volumes/IshaVerbat/Isha/TCR/filtered\\_for\\_tcrdist\"  # directory with patient files\n",
    "OUTFILE    = \"tcrdist\\_output.csv\"\n",
    "N_CPUS     = min(multiprocessing.cpu_count(), 8)\n",
    "\n",
    "# set wd\n",
    "\n",
    "os.chdir(\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist\")\n",
    "\n",
    "MAX_FILES = 4\n",
    "\n",
    "print(' - Set parameters.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "# Helper functions\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "GENE_PAT = re.compile(r\"\\[*!].*\\$\")\n",
    "\n",
    "def _normalize_gene(s: str) -> str:\n",
    "\n",
    " \"\"\" Strip prefixes / allele info to IMGT-like gene names understood by tcrdist3,\n",
    "remove leading zeros from gene numbers (e.g. TRBV05 → TRBV5),\n",
    "and from any numeric subfields (e.g. TRBV5-05 → TRBV5-5) \"\"\"\n",
    "\n",
    "if pd.isna(s):\n",
    "return np.nan\n",
    "\n",
    "```\n",
    "print(f'Processing: {s}')\n",
    "# Standardize prefix\n",
    "s = str(s).replace(\"TCR\", \"TR\")  # e.g. TCRBV05-05 → TRBV05-05\n",
    "\n",
    "# Remove leading zeros from primary gene number (e.g. TRBV05 → TRBV5)\n",
    "s = re.sub(\n",
    "    r\"^([A-Z]+)(0+)(\\d+)\",\n",
    "    lambda m: f\"{m.group(1)}{int(m.group(2) + m.group(3))}\",\n",
    "    s,\n",
    ")\n",
    "\n",
    "# Remove leading zeros from any subfield after hyphens (e.g. -05 → -5)\n",
    "s = re.sub(\n",
    "    r\"-(0+)(\\d+)\",\n",
    "    lambda m: f\"-{int(m.group(1) + m.group(2))}\",\n",
    "    s,\n",
    ")\n",
    "\n",
    "# Normalize allele information to *01\n",
    "if \"*\" in s:\n",
    "    s = re.sub(r\"[*!].*$\", \"*01\", s)\n",
    "else:\n",
    "    s = s + \"*01\"\n",
    "\n",
    "# Fix adaptive errors in V gene names - they added in -01 for families with only one gene. \n",
    "# Get the number from the string\n",
    "\n",
    "if not re.match(r'^TRB[JV]\\d+(-\\d+)?\\*\\d+$', s):\n",
    "    excluded_genes.append(s)\n",
    "    print(f\"Excluded malformed gene: {s}\")\n",
    "    return np.nan\n",
    "\n",
    "number = re.search(r'TRB[JV](\\d+)', s).group(1)\n",
    "\n",
    "if number in no_hyphen_numbers and s not in trb_v_list:\n",
    "    print(f'{s}: no match')\n",
    "    s = re.sub(r'-\\d+(?=\\*)', '', s)\n",
    "    print(f'Changed to {s}')\n",
    "\n",
    "return s\n",
    "```\n",
    "\n",
    "def prepare_df(df: pd.DataFrame) -> pd.DataFrame:\n",
    "\"\"\"Bring an arbitrary clonotype table to tcrdist3‑ready shape.\"\"\"\n",
    "# Remove d\\_gene column\n",
    "if \"d_gene\" in df.columns:\n",
    "df.drop(columns=[\"d_gene\"], inplace=True)\n",
    "\n",
    "```\n",
    "rename_map = {\n",
    "    \"amino_acid\": \"cdr3_b_aa\",\n",
    "    \"cdr3s_aa\": \"cdr3_b_aa\",\n",
    "    \"v_call\": \"v_b_gene\",\n",
    "    \"j_call\": \"j_b_gene\",\n",
    "}\n",
    "\n",
    "df = df.rename(columns=rename_map)\n",
    "need = [\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"]\n",
    "df = df[need].copy()\n",
    "df.dropna(inplace=True)\n",
    "\n",
    "df[\"v_b_gene\"] = df[\"v_b_gene\"].apply(_normalize_gene)\n",
    "df[\"j_b_gene\"] = df[\"j_b_gene\"].apply(_normalize_gene)\n",
    "\n",
    "\n",
    "df = df.drop_duplicates().reset_index(drop=True)\n",
    "return df\n",
    "\n",
    "\n",
    "print('Defined functions.')\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "# Load the CMV‑associated search set once\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "print(\"Loading high‑confidence CMV search set …\")\n",
    "high_raw  = highconf\n",
    "high_df   = prepare_df(high_raw)\n",
    "print(high_df.head())\n",
    "\n",
    "print(f\" → {len(high_df):,} unique TRB clonotypes ready for search.\")\n",
    "\n",
    "tr_search = TCRrep(cell_df=high_df, organism=\"human\", chains=[\"beta\"], compute_distances=False)\n",
    "\n",
    "tr_search.cpus = N_CPUS\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "# Prepare output frame – one row per patient\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "cols = [\"patient_id\"] + [f\"dist_{d}\" for d in range(MAX_DIST + 1)]\n",
    "results = pd.DataFrame(columns=cols)\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "# Main loop over patient repertoires\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "file_names = sorted(fn for fn in os.listdir(PATIENT_DIR) if fn.endswith(('.tsv', '.csv')) and fn.startswith('P'))\n",
    "\n",
    "for idx, fname in enumerate(file_names, 1):\n",
    "if idx > MAX_FILES:\n",
    "print(f\"Reached maximum number of files to process ({MAX_FILES}).\")\n",
    "break\n",
    "patient_id = os.path.splitext(fname)[0]\n",
    "print(f\"Processing ({idx}/{len(file_names)}): {patient_id}\")\n",
    "\n",
    "```\n",
    "# Load & clean bulk repertoire\n",
    "rep_raw = pd.read_csv(os.path.join(PATIENT_DIR, fname), sep=\"\\t\", low_memory=False)\n",
    "rep_df  = prepare_df(rep_raw)\n",
    "\n",
    "if rep_df.empty:\n",
    "    print(f\" ! No usable sequences in {patient_id} – skipping.\")\n",
    "    continue\n",
    "\n",
    "tr_bulk = TCRrep(cell_df=rep_df, organism=\"human\", chains=[\"beta\"], compute_distances=False)\n",
    "\n",
    "\n",
    "\n",
    "# Compute sparse rectangular distances (rows=search, cols=bulk)\n",
    "chunk_size = get_safe_chunk(tr_search.clone_df.shape[0], tr_bulk.clone_df.shape[0])\n",
    "tr_search.compute_sparse_rect_distances(\n",
    "    df=tr_search.clone_df,\n",
    "    df2=tr_bulk.clone_df,\n",
    "    radius=RADIUS,\n",
    "    chunk_size=chunk_size,\n",
    ")\n",
    "\n",
    "# Histogram distances ≤ RADIUS\n",
    "bins = np.zeros(MAX_DIST + 1, dtype=int)\n",
    "for d in tr_search.rw_beta.data.astype(int):\n",
    "    bins[d] += 1\n",
    "\n",
    "# Append to results\n",
    "row = {\"patient_id\": patient_id, **{f\"dist_{d}\": bins[d] for d in range(MAX_DIST + 1)}}\n",
    "results = pd.concat([results, pd.DataFrame([row])], ignore_index=True)\n",
    "\n",
    "# Persist after each patient\n",
    "results.to_csv(OUTFILE, index=False)\n",
    "print(f\"   ↳ saved ➜ {OUTFILE}\")\n",
    "\n",
    "# Free distance matrix to keep RAM stable before next iteration\n",
    "del tr_search.rw_beta\n",
    "```\n",
    "\n",
    "print(\"✔︎ All patient repertoires processed.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d05de488",
   "metadata": {},
   "source": [
    "### Edited to be freq. USE THIS IT WORKS (i think)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "92838937",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "↪ Resuming from checkpoint — 99 patients already processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing (1/292): P00002.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.692 of 73706 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "37d1c62782ca4af1a893cacfcee0ccfe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 14220\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1755\n",
      "  • Example matched freqs (first 10): [2.66106306e-06 5.32212612e-06 6.20914714e-06 1.68533994e-05\n",
      " 1.68533994e-05 0.00000000e+00 1.24182943e-05 1.50793573e-05\n",
      " 2.66106306e-06 2.66106306e-06]\n",
      "Processing (2/292): P00003.slim\n",
      "↪ Already processed P00003.slim – skipping.\n",
      "Processing (3/292): P00004.slim\n",
      "↪ Already processed P00004.slim – skipping.\n",
      "Processing (4/292): P00006.slim\n",
      "↪ Already processed P00006.slim – skipping.\n",
      "Processing (5/292): P00007.slim\n",
      "↪ Already processed P00007.slim – skipping.\n",
      "Processing (6/292): P00008.slim\n",
      "↪ Already processed P00008.slim – skipping.\n",
      "Processing (7/292): P00010.slim\n",
      "↪ Already processed P00010.slim – skipping.\n",
      "Processing (8/292): P00011.slim\n",
      "↪ Already processed P00011.slim – skipping.\n",
      "Processing (9/292): P00014.slim\n",
      "↪ Already processed P00014.slim – skipping.\n",
      "Processing (10/292): P00017.slim\n",
      "↪ Already processed P00017.slim – skipping.\n",
      "Processing (11/292): P00019.slim\n",
      "↪ Already processed P00019.slim – skipping.\n",
      "Processing (12/292): P00020.slim\n",
      "↪ Already processed P00020.slim – skipping.\n",
      "Processing (13/292): P00021.slim\n",
      "↪ Already processed P00021.slim – skipping.\n",
      "Processing (14/292): P00024.slim\n",
      "↪ Already processed P00024.slim – skipping.\n",
      "Processing (15/292): P00029.slim\n",
      "↪ Already processed P00029.slim – skipping.\n",
      "Processing (16/292): P00030.slim\n",
      "↪ Already processed P00030.slim – skipping.\n",
      "Processing (17/292): P00031.slim\n",
      "↪ Already processed P00031.slim – skipping.\n",
      "Processing (18/292): P00032.slim\n",
      "↪ Already processed P00032.slim – skipping.\n",
      "Processing (19/292): P00036.slim\n",
      "↪ Already processed P00036.slim – skipping.\n",
      "Processing (20/292): P00037.slim\n",
      "↪ Already processed P00037.slim – skipping.\n",
      "Processing (21/292): P00038.slim\n",
      "↪ Already processed P00038.slim – skipping.\n",
      "Processing (22/292): P00040.slim\n",
      "↪ Already processed P00040.slim – skipping.\n",
      "Processing (23/292): P00044.slim\n",
      "↪ Already processed P00044.slim – skipping.\n",
      "Processing (24/292): P00045.slim\n",
      "↪ Already processed P00045.slim – skipping.\n",
      "Processing (25/292): P00048.slim\n",
      "↪ Already processed P00048.slim – skipping.\n",
      "Processing (26/292): P00049.slim\n",
      "↪ Already processed P00049.slim – skipping.\n",
      "Processing (27/292): P00054.slim\n",
      "↪ Already processed P00054.slim – skipping.\n",
      "Processing (28/292): P00055.slim\n",
      "↪ Already processed P00055.slim – skipping.\n",
      "Processing (29/292): P00056.slim\n",
      "↪ Already processed P00056.slim – skipping.\n",
      "Processing (30/292): P00058.slim\n",
      "↪ Already processed P00058.slim – skipping.\n",
      "Processing (31/292): P00059.slim\n",
      "↪ Already processed P00059.slim – skipping.\n",
      "Processing (32/292): P00063.slim\n",
      "↪ Already processed P00063.slim – skipping.\n",
      "Processing (33/292): P00065.slim\n",
      "↪ Already processed P00065.slim – skipping.\n",
      "Processing (34/292): P00066.slim\n",
      "↪ Already processed P00066.slim – skipping.\n",
      "Processing (35/292): P00067.slim\n",
      "↪ Already processed P00067.slim – skipping.\n",
      "Processing (36/292): P00072.slim\n",
      "↪ Already processed P00072.slim – skipping.\n",
      "Processing (37/292): P00073.slim\n",
      "↪ Already processed P00073.slim – skipping.\n",
      "Processing (38/292): P00077.slim\n",
      "↪ Already processed P00077.slim – skipping.\n",
      "Processing (39/292): P00079.slim\n",
      "↪ Already processed P00079.slim – skipping.\n",
      "Processing (40/292): P00083.slim\n",
      "↪ Already processed P00083.slim – skipping.\n",
      "Processing (41/292): P00084.slim\n",
      "↪ Already processed P00084.slim – skipping.\n",
      "Processing (42/292): P00085.slim\n",
      "↪ Already processed P00085.slim – skipping.\n",
      "Processing (43/292): P00086.slim\n",
      "↪ Already processed P00086.slim – skipping.\n",
      "Processing (44/292): P00088.slim\n",
      "↪ Already processed P00088.slim – skipping.\n",
      "Processing (45/292): P00091.slim\n",
      "↪ Already processed P00091.slim – skipping.\n",
      "Processing (46/292): P00093.slim\n",
      "↪ Already processed P00093.slim – skipping.\n",
      "Processing (47/292): P00094.slim\n",
      "↪ Already processed P00094.slim – skipping.\n",
      "Processing (48/292): P00097.slim\n",
      "↪ Already processed P00097.slim – skipping.\n",
      "Processing (49/292): P00098.slim\n",
      "↪ Already processed P00098.slim – skipping.\n",
      "Processing (50/292): P00099.slim\n",
      "↪ Already processed P00099.slim – skipping.\n",
      "Processing (51/292): P00106.slim\n",
      "↪ Already processed P00106.slim – skipping.\n",
      "Processing (52/292): P00107.slim\n",
      "↪ Already processed P00107.slim – skipping.\n",
      "Processing (53/292): P00110.slim\n",
      "↪ Already processed P00110.slim – skipping.\n",
      "Processing (54/292): P00111.slim\n",
      "↪ Already processed P00111.slim – skipping.\n",
      "Processing (55/292): P00112.slim\n",
      "↪ Already processed P00112.slim – skipping.\n",
      "Processing (56/292): P00116.slim\n",
      "↪ Already processed P00116.slim – skipping.\n",
      "Processing (57/292): P00120.slim\n",
      "↪ Already processed P00120.slim – skipping.\n",
      "Processing (58/292): P00125.slim\n",
      "↪ Already processed P00125.slim – skipping.\n",
      "Processing (59/292): P00128.slim\n",
      "↪ Already processed P00128.slim – skipping.\n",
      "Processing (60/292): P00130.slim\n",
      "↪ Already processed P00130.slim – skipping.\n",
      "Processing (61/292): P00131.slim\n",
      "↪ Already processed P00131.slim – skipping.\n",
      "Processing (62/292): P00132.slim\n",
      "↪ Already processed P00132.slim – skipping.\n",
      "Processing (63/292): P00133.slim\n",
      "↪ Already processed P00133.slim – skipping.\n",
      "Processing (64/292): P00135.slim\n",
      "↪ Already processed P00135.slim – skipping.\n",
      "Processing (65/292): P00137.slim\n",
      "↪ Already processed P00137.slim – skipping.\n",
      "Processing (66/292): P00138.slim\n",
      "↪ Already processed P00138.slim – skipping.\n",
      "Processing (67/292): P00140.slim\n",
      "↪ Already processed P00140.slim – skipping.\n",
      "Processing (68/292): P00141.slim\n",
      "↪ Already processed P00141.slim – skipping.\n",
      "Processing (69/292): P00142.slim\n",
      "↪ Already processed P00142.slim – skipping.\n",
      "Processing (70/292): P00143.slim\n",
      "↪ Already processed P00143.slim – skipping.\n",
      "Processing (71/292): P00144.slim\n",
      "↪ Already processed P00144.slim – skipping.\n",
      "Processing (72/292): P00148.slim\n",
      "↪ Already processed P00148.slim – skipping.\n",
      "Processing (73/292): P00150.slim\n",
      "↪ Already processed P00150.slim – skipping.\n",
      "Processing (74/292): P00151.slim\n",
      "↪ Already processed P00151.slim – skipping.\n",
      "Processing (75/292): P00152.slim\n",
      "↪ Already processed P00152.slim – skipping.\n",
      "Processing (76/292): P00161.slim\n",
      "↪ Already processed P00161.slim – skipping.\n",
      "Processing (77/292): P00163.slim\n",
      "↪ Already processed P00163.slim – skipping.\n",
      "Processing (78/292): P00165.slim\n",
      "↪ Already processed P00165.slim – skipping.\n",
      "Processing (79/292): P00166.slim\n",
      "↪ Already processed P00166.slim – skipping.\n",
      "Processing (80/292): P00167.slim\n",
      "↪ Already processed P00167.slim – skipping.\n",
      "Processing (81/292): P00170.slim\n",
      "↪ Already processed P00170.slim – skipping.\n",
      "Processing (82/292): P00171.slim\n",
      "↪ Already processed P00171.slim – skipping.\n",
      "Processing (83/292): P00173.slim\n",
      "↪ Already processed P00173.slim – skipping.\n",
      "Processing (84/292): P00177.slim\n",
      "↪ Already processed P00177.slim – skipping.\n",
      "Processing (85/292): P00178.slim\n",
      "↪ Already processed P00178.slim – skipping.\n",
      "Processing (86/292): P00182.slim\n",
      "↪ Already processed P00182.slim – skipping.\n",
      "Processing (87/292): P00191.slim\n",
      "↪ Already processed P00191.slim – skipping.\n",
      "Processing (88/292): P00193.slim\n",
      "↪ Already processed P00193.slim – skipping.\n",
      "Processing (89/292): P00194.slim\n",
      "↪ Already processed P00194.slim – skipping.\n",
      "Processing (90/292): P00196.slim\n",
      "↪ Already processed P00196.slim – skipping.\n",
      "Processing (91/292): P00198.slim\n",
      "↪ Already processed P00198.slim – skipping.\n",
      "Processing (92/292): P00200.slim\n",
      "↪ Already processed P00200.slim – skipping.\n",
      "Processing (93/292): P00201.slim\n",
      "↪ Already processed P00201.slim – skipping.\n",
      "Processing (94/292): P00203.slim\n",
      "↪ Already processed P00203.slim – skipping.\n",
      "Processing (95/292): P00205.slim\n",
      "↪ Already processed P00205.slim – skipping.\n",
      "Processing (96/292): P00208.slim\n",
      "↪ Already processed P00208.slim – skipping.\n",
      "Processing (97/292): P00209.slim\n",
      "↪ Already processed P00209.slim – skipping.\n",
      "Processing (98/292): P00211.slim\n",
      "↪ Already processed P00211.slim – skipping.\n",
      "Processing (99/292): P00212.slim\n",
      "↪ Already processed P00212.slim – skipping.\n",
      "Processing (100/292): P00214.slim\n",
      "↪ Already processed P00214.slim – skipping.\n",
      "Processing (101/292): P00216.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "113\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.760 of 89493 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "966874ac3c0d47528b9883dcc7d3bbdc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 23363\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.2088\n",
      "  • Example matched freqs (first 10): [3.40120147e-05 1.13373382e-05 2.26746765e-05 1.70060074e-05\n",
      " 5.66866912e-06 7.08583641e-06 1.13373382e-05 1.13373382e-05\n",
      " 7.08583641e-06 1.13373382e-05]\n",
      "Processing (102/292): P00217.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.2537 of 201367 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cc9ecc69a4594dd4b80eca584bbd0d2f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 42756\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1627\n",
      "  • Example matched freqs (first 10): [3.34986762e-06 3.22579845e-06 1.36476088e-06 8.79650423e-05\n",
      " 3.72207513e-07 1.36476088e-06 0.00000000e+00 3.22579845e-06\n",
      " 3.97021347e-06 5.21090518e-06]\n",
      "Processing (103/292): P00221.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "89\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.883 of 113268 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6e74acef4f1f452cb9560494b450e85f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 24848\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1701\n",
      "  • Example matched freqs (first 10): [3.75818236e-06 2.60181856e-06 5.78181901e-07 3.46909141e-06\n",
      " 3.75818236e-06 1.15636380e-06 6.36000091e-06 3.75818236e-06\n",
      " 3.75818236e-06 7.22727377e-06]\n",
      "Processing (104/292): P00224.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.1985 of 283086 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "43901df00842432a81ee02caf4440af4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 60732\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1766\n",
      "  • Example matched freqs (first 10): [1.84177354e-06 3.99050934e-06 2.45569806e-06 7.67405642e-07\n",
      " 9.97627335e-06 4.60443385e-07 1.07436790e-06 7.67405642e-07\n",
      " 0.00000000e+00 7.67405642e-07]\n",
      "Processing (105/292): P00225.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.1137 of 151201 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6c89d972fe284af59bbe65ecc4bc5ab2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 29612\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1726\n",
      "  • Example matched freqs (first 10): [1.44315619e-06 2.99732439e-06 2.66428835e-06 2.22024029e-07\n",
      " 3.10833640e-06 2.88631237e-06 3.33036043e-07 3.44137245e-06\n",
      " 3.21934842e-06 2.66428835e-06]\n",
      "Processing (106/292): P00229.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.2917 of 259189 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bd4b40ee606f4e4f952deb6a55fd4e0c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 57090\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1694\n",
      "  • Example matched freqs (first 10): [8.48571737e-07 1.90928641e-06 4.24285869e-06 6.36428803e-07\n",
      " 2.33357228e-06 1.48500054e-06 1.27285761e-06 4.24285869e-06\n",
      " 3.81857282e-06 6.36428803e-07]\n",
      "Processing (107/292): P00230.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.2943 of 256935 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9cce6492e30d49cfb5f7c171115ac65e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 59702\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1965\n",
      "  • Example matched freqs (first 10): [7.19828652e-07 7.19828652e-07 7.19828652e-07 1.11573441e-05\n",
      " 2.87931461e-06 5.39871489e-07 1.97952879e-06 1.61961447e-06\n",
      " 1.79957163e-06 1.25970014e-06]\n",
      "Processing (108/292): P00232.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.957 of 164969 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "92ed87e709214b4bb05d171880c70efb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 33920\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1882\n",
      "  • Example matched freqs (first 10): [4.09955300e-07 4.50950830e-06 3.00633887e-06 4.78281183e-06\n",
      " 1.31185696e-05 6.83258833e-07 5.73937420e-06 8.19910600e-07\n",
      " 3.27964240e-06 1.63982120e-06]\n",
      "Processing (109/292): P00233.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.3797 of 283869 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d5a771e3f7094e3ba1555161dd8d038a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SANITY CHECK ===\n",
      "Total non-zero entries (matches): 58801\n",
      "  • Distance range: -1 to 50\n",
      "  • Sum of matched frequencies: 0.1809\n",
      "  • Example matched freqs (first 10): [1.75010140e-06 5.65417377e-06 4.17331873e-06 3.50020281e-06\n",
      " 6.05804332e-06 6.73115925e-07 1.88472459e-06 0.00000000e+00\n",
      " 3.63482599e-06 2.42321733e-05]\n",
      "Processing (110/292): P00239.slim\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency\n",
      "\n",
      "  self._validate_cell_df()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:791: UserWarning: No 'count' column provided; count column set to 1\n",
      "  warnings.warn(\"No 'count' column provided; count column set to 1\")\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:165: UserWarning: db_file must be 'alphabeta_gammadelta_db.tsv' or 'alphabeta_db.tsv' or 'gammadelta_db.tsv' unless you have built tcrdist3 from scratch\n",
      "  self._validate_db_file()\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:500: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f0 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:504: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f1 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV13-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV7-5*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:508: UserWarning: TRBV22-1*01 gene was not recognized in reference db no cdr seq could be inferred\n",
      "  f2 = lambda v : self._map_gene_to_reference_seq2(gene = v,\n",
      "/Users/ishaharris/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:190: UserWarning: Not all cells/sequences could be grouped into clones.2391 of 208549 were not captured. This occurs when any of the values in the index columns are null or missing for a given sequence. To see entries with missing values use: tcrdist.repertoire.TCRrep._show_incomplete()\n",
      "  self.deduplicate()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-107:\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 204\u001b[0m\n\u001b[1;32m    199\u001b[0m \u001b[38;5;66;03m## 5.3 Compute sparse distances\u001b[39;00m\n\u001b[1;32m    200\u001b[0m chunk_size \u001b[38;5;241m=\u001b[39m get_safe_chunk(\n\u001b[1;32m    201\u001b[0m     tr_search\u001b[38;5;241m.\u001b[39mclone_df\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m],\n\u001b[1;32m    202\u001b[0m     tr_bulk\u001b[38;5;241m.\u001b[39mclone_df\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    203\u001b[0m )\n\u001b[0;32m--> 204\u001b[0m tr_search\u001b[38;5;241m.\u001b[39mcompute_sparse_rect_distances(\n\u001b[1;32m    205\u001b[0m     df\u001b[38;5;241m=\u001b[39mtr_search\u001b[38;5;241m.\u001b[39mclone_df,\n\u001b[1;32m    206\u001b[0m     df2\u001b[38;5;241m=\u001b[39mtr_bulk\u001b[38;5;241m.\u001b[39mclone_df,\n\u001b[1;32m    207\u001b[0m     radius\u001b[38;5;241m=\u001b[39mRADIUS,\n\u001b[1;32m    208\u001b[0m     chunk_size\u001b[38;5;241m=\u001b[39mchunk_size,\n\u001b[1;32m    209\u001b[0m )\n\u001b[1;32m    211\u001b[0m \u001b[38;5;66;03m## 5.4 Build histogram\u001b[39;00m\n\u001b[1;32m    212\u001b[0m coo   \u001b[38;5;241m=\u001b[39m tr_search\u001b[38;5;241m.\u001b[39mrw_beta\u001b[38;5;241m.\u001b[39mtocoo()\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:378\u001b[0m, in \u001b[0;36mTCRrep.compute_sparse_rect_distances\u001b[0;34m(self, df, df2, radius, chunk_size)\u001b[0m\n\u001b[1;32m    355\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_sparse_rect_distances\u001b[39m(\u001b[38;5;28mself\u001b[39m, df\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, df2\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, radius\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m50\u001b[39m, chunk_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m100\u001b[39m):\n\u001b[1;32m    356\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;124;03m    Computes sparse rectangular distance as a scipy.sparse csr matrix, only retaining distances <= radius. \u001b[39;00m\n\u001b[1;32m    358\u001b[0m \u001b[38;5;124;03m    This function is crucial when working with bulk data. Values > radius are zeros and not \u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    376\u001b[0m \u001b[38;5;124;03m        to tune chunk size to the scale of the task.\u001b[39;00m\n\u001b[1;32m    377\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 378\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_rect_distances(pw_dist_func\u001b[38;5;241m=\u001b[39mcompute_pws_sparse, df\u001b[38;5;241m=\u001b[39mdf, df2\u001b[38;5;241m=\u001b[39mdf2, radius\u001b[38;5;241m=\u001b[39mradius, chunk_size\u001b[38;5;241m=\u001b[39mchunk_size)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/repertoire.py:398\u001b[0m, in \u001b[0;36mTCRrep._rect_distances\u001b[0;34m(self, pw_dist_func, df, df2, **kwargs)\u001b[0m\n\u001b[1;32m    396\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_distance_attributes(d \u001b[38;5;241m=\u001b[39m pw_alpha, chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124malpha\u001b[39m\u001b[38;5;124m'\u001b[39m,prefix \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrw\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    397\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchains:\n\u001b[0;32m--> 398\u001b[0m     pw_beta \u001b[38;5;241m=\u001b[39m pw_dist_func(\n\u001b[1;32m    399\u001b[0m         df \u001b[38;5;241m=\u001b[39m df,\n\u001b[1;32m    400\u001b[0m         df2 \u001b[38;5;241m=\u001b[39m df2, \n\u001b[1;32m    401\u001b[0m         metrics \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmetrics_b, \n\u001b[1;32m    402\u001b[0m         weights \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mweights_b, \n\u001b[1;32m    403\u001b[0m         kargs   \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkargs_b, \n\u001b[1;32m    404\u001b[0m         cpu     \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcpus, \n\u001b[1;32m    405\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_distance_attributes(d \u001b[38;5;241m=\u001b[39m pw_beta, chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m'\u001b[39m,prefix \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrw\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    407\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgamma\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchains:\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/tcrdist/rep_funcs.py:244\u001b[0m, in \u001b[0;36mcompute_pws_sparse\u001b[0;34m(df, metrics, weights, kargs, radius, df2, cpu, chunk_size, store, pm_pbar)\u001b[0m\n\u001b[1;32m    241\u001b[0m         seqs2 \u001b[38;5;241m=\u001b[39m df2[k]\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m    243\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"With reexapnd = False, returns: pw_mat, uind_i1, uind_i2\"\"\"\u001b[39;00m\n\u001b[0;32m--> 244\u001b[0m     pwmat, ind1, ind2 \u001b[38;5;241m=\u001b[39m pw\u001b[38;5;241m.\u001b[39mapply_pairwise_rect(metric\u001b[38;5;241m=\u001b[39mmetrics[k], \n\u001b[1;32m    245\u001b[0m                                                 seqs1\u001b[38;5;241m=\u001b[39mdf[k]\u001b[38;5;241m.\u001b[39mvalues, \n\u001b[1;32m    246\u001b[0m                                                 seqs2\u001b[38;5;241m=\u001b[39mseqs2, \n\u001b[1;32m    247\u001b[0m                                                 ncpus\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mmin\u001b[39m(cpu, \u001b[38;5;241m2\u001b[39m),\n\u001b[1;32m    248\u001b[0m                                                 uniqify\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, \n\u001b[1;32m    249\u001b[0m                                                 reexpand\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    250\u001b[0m                                                 \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkargs[k])\n\u001b[1;32m    252\u001b[0m     components[k] \u001b[38;5;241m=\u001b[39m (pwmat \u001b[38;5;241m*\u001b[39m weights[k], ind1, ind2) \n\u001b[1;32m    254\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Can't do this because it will be huge. Also can't compute list of\u001b[39;00m\n\u001b[1;32m    255\u001b[0m \u001b[38;5;124;03mpotential D < radius because that also could be too large.\u001b[39;00m\n\u001b[1;32m    256\u001b[0m \u001b[38;5;124;03mBut need to pre-compute these non-CDR3 because otherwise the CDR3\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    259\u001b[0m \u001b[38;5;124;03mSolution: chunk the computation here. Only compute subsets of potential\u001b[39;00m\n\u001b[1;32m    260\u001b[0m \u001b[38;5;124;03mseqs pairs sparsely and spread across processors.\"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/pwseqdist/pairwise.py:149\u001b[0m, in \u001b[0;36mapply_pairwise_rect\u001b[0;34m(metric, seqs1, seqs2, ncpus, use_numba, uniqify, reexpand, alphabet, *args, **kwargs)\u001b[0m\n\u001b[1;32m    146\u001b[0m     \u001b[38;5;66;03m# , prefer='threads', require='sharedmem'\u001b[39;00m\n\u001b[1;32m    147\u001b[0m     \u001b[38;5;66;03m# dists = Parallel(n_jobs=ncpus)(delayed(metric)(pw_i, seqs_mat, seqs_L, *numba_args) for pw_i in chunked_indices)\u001b[39;00m\n\u001b[1;32m    148\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m multiprocessing\u001b[38;5;241m.\u001b[39mPool(ncpus) \u001b[38;5;28;01mas\u001b[39;00m pool:\n\u001b[0;32m--> 149\u001b[0m         dists \u001b[38;5;241m=\u001b[39m parmap\u001b[38;5;241m.\u001b[39mmap(metric,\n\u001b[1;32m    150\u001b[0m                            chunked_indices,\n\u001b[1;32m    151\u001b[0m                            seqs_mat,\n\u001b[1;32m    152\u001b[0m                            seqs_L,\n\u001b[1;32m    153\u001b[0m                            \u001b[38;5;241m*\u001b[39margs,\n\u001b[1;32m    154\u001b[0m                            \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    155\u001b[0m                            pm_parallel\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    156\u001b[0m                            pm_pool\u001b[38;5;241m=\u001b[39mpool)\n\u001b[1;32m    157\u001b[0m     urect \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mconcatenate(dists)\n\u001b[1;32m    158\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/parmap/parmap.py:333\u001b[0m, in \u001b[0;36mmap\u001b[0;34m(function, iterable, *args, **kwargs)\u001b[0m\n\u001b[1;32m    307\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmap\u001b[39m(function, iterable, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    308\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"This function is equivalent to:\u001b[39;00m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;124;03m     >>> [function(x, args[0], args[1],...) for x in iterable]\u001b[39;00m\n\u001b[1;32m    310\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    331\u001b[0m \u001b[38;5;124;03m    :type pm_pbar: bool, dict or callable\u001b[39;00m\n\u001b[1;32m    332\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 333\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _map_or_starmap(function, iterable, args, kwargs, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmap\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/site-packages/parmap/parmap.py:267\u001b[0m, in \u001b[0;36m_map_or_starmap\u001b[0;34m(function, iterable, args, kwargs, map_or_starmap)\u001b[0m\n\u001b[1;32m    261\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    262\u001b[0m     result \u001b[38;5;241m=\u001b[39m pool\u001b[38;5;241m.\u001b[39mmap_async(\n\u001b[1;32m    263\u001b[0m         func_star,\n\u001b[1;32m    264\u001b[0m         \u001b[38;5;28mzip\u001b[39m(repeat(function), iterable, repeat(\u001b[38;5;28mlist\u001b[39m(args)), repeat(kwargs)),\n\u001b[1;32m    265\u001b[0m         chunksize,\n\u001b[1;32m    266\u001b[0m     )\n\u001b[0;32m--> 267\u001b[0m     output \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mget()\n\u001b[1;32m    268\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m    269\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m close_pool:\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/multiprocessing/pool.py:768\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 768\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwait(timeout)\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mready():\n\u001b[1;32m    770\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTimeoutError\u001b[39;00m\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/multiprocessing/pool.py:765\u001b[0m, in \u001b[0;36mApplyResult.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwait\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 765\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event\u001b[38;5;241m.\u001b[39mwait(timeout)\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/threading.py:629\u001b[0m, in \u001b[0;36mEvent.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    627\u001b[0m signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_flag\n\u001b[1;32m    628\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m signaled:\n\u001b[0;32m--> 629\u001b[0m     signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cond\u001b[38;5;241m.\u001b[39mwait(timeout)\n\u001b[1;32m    630\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m signaled\n",
      "File \u001b[0;32m~/Projects/TCR/TCR-Isha/tcr_env/lib/python3.11/threading.py:327\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    326\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 327\u001b[0m         waiter\u001b[38;5;241m.\u001b[39macquire()\n\u001b[1;32m    328\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    329\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "\"\"\"\n",
    "Compute TCRdist between a panel of high-confidence CMV-associated TRB clonotypes and each patient's bulk TCRB repertoire.\n",
    "Uses tcrdist3 for biochemically-weighted distances, and writes out one CSV every 10 patients.\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import re\n",
    "import multiprocessing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tcrdist.repertoire import TCRrep\n",
    "from tcrdist.breadth import get_safe_chunk\n",
    "import pickle\n",
    "\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 0. IMPORT PICKLES \n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "PICKLE_DIR = \"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/pickles/\"  # ← adjust to your actual directory\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/trb_v_no_hyphen_list.pkl\", \"rb\") as f:\n",
    "    trb_v_no_hyphen_list = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/trb_v_hyphen_list.pkl\", \"rb\") as f:\n",
    "    trb_v_hyphen_list = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/j_no_hyphen_numbers.pkl\", \"rb\") as f:\n",
    "    j_no_hyphen_numbers = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/j_hyphen_numbers.pkl\", \"rb\") as f:\n",
    "    j_hyphen_numbers = pickle.load(f)\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 1. CONSTANTS & CONFIGURATION\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "MAX_DIST    = 50        # highest TCRdist bucket to tally (inclusive)\n",
    "RADIUS      = MAX_DIST  # distances > RADIUS are discarded by tcrdist3\n",
    "HIGHCONF    = highconf  # your preloaded DataFrame of CMV-associated clonotypes\n",
    "PATIENT_DIR = \"/Volumes/IshaVerbat/Isha/TCR/filtered_for_tcrdist\"\n",
    "OUTFILE     = \"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/tcrdist_output_freqs.csv\"\n",
    "N_CPUS      = min(multiprocessing.cpu_count(), 8)\n",
    "MAX_FILES   = 292       # for testing; remove or increase as needed\n",
    "\n",
    "######## RESUMING AUTOMATICALLY \n",
    "\n",
    "# Load previously saved results if resuming\n",
    "if os.path.exists(OUTFILE):\n",
    "    saved_df = pd.read_csv(OUTFILE)\n",
    "    processed_ids = set(saved_df['patient_id'])\n",
    "    results_buffer = saved_df.to_dict('records')\n",
    "    print(f\"↪ Resuming from checkpoint — {len(processed_ids)} patients already processed.\")\n",
    "else:\n",
    "    print('No checkpoint found, starting fresh.')\n",
    "    processed_ids = set()\n",
    "    results_buffer = []\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 2. HELPER FUNCTIONS\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "excluded_genes = []\n",
    "\n",
    "def _normalize_gene(s: str) -> str:\n",
    "    \"\"\"Strip prefixes/allele info to IMGT-like gene names, remove leading zeros.\"\"\"\n",
    "    if pd.isna(s):\n",
    "        return np.nan\n",
    "    s = str(s).replace(\"TCR\", \"TR\")\n",
    "    # remove leading zeros in main and subfields\n",
    "    # Remove leading zeros from primary gene number (e.g. TRBV05 → TRBV5)\n",
    "    s = re.sub(\n",
    "        r\"^([A-Z]+)(0+)(\\d+)\",\n",
    "        lambda m: f\"{m.group(1)}{int(m.group(2) + m.group(3))}\",\n",
    "        s,\n",
    "    )   \n",
    "\n",
    "    s = re.sub(\n",
    "        r\"-(0+)(\\d+)\",\n",
    "        lambda m: f\"-{int(m.group(1) + m.group(2))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "\n",
    "    # normalize allele to *01\n",
    "    if \"*\" in s:\n",
    "        s = re.sub(r\"[*!].*$\", \"*01\", s)\n",
    "    else:\n",
    "        s = s + \"*01\"\n",
    "    # validate final form\n",
    "    if not re.match(r\"^TRB[JV]\\d+(-\\d+)?\\*\\d+$\", s):\n",
    "        excluded_genes.append(s)\n",
    "        return np.nan\n",
    "\n",
    "    number = re.search(r'TRB[JV](\\d+)', s).group(1)\n",
    "\n",
    "    # check if gene starts with TRBJ or TRBV\n",
    "    if s.startswith('TRBJ'):\n",
    "        if number in j_no_hyphen_numbers and s not in trb_j_list:\n",
    "           \n",
    "            s = re.sub(r'-\\d+(?=\\*)', '', s)\n",
    "            \n",
    "    elif s.startswith('TRBV'):\n",
    "        if number in v_no_hyphen_numbers and s not in trb_v_list:\n",
    "            \n",
    "            s = re.sub(r'-\\d+(?=\\*)', '', s)\n",
    "            \n",
    "\n",
    "    \n",
    "    \n",
    "    return s\n",
    "\n",
    "def prepare_df(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Bring any clonotype table to tcrdist3-ready shape.\"\"\"\n",
    "    df = df.copy()\n",
    "    if \"d_gene\" in df.columns:\n",
    "        df = df.drop(columns=[\"d_gene\"])\n",
    "    # rename alternate columns\n",
    "    rename_map = {\n",
    "        \"amino_acid\": \"cdr3_b_aa\",\n",
    "        \"cdr3s_aa\":  \"cdr3_b_aa\",\n",
    "        \"v_call\":    \"v_b_gene\",\n",
    "        \"j_call\":    \"j_b_gene\",\n",
    "    }\n",
    "    df = df.rename(columns=rename_map)\n",
    "    # detect frequency\n",
    "    has_freq = \"productive_frequency\" in df.columns\n",
    "    use_cols = [\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"] + ([\"productive_frequency\"] if has_freq else [])\n",
    "    df = df[use_cols].dropna(subset=[\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"])\n",
    "    if not has_freq:\n",
    "        df[\"productive_frequency\"] = 1\n",
    "    # normalize gene names\n",
    "    df[\"v_b_gene\"] = df[\"v_b_gene\"].apply(_normalize_gene)\n",
    "    df[\"j_b_gene\"] = df[\"j_b_gene\"].apply(_normalize_gene)\n",
    "    # drop any failed normalizations\n",
    "    df = df.dropna(subset=[\"v_b_gene\", \"j_b_gene\"]).reset_index(drop=True)\n",
    "    # dedupe & aggregate frequencies\n",
    "    df = df.groupby(\n",
    "        [\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"],\n",
    "        as_index=False\n",
    "    ).agg({\"productive_frequency\": \"sum\"})\n",
    "    return df\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 3. LOAD CMV-ASSOCIATED SEARCH SET ONCE\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "high_df = prepare_df(HIGHCONF)\n",
    "tr_search = TCRrep(\n",
    "    cell_df=high_df,\n",
    "    organism=\"human\",\n",
    "    chains=[\"beta\"],\n",
    "    compute_distances=False\n",
    ")\n",
    "tr_search.cpus = N_CPUS\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 4. PREPARE OUTPUT BUFFER & FILE LIST\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "results_buffer = []\n",
    "outfile = OUTFILE\n",
    "file_names = sorted(\n",
    "    fn for fn in os.listdir(PATIENT_DIR)\n",
    "    if fn.endswith(('.tsv', '.csv')) and fn.startswith('P')\n",
    ")\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 5. MAIN LOOP OVER PATIENT REPERTOIRES\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "for idx, fname in enumerate(file_names, 1):\n",
    "    if idx > MAX_FILES:\n",
    "        print(f\"Reached maximum number of files to process ({MAX_FILES}).\")\n",
    "        break\n",
    "\n",
    "    patient_id = os.path.splitext(fname)[0]\n",
    "    print(f\"Processing ({idx}/{len(file_names)}): {patient_id}\")\n",
    "\n",
    "    if patient_id in processed_ids: \n",
    "        print(f'↪ Already processed {patient_id} – skipping.')\n",
    "        continue\n",
    "\n",
    "    ## 5.1 Load & clean\n",
    "    rep_raw = pd.read_csv(\n",
    "        os.path.join(PATIENT_DIR, fname),\n",
    "        sep=\"\\t\", low_memory=False\n",
    "    )\n",
    "    rep_df = prepare_df(rep_raw)\n",
    "    if rep_df.empty:\n",
    "        print(f\" ! No usable sequences in {patient_id} – skipping.\")\n",
    "        continue\n",
    "\n",
    "    ## 5.2 Build TCRrep for bulk\n",
    "    tr_bulk = TCRrep(\n",
    "        cell_df=rep_df,\n",
    "        organism=\"human\",\n",
    "        chains=[\"beta\"],\n",
    "        compute_distances=False\n",
    "    )\n",
    "    tr_bulk.cpus = N_CPUS\n",
    "\n",
    "    ## 5.3 Compute sparse distances\n",
    "    chunk_size = get_safe_chunk(\n",
    "        tr_search.clone_df.shape[0],\n",
    "        tr_bulk.clone_df.shape[0]\n",
    "    )\n",
    "    tr_search.compute_sparse_rect_distances(\n",
    "        df=tr_search.clone_df,\n",
    "        df2=tr_bulk.clone_df,\n",
    "        radius=RADIUS,\n",
    "        chunk_size=chunk_size,\n",
    "    )\n",
    "\n",
    "    ## 5.4 Build histogram\n",
    "    coo   = tr_search.rw_beta.tocoo()\n",
    "    dists = coo.data.astype(int)\n",
    "    cols  = coo.col\n",
    "    freqs = tr_bulk.clone_df['productive_frequency'].values\n",
    "\n",
    "    # SANITY CHECK:\n",
    "    print(\"=== SANITY CHECK ===\")\n",
    "    print(f\"Total non-zero entries (matches): {coo.nnz}\")\n",
    "    if coo.nnz > 0:\n",
    "        print(f\"  • Distance range: {dists.min()} to {dists.max()}\")\n",
    "        matched_freqs = freqs[cols]\n",
    "        print(f\"  • Sum of matched frequencies: {matched_freqs.sum():.4f}\")\n",
    "        print(f\"  • Example matched freqs (first 10): {matched_freqs[:10]}\")\n",
    "    else:\n",
    "        print(\"  ⚠️ No matches found at any distance!\") \n",
    "\n",
    "    bins  = np.zeros(MAX_DIST + 1, dtype=float)\n",
    "    np.add.at(bins, dists, freqs[cols])\n",
    "\n",
    "    ## 5.5 Buffer the result\n",
    "    row = {\"patient_id\": patient_id}\n",
    "    row.update({f\"dist_{d}\": (bins[d]) for d in range(MAX_DIST + 1)})\n",
    "    results_buffer.append(row)\n",
    "\n",
    "    ## 5.6 Checkpoint every 10 patients\n",
    "    if idx % 10 == 0:\n",
    "        pd.DataFrame(results_buffer).to_csv(outfile, index=False)\n",
    "        print(f\"   ↳ checkpoint saved ({idx} patients) ➜ {outfile}\")\n",
    "\n",
    "    ## 5.7 Clean up sparse matrix\n",
    "    del tr_search.rw_beta\n",
    "\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 6. FINAL WRITE & WRAP UP\n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "pd.DataFrame(results_buffer).to_csv(outfile, index=False)\n",
    "print(f\"✔︎ All patient repertoires processed — final results saved to {outfile}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f1290fe",
   "metadata": {},
   "source": [
    "### Try to write it myself"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a326d88",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 36\u001b[0m\n\u001b[1;32m     32\u001b[0m         s \u001b[38;5;241m=\u001b[39m s \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m*01\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     34\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m s\n\u001b[0;32m---> 36\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprepare_df\u001b[39m(df: pd\u001b[38;5;241m.\u001b[39mDataFrame) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame:\n\u001b[1;32m     37\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Bring an arbitrary clonotype table to tcrdist3‑ready shape.\"\"\"\u001b[39;00m\n\u001b[1;32m     38\u001b[0m     \u001b[38;5;66;03m# Remove d_gene column\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "HLA = 'A02'\n",
    "PATIENT_DIR = f\"/Volumes/IshaVerbat/Isha/TCR/Filtered_HLA-{HLA}\"  # directory with patient files\n",
    "# Prepare df function\n",
    "\n",
    "def _normalize_gene(s: str) -> str:\n",
    "    \"\"\"Strip prefixes / allele info to IMGT-like gene names understood by tcrdist3,\n",
    "    remove leading zeros from gene numbers (e.g. TRBV05 → TRBV5),\n",
    "    and from any numeric subfields (e.g. TRBV5-05 → TRBV5-5).\"\"\"\n",
    "    if pd.isna(s):\n",
    "        return np.nan\n",
    "\n",
    "    # Standardize prefix\n",
    "    s = str(s).replace(\"TCR\", \"TR\")  # e.g. TCRBV05-05 → TRBV05-05\n",
    "\n",
    "    # Remove leading zeros from primary gene number (e.g. TRBV05 → TRBV5)\n",
    "    s = re.sub(\n",
    "        r\"^([A-Z]+)(0+)(\\d+)\",\n",
    "        lambda m: f\"{m.group(1)}{int(m.group(2) + m.group(3))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Remove leading zeros from any subfield after hyphens (e.g. -05 → -5)\n",
    "    s = re.sub(\n",
    "        r\"-(0+)(\\d+)\",\n",
    "        lambda m: f\"-{int(m.group(1) + m.group(2))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Normalize allele information to *01\n",
    "    if \"*\" in s:\n",
    "        s = re.sub(r\"[*!].*$\", \"*01\", s)\n",
    "    else:\n",
    "        s = s + \"*01\"\n",
    "\n",
    "    return s\n",
    "\n",
    "def prepare_df(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Bring an arbitrary clonotype table to tcrdist3‑ready shape.\"\"\"\n",
    "    # Remove d_gene column\n",
    "    if \"d_gene\" in df.columns:\n",
    "        df.drop(columns=[\"d_gene\"], inplace=True)\n",
    "    \n",
    "    rename_map = {\n",
    "        \"amino_acid\": \"cdr3_b_aa\",\n",
    "        \"cdr3s_aa\": \"cdr3_b_aa\",\n",
    "        \"v_call\": \"v_b_gene\",\n",
    "        \"j_call\": \"j_b_gene\",\n",
    "    }\n",
    "\n",
    "    df = df.rename(columns=rename_map)\n",
    "    need = [\"cdr3_b_aa\", \"v_b_gene\", \"j_b_gene\"]\n",
    "    df = df[need].copy()\n",
    "    df.dropna(inplace=True)\n",
    "\n",
    "    df[\"v_b_gene\"] = df[\"v_b_gene\"].apply(_normalize_gene)\n",
    "    df[\"j_b_gene\"] = df[\"j_b_gene\"].apply(_normalize_gene)\n",
    "  \n",
    "\n",
    "    df = df.drop_duplicates().reset_index(drop=True)\n",
    "    return df\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "rep_raw = pd.read_csv(os.path.join(PATIENT_DIR, fname), sep=\"\\t\", low_memory=False)\n",
    "df_bulk  = prepare_df(rep_raw)\n",
    "\n",
    "tr_bulk = TCRrep(cell_df = df_bulk,\n",
    "                 organism = \"human\",\n",
    "                 chains = [\"beta\"],\n",
    "                 #db_file = db,\n",
    "                 compute_distances = False)\n",
    "\n",
    "t = tr_bulk.show_incomplete()\n",
    "\n",
    "df_search = tr_bulk.clone_df.copy()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "0531a53e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ishaharris/Projects/TCR/TCR-Isha/data/Repertoires\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99280e6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "9fdbf451",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def _normalize_gene(s: str) -> str:\n",
    "    \"\"\"Strip prefixes / allele info to IMGT-like gene names understood by tcrdist3,\n",
    "    remove leading zeros from gene numbers (e.g. TRBV05 → TRBV5),\n",
    "    and from any numeric subfields (e.g. TRBV5-05 → TRBV5-5).\"\"\"\n",
    "    if pd.isna(s):\n",
    "        return np.nan\n",
    "\n",
    "    # Standardize prefix\n",
    "    s = str(s).replace(\"TCR\", \"TR\")  # e.g. TCRBV05-05 → TRBV05-05\n",
    "\n",
    "    # Remove leading zeros from primary gene number (e.g. TRBV05 → TRBV5)\n",
    "    s = re.sub(\n",
    "        r\"^([A-Z]+)(0+)(\\d+)\",\n",
    "        lambda m: f\"{m.group(1)}{int(m.group(2) + m.group(3))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Remove leading zeros from any subfield after hyphens (e.g. -05 → -5)\n",
    "    s = re.sub(\n",
    "        r\"-(0+)(\\d+)\",\n",
    "        lambda m: f\"-{int(m.group(1) + m.group(2))}\",\n",
    "        s,\n",
    "    )\n",
    "\n",
    "    # Normalize allele information to *01\n",
    "    if \"*\" in s:\n",
    "        s = re.sub(r\"[*!].*$\", \"*01\", s)\n",
    "    else:\n",
    "        s = s + \"*01\"\n",
    "\n",
    "    return s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "8483d5df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TRBV5-5*01'"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_normalize_gene('TCRBV05-05')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "46a33237",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3_b_aa</th>\n",
       "      <th>v_b_gene</th>\n",
       "      <th>j_b_gene</th>\n",
       "      <th>cdr1_b_aa</th>\n",
       "      <th>cdr2_b_aa</th>\n",
       "      <th>pmhc_b_aa</th>\n",
       "      <th>count</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cdr3_b_aa, v_b_gene, j_b_gene, cdr1_b_aa, cdr2_b_aa, pmhc_b_aa, count, clone_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_search"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9114e8",
   "metadata": {},
   "source": [
    "### Updating v gene info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "54158124",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['taxonId', 'speciesNames', 'genes', 'meta', 'sequenceFragments'], dtype='object')\n",
      "(23, 5)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   taxonId           speciesNames  \\\n",
      "0     7955            [zebrafish]   \n",
      "1     8022        [rainbow_trout]   \n",
      "2     8030               [salmon]   \n",
      "3     9258             [platypus]   \n",
      "4     9541  [macaca fascicularis]   \n",
      "\n",
      "                                               genes  \\\n",
      "0  [{'baseSequence': 'file://zebrafish_V_IGH.fast...   \n",
      "1  [{'baseSequence': 'file://rainbow_trout_V_IGH....   \n",
      "2  [{'baseSequence': 'file://salmon_V_IGH.fasta#I...   \n",
      "3  [{'baseSequence': 'file://platypus_V_IGH.fasta...   \n",
      "4  [{'baseSequence': 'file://macacaFascicularis_V...   \n",
      "\n",
      "                                                meta  \\\n",
      "0  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
      "1  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
      "2  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
      "3  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
      "4  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
      "\n",
      "                                   sequenceFragments  \n",
      "0  [{'uri': 'file://zebrafish_C_IGI.fasta#IGIC1*0...  \n",
      "1  [{'uri': 'file://rainbow_trout_C_IGH.fasta#IGH...  \n",
      "2  [{'uri': 'file://salmon_C_IGH.fasta#IGHD*01', ...  \n",
      "3  [{'uri': 'file://platypus_C_IGH.fasta#IGHA1*01...  \n",
      "4  [{'uri': 'file://macacaFascicularis_C_IGH.fast...  \n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_json(\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/imgt.202312-3.sv8.json\")\n",
    "# df\n",
    "\n",
    "# import json\n",
    "# import pandas as pd\n",
    "\n",
    "# Load the JSON data\n",
    "with open (\"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/imgt.202312-3.sv8.json\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "# Convert to a DataFrame\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "# Confirm structure\n",
    "print(df.columns)\n",
    "print(df.shape)\n",
    "print(df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "b79bdd85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Explode the genes column\n",
    "df_expanded = df.explode('genes').reset_index(drop=True)\n",
    "\n",
    "# Normalize the dictionary in each gene row\n",
    "genes_flat = pd.json_normalize(df_expanded['genes'])\n",
    "\n",
    "# Merge flattened gene info back into the main df\n",
    "result = pd.concat([df_expanded.drop(columns=['genes']), genes_flat], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "84971516",
   "metadata": {},
   "outputs": [],
   "source": [
    "aliases = ['homosapiens', 'hs', 'hsa', 'human']\n",
    "\n",
    "human_genes = result[result['speciesNames'].apply(\n",
    "    lambda names: any(alias.lower() in [n.lower() for n in names] for alias in aliases)\n",
    ")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "502c8e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_genes\n",
    "\n",
    "#save human_genes to a csv\n",
    "human_genes.to_csv('/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/human_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "ed5c96ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save human_genes to a pickle\n",
    "human_genes.to_pickle('/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/human_genes.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "b43fd788",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read human_genes pickle\n",
    "human_genes = pd.read_pickle('/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/human_genes.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "03640b36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxonId</th>\n",
       "      <th>speciesNames</th>\n",
       "      <th>meta</th>\n",
       "      <th>sequenceFragments</th>\n",
       "      <th>baseSequence</th>\n",
       "      <th>name</th>\n",
       "      <th>geneType</th>\n",
       "      <th>isFunctional</th>\n",
       "      <th>chains</th>\n",
       "      <th>type</th>\n",
       "      <th>...</th>\n",
       "      <th>anchorPoints.CDR3Begin</th>\n",
       "      <th>anchorPoints.VEnd</th>\n",
       "      <th>anchorPoints.DBegin</th>\n",
       "      <th>anchorPoints.DEnd</th>\n",
       "      <th>anchorPoints.JBegin</th>\n",
       "      <th>anchorPoints.FR4Begin</th>\n",
       "      <th>anchorPoints.FR4End</th>\n",
       "      <th>anchorPoints.CBegin</th>\n",
       "      <th>anchorPoints.CExon1End</th>\n",
       "      <th>anchorPoints.CEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2552</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_V_IGH.fasta#IGHV1-2*01</td>\n",
       "      <td>IGHV1-2*01</td>\n",
       "      <td>V</td>\n",
       "      <td>True</td>\n",
       "      <td>[IGH]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>285.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2553</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_V_IGH.fasta#IGHV1-2*02</td>\n",
       "      <td>IGHV1-2*02</td>\n",
       "      <td>V</td>\n",
       "      <td>True</td>\n",
       "      <td>[IGH]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>285.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2554</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_V_IGH.fasta#IGHV1-2*04</td>\n",
       "      <td>IGHV1-2*04</td>\n",
       "      <td>V</td>\n",
       "      <td>True</td>\n",
       "      <td>[IGH]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>285.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2555</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_V_IGH.fasta#IGHV1-2*05</td>\n",
       "      <td>IGHV1-2*05</td>\n",
       "      <td>V</td>\n",
       "      <td>True</td>\n",
       "      <td>[IGH]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>285.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2556</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_V_IGH.fasta#IGHV1-2*06</td>\n",
       "      <td>IGHV1-2*06</td>\n",
       "      <td>V</td>\n",
       "      <td>True</td>\n",
       "      <td>[IGH]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>285.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3827</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_C_TRG.fasta#TRGC2*04</td>\n",
       "      <td>TRGC2*04</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRG]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3828</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_C_TRG.fasta#TRGC2*05</td>\n",
       "      <td>TRGC2*05</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRG]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3829</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_C_TRG.fasta#TRGC2*06</td>\n",
       "      <td>TRGC2*06</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRG]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3830</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_C_TRG.fasta#TRGC2*07</td>\n",
       "      <td>TRGC2*07</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRG]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3831</th>\n",
       "      <td>9606</td>\n",
       "      <td>[homosapiens, hs, hsa, human]</td>\n",
       "      <td>{'comments': ['Imported from: /Volumes/Data/Pr...</td>\n",
       "      <td>[{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...</td>\n",
       "      <td>file://human_C_TRG.fasta#TRGC2*08</td>\n",
       "      <td>TRGC2*08</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRG]</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1280 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxonId                   speciesNames  \\\n",
       "2552     9606  [homosapiens, hs, hsa, human]   \n",
       "2553     9606  [homosapiens, hs, hsa, human]   \n",
       "2554     9606  [homosapiens, hs, hsa, human]   \n",
       "2555     9606  [homosapiens, hs, hsa, human]   \n",
       "2556     9606  [homosapiens, hs, hsa, human]   \n",
       "...       ...                            ...   \n",
       "3827     9606  [homosapiens, hs, hsa, human]   \n",
       "3828     9606  [homosapiens, hs, hsa, human]   \n",
       "3829     9606  [homosapiens, hs, hsa, human]   \n",
       "3830     9606  [homosapiens, hs, hsa, human]   \n",
       "3831     9606  [homosapiens, hs, hsa, human]   \n",
       "\n",
       "                                                   meta  \\\n",
       "2552  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "2553  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "2554  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "2555  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "2556  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "...                                                 ...   \n",
       "3827  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "3828  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "3829  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "3830  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "3831  {'comments': ['Imported from: /Volumes/Data/Pr...   \n",
       "\n",
       "                                      sequenceFragments  \\\n",
       "2552  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "2553  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "2554  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "2555  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "2556  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "...                                                 ...   \n",
       "3827  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "3828  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "3829  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "3830  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "3831  [{'uri': 'file://human_C_IGH.fasta#IGHA1*01', ...   \n",
       "\n",
       "                             baseSequence        name geneType  isFunctional  \\\n",
       "2552  file://human_V_IGH.fasta#IGHV1-2*01  IGHV1-2*01        V          True   \n",
       "2553  file://human_V_IGH.fasta#IGHV1-2*02  IGHV1-2*02        V          True   \n",
       "2554  file://human_V_IGH.fasta#IGHV1-2*04  IGHV1-2*04        V          True   \n",
       "2555  file://human_V_IGH.fasta#IGHV1-2*05  IGHV1-2*05        V          True   \n",
       "2556  file://human_V_IGH.fasta#IGHV1-2*06  IGHV1-2*06        V          True   \n",
       "...                                   ...         ...      ...           ...   \n",
       "3827    file://human_C_TRG.fasta#TRGC2*04    TRGC2*04        C          True   \n",
       "3828    file://human_C_TRG.fasta#TRGC2*05    TRGC2*05        C          True   \n",
       "3829    file://human_C_TRG.fasta#TRGC2*06    TRGC2*06        C          True   \n",
       "3830    file://human_C_TRG.fasta#TRGC2*07    TRGC2*07        C          True   \n",
       "3831    file://human_C_TRG.fasta#TRGC2*08    TRGC2*08        C          True   \n",
       "\n",
       "     chains     type  ... anchorPoints.CDR3Begin  anchorPoints.VEnd  \\\n",
       "2552  [IGH]  primary  ...                  285.0              296.0   \n",
       "2553  [IGH]  primary  ...                  285.0              296.0   \n",
       "2554  [IGH]  primary  ...                  285.0              296.0   \n",
       "2555  [IGH]  primary  ...                  285.0              296.0   \n",
       "2556  [IGH]  primary  ...                  285.0              296.0   \n",
       "...     ...      ...  ...                    ...                ...   \n",
       "3827  [TRG]  primary  ...                    NaN                NaN   \n",
       "3828  [TRG]  primary  ...                    NaN                NaN   \n",
       "3829  [TRG]  primary  ...                    NaN                NaN   \n",
       "3830  [TRG]  primary  ...                    NaN                NaN   \n",
       "3831  [TRG]  primary  ...                    NaN                NaN   \n",
       "\n",
       "      anchorPoints.DBegin  anchorPoints.DEnd  anchorPoints.JBegin  \\\n",
       "2552                  NaN                NaN                  NaN   \n",
       "2553                  NaN                NaN                  NaN   \n",
       "2554                  NaN                NaN                  NaN   \n",
       "2555                  NaN                NaN                  NaN   \n",
       "2556                  NaN                NaN                  NaN   \n",
       "...                   ...                ...                  ...   \n",
       "3827                  NaN                NaN                  NaN   \n",
       "3828                  NaN                NaN                  NaN   \n",
       "3829                  NaN                NaN                  NaN   \n",
       "3830                  NaN                NaN                  NaN   \n",
       "3831                  NaN                NaN                  NaN   \n",
       "\n",
       "      anchorPoints.FR4Begin  anchorPoints.FR4End  anchorPoints.CBegin  \\\n",
       "2552                    NaN                  NaN                  NaN   \n",
       "2553                    NaN                  NaN                  NaN   \n",
       "2554                    NaN                  NaN                  NaN   \n",
       "2555                    NaN                  NaN                  NaN   \n",
       "2556                    NaN                  NaN                  NaN   \n",
       "...                     ...                  ...                  ...   \n",
       "3827                    NaN                  NaN                  0.0   \n",
       "3828                    NaN                  NaN                  0.0   \n",
       "3829                    NaN                  NaN                  0.0   \n",
       "3830                    NaN                  NaN                  0.0   \n",
       "3831                    NaN                  NaN                  0.0   \n",
       "\n",
       "      anchorPoints.CExon1End  anchorPoints.CEnd  \n",
       "2552                     NaN                NaN  \n",
       "2553                     NaN                NaN  \n",
       "2554                     NaN                NaN  \n",
       "2555                     NaN                NaN  \n",
       "2556                     NaN                NaN  \n",
       "...                      ...                ...  \n",
       "3827                   330.0              330.0  \n",
       "3828                   330.0              330.0  \n",
       "3829                   330.0              330.0  \n",
       "3830                   330.0              330.0  \n",
       "3831                   330.0              330.0  \n",
       "\n",
       "[1280 rows x 26 columns]"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9860649b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "# 0. IMPORT PICKLES \n",
    "# ──────────────────────────────────────────────────────────────────────────────\n",
    "\n",
    "PICKLE_DIR = \"/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/pickles/\"  # ← adjust to your actual directory\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/trb_v_no_hyphen_list.pkl\", \"rb\") as f:\n",
    "    trb_v_no_hyphen_list = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/trb_v_hyphen_list.pkl\", \"rb\") as f:\n",
    "    trb_v_hyphen_list = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/j_no_hyphen_numbers.pkl\", \"rb\") as f:\n",
    "    j_no_hyphen_numbers = pickle.load(f)\n",
    "\n",
    "with open(f\"{PICKLE_DIR}/j_hyphen_numbers.pkl\", \"rb\") as f:\n",
    "    j_hyphen_numbers = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3caca253",
   "metadata": {},
   "source": [
    "### Analyse data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "994d49c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the df\n",
    "tcrdist_freqs = pd.read_csv('/Users/ishaharris/Projects/TCR/TCR-Isha/data/tcrdist/tcrdist_output_freqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "94c84336",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>dist_0</th>\n",
       "      <th>dist_1</th>\n",
       "      <th>dist_2</th>\n",
       "      <th>dist_3</th>\n",
       "      <th>dist_4</th>\n",
       "      <th>dist_5</th>\n",
       "      <th>dist_6</th>\n",
       "      <th>dist_7</th>\n",
       "      <th>dist_8</th>\n",
       "      <th>...</th>\n",
       "      <th>dist_42</th>\n",
       "      <th>dist_43</th>\n",
       "      <th>dist_44</th>\n",
       "      <th>dist_45</th>\n",
       "      <th>dist_46</th>\n",
       "      <th>dist_47</th>\n",
       "      <th>dist_48</th>\n",
       "      <th>dist_49</th>\n",
       "      <th>dist_50</th>\n",
       "      <th>cmv_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006941</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.042541</td>\n",
       "      <td>0.001035</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.056697</td>\n",
       "      <td>0.006870</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>Positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P00004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011304</td>\n",
       "      <td>0.001454</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.027303</td>\n",
       "      <td>0.001179</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.077460</td>\n",
       "      <td>0.004915</td>\n",
       "      <td>0.003292</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P00006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.229641e-05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009139</td>\n",
       "      <td>0.001829</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.021875</td>\n",
       "      <td>0.001605</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.071586</td>\n",
       "      <td>0.005188</td>\n",
       "      <td>0.001073</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P00007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009870</td>\n",
       "      <td>0.000868</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.016478</td>\n",
       "      <td>0.001278</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.127134</td>\n",
       "      <td>0.005021</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>Positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012171</td>\n",
       "      <td>0.001581</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.024982</td>\n",
       "      <td>0.001344</td>\n",
       "      <td>0.000234</td>\n",
       "      <td>0.079403</td>\n",
       "      <td>0.005069</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>P00208</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010777</td>\n",
       "      <td>0.001513</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.001770</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.088760</td>\n",
       "      <td>0.004344</td>\n",
       "      <td>0.000995</td>\n",
       "      <td>Positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>P00209</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011378</td>\n",
       "      <td>0.001408</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.032534</td>\n",
       "      <td>0.001889</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.092358</td>\n",
       "      <td>0.004639</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>P00211</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011271</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.029476</td>\n",
       "      <td>0.001934</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>0.084483</td>\n",
       "      <td>0.004826</td>\n",
       "      <td>0.001069</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>P00212</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.967240e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010182</td>\n",
       "      <td>0.001408</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.020836</td>\n",
       "      <td>0.001069</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>0.064332</td>\n",
       "      <td>0.004129</td>\n",
       "      <td>0.000481</td>\n",
       "      <td>Positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>P00214</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010250</td>\n",
       "      <td>0.001263</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.023482</td>\n",
       "      <td>0.001288</td>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.076963</td>\n",
       "      <td>0.004139</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient_id  dist_0  dist_1  dist_2    dist_3        dist_4  dist_5  \\\n",
       "0      P00003     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "1      P00004     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "2      P00006     0.0     0.0     0.0  0.000000  1.229641e-05     0.0   \n",
       "3      P00007     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "4      P00008     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "..        ...     ...     ...     ...       ...           ...     ...   \n",
       "94     P00208     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "95     P00209     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "96     P00211     0.0     0.0     0.0  0.000000  0.000000e+00     0.0   \n",
       "97     P00212     0.0     0.0     0.0  0.000000  6.967240e-07     0.0   \n",
       "98     P00214     0.0     0.0     0.0  0.000005  0.000000e+00     0.0   \n",
       "\n",
       "      dist_6  dist_7  dist_8  ...   dist_42   dist_43   dist_44   dist_45  \\\n",
       "0   0.000000     0.0     0.0  ...  0.006941  0.000984  0.000048  0.042541   \n",
       "1   0.000000     0.0     0.0  ...  0.011304  0.001454  0.000071  0.027303   \n",
       "2   0.000000     0.0     0.0  ...  0.009139  0.001829  0.000076  0.021875   \n",
       "3   0.000004     0.0     0.0  ...  0.009870  0.000868  0.000041  0.016478   \n",
       "4   0.000000     0.0     0.0  ...  0.012171  0.001581  0.000099  0.024982   \n",
       "..       ...     ...     ...  ...       ...       ...       ...       ...   \n",
       "94  0.000000     0.0     0.0  ...  0.010777  0.001513  0.000138  0.028569   \n",
       "95  0.000011     0.0     0.0  ...  0.011378  0.001408  0.000056  0.032534   \n",
       "96  0.000000     0.0     0.0  ...  0.011271  0.001406  0.000126  0.029476   \n",
       "97  0.000004     0.0     0.0  ...  0.010182  0.001408  0.000057  0.020836   \n",
       "98  0.000010     0.0     0.0  ...  0.010250  0.001263  0.000067  0.023482   \n",
       "\n",
       "     dist_46   dist_47   dist_48   dist_49   dist_50  cmv_status  \n",
       "0   0.001035  0.000277  0.056697  0.006870  0.000530    Positive  \n",
       "1   0.001179  0.000300  0.077460  0.004915  0.003292    Negative  \n",
       "2   0.001605  0.000308  0.071586  0.005188  0.001073    Negative  \n",
       "3   0.001278  0.000313  0.127134  0.005021  0.000318    Positive  \n",
       "4   0.001344  0.000234  0.079403  0.005069  0.000505    Negative  \n",
       "..       ...       ...       ...       ...       ...         ...  \n",
       "94  0.001770  0.000244  0.088760  0.004344  0.000995    Positive  \n",
       "95  0.001889  0.000284  0.092358  0.004639  0.000656    Negative  \n",
       "96  0.001934  0.000381  0.084483  0.004826  0.001069    Negative  \n",
       "97  0.001069  0.000320  0.064332  0.004129  0.000481    Positive  \n",
       "98  0.001288  0.000304  0.076963  0.004139  0.000656    Negative  \n",
       "\n",
       "[99 rows x 53 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add CMV annotation\n",
    "## Add CMV status\n",
    "\n",
    "tcrdist_freqs['patient_id'] = tcrdist_freqs['patient_id'].str.split('.').str[0]\n",
    "pos_names_stripped = [name.split('.')[0] for name in pos_names]\n",
    "neg_names_stripped = [name.split('.')[0] for name in neg_names]\n",
    "\n",
    "\n",
    "tcrdist_freqs['cmv_status'] = np.where(\n",
    "    tcrdist_freqs['patient_id'].isin(pos_names_stripped),\n",
    "    'Positive',\n",
    "    np.where(tcrdist_freqs['patient_id'].isin(neg_names_stripped), 'Negative', 'unknown')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3d758c8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cmv_status\n",
      "Negative    56\n",
      "Positive    43\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# count number of pos and neg cmv\n",
    "cmv_counts = tcrdist_freqs['cmv_status'].value_counts()\n",
    "print(cmv_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "97f1157c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cmv_status</th>\n",
       "      <th>dist_0</th>\n",
       "      <th>dist_1</th>\n",
       "      <th>dist_2</th>\n",
       "      <th>dist_3</th>\n",
       "      <th>dist_4</th>\n",
       "      <th>dist_5</th>\n",
       "      <th>dist_6</th>\n",
       "      <th>dist_7</th>\n",
       "      <th>dist_8</th>\n",
       "      <th>...</th>\n",
       "      <th>dist_41</th>\n",
       "      <th>dist_42</th>\n",
       "      <th>dist_43</th>\n",
       "      <th>dist_44</th>\n",
       "      <th>dist_45</th>\n",
       "      <th>dist_46</th>\n",
       "      <th>dist_47</th>\n",
       "      <th>dist_48</th>\n",
       "      <th>dist_49</th>\n",
       "      <th>dist_50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Negative</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>6.780825e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.011256</td>\n",
       "      <td>0.001559</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.025250</td>\n",
       "      <td>0.001581</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>0.083733</td>\n",
       "      <td>0.004938</td>\n",
       "      <td>0.000872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Positive</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.023595e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.012143</td>\n",
       "      <td>0.002018</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.028754</td>\n",
       "      <td>0.002730</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>0.082903</td>\n",
       "      <td>0.004895</td>\n",
       "      <td>0.000624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  cmv_status  dist_0  dist_1  dist_2    dist_3    dist_4  dist_5    dist_6  \\\n",
       "0   Negative     0.0     0.0     0.0  0.000002  0.000004     0.0  0.000008   \n",
       "1   Positive     0.0     0.0     0.0  0.000039  0.000004     0.0  0.000006   \n",
       "\n",
       "         dist_7  dist_8  ...   dist_41   dist_42   dist_43   dist_44  \\\n",
       "0  6.780825e-08     0.0  ...  0.000037  0.011256  0.001559  0.000102   \n",
       "1  5.023595e-07     0.0  ...  0.000032  0.012143  0.002018  0.000087   \n",
       "\n",
       "    dist_45   dist_46   dist_47   dist_48   dist_49   dist_50  \n",
       "0  0.025250  0.001581  0.000376  0.083733  0.004938  0.000872  \n",
       "1  0.028754  0.002730  0.000386  0.082903  0.004895  0.000624  \n",
       "\n",
       "[2 rows x 52 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list of distance columns\n",
    "distance_cols = [f'dist_{i}' for i in range(51)]\n",
    "\n",
    "# group by cmv_status and take mean of each distance column\n",
    "avg_distances = (\n",
    "    tcrdist_freqs\n",
    "    .groupby('cmv_status')[distance_cols]\n",
    "    .mean()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "avg_distances\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9b4a7c07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  cmv_status  distance  mean_freq\n",
      "0   Negative         0        0.0\n",
      "1   Positive         0        0.0\n",
      "2   Negative         1        0.0\n",
      "3   Positive         1        0.0\n",
      "4   Negative         2        0.0\n"
     ]
    }
   ],
   "source": [
    "avg_long = avg_distances.melt(id_vars='cmv_status', \n",
    "                              value_vars=distance_cols, \n",
    "                              var_name='distance', \n",
    "                              value_name='mean_freq')\n",
    "\n",
    "# optionally convert 'dist_0' → 0\n",
    "avg_long['distance'] = avg_long['distance'].str.replace('dist_', '').astype(int)\n",
    "\n",
    "print(avg_long.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2f0b4af4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnYAAAHWCAYAAAD6oMSKAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQABAABJREFUeJzsnXecFPX9/5+zvV9v1KMjgqIIdoSoEKxRiQ0VA5pESexR/JpoNBrLT41JRJIYARMbsZcgiokKsVEERVFA4GjX6/Y+vz9mdu+Wa7t7u9xxfp6Pxz1gZz8zn8/O7O689l0lWZZlBAKBQCAQCASHPJreXoBAIBAIBAKBIDMIYScQCAQCgUDQTxDCTiAQCAQCgaCfIISdQCAQCAQCQT9BCDuBQCAQCASCfoIQdgKBQCAQCAT9BCHsBAKBQCAQCPoJQtgJBAKBQCAQ9BOEsBMIBAKBQCDoJwhhJ+hTSJKU1N8HH3wAQE1NDQsXLmTChAnYbDZMJhOjRo3i+uuvZ/v27fHj/va3v03YX6/XM2TIEK6++mqqq6uTWtuVV17Z6XreeuutbJyO7y3Tpk1j/PjxWZ+nvLw8fg01Gg05OTkcdthhXHHFFbz77rsd7iNJEr/97W9TmmfFihUp7yNQKC8v58orr0xqrNPp5L777uOYY47B4XBgNBopLy9n3rx5fP755/Fxy5Yta/dd0hZZlhk5ciSSJDFt2jQAXn/9dSRJ4i9/+Uun869atQpJknj00Ue7XOdnn33Geeedx5AhQzAajZSUlHD88cdz8803J4x74oknWLZsWVKvvTN+//vf89prr/XoGIJDC11vL0AgaMsnn3yS8Ph3v/sd77//Pv/9738Tto8bN461a9dy1llnIcsyv/jFLzj++OMxGAxs3bqVZ555hilTptDU1JSw38qVK8nJycHtdvPuu+/yyCOP8PHHH7Np0yb0en236zObze3WAjB27Ng0Xq2gL3DiiSfy8MMPA+B2u9m6dSsvvPACM2fO5IILLuD5559PeG988sknDBo0KKU5VqxYwaJFi4S4yyI7duxgxowZ1NbW8vOf/5y7774bm81GRUUF//rXv5g0aRLNzc3k5OTE97Hb7Tz11FNx8Rbjww8/ZMeOHdjt9vi2M888k9LSUpYsWcLPf/7zDtewdOlS9Ho9l19+eafr/Pe//80555zDtGnTeOihhygrK6Oqqor169fzwgsv8Mgjj8THPvHEExQWFiYtbDvi97//PbNnz+ZHP/pR2scQHGLIAkEfZu7cubLVam23vaWlRS4tLZUHDx4s7927t8N9X3zxxfj/77rrLhmQ6+rqEsb85Cc/kQH5v//9b9pr6QqPx5PSeIHCKaecIh9++OFZn2fo0KHymWee2eFzsffMrbfe2uN5FixYIIuv2/QYOnSoPHfu3C7HhMNhecKECbLD4ZA3b97c4ZgVK1bEP49Lly6VAfmqq66SzWaz3NLSkjD2sssuk48//nj58MMPl0855ZT49ltvvVUGOpyjqalJNplM8gUXXNDlWqdOnSqPGDFCDoVC7Z6LRCIJjw+cPx2sVmu350/QvxCuWMEhyZNPPkl1dTUPPfRQp9aT2bNnd3ucY445BlBcuj0l5u79/PPPmT17Nnl5eYwYMQJQXDtPPPEEEydOxGw2k5eXx+zZs9m5c2fCMWRZ5qGHHmLo0KGYTCaOPvpo3n77baZNm5ZgVYi5kioqKhL2/+CDDzp0L7333nuceuqpOBwOLBYLJ554Iv/5z386XP/XX3/NJZdcQk5ODiUlJcybN4+WlpaEsdFolD//+c/x15Obm8txxx3HG2+8AcD8+fPJz8/H6/W2O08/+MEPOPzww5M6p2vWrOG4447DbDYzcOBAfvOb3xCJROLnatSoUcycObPdfm63m5ycHBYsWJDUPB3x29/+lsMPP5zHH38cv98f336gK9br9XLLLbcwbNgwTCYT+fn5HHPMMTz//POA4sJftGhRfN/YX+zaLVq0iKlTp1JcXIzVamXChAk89NBDhEKhhPXE3NPr1q3j5JNPxmKxMHz4cB544AGi0WjC2ObmZm6++WaGDx+O0WikuLiYM844g2+//TY+JhgMcu+99zJ27FiMRiNFRUX85Cc/oa6urttzs379ei6++GLKy8sxm82Ul5dzySWXsHv37oRxsffp+++/zzXXXENhYSEFBQWcf/75VFZWJowNhULceuutlJaWYrFYOOmkk1i7dm23awF47bXX2Lx5M7fffnunLvxZs2ZhsVgStl1yySUA8WsF0NLSwssvv8y8efPaHWP+/PmAYpk7kOeffx6/39/hfm1paGigsLAQna69w0yjab0ll5eX8/XXX/Phhx/G3zPl5eUA+P1+br75ZiZOnEhOTg75+fkcf/zxvP766wnHkyQJj8fD008/HT9G7Hsk9nk/kI6+W/773/8ybdo0CgoKMJvNDBkyhAsuuKDDz7eg9xHCTnBI8u6776LVajn77LN7dJxdu3YBMHr06KT3CYfDCX8xoRHj/PPPZ+TIkbz44ovxeJyf/exn3HDDDZx22mm89tprPPHEE3z99deccMIJCaLy7rvv5rbbbuP000/ntdde45prruHqq69m69atab/GZ555hhkzZuBwOHj66af517/+RX5+PjNnzmwn7gAuuOACRo8ezcsvv8zChQt57rnnuPHGGxPGXHnllVx//fVMnjyZ5cuX88ILL3DOOefEbwbXX389TU1NPPfccwn7bdmyhffffz8pwVVdXc3FF1/MnDlzeP3115k9ezb33nsv119/PaDctH75y1+yatWqhHhKgH/84x84nc4eCTuAs88+G6/Xy/r16zsdc9NNN7F48WKuu+46Vq5cyT//+U9+/OMf09DQAMBvfvOb+I+MTz75JP5XVlYGKC7ESy+9lH/+85+89dZbzJ8/n//3//4fP/vZzzo8J3PmzOGyyy7jjTfeYNasWdx+++0888wz8TEul4uTTjqJv/71r/zkJz/hzTff5C9/+QujR4+mqqoKUIT5ueeeywMPPMCll17Kv//9bx544AFWrVrFtGnT8Pl8XZ6XiooKxowZw2OPPcY777zDgw8+SFVVFZMnT6a+vr7d+Kuuugq9Xs9zzz3HQw89xAcffMBll12WMObqq6/m4Ycf5oorruD111/nggsu4Pzzz28XTtERsXjIVN2NDoeD2bNns2TJkvi2559/Ho1Gw0UXXdRu/OjRoznppJN45pln2gnvpUuXMnDgwA5/aLTl+OOP57PPPuO6667js88+a3ecGK+++irDhw/nqKOOir9nXn31VQACgQCNjY3ccsstvPbaazz//POcdNJJnH/++fzjH/+IH+OTTz7BbDZzxhlnxI/xxBNPJH1+QLnWZ555JgaDgSVLlrBy5UoeeOABrFYrwWAwpWMJDhK9bDEUCLqkM/fn2LFj5dLS0qSPE3OrVVdXy6FQSG5qapL/9a9/yVarVb7kkkuSXgvQ7u/EE09MmOPOO+9M2O+TTz6RAfmRRx5J2L53717ZbDbHXX0xV855552XMO6jjz6SgQSXTMyVtGvXroSx77//vgzI77//vizLiis4Pz9fPvvssxPGRSIR+cgjj5SnTJnS7hw99NBDCWOvvfZa2WQyydFoVJZlWV69erUMyHfccUeX5+uUU06RJ06cmLDtmmuukR0Oh+xyubrdF5Bff/31hO1XX321rNFo5N27d8uyLMtOp1O22+3y9ddfnzBu3Lhx8vTp07ucQ5a7dsXKsiwvXrxYBuTly5fHtwHyXXfdFX88fvx4+Uc/+lGX8yTrio1EInIoFJL/8Y9/yFqtVm5sbIw/Fzsnn332WcI+48aNk2fOnBl/fM8998iAvGrVqk7nef7552VAfvnllxO2r1u3TgbkJ554otu1tiUcDstut1u2Wq3yH//4x/j22Pv02muvTRj/0EMPyYBcVVUly7Isf/PNNzIg33jjjQnjnn32WRno1pX4wx/+UAZkv9+f1Hpj61q3bl38M/PVV1/JsizLkydPlq+88kpZljt2hcb2feWVV+Lbvvrqq6Q+E7Isy/X19fJJJ50U//7Q6/XyCSecIN9///3tPhfJumLD4bAcCoXk+fPny0cddVTCc525YmOf9wM58LvlpZdekgF506ZN3a5D0DcQFjvB94rS0lL0ej15eXlceOGFTJo0iaeffjrp/c1mM+vWrUv4e+qppxLGXHDBBQmP33rrLSRJ4rLLLkuw9JWWlnLkkUfG3aaffPIJfr+fOXPmJOx/wgknMHTo0LRe78cff0xjYyNz585NmDsajfLDH/6QdevW4fF4EvY555xzEh4fccQR+P1+amtrAXj77bcBurWGXX/99WzatImPPvoIUDIW//nPfzJ37lxsNlu3a7fb7e3WcumllxKNRlm9enV8zE9+8hOWLVsWfx3//e9/2bJlC7/4xS+6naM7ZFnudsyUKVN4++23WbhwIR988EG31q4D2bhxI+eccw4FBQVotVr0ej1XXHEFkUiEbdu2JYwtLS1lypQpCduOOOKIBBfo22+/zejRoznttNM6nfOtt94iNzeXs88+O+F9MXHiREpLSzvMFG2L2+3mtttuY+TIkeh0OnQ6HTabDY/HwzfffNNufEfvKSC+7vfffx+g3Xv/wgsv7NBlmUlOOeUURowYwZIlS9i8eTPr1q3r0p164YUXYrfbE6x8S5YsQZIkfvKTn3Q7X0FBAWvWrGHdunU88MADnHvuuWzbto3bb7+dCRMmdGjx7IgXX3yRE088EZvNhk6nQ6/X89RTT3V4/nvCxIkTMRgM/PSnP+Xpp59uFz4i6HsIYSc4JBkyZAh1dXXtREl3vPfee6xbt4533nmHCy64gNWrV/PLX/4y6f01Gg3HHHNMwt+YMWMSxsRcbDFqamqQZZmSkhL0en3C36effhr/Io+57kpLS9vN29G2ZIi5eWfPnt1u7gcffBBZlmlsbEzYp6CgIOGx0WgEiAuWuro6tFptt2s699xzKS8vj8eXxcRXsu7RkpKSdttic8bOFcAvf/lLXC4Xzz77LACPP/44gwYN4txzz01qnq6ICY8BAwZ0OuZPf/oTt912G6+99hrTp08nPz+fH/3oR+3cwx2xZ88eTj75ZPbv388f//jH+A0/ds4OFIkHXhtQrk/bcXV1dd1m7dbU1NDc3IzBYGj3vqiuru5WXFx66aU8/vjjXHXVVbzzzjusXbuWdevWUVRU1KGw7e491dl7X6fTdfiaD2TIkCFAa2hFKsQE2TPPPBN3WZ988smdjrdYLFx88cWsXLmS6upqwuEwzzzzTFwgJssxxxzDbbfdxosvvkhlZSU33ngjFRUVPPTQQ93u+8orr3DhhRcycOBAnnnmGT755JO4IG0bD5oJRowYwXvvvUdxcTELFixgxIgRjBgxgj/+8Y8ZnUeQOUS5E8EhycyZM3n33Xd58803ufjii5Pe78gjj6SwsBCA008/nZkzZ/K3v/2N+fPnM3ny5Iys7cCA5MLCQiRJYs2aNfEbWlti22I3sI7q6lVXV8cDpwFMJhOgxNq05cAbcuy1/vnPf+a4447rcL0dCaiuKCoqIhKJUF1d3U7EtkWj0bBgwQL+7//+j0ceeYQnnniCU089tZ0Q7oyOElpi56btzX7kyJHMmjWLRYsWMWvWLN544w3uvvtutFptSq/rQGRZ5s0338RqtcaTbDrCarVy9913c/fdd1NTUxO33p199tkJyQod8dprr+HxeHjllVcSrLKbNm1Ke91FRUXs27evyzGxJIaVK1d2+HzbMh8H0tLSwltvvcVdd93FwoUL49tjcV/p0Pa9P3DgwPj2cDicIOI7I/Y5fu211xLWlCxXXnkld955J3/5y1+47777uh0/f/58nnzySf7xj38wevRoamtrE8qUpIper+euu+7iD3/4A1999VW345955hmGDRvG8uXLE75vDvw+6Iq23yFtv5c6EvUnn3wyJ598MpFIhPXr1/PnP/+ZG264gZKSkpS+fwUHB2GxExySzJ8/n9LSUm699Vb279/f4ZhXXnmly2NIksSiRYvQarX8+te/zsYyAeK19vbv39/O2nfMMccwYcIEAI477jhMJlPc8hTj448/bpdtGBN5X375ZcL2WFZqjBNPPJHc3Fy2bNnS4dzHHHMMBoMhpdcza9YsABYvXtzt2KuuugqDwcCcOXPYunVrSu5Rl8vV7vU899xzaDQapk6dmrD9+uuv58svv2Tu3LlotVquvvrqpOfpjLvvvpstW7Zw/fXXx2+C3VFSUsKVV17JJZdcwtatW+NZgwdaqGLEbsptb6yyLPPkk0+mve5Zs2axbdu2DustxjjrrLNoaGggEol0+J7oSnxLkoQsy+1+pPz9739vl0iULLFMzQPf+//6178Ih8Pd7n/uuecyYcIE7r///k6F0TvvvNNpFufAgQP51a9+xdlnn83cuXO7ne/YY49l/PjxLF26lKVLl5KTk9MuBKMzYgksBxJzoba1Dh9ojY0hSRIGgyFB1FVXV7fLiu3qGJ19h7z55pudrl2r1XLsscfGLcptiz4L+g7CYic4JMnJyeH111/nrLPO4qijjkooULx9+3aeeeYZvvjiC84///wujzNq1Ch++tOf8sQTT/C///2Pk046KeNrPfHEE/npT3/KT37yE9avX8/UqVOxWq1UVVXxv//9jwkTJnDNNdeQl5fHLbfcwr333stVV13Fj3/8Y/bu3ctvf/vbdi6qyZMnM2bMGG655RbC4TB5eXm8+uqr/O9//0sYZ7PZ+POf/8zcuXNpbGxk9uzZFBcXU1dXxxdffEFdXV1SAq0tJ598Mpdffjn33nsvNTU1nHXWWRiNRjZu3IjFYklwbefm5nLFFVewePFihg4dmlIWc0FBAddccw179uxh9OjRrFixgieffJJrrrkm7nqLcfrppzNu3Djef/99LrvsMoqLi5Oep7m5mU8//RQAj8cTL1C8Zs0aLrzwQu6+++4u9z/22GM566yzOOKII8jLy+Obb77hn//8J8cff3y8vEZMvD/44IPMmjULrVbLEUccwemnn47BYOCSSy7h1ltvxe/3s3jx4qQyQTvjhhtuYPny5Zx77rksXLiQKVOm4PP5+PDDDznrrLOYPn06F198Mc8++yxnnHEG119/PVOmTEGv17Nv3z7ef/99zj33XM4777wOj+9wOJg6dSr/7//9PwoLCykvL+fDDz/kqaeeIjc3N601H3bYYVx22WU89thj6PV6TjvtNL766isefvhhHA5Ht/trtVpeffVVZsyYwfHHH88111zD9OnTsVqt7N69m5deeok333yzy/P6wAMPpLTmefPmcdNNN7F161Z+9rOfYTabk9pv5syZDBo0iLPPPpuxY8cSjUbZtGkTjzzyCDabLZ71Dcr75oUXXmD58uUMHz4ck8nEhAkTOOuss3jllVe49tprmT17Nnv37uV3v/sdZWVl7UIAJkyYwAcffMCbb75JWVkZdrudMWPGcMYZZ5Cfn8/8+fO555570Ol0LFu2jL179ybs/5e//IX//ve/nHnmmQwZMgS/3x+PL+wqjlPQi/Ri4oZA0C3dFQWurq6Wb7vtNvnwww+XLRaLbDQa5ZEjR8o/+9nPEoqIdlagWJZluaamRrbZbN1mUXa3lq7mkGVZXrJkiXzsscfKVqtVNpvN8ogRI+QrrrhCXr9+fXxMNBqV77//fnnw4MGywWCQjzjiCPnNN9+UTznllHbZcdu2bZNnzJghOxwOuaioSP7lL38p//vf/07Iio3x4Ycfymeeeaacn58v6/V6eeDAgfKZZ56ZVBHnjjJwI5GI/Ic//EEeP368bDAY5JycHPn444+X33zzzXav+4MPPpAB+YEHHuj03B1IrEDxBx98IB9zzDGy0WiUy8rK5P/7v//rsLCrLMvyb3/7WxmQP/3006TnGTp0aDw7UZIk2WazyWPGjJEvv/xy+Z133ulwHw7Iil24cKF8zDHHyHl5ebLRaJSHDx8u33jjjXJ9fX18TCAQkK+66iq5qKhIliQp4Xy++eab8pFHHimbTCZ54MCB8q9+9Sv57bffbncdOyvaPHfuXHno0KEJ25qamuTrr79eHjJkiKzX6+Xi4mL5zDPPlL/99tv4mFAoJD/88MPxuW02mzx27Fj5Zz/7mbx9+/Yuz9u+ffvkCy64QM7Ly5Ptdrv8wx/+UP7qq6/aFRNum33algOzt2Pn6Oabb5aLi4tlk8kkH3fccfInn3ySVIHiGM3NzfLvfvc7+eijj5ZtNpus1+vlIUOGyJdddpn80UcfdbuuA+kqK7Wurk42GAwyIK9duzap9cmyLC9fvly+9NJL5VGjRiWs8fLLL5e3bNmSMLaiokKeMWOGbLfbZSDhOj/wwANyeXm5bDQa5cMOO0x+8sknO8x03bRpk3ziiSfKFoulXXb92rVr5RNOOEG2Wq3ywIED5bvuukv++9//nvD+/OSTT+TzzjtPHjp0qGw0GuWCggL5lFNOkd94442kX7Pg4CLJchJpXwKBoFeJuaq6y1bsi9x8880sXryYvXv3JhUIny7HHHMMkiSxbt26rM0hEAgEfR3hihUIBFnh008/Zdu2bTzxxBP87Gc/y4qoczqdfPXVV7z11lts2LAhXsBVIBAIvq8IYScQCLJCLMbsrLPO4t57783KHJ9//jnTp0+noKCAu+66SzQ6FwgE33uEK1YgEAgEAoGgnyDKnQgEAoFAIBD0E4SwEwgEAoFAIOgnCGEnEAgEAoFA0E8QyRPdEI1GqaysxG63t2sVJRAIBAKBQJBtZFnG5XIxYMAANJqubXJC2HVDZWUlgwcP7u1lCAQCgUAg+J6zd+9eBg0a1OUYIey6IdYMe+/evUm1tkmHUCjEu+++y4wZM9Dr9VmZQ5Ac4lr0LcT16DuIa9G3ENej73AwroXT6WTw4MFxTdIVQth1Q8z96nA4sirsLBYLDodDfEB7GXEt+hbievQdxLXoW4jr0Xc4mNcimZAwkTwhEAgEAoFA0E8Qwk4gEAgEAoGgnyCEnUAgEAgEAkE/QcTYZYhIJEIoFEpr31AohE6nw+/3E4lEMrwyQSr0lWthMBi6TWkXCAQCgeBAvhfC7rzzzuODDz7g1FNP5aWXXsrosWVZprq6mubm5h4do7S0lL1794paeb1MX7kWGo2GYcOGYTAYem0NAoFAIDj0+F4Iu+uuu4558+bx9NNPZ/zYMVFXXFyMxWJJSwxEo1Hcbjc2m01YaXqZvnAtYkWxq6qqGDJkiBD7AoFAIEia74Wwmz59Oh988EHGjxuJROKirqCgIO3jRKNRgsEgJpNJCLtepq9ci6KiIiorKwmHw6KUgUAgEAiSptdVxOrVqzn77LMZMGAAkiTx2muvtRvzxBNPMGzYMEwmE5MmTWLNmjUHf6EdEIups1gsvbwSQX8j5oIVMZcCgUAgSIVeF3Yej4cjjzySxx9/vMPnly9fzg033MAdd9zBxo0bOfnkk5k1axZ79uyJj5k0aRLjx49v91dZWXlQXoNwlQkyjXhPCQQCgSAdet0VO2vWLGbNmtXp848++ijz58/nqquuAuCxxx7jnXfeYfHixdx///0AbNiwIWPrCQQCBAKB+GOn0wko1rkDs15DoRCyLBONRolGo2nPKcty/N+eHEfQc/rKtYhGo8iyTCgUQqvV9to6epvYZy7djHNB5hDXom8hrkff4WBci1SO3evCriuCwSAbNmxg4cKFCdtnzJjBxx9/nJU577//fu6+++5229999912LledTkdpaSlut5tgMNjjuV0uV4+PIcgM3V2Ls846iwkTJsR/XGSaYDCIz+dj9erVhMPhrMxxKLFq1areXoJARVyLvoW4Hn2HbF4Lr9eb9Ng+Lezq6+uJRCKUlJQkbC8pKaG6ujrp48ycOZPPP/8cj8fDoEGDePXVV5k8eXKHY2+//XZuuumm+ONY490ZM2a06xXr9/vZu3cvNpsNk8mUwitLRJZlXC4Xdrv9oLngqqur+f3vf8+KFSvYv38/xcXFHHnkkVx//fWceuqpAAwfPpzdu3fz7LPPcvHFFyfsP2HCBLZs2cJTTz3FpZdeyqBBg7j++uu544472s31wAMP8Oijj7Jv375uy3fcfffd3HPPPfz0pz9l8eLF8e2bNm1i0qRJ7Nixg/Ly8p6fgE448FrEyuQ0NDSQm5sbH/faa6+h1+uTasicDn6/H7PZzNSpU3v03jrUCYVCrFq1itNPP10kkfQy4lr0LcT16DscjGsR8x4mQ58WdjEOFDuyLKckgN55552kxxqNRoxGY7vter2+3QWLRCJIkoRGo+lRBmXM5Rc7VrapqKjgxBNPJDc3l4ceeogjjjiCUCjEO++8wy9/+Uu+/fbb+NjBgwfz9NNPc+mll8a3ffrpp1RXV2O1WtFoNJhMJi677DKefvppfv3rX7e7NsuWLePyyy9PSqBIkoTJZGLJkiXcfPPNjB49GiB+Xnp6rrvjwGvR2byFhYVZW0NsPkmSOnzffR8R56HvIK5F30Jcj75DNq9FKsft9eSJrigsLESr1bazztXW1raz4mWaRYsWMW7cuE4te4cy1157LZIksXbtWmbPns3o0aM5/PDDuemmm/j0008Txs6ZM4cPP/yQvXv3xrctWbKEOXPmoNO1/i6YP38+O3bsYPXq1Qn7r1mzhu3btzN//vyk1zdmzBimT5/Or3/96y7HbdmyhTPOOAObzUZJSQmXX3459fX18eddLhdz5szBarVSVlbGH/7wB6ZNm8YNN9wQH/PMM89wzDHHYLfbKS0tZc6cOdTV1QGKAJ4+fToAeXl5SJLElVdeCZBwnNtvv53jjjuu3fqOOOII7rrrrvjjpUuXcthhh2EymRg7dixPPPFE0udEIBAIBJkhGpVxB/pviEufFnYGg4FJkya181uvWrWKE044IatzL1iwgC1btrBu3bqk95FlGW8wnNafLxhJe19vMBwP+u+OxsZGVq5cyYIFC7Bare2eb+tuBMXtPXPmzHhxZ6/Xy/Lly5k3b17CuAkTJjB58mSWLl2asH3JkiVMmTKF8ePH88EHHyBJEhUVFd2u84EHHuDll1/u9PxXVVVxyimnMHHiRNavX8/KlSupqanhwgsvjI+56aab+Oijj3jjjTdYtWoVa9as4fPPP084TjAY5He/+x1ffPEFr732GhUVFVx77bWAYq18+eWXAdi6dStVVVX88Y9/bLeWOXPm8Nlnn7Fjx474tq+//prNmzczZ84cAJ588knuuOMO7rvvPr755ht+//vf85vf/CYrRbMFAoFA0Dlv/PXXbP398eytrOrtpWSFXnfFut1uvvvuu/jjXbt2sWnTJvLz8xkyZAg33XQTl19+OccccwzHH388f/vb39izZw8///nPe3HVHeMLRRh3Z/Ju30yy5Z6ZWAzdX87vvvsOWZYZO3Zs0seeN28eN998M3fccQcvvfQSI0aMYOLEiR2Ou+WWW3j88cex2Wy43W5efPFFHn30UUCp9zdmzJikTMpHH300F154IQsXLuQ///lPu+cXL17M0Ucfze9///v4tiVLljB48GC2bdtGWVkZTz/9NM8991w8ZnDp0qUMGDCg3ZpjDB8+nMcee4zjjjsOt9uNw+EgPz8fgOLi4naiN8b48eM54ogjeO655/jNb34DwLPPPsvkyZPjruTf/e53PPLII5x//vkADBs2jC1btvDXv/6VuXPndns+BAKBQJAZjqp9laHSfj77fAWDByTvTTpU6HWL3fr16znqqKM46qijAMXKctRRR3HnnXcCcNFFF/HYY49xzz33MHHiRFavXs2KFSsYOnRoVtfVX12xMcteKjGKZ555Jm63m9WrV7NkyZJ21roYl1xyCdFolOXLlwNKDUJZluOJF1OmTOHbb79l4MCBSc177733smbNGt599912z23YsIH3338fm80W/4uJ1R07drBz505CoRBTpkyJ75OTk8OYMWMSjrNx40bOPfdchg4dit1u5wc/+AFAQp3EZJgzZw7PPvssoJzj559/Pm6tq6urY+/evcyfPz9hvffee2+ClU8gEAgE2ccuK4kIwbr++f3b6xa7adOmdetGvPbaa+PusYPFggULWLBgAU6nk5ycnKT2Meu1bLlnZspzRaNRXE4Xdoc97cQAsz65WmejRo1CkiS++eYbfvSjHyW1j06n4/LLL+euu+7is88+49VXX+1wXE5ODrNnz2bp0qXMnz+fpUuXMnv27HbZxMkyYsQIrr76ahYuXMhTTz2V8Fw0GuXss8/mwQcfbLdfWVkZ27dvBzpOvInh8XiYMWMGM2bM4JlnnqGoqIiKigpmzZqVcvmaSy+9lIULF/L555/j8/nYu3dvXNDGEjKefPJJjj322IT9vs816gQCgeBgEw6HyZHdIIGueVdvLycr9Lqw609IkpSUO/RAotEoYYMWi0GX9azY/Px8Zs6cyaJFi7juuuvaxdk1Nzd36HKcN28eDz/8MBdddBF5eXmdHn/+/PlMmzaNt956i48++ijBVZoOd955JyNGjOCFF15I2H700Ufz8ssvU15enpDEEWPEiBHo9XrWrl3L4MGDASVdfPv27ZxyyikAfPvtt9TX1/PAAw/Ex6xduzbhOMm29ho0aBBTp07l2Wefxefzcdppp8UTfEpKShg4cCA7d+6MW/EEAoFAcPBxNjeQLyk/8G3evd2MPjTpdVdsX6W/umJB6b0biUSYMmUKL7/8Mtu3b+ebb77hT3/6E8cff3yH+xx22GHU19e3S444kFNOOYWRI0dyxRVXMHLkSKZOnRp/bu3atYwdO5b9+/cnvdaSkhJuuukm/vSnPyVsX7BgAY2NjVxyySWsXbuWnTt38u677zJv3jwikQh2u525c+fyq1/9ivfff5+vv/6aefPmxcuIAAwZMgSDwcCf//xndu7cyRtvvMF9992XMM/QoUORJIm33nqLuro63G53p2udM2cOL7zwAi+++CKXXXZZwnO//e1vuf/++/njH//Itm3b2Lx5M0uXLo3HHwoEAoEg+7iaauP/Lw4lfy86lBDCrhPSyYo9VBg2bBiff/4506dP5+abb2b8+PGcfvrp/Oc//0koCnwgBQUFmM3mbo8/b948mpqa2sXieb1etm7dmnLblV/96lfYbLaEbQMGDOCjjz4iEokwc+ZMxo8fz/XXX09OTk7c6vnoo49y/PHHc9ZZZ3Haaadx4oknxsuNABQVFbFs2TJefPFFxo0bxwMPPMBDDz2UMM/AgQO5++67WbhwISUlJfziF7/odJ0//vGPaWhowOv1tnNzX3XVVfz9739n2bJlTJgwgVNOOYVly5YxbNiwlM6FQCAQCNLH29wq7EpopLmlufcWkyUkOdk6Gd9TYjF2LS0tHXae2LVrF8OGDetRd4BoNIrT6cThcByUAsXfVzweDwMHDuSRRx7ptK5eX7kWmXpvHeqEQiFWrFjBGWecIYqw9jLiWvQtxPVIj43/Wc5Ra34af7z1vHcZc+SxXezRPQfjWnSlRQ5EqAhBv2Xjxo08//zz7Nixg88//zwe33buuef28soEAoFA0BsEXfUJj5v3f9vJyEMXkTzRCYsWLWLRokXdBs0L+jYPP/wwW7dujRe7XrNmTdbbgQkEAoGgbxL1NCY87o8lT4Sw64R0yp0I+hZHHXUUGzZs6O1lCAQCgaCPIHsbEh5r+2HJE+GKFQgEAoFA8L1A428CoEYqAsDmSa0Y/aGAEHYCgUAgEAi+F+gCLQBU28cDUBjsfyVPhLATCAQCgUDwvcAQbAYgPOBoAErletweTy+uKPMIYdcJ/blAsUAgEAgE30csEcViZygdhwcTGkmmanf/yowVwq4T+nOBYoFAIBAIvo9YI04ATI5CanQDAGjZt7U3l5RxhLATCAQCgUDwvcAhuwCw5hbjNCs9wv013/XmkjKOEHaC7xWSJPHaa691Oebaa6/lvPPOOzgLEggEAsFBwef1YpUCANjziwnlKC0dNU39q+SJEHbfU6qrq/nlL3/J8OHDMRqNDB48mLPPPpv//Oc/8THl5eVIksQLL7zQbv/DDz8cSZJYtmwZwWCQwsJC7r333g7nuv/++yksLCQYDHa7rt/+9rdIkoQkSWi1WgYPHsxVV11FXV1d+i+2DVVVVcyaNQuAiooKJEli06ZN7da7dOnSjMwnEAgEgr5BS2MNABFZwubIQ1c4AgCLZ3dvLivjCGHXCf05eaKiooJJkybx3//+l4ceeojNmzezcuVKpk+fzoIFCxLGDh48uJ3I+fTTT6mursZqtQJgMBi47LLLWLZsGR21Hl66dCmXX345BoMhqfUdfvjhVFVVsWfPHhYvXsybb77JFVdckearTaS0tBSj0djlmJycHHJzczMyn0AgEAj6Bp7mWgCckh1Jo8U2YDQABYHK3lxWxhHCrhP6c/LEtddeiyRJrF27ltmzZzN69GgOP/xwbrrpJj799NOEsXPmzOHDDz9k79698W1Llixhzpw56HStjUvmz5/Pjh07WL16dcL+a9asYfv27cyfPz/p9el0OkpLSxk4cCBnnXUW1113He+++y4+n49oNMo999zDoEGDMBqNTJw4kZUrV8b3DQaD/OIXv6CsrAyTyUR5eTn3339//Pm2rthhwxQz/FFHHYUkSUybNi1+fmKu2L/+9a8MHDiQaDSasMZzzjmHuXPnxh+/+eabTJo0CZPJxPDhw7n77rsJh8NJv2aBQCAQZBdfs+L5cWnsABQPPQyAUrkWn8/fa+vKNELYZRJZhqAnvb+QN/19gx5l7iRobGxk5cqVLFiwIG5xa8uBlqqSkhJmzpzJ008/DYDX62X58uXMmzcvYdyECROYPHlyO+vekiVLmDJlCuPHj+eDDz5AkiQqKiqSP6eA2WwmGo0SDof54x//yCOPPMLDDz/Ml19+ycyZMznnnHPYvn07AH/605944403+Ne//sXWrVt55plnKC8v7/C4a9euBeC9996jqqqKV155pd2YH//4x9TX1/P+++/HtzU1NfHOO+8wZ84cAN555x0uu+wyrrvuOrZs2cJf//pXli1bxn333ZfS6xQIBAJB9gi46gHwapU2oTlFg/GjRy9FqNqzvTeXllFEr9hMEvLC7wekvJsGyO3p3P9XCYb2Qu1AvvvuO2RZZuzYsUkfet68edx8883ccccdvPTSS4wYMYKJEyd2OO6WW27h8ccfx2az4Xa7efHFF3n00UcBsFgsjBkzBr1en/Tc3377LYsXL2bKlCnY7XYefvhhbrvtNi6++GIAHnzwQd5//30ee+wxFi1axJ49exg1ahQnnXQSkiQxdOjQTo9dVKS0lCkoKKC0tBSgnWUuPz+fH/7whzz33HOceuqpALz44ovk5+fHH993330sXLgwbsEbPnw4v/vd77j11lu56667kn6tAoFAIMgeIbfSJ9avV/u/azTUaAcwNLKbxn3fMnzMhF5cXeYQFrvvGbEYOEmSkt7nzDPPxO12s3r1apYsWdLOWhfjkksuIRqNsnz5cgCWL1+OLMtxETZlyhS+/fZbBg4c2OV8mzdvxmazYTabGTduHIMHD+bZZ5/F6XRSWVnJiSeemDD+xBNP5JtvvgHgyiuvZNOmTYwZMybuwu0pc+bM4eWXXyYQULKpnn32WS6++GK0Wi0AGzZs4J577sFms8X/rr76aqqqqvB6vT2eXyAQCAQ9R/Yowi5kyI1vazEPAsBfIyx2go7QWxTLWYpEo1GcLhcOux2NJk2trbckNWzUqFFIksQ333zDj370o6T20el0XH755dx111189tlnvPrqqx2Oy8nJYfbs2SxdupT58+ezdOlSZs+ejcPhSPZVADBmzBjeeOMNtFotAwYMiCc7OJ1KYckDRaksy/FtRx99NLt27eLtt9/mvffe48ILL+S0007jpZdeSmkNbTn77LOJRqP8+9//ZvLkyaxZsyZuhQTl+t19992cf/757fY1mUxpzysQCASCDOJrBCBqyotvCjrKwf0RNPafkidC2GUSSUrKHdqOaBT0EWXfdIVdkuTn5zNz5kwWLVrEdddd1y7Orrm5ucOM0Hnz5vHwww9z0UUXkZeX1+75GPPnz2fatGm89dZbfPTRR/z+979PeY0Gg4GRI0e22+5wOBgwYAD/+9//mDp1anz7xx9/zJQpUxLGXXTRRVx00UXMnj2bH/7whzQ2NpKfn99uHoBIJNLlesxmM+effz7PPvss3333HaNHj2bSpEnx548++mi2bt3a4ZoFAoFA0DfQ+psBkM2t9wJNwQioBIu7/5Q8EcLue8gTTzzBCSecwJQpU7jnnns44ogjCIfDrFq1isWLF8fdmm057LDDqK+vx2Lp2jJ4yimnMHLkSK644gpGjhyZIMDWrl3LFVdcwX/+859u3bGd8atf/Yq77rorHue3dOlSNm3axLPPPgvAH/7wB8rKypg4cSIajYYXX3yR0tLSDsVqcXExZrOZlStXMmjQIEwmE3a7vcN558yZw9lnn83XX3/NZZddlvDcnXfeyVlnncXgwYP58Y9/jEaj4csvv2Tz5s2d1vYTCASC7z3+Fti7DoZPA2325Ygh2AyAxtoq7Kxlo2Az5Af2ZX3+g4WIseuE/lzHbtiwYXz++edMnz6dm2++mfHjx3P66afzn//8h8WLF3e6X0FBAWazudvjz5s3j6ampnaxeF6vl61btxIKhdJe+3XXXcfNN9/MzTffzIQJE1i5ciVvvPEGo0aNAsBms/Hggw9yzDHHMHnyZCoqKlixYkWHLm6dTsef/vQn/vrXvzJgwADOPffcTuf9wQ9+QH5+Plu3buXSSy9NeG7mzJm89dZbrFq1ismTJ3Pcccfx6KOPdpm4IRAIBN973rsbnr0Avko/VCYVTOFmAPS2gvi2wiFKyZOyaA2BJIroHwpIckcVZQVxnE4nOTk5tLS0tIsV8/v97Nq1i2HDhvUolioajeJ0OnE4HOnH2AkyQl+5Fpl6bx3qhEIhVqxYwRlnnJFSNrUg84hr0bfoF9fjn+fBjv/C5KvhzIezPt2eu8cxRN7PlhnPMe6EMwGQI2HC9xSjlyLsvvwzho5IvmJEjINxLbrSIgciVIRAIBAIBAdyQOkjQRbwNSv/1rYP/8kGdllJwDPnFMW3SVod1doyABr3Hpx1ZBsh7AQCgUAgaIuvGf5wOLzys95eSb+mpVFp8RWq3pL1uaKRCA7ZDYA9rzjhuWaTUvLEW/1d1tdxMBDCTiAQCASCNvj2bgJXJcGvXk+6q48gdbTBFgD0gUZw12V1LldLE1pJuZb2/ERhF3Co8dCNO7K6hoOFEHYCgUAgELRh//49ABiiPvA19fJq+inRKJaou/VxXXbdoK6mGgC8shGjKbG6g1QwHACTa09W13CwEMJOIBAIBII2RFyt1iNXzc5eXEk/JuBEQxtraO23WZ3O06y4fZ1S+8QDS8loAPL8e7O6hoOFEHYZ4MD+ogJBTxHJ6gJB7xH11Mf/31LVP9xzfY4DLKFybXbj7PwtyjX1aNvXKi0YqmTClkWrCYXDWV3HwUAUKO4BBoMBjUZDZWUlRUVFGAyGlHqwxohGowSDQfx+vyh30sv0hWshyzJ1dXVIknToljEQCA5l2gg7T21F762jHxP1NidYlsLVW8jmt13QpVxTny6n3XOFA0YSljWYpSD79lcwaOih3UVICLseoNFoGDZsGFVVVVRWpt4jNoYsy/h8Psxmc1rCUJA5+sq1kCSJQYMGodVqe20NAsH3FZ2/If7/SGP/iLvqa/ic9ViBgKzDKIWR6r5VElWy9L0b8SjXNKBvL+w0egPV2mIGRKup3/OtEHb9lUWLFrFo0aJu+4gaDAaGDBlCOBzudmxnhEIhVq9ezdSpU4WFppfpK9dCr9cLUScQ9BIGf2P8/1pX/4i76mt4WuqwAlvkco5gB7pgC7hrwF6anQm9yjUNG3M7fLrJOIgBvmq81duzM/9BRAi7TliwYAELFiyIV3vuipjLLF0hoNVqCYfDmEwmIex6GXEtBAKBMdQa/2X2pu+NEXROwKUIrVo5lwpKGSFVKYWKsyTsJL9yTaPm/A6f99mHgG890YZDP1lGBHQJBAKBQNAGS6g5/v/cYHXvLaQfE3IrrtFm2cZ2WSkQnM0OFLqAIuwkS8fCjvwRABidFVlbw8FCCDuBQCAQCGJEI9iizvhDh+yCgLuLHQTpEPEqQsuJlW3yQGVjFmvZGYLNAGitBR0+bykZBUCuf1/W1nCwEMJOIBAIBIIYviY0KCWs3LIJgGBjRS8uqH8iq8JOZ8tne1Sx2MlZrGVnDiti3WAv7PD5vMFjACiLVBGJHNolzISwEwgEAoEghlrqpFm2sldWWk817T/04676GpK/GYCcvCK2MxgAufabrLVws0UUYWd0dCzsioeMISpL2CQftTWHttVOCDuBQCAQCGJ4FWHXIDvYTxEA7tpdvbmifole7ROrs+UTyBlGSNaiCbrAmZ1kFYesCDtrbnGHz2sNZuo0iuir353d9mbZRgg7gUAgEAhUom5F2DVix2MqAyDYsLs3l9Qv0YcUoaWz5DOoMJcKWc2GzUICRdDvwyIFAHDkdyzsABqMikvYXXVolzwRwk4gEAgEApWAU2kW3yDnQO4QADQtopZdpjHFY97yKS/IbgKFs1G5phFZwp7TcfIEgM+mXO9I/aHdRk4IO4FAIBAIVIItSrP4ZsmBPl+50Rs9+3tzSf0SS8QFgMlRSHmhtU3Jk8wnULialGvqlGxouij8Hs0bBoDhEC95IoSdQCAQCAQqEVcdAB5dLuaicgByAlW9uKJ+SDiAGT8A1pxChhVa2KZmxmbDYudrUa6pS3J0Oc6kljzJ8R3aFtp+L+z27t3LtGnTGDduHEcccQQvvvhiby9JIBAIBH2UqJoV69fn4SgdDkBetBHCgd5cVnap/Raeuwj2bzg48/maAYjKEvbcfIYWWNkqq5mxdVshmtlyIwGXUgzZo+ta2OUNGgtASbgSOUvZuQeDfi/sdDodjz32GFu2bOG9997jxhtvxOPx9PayBAKBQNAHkdSs2IAxj+LSQfhkAwDR5kO7BEaXfLkctq2Ejx8/KNO1Fie2kGs1MTjPwl5KCMpapKAbMhzTGHKp11TXdXvQ4iFKLbtcyUNd7aHbcaTfC7uysjImTpwIQHFxMfn5+TQ2Nna9k0AgEAi+l+j8inUnZCqkNNdMpawE27dUH9oB9V3RUKeIGH/FZwdlPk+z4hptlm3kmPUYdBpK8xzslAcoA+oyG2cX9Sj3/KAht8txBoudOklpOVZ7CJc86XVht3r1as4++2wGDBiAJEm89tpr7cY88cQTDBs2DJPJxKRJk1izZk1ac61fv55oNMrgwYN7uGqBQCAQ9EcMAUUEyOYC9FoNddoSAJzV/beWXW2dkjVq8uwHV/YtVV6nYkFzSzb0WkWGDC2wsD2WGZvpkic+5ZpGTHndDm0wKOLSXbkts2s4iPS6sPN4PBx55JE8/njHJuDly5dzww03cMcdd7Bx40ZOPvlkZs2axZ49e+JjJk2axPjx49v9VVa2FjpsaGjgiiuu4G9/+1vWX5NAIBAIDkGiUYxq4VzJphSrdaq17AL1Fb21qqyjU18zAHvXZn0+v1ONedPa49uGFVrbJFBk1mKn9SuuX8zdCzuPdSgA4frvMrqGg4mutxcwa9YsZs2a1enzjz76KPPnz+eqq64C4LHHHuOdd95h8eLF3H///QBs2NB1wGcgEOC8887j9ttv54QTTuh2bCDQGiTrdCq1dkKhEKFQKKnXlCqx42br+ILkEdeibyGuR9/he3EtvI3oiQAgmfMJhUIELAPAB9GmPX3qtWfyeuiDzvj/vTs+QT+q83tyJvCrFju/zh5f/+A8E+vVBIpozRYiGTzX+qAi7GRzXrfnK5JTDo2ga65I+twejM9GKsfudWHXFcFgkA0bNrBw4cKE7TNmzODjjz9O6hiyLHPllVfygx/8gMsvv7zb8ffffz933313u+3vvvsuFosluYWnyapVq7J6fEHyiGvRtxDXo+/Qn6+FzV/JqUCLbGHfnt2sWFFBbUBNnmjYwYoVK3p3gR2QiesxOdgc/3/DV+/xefS4Hh+zK+w7v2Es4Awb4ue0rklim1rLLlrzDSv+/RZImXEqDvcpFsJ9tU7qurmGEZfEFMDqrkj5emfzs+H1epMe26eFXX19PZFIhJKSkoTtJSUlVFcnFwfw0UcfsXz5co444oh4/N4///lPJkyY0OH422+/nZtuuin+2Ol0MnjwYGbMmIHD0XWqdLqEQiFWrVrF6aefjl6vz8ocguQQ16JvIa5H3+H7cC2kPR/DN0qf2MkTJ3DG5EG8904Y1i+mSG5i1Bln9PYS42Tyevg3XQtqdY+y0G7OmHkaaA0ZWGXHbF32b3CBPqeUU9Vzeli9h79/u4aArMdIkDNOOBzUgsE9pfKL/wMZxhxxDONO6Poa7tuSD68+zgC5ljGzZiFJUrfHPxifjZj3MBn6tLCLceCJlWU5qZMNcNJJJxFNoSaO0WjEaDS2267X67P+ZXYw5hAkh7gWfQtxPfoOB/taNHmCaCSJHMtBmDPQDEAjDvJtJvR6PTkDRgKQF6lDp9WApvPOBb1Bj69HNIpWVkqABWUthmgAGrbCwKMztML2aANKTJ9szouvfVixA0mjZYc8gHHSbvSN30Hx6IzMZ48qosiaX9rtuSobPh6AAqmFRpeT/ILCpOfJ5mcjleP2evJEVxQWFqLVattZ52pra9tZ8TLNokWLGDduHJMnT87qPAKBQCDomFAkyozHVjPjsQ8JRzJbtLZD1OLEjbIdh1mxexSWDSUka9ERAVc/7EARcKJRzXVro0qBXvaty+qUGjVZQ2POjW/TazUMyjNnvGesHI3ikN0A2PKKux1vsufRhOKdqz5ES570aWFnMBiYNGlSO7/1qlWruk2C6CkLFixgy5YtrFuX3Te4QCAQCDqmyRvE6XLR5HRT6zoInR+8SixWg+zAYVIsJAPzbVTLSm0zb23/K3kScCuv2Scb+Cx6GADBik+zOqdBFXZaa37C9qEFbTJjM9Qz1uNqQicpPwoc+d0LO4A6gyIuXYdoyZNeF3Zut5tNmzaxadMmAHbt2sWmTZvi5Uxuuukm/v73v7NkyRK++eYbbrzxRvbs2cPPf/7zXly1QCAQCLKNyxvgXcOtrDTcRlWTO/sTepTCuQ04cJgVYWc16qjWKIKguWpn9tdwkPG0KFbKFqzstyux55E92TVomMKKa1RvK0jYPqzAwjY1MzZTFjtng1KjzysbMZutSe3jtgwBIFh3aJY86fUYu/Xr1zN9+vT441jiwty5c1m2bBkXXXQRDQ0N3HPPPVRVVTF+/HhWrFjB0KFDs7quRYsWsWjRIiKRSFbnEQgEAkHH+FpqGa+pBeA/NRUwrCir80XddWiARtlBjrk1pqnZUArBr/HV9T+Lna9Fsdi5JRvaQccQ3S5h9uwFdy3YkrNwpYo54gLA5Ei02JUXWnlfjtWy2wbRSI9jGr1qlwunZMeSZGx+JHcYNIO+uaJHc/cWvW6xmzZtGrIst/tbtmxZfMy1115LRUUFgUCADRs2MHXq1KyvS7hiBQKBoHcJOFvbP/pqsi+qIm7FetUg27GbWu0efovSjSDStKfD/Q5l/C5F2Hk1NkYNKYuXHMlaoWJZxiYrws7sSBTq5YVW9spF+DFAJACNPb/mPqci7Nza5KtaGIpHAGDzHprXu9eFnUAgEAgEHRF0twq7SGNF1ueT1eQJjy433uoKIOpQ3IM61/6sr+FgE3Qp59iny2HCwFw+jypZwFlLoAi60aHEvNlyEzNOywusyGjYkcEEiqBLEXa+FISdY+AYAIpDh+b1FsKuE0RWrEAgEPQuIU9z/P86196szyd5FWEXMCTGfunylZgri6+y3T6HOmGv0pUhqLdz+EAHG+VRAIR2f5aV+UJqskZA1pN7QG3YQXlmtBqJb6OxnrE9T6CIqD8OAobcpPcpKR8HQDFNNDc39XgNBxsh7DpBuGIFAoGgd4mJDgCLZ192J4tG0fnVZvHmRGFnLVEK5eYFa0CWs7uOg0xUPcdhQw4Ok566nCMA0FRthEjmW2S52yRrOMyJtdn0Wg2D88xsi2YugUL2Ktc0bMxNeh9LThEt2ACo2b21x2s42AhhJxAIBII+iexrbU6fF8xyDTl/M5KsJMvJlsRm8XllirAzEojXuusvSP5mAGRjDgA5g8bRIlvQRvxQ83XG5/M2K+fPKdmUgs8HMLTA2hrnV9tzYSf5FGEXNeV1MzKROr0SV9myXwg7gUAgEAgygqR2ggAoidZkt0ixWsPOKZuxHFAWY0BBLjVyLgChxt3ZW0MvoFG7QKAWCx4/KI+NUcUdm404O58zlqxh7/D5YYVWtseEXf32HlsNdep7SDKnJuxch3DJEyHsOkHE2AkEAkHvIgVaLXYlNFHblHy/zJSJd51wtHMRFtoM7EfJ4GzpZ7Xs9LEuEBal9MiEQTlsjCVQZCEzNqhm4fp1HQu78gIL++UC/JIJoiFo7Nn5NoSaAdAcUDOvO8I5Skk1bdOhV+JGCLtOEDF2AoFA0Lvogq74/zWSTGNlFkVV2+LEpsQSr5Ik0aQvBcDdz7pPGELKOdZbFYvW4QMcfK4mUESyIOxakzVyOny+vFDJjN2tUePseuiONYUU4WqwJd/zFUBfpIjbQ7HkiRB2AoFAIOiT6EOuhMeemh3Zm8zbarHLMbdvuO41KzFX/c0Va44oVlCDatGym/Q05x9BVJbQNleAuy6j88WTGQydCLsCxQ3+dUg539T1LDPWGlVen9GRWnFrx4DRABQGD72SJ0LYCQQCgaBPYgwnCrtQfRatZZ5Yn1h7O1csQNimlODQtmS/7MrBxBpVWrWZHa2uyuGDBvCdrAqrTMfZ+ZoBiJpyO3x6UJ4ZnUZiSySWQLGlR9M51GLI1tzULHZF5Urf3BK5AbfnILSzyyBC2HWCiLETCASC3iXWeqpG7dUqtWTRLRaz2OHAYWov7DR5SjC9yduPatlFI1jxAmDNbRV2Ewbm8HmWEig0sYSYTpIZdFoNg/MtrQkUPahlFw76seIHwJZXktK+9vwBeDChkWQqd2amb+3BQgi7ThAxdgKBQNC7mFVrUp1V6QRgdGexlp2ntZ2Yw9y+jbq5WCl5khOozt4aDjJRXwsalLp8ttxWV+WEgTnxOLtMCzuDmqyhtXSepTq0wMK2qCrsGndAOJjWXM5Gpc9wRJbIyUvNYockUaNXxHz1dxvTmr+3EMJOIBAI0mD1tjpm/XENX+5r7u2l9FtssgeAQOHhADj8WYx3iiVPyDkdumJzSocDYJXd4G9p9/yhiEetKeeVjeTYWku8HD4wh02qsJP3bYBIOGNzGlT3ut6W3+mY8gIrVeTj11ohGoaG9EqOuJsUYeeUbOh07cV6t/vnK++74F4h7AQCgaDf8+YXlXxT5eTfX2a5cO73lFA4gh1F2BkHHwVAYTiL1jK1jl0j9g5dsWVFhTTJSjcCufnQy5TsiFgXCCdWDLpWOWAz6ojmj8IpW5DC3h7HubXFoiZrGO2dlx8ZVmgFJPbrFItZuh0ovC2KsHNJHZdW6Q7zEOV952jO3Os/GAhhJxAIBGlgc27n97onCTb2j5t8X8PldmGQlE4QhaOUWOd8nIR9WaplF3fFdpwVW5pjYp+suPNcNf2j5EmsWLBHY2v33ITB+WyKjlAe7Mtc2ZOOkjUOZGiBBYCt0Z6VPAk4lWvq0Tq6GdkxAw47FoDhkZ3UtvjSOkZvIISdQCAQpMEJja9yqe59xte83ttL6Zd4VWtSWNZQNGAYLbLiKmyqzEInAFlGjlns5I6TJww6DfVaJQDfWd0/ihT71WLBXm17i9b4gTlsjMXZ7c1QnF0kjE1N1mgb03cgisUONvmV2oHpCrtQvBhyblr7WwcfSQQNhZKTzVsPndZiQth1gsiKFQgEXWEMKZYjq7//BNP3JXxOpd6ZW7Ki1Wqo1iiiKiudH/wtSFGldVWjZMdu6jgey20qAyBY3z9q2YU8SrFgv669ReuIQZnPjA2q8wE4ukhmGJgbK3milJhJt5ZdVC1hE+ykZl636M3Um5QOFHXbPkvvGL2AEHadILJiBQJBVxgiSvyXI5TZAq4ChYBbEQEeSbXUGRTrTaAuC8JOdcO6ZDMGowWNRupwWFCtZUeWatnJssxnOxto8fWsP2qyRDyKeA7p2wu7cWUONsmqK7ZxR7zOX09wxZIZZDMOi7nTcTqthiH5bTNjd0LIn/qEPuX1RUyp9YltS6BwAgDRyi/TPsbBRgg7gUAgSAOTKuyKog14g5nLGhQoxKw7MTehx6KIqmhjReYni3ed6DhxIk6uEvNlcGcnO3fN9nou+tunnPP4/6hszn5Ml6wWC450YNGyGnUUF5XyXTRzhYo9qnvdLdk6Fc8xhhZYqCWXoM4OchQatqc8n8avvIdkc+cZuN1hH3Y0AIXub/GHImkf52AihJ1AIBCkgUlWYoVKpEZqnYFeXk3/I+xpBiCgVQL7w3ZFVOldWbCWedoUJ+4gcSKGqbAcAHsgO5nQ3+2v5Q7dM5Q0buDiv32adXGn8TcDIJtzO3w+sVBxzxMofC2K1c+t6T5LtVzNjK0xKfUD04mz0webAZAs6Qu73OGTABgnVfDF3ua0j3MwEcJOIBAIUiQQjsSDwB2Sj7qG+l5eUf8j6os1i1fchFJ+OQA2bxasZd42xYk7ia8DsJeqRYojTRDKvOjK3/MOV+tW8LThQfKbvuTiv33K/iyKO61aLFjqpL3X+AwXKg64lPPs7yBZ40BiCRQ7JbXkSRrCzhhSXl9XNfO6Qyo9AoBBUj2bvzs0sqGFsBMIBIIUcfnD2Gi94brq+lf/0D6B6iYM6xURYC5SRFVeKAvJKm2KE3dU6iRGcXEZHtmoPGjJfBcMrVN5H5mlIMuMD0PTLi7+2ydZE3f6oJIApLV2HIN2xKAcNkZHKg/2fw7Rnrkiw6p7PdBBTN+BDC1QhN1XIdUVnEYChSWsCDuDvfMM3G4x5+I0K7F+zTvWp3+cg4gQdgKBQJAiLl8oQdj5G0Qtu0wjBRTRETEqIsBRqgTyW2UP+Jo63S8tPG2KE3ch7AbmWdgnKyLBX1eR2TUARq8iWmUkcnHyrOn/4W6s4eK/fcK+Jm/m5wsr59hg7diiNW6Agx0MwiWbIejucaHiqFdN1jDkdjt2mCrs1nqUPsHpWOxsUeX1mXNSbCd2ANESJYFCW7uZaFTu0bEOBkLYdYIodyIQCDrD7XGhk6Lxx+HmLLa6+p6iUYWdbMwFoKSwgDpZEXmRTCdQeFuLE3eVPGE36anRKMKupXpHZtcAWANK1mjlxBsgZwiD5Ur+afkDtY0tXPLkpxkXd+aI0t7LYO9Y2FkMOoYXO9oUKu6ZOzaWrBE1dl9+ZECuCb1W4uuQmoncVAHBFF6/LOOQlWLIttziFFeaiH2YEmc3PLyTnfXuHh3rYCCEXSeIcicCgaAzfK7mhMeSS9SyyzQ6tU4gJkUEFNmN7JOVG7SzKsOiytMmK9bcdU/RFrXsiq++IqNLkGWZ3IiyDsOQSXDZS2DKZXx0K3+z/oX9jR4u/ltmxZ1NFT5dWbSUQsWqO7aHhYq1gWblP+buy4/otBoG51uox0HImAfIUJ98kWCfqzn+48uW3zNhpx0wEYDx0i427M6wtTgLCGEnEAgEKeJ3J365x1xogsxhCCnWJI0lFwCtRqJep4gqb02Ga9l5W7Niu4qxAwhYFQuS3JTZuMomb4gSFFdlbskQKBoDFz8HWgOnRD7lIdsL7GvycfHfPmVvY8/FnRwJY493gehc2B2RwcxYXUCJedNYkqsrV16gZMY2WlWLYU3yrmBXUw0APtmA3ZZer9g4ZUcCMEyq5ssdfd86L4SdQCAQpEjA05LwOOZCE2QOk+om1LURAS6TEkgfashwdqJqsavvxhULEM1Ryq7oXJlNnqhpbKFQUqyU+jy1R2r5ifCjxQDMDr/FLY734uJuX1PPEir8rjZdILpo7zVhUA6bYgkUDd+BGieXDgY1pk/XSUzfgZSrcXYVelXYVX2R9FxutRhyi2RHkrqumdcttmICpmI0koxz96aeHesgIISdQCAQpEjIm9iIPjcsuk9kGrPaLN5oy41v89uU7ERtSwaTVWS5jSu26zp2AIZ8pfyG1ZfZWnbNtcprCqJPdFVOmA2n3wPAguBS5uZ+wf5mH5ctWUdjD8onulqU96xHNmIxmzodN64sB6dkZ0dUaafGvvQzQ81hRawb7QVJjR9WaAHgy+hwZUPlxqTn8juV1+fWdJ+BmwyaAUrZkzznNzR6ghk5ZrYQwk4gEAhSJOJTLHYBSbkhFtOIO9C/u0+EI1ECB7HwvjUu7FpFgJyj9O00eTJoLQs4Qe0T24Cjyzp2ANYSRWTkhOsgkrlr7lFL5jTpiuBAC9MJ18Hkq5CQ+W3oMc7K28P+Zj9v703/Fu5pVtuoSbYuLVpmg5ZRxXY2yj13x1rjWarJCbtYyZOPvaoFs/rLpM95wKlkOvs66IObDvpBRwFwuLS7z8fZCWEnEAgEKRL1KTeoRrMiNIqkFmqanF3tcsjz02c2ctcG7cGxVsgyNllp2WZytLrt9Gotu5xAlWJpywSqtc6LiQAGcixdW+wKSgcTkHVoiYIzc/FWwSZFrHqMHbhFJQlmPQSjZyFFAjwaeYBhUhV1/vRdjH6X4lL1amzdjp0wqG2cXZoJFLKMXb2mVkdy5UdiRYo/bslFNtgh7E+6nl3YrRZD1uemvtaOUAsVj9f0/QQKIewEAoEgRWS1FIfHMoggioWnuaZ/FynetK8FX0Ri076W7gf3kIjfhVZShJuljbCzl5QTlSUMciBeVLjHeBXLToNaSqW7GLtBeVYqZcXiFG7KnEtYdlYCELSUdjxAo4XZT8GAozEEm1mmfxBdMP1rEVCFnS+JLhATBrYpVLxvQ1qFiv0+N0ZJsYza8pIrGDwg14xeKxEIQ6BYqSWXrDtWVmMBw0nUzEuKMkXYjZL28UVFTWaOmSWEsBMIBIIUkYKKmzBqdNCsVawPnrr+W6Q4GpXjruYdddmv4+VV3WgBWYfd1mpRKslzUI0af9a0OzOTxbtOKAKnuxi7QpuRKhRh4sxgLTudW4nZk+0DOh9ksMKl/yLkGMJQTS3zoy+lPV/YowifgL77mnITBuWwVR6MBxMEXWl1gXA1Kec5LGtwOHKT2kerkRicr8TZNdjHKRuTFHaST3l90SRKqyRF7lAixhwMUgTv/q8JhA9iXEKKCGEnEAgEKaJRhZ1ktOMyKDWyAk2ZbzHVV3AFwlyoeZ/f6Zawszb7LmevU7kpu7Bi1LfGvA3IMbNXrWUXbarIzGSe1uLEWo2E1aDtcrhGI9FkKFF2rc3QGgCzX8ni1OYO7HqgrYjItDsAGCZVph3bGfEq7sSQofsYtHFlDiSNlk0RNTt1b+pxdu5mRdg5JRuSJnnpEetAsdMwWtmQpLDTqTXzZHP6fWITkCQ0qtVujLyTryv7buiFEHadIDpPCASCztCGFGGnMTkIWJSbvNxS2ZtLyipOX4ibdS9yue49dFWfZ30+v0ux2Lkla8L2IruR/WpLL09Nhqxl3jbFiU26pEpjeM2KVS3SmDkrbU5IEXaWgsHdjjXlK9nBhbRQ704zNTbWBcLQvcXOpNcyqtjGZlnNTk2jvZe3Rb2mmtRqypWrcXZfRpT4Smq+gnD3cZ6GYDMAWmtyiRrJIKn17A6XKthQ0Xfj7ISw6wTReUIgEHSGPqwIO63FQcSmlIHQejJb/qIv4fSHcKAEvhubv0POVOJCJwTVAtAHBvZrNRKNBuV8B+syVMtO7RPbQPelTmJEHIqw0jgzE1cZCEcokBUrpb14aPc72JQfE0VSM7Wu9ISdJmbRMuUmNX7CwBz2qNZSmlMXtAFVrPvSFHYb3blKF5JIMKmetaawEn9oSLK0SlLEhJ1mN+t3p1/PL9sIYScQCAQpYowoIkdvyUGTo1hvzL6+HVDdE5wePyY18H1geF/aYiJZQh5F2Pm07TM2vRZFVMlpiIsOicfYdV+cOIY2TxFfFm9mrLS1LT5KUF6zo7h7ix02RWA5JB9NTeklUKTaBWLCoBz2yWo2a3Pq8Y0hdyymL7XyI+UFSoxdRaMPBiglR5Jxx1ojiqvUYE8uUSMp1MzYcdJuNlY0ZP0HTroIYScQCAQpYoqq1itLLsZ85UZsC/bfIsVed3P8/yOkSr6rzW4CRcSrzBfUtbfuRBzK+Ta4MpSF7G1bnLjrGnYxLMXlADiCNRCN9ngJDTX70UsRImiQbJ1kxbbF6CAoGQHwNKUnLlu7QCQp7AbmxOMb5eY9KZebaY3p697125ZY94k9DV6iZckLO7usvD5LF+3SUqZwFLLOjEUKYPfuZk8GWrtlAyHsBAKBIAVCkSgWWWnnZLTlYCtUOhEUROr67C/4nuJzt1qFDoawi6rxX0F9e2GnyS8HwOqvSqvsRjtiyRPYk7bY5ZWWE5ElDIQyUnbFVadYwFo0eaBNQlxKEm69khQQaEqvT7FJ7QJhsCWXXHBYmYNqSRFJUtANvtRizGQ1SzVizE1pvwG5ZgxaDcFIlMYcNTO2alOX+0RDAaz4AbDnFac0X5dotEil4wElzm59H42zE8JOIBAIUsDlD2NDEXZmWw65ZWqRYppw+vp2q6F0CbbpjTtYqqWiuiG7E/qV+cIdWHdsRYMJylq0chicGXCFqnXsGmUHOUnG2A3Id1CNIojkNNySB+JrUKyPTkPybsOAURFZEVd6IQCWSKy9V3LCzqTXUl5SQI2cq2xIMStZo15Tkozpi6GUPDEDsEs/RtlYswVC/k73ifWJjcoSOXkZtNhBmzi7CjbsEcJOIBAIDnlc/hA2SRF2OnMOptwBRJEwSBEaajLXiaAv0VbYaSUZT9W2rM6nUeO/ZGP7eKzSXBuV8VivHsbZtekT25BEn9gYZbkm9qtr8NT2PIkj2qy8b/ymkqT3CVsUEajx1KY1p01WrK6WnOTF5ISBjrg7NtVzr1OzVKUkY/raMrJYibVc12wBS4HSAq72607Hu1Rh58SK0WBIeb4uUePs+nJmrBB2AoFAkAIufxi7arHDaAetniZJuVm11PbPIsUhf2LNLk3j9qzOpw0q1iTZ2N5iV5ZrYq9a8iSdIP4EAi6IKIkgyfSJjWHUaWnQKiLMVd1zYadxK+7UsJphnRRqAoXel7orOBQKYld/nNhTiEEbVWxPO4HCEEwtpq8tJ4xQ5lyzvSGpBApvsyrsUszATYqyWGuxCrbVOmnxhTI/Rw8Rwk4gEAhSwOXxxFsjYVRuHC065cbjq++fwi7qcyU8LvLvpsWbvRuaPqSIAI05t91zZTkm9qnCLtpY0bOJ1MSJgGTCjzFpix2A26yWXWnsuSvW6FOEXSzDOhl0Ocr85mB9yvO5mlr3SUXYlRda07bYmWJZqrbUy49MHa1c7/W7GwkWK67QroRdrA6iR5NaBm5SFI8DjY5cycMAuZ6NfdAdK4SdQCAQpEDbRAIMirDzqC60UHP/dMVGA4nCboSmku/qXJ2M7jlGNbBfa8lt91yRzch+taVXoL6H1jK1hp1To1gGk42xAwjZlA4RUkvPs3NtAcXCFMuwTgZzriLsHOFGItHUknbcLYqw82BCq0/eVTms0BoX1amWm7FEFNev2ZG6sCsvsDA430woIvONRu1+Ubmp0/Eht/L6/Em0S0sZnRGKDgPUOLvdQtgJBALBIY1fjTcLSMZ4BmNI7T4hZSKYvy+iCrtYiY2RWc6MNamB/foO3HY6rQanSRFVkYaKnk2kWuyaUSw7yWbFAki5Sja00d0zMS/LMvkRZR2xDOtksOQr1r1CqYWGFLtPeGJdIKT2dQK7Yki+JR7fmMq5l2UZu6xcU3NO6skMkiQxdZQiKN9rVluu1X4DwY7LjURiNfMMuSnPlRSqO1YIO4FAIOgHBD3NAPg1re2uZIdyk9V7+2eRYk1QqdtXbSwHYLhUxXc12euVaVHrBHZWiiNgVyxb2p52flBLldTLqrBLso4dgKmoXNknUJ1yTbe2tHiDFKMIkdzSJLpOqGjsiks0ne4TsS4Q3hRj0Aw6DUG1jqDUknwtO38wjANFhDny0ys/EnPH/nuXrHTekCNKe7GOUDOdI8bU4/mSok1rsU17mwlHel7LMJP0e2HncrmYPHkyEydOZMKECTz55JO9vSSBQHAIE/YqgiaobRV2uhzFimAN9E9hF+uN22gcQkTSYZECNFZVZGeyaBSLrIgAo73jG3PMWmbyVSfVN7RT1IzYuqhiuUrFYucoUXqXmmUv+JvTXkJtfS1WSRFmRrUHbDLIsbZitFDn9KU0Z1C1aPk7KADdHZaicqKyhDbiT7qGX3NTPRpJEYGWNFyxAMePKECrkdjZ4MVXqFjMOouz06jXQzZnSdipmbETtBV4gxG+rc5uXcdU6ffCzmKx8OGHH7Jp0yY+++wz7r//fhoaslyDSSAQ9FsiPsUVG9K1urFMhYoVwxHqn90ndGoLtYjOSsCuWJWidVkqeRJwoiEmAjq+MVvzB+CTDUjI0JMYN9WyUx1WrmUqMXZlhflxS19Pyq401yj7OiU76M3J72hVLFh6KUJLY2olT8Jqy7ZAGjFoQwpzqEa9Lkm+bnez8rnwYkLSGVOeExTRffSQXAC260YqGzsRdrqA6h41J1ejL2VKxwMSJTRRQAuf723Ozjxp0u+FnVarxWJRes35/X4ikUi/rQ4vEAiyT9SvxApF9K3CLqdYsSAVRvtu/8h0iUZljBE1lklvQlM0GoBc7058wQx0fjgAWe1o4JMN2K0dx4CV5ZrjQfw9KnnSpoYdkFJW7MA8c7z0R6A+/TX46hVhGsusThqtAacaI+dLsa1YVG3ZFjGknjWamBmb3Ov2qskaqcb0HUgszu5Dt2rZ7ETYmULKjy+dLcPFiWMY7VCgJHEcrqng893N2ZknTXpd2K1evZqzzz6bAQMGIEkSr732WrsxTzzxBMOGDcNkMjFp0iTWrFmT0hzNzc0ceeSRDBo0iFtvvZXCwixdbIFA0P8JKK7YqKH1JpWnxkbZJD/NTY29sqxs4QqEsap1+ySdGX2JUv1/BJXsqMu8C8rvVoSdEwv2TurKlea0rWXXgxIzsT6x2DFoNRh1yd8SHSYdNZIicJzVO9NeQqhpn7IUY+qxZy5NLgDBpqqU9pNUV2W0gzqB3VFeaG2tZdeUnLCLx/Rpe1ZXLhZn93KVeu3rtkKg/XvQHFGEnTFNt29SqO7Y8VLf60DR68LO4/Fw5JFH8vjjj3f4/PLly7nhhhu444472LhxIyeffDKzZs1iz57WD/OkSZMYP358u7/KSuVXTG5uLl988QW7du3iueeeo6amf8bBCASCg0AgVjy39SZltDhwosTcNWWgYG1fwukLxVuoRbUm5ELFYjdCqsqKsPM5FWHcIlsx67UdjinLMbcKuyTFRYfEkydycJh1SJKU9K6SJOEylgIQqK9Ifw1qJnXQWpryrh6tIsxkd2r3NG2gWflPGjFowwut7FMtdsmWPAmqYj2g61ldufEDc8iz6KkI2AhaygAZqr9sN86m1swzOZLvqpEybVqLVTsDNKWWv5JVkk8ByhKzZs1i1qxZnT7/6KOPMn/+fK666ioAHnvsMd555x0WL17M/fffD8CGDRuSmqukpIQjjjiC1atX8+Mf/7jDMYFAgECg9Qo5ncobJBQKEQplpyBn7LjZOr4gecS16Fv0xeuhCapixmBLWFeDphBH1ENLdQWhUUf20uoyT6Pbh1VS+nKGtSbCOcPQodSyW1PlJHR4BpusA97mOvIBj8ZGOBzucEyRVRd3xUYadhFN8/2h89QjAY2yHYdJl/L7zG8dAEGINu1J+z2q8yjWtqi1NKVjhEIh/LocCILGU5PSvrpgrG9rTsrrLrbqqFTrCPprd6BLYv+QSxHQAb2jx5/lE4YX8O+vqtltHM0obxWRveuJDpjcOkCWsctukMBkz8/ad4dUdDg64Cj9HgjBLpeU1e+plK5v1laRAYLBIBs2bGDhwoUJ22fMmMHHH3+c1DFqamowm804HA6cTierV6/mmmuu6XT8/fffz913391u+7vvvhuP1csWq1atyurxBckjrkXfok9dDzUGrLrRzRcrVsQ3D8TBMGDHl5+yx5V8rFZfZ3uLxHmqxS6sNfOfL/ZwJlAsNbPpyy9YEcxsEoVl/3oGAW7ZzIo257ctEZm4sGus+JKPOxnXJbLMWa5atCjtxLR+T6fzdUZNQEkEkBt2pLxvjCK3YrHb1xKmIsVjDFSTHzSuqpTmH+JXrKL7apupTGPdTp3i4vRVb2NNEvuH9ytW7OaAJu3zFMPhkwAtH7YUMQqo3LCCzxvalIkJeTlXUmI/136+mS+/zk6SjyHsYhYwIFqFHS87Xaasfk95vR3X7OuIPi3s6uvriUQilJQkNkYuKSmhuro6qWPs27eP+fPnI8sysizzi1/8giOOOKLT8bfffjs33XRT/LHT6WTw4MHMmDEDhyML7UlQlPiqVas4/fTT0ev7zw3hUERci75FMtejosHDHa9t4WdThzF1VPbjZ1duWgoyDBkxlkk/PCO+/fNdy6F5MwMcGiafcUYXRzi0WLWlFusO1WKnMTH9h+cQ2P4bjL5a8qUWzjjjgozOV/Hvr6EWQoYczujiPL72VQ2EIVdu7nJcpwTdaDcpVpBG2cGk0kLOOGNSSof4l8YEG/5ASbSGM2bNghRcuTF2brwTgMOOPpHBU5J/HaFQiM1L3wYgFyczUzgHu7+4E2QYM2ESo05I/dytqQEqISfSyBmzfghS11FdH+1+C/xgLRrM8T38bBzt9PP8/1vNGv8IrjLAIKmW0jbHbNi3Db4Cv6zn3HN/hFaT+jVJFnn375Gc+zlM2s0u19is3jdi3sNk6NPCLsaBcQ+yLCcdCzFp0iQ2bdqU9FxGoxGj0ciiRYtYtGgRkYii/PV6fdZv9AdjDkFyiGvRt+jqevx3awNrK5rIsxo4dVwKTdTTxBj1gKTUWGu7pqitDJpB56nuV+8dTygaj7ELa83o9XoihWNgby0W507QaNFrMxiurWYdh/T2Ls9jJGcINIDe3wByCAwpelTU1nBhjQkvRnIshpSvm33gGELrtZjwga8WcpKvQwcQikQpkutBgvwBw1OeXzYpFru8aDPBqITVmNwt3RpVwgmsuUVpvVdzy4YR3q9BFw2i9TeCo+vPnU7t/au15vX4szG4QM+YEjuba5Q6glLjDvQRL6jnIqDW6GuR7JQYk2+XlhZlR4JzP+M1Faz3jCUYlbBk6bOfynnr9eSJrigsLESr1bazztXW1raz4mWaBQsWsGXLFtatW5fVeQQCQc9wB5Q4rFSr76dDOBLFHFWL51pzE59UG7ibfP0rOcvpC2KTWi12AMbSsQAMYz+7G5J3ESWD7GsGIKTv2kOSk1+EU1bFXDqZsWqpE68+F5BSKnUSY3BhDhWymvRQ923K+9c2tZAvKSLLUZx814kYETWrNZXuE7IsY5OVuoTW3PSSC4YW5VCNWiMuiXNvUGP6tJbMZKlOHV1IIw4a9eq5r/oi/pxPrZnn1mTHw5aAmkBxjHEvUSS+2NfSzQ4Hhz4t7AwGA5MmTWrnt161ahUnnHBCL61KIBD0JVx+Vdg5sy/s3IEwdkmxXplsiaUiDHmKtcYWTK1YbF/H5251AYW0SgFdKZ4Zm/mesVJAuTl2V2Ot1NHDWnaqsHNrc4HUihPHGFlsY7usdB3xVX6T8v6NVcq6/RjQWFLPUA3ocgEl3rEuSWHn9vri72FbbnqhC8MKrOyNxmrZdS/sTGHlPWSwZ6Zg8MlqPbtN4XJlQ5t6dgGXKti1B0HYqSVPTrZXcsuEMFPKs9TpIkV6Xdi53W42bdoUd5fu2rWLTZs2xcuZ3HTTTfz9739nyZIlfPPNN9x4443s2bOHn//851ld16JFixg3bhyTJ0/ufrBAIOg1Yha7Olcg68WBXf5w3C2pMycKO2uRUqQ4L1yf1TUcbAIeRWhF0RKVVPFTOApQhF2mS55o1DqBMTdjZ5S1rWWXTskTtYadU6PMk0o7sRh2k55qg2Jpc+/rpG9pF7jrlPtck7Ywrfg8v5o8kSt5qG9OLgbL2dT6/jR10ou3O4YVtdayizR2X97HHFHc68YMCbspw/Ix6jSsC5YrG9oIu7BbqZnnT6OrRsqUKcLO5tzBUEsQXSZDEnpAr8fYrV+/nunTp8cfxxIX5s6dy7Jly7joootoaGjgnnvuoaqqivHjx7NixQqGDk3dbJ0KCxYsYMGCBTidTnJyDsIbRCAQpIVbtdgFI1FafCFyLdmLq3H6QxSo1g6MicVWc0tUYYeTaNCPxmDKyhqiUZkWX4g8a5bjh1RCagu1oM7SKj6KlCLFQ6UadlZntiCzLqgKlO6EXW7bIsXpW+yaiHWdSO92GMgdCQ0g121Nfd9GpeuEy1BMOtGhIa2VMDp0hHE3VALl3e7jaVGEjwczVm16r7nMYaJKUsKhvLW76KrssNym/Ig1JzN15Ux6LccOL+DL75Q4u7bCTvYory9kOAjWM8dAsBQgeRtw+PZlf74k6XV5OW3atHjGatu/ZcuWxcdce+21VFRUEAgE2LBhA1OnTu29BQsEgj5Fruc73jYsZKZmLTVZdse2tdgdKOwKCkvxy4rVp7m2B90QuuHON77imPve49WNB+dGEvapLdR01taN9jLCOgs6KYqn+ruMzmdQ3XaabornKha71FpbJRAvTqwKuzQsdgC60sMAsDu/gxQtxtEWpdSJ35xmzLgk4dYrVjB/c3LdJ7xORfj0pL2XRiPhtyku6HBDRdfzBSM4UGL60nX9dsTUUYVsjqrCrqkCvMoPDEktRxQxHQRhJ0lxd2yOrweFsjNMrws7gUAg6AmHuT7hMM0eLtCuodblz+pcLm8gnkiAMTGGR6/TUicpweEtNRVZW8PGPc1EojK3vbSZtbuy374s0kFvXCSJcL7ijtU3fkc0mjkXuDGsuHZ11q5vzKVtuk/IabliFYFTF1UEejoxdgD5g8cRlSXF3aiKxWTRutXixLb0s7n9RkUshVuSKwEWa+/l62F7LylX8ZppnXu7HNfkdGGRlB9cpgy2+DpldBFObOyWVVFctQkATQ+6aqSFmkCR4604OPMlQcrCbtmyZSkVyjtUETF2AsGhgT6kCI9CqSXrCRQ+T5usN2P7G2OT2sjdW589a1qjJwgoruef/XM9FfWerM0FgF+NeTNYEzYb1J6xg6P7qGzxZWw6s1qKQ2/L7XJcsd0Y74CQlrBTXbHVIeV1pZMVCzB8QCF7YpbDFDNjYxnUGjWjOh3CFmVuKcm2YiGP8mMgoOuZsDOXDAfA6q+CaKTTce5m5TxHkZDS6E3bGSOLbZQ6THwZTXTHGoLNAGhsWewT2xY1zi73ULbY3X777ZSWljJ//vykuz8ciohyJwLBoYFBtfAU0ZL1kiexRIIQetAZ2z3vURu5B5uyI+xkWaZBFXZDCyw0eUPMW7aOFm/2WhlJIVU4GhKFgKZIzYzV7M9cZmwkjFlWRKLR1rXFRa/V4Lcq7kBNoAXUMilJoyZPVMaEnSm9eLNRJfZ4Zqy3cktK+9rVDGpjweC05gbAprzn9L7krIURj+KqDHaTddwdRWVDCclatHIYXJ27gT0tavaxZANN5pyEkiQxdXQhX0YVgRkTdqaQ8hnVHyxhV6pY7By+vRDtuAXewSbls7xv3z6eeeYZmpqamD59OmPHjuXBBx9MuhOEQCAQZBJTRBEVxVIztc7MWY46IuRpBiCg7bgYbtCi1tVSG7tnGm8wQjAcBeAf86YwMNfMznoPP3tmfXx7pon1xtWYDojJKlQsdhkteRJozew0J5FBmZObH4+RS7mWnRpkvz/YM4udzaijRs2Mde1JPjNWlmXyIsoa7GpGdTroHIor0hRILhtb9jcDEDHmpj0nwNAiB5WyKp66OPd+1fXr1aQf09cZU0cXsVmOCTullp0logg7oz0ziRrdkj8c2WBFK4egfvvBmbMbUhZ2Wq2Wc845h1deeYW9e/fy05/+lGeffZYhQ4Zwzjnn8PrrrxONZucLRiA41Hj4na3c9tKXWS/D8X0lGpUxRxWLklEK4WrJbsxZ2KsIj6Cu45tULFYq1tg90zR6gpTQyJn69QzJNfLUlcdgM+r4dGcjd7y6OePvs0hURhdWzq/WdICFJ17LrorvalyZmVANfHfLJuyW7rOKB+SY0q9lF0ueoGfJEwC+3JEAyCm4Yp3eAEUorze/rDztuU25ynvOHm4kkkSso6RaNuUeCrthhdZ4jGNXCRTBWEyfLvN15U4aWcgWuVx50LIHPPXYZeW9aM5QBm63aDREJ1zMzsLTOrTi9wY9sosWFxdz4okncvzxx6PRaNi8eTNXXnklI0aM4IMPPsjQEnsHEWMn6CnRqMwTH3zH8vV72deUXUvS9xVvKIJNaj23yQaQp0tYjTcL66wdPq/NVYoUm/3ZKVLc4Alyj34Zi7SPIv37RsaW2Hn80qPQSPDihn0s/nBHRudz+8Px86uzHHBjzh9GVNJil3w0VGcoC9ivWFtasGJPQmiVplvLLuiBsPK6GmUHZr0Wgy7926GuRMmMtTp3Jr1Pfc1edFKUMJq4OEsHS74Sn1ckNdPg6T4UQacWgMacm/acoMQ4VmsUN7CzuvP3XViN6euuk0g65FoMDB80gB1R5fzJez+LZ63b84szPl9nRH/4IJsHXwH5ww/anF2R1ju5pqaGhx9+mMMPP5xp06bhdDp566232LVrF5WVlZx//vnMnTs302s9qIgYO0FP8YUixH5AH4x2V99H3P4wdtqIZnd2hZ2sCruovmOLnSlfEXY5odSyI5Ol0ROgXFJf4+f/gP/+jmljirn7nMMBeGjlVlZszpy10OkPYUXJAtYemCyiMxJyqC7Ehm0ZsRYG1fgvp2zBnkTMW1m6Fjs1cSKqNeLBlHYNuxi5Q5Tzbw83xK2O3eGsVtbbJOWDRpv23BqH4v5XQhG6/57Rq31bu8s67g5JkvCYlfe7v7bzIsVRr3I+wobcHs3XGVNHFfKl6o4NbfuPMqcskZOXudIqhxopC7uzzz6bwYMHs2zZMq6++mr279/P888/z2mnnQaA2Wzm5ptvZu/erlOgBYL+jifYGkhb7xbCLhu4A2HstGbpa73ZEVRxAoqbJ2roWNjZ1X6fedHGLjMF06XBHaRAatNhYM0j8OlfuPz4cuadqGQH3rh8Exv3JCcuuqPFF+q0bh+ArljpGVsa3BNP6ugJfpcq7LBiNSQj7MyttexSsdipiRMhYx4g9cgNCzB8YCn7Y/FmdduSW0KDco9s0fdMgMhW5fUX0UJdEuV+Yu299Naed4GQc9Skjy5Edcz1GzXl9ni+jpg6uojNagKFtOO/gGLxNRsPTgHvvkjKwq64uJgPP/yQr776ihtuuIH8/PZvjrKyMnbt6r7NiEDQn/EGWm/sQthlB6V3a6uwy400xluMZQMpqMaSdSByAApKBhKRJXREiLgy745tdnvJR13DsWpbxZW3wZcvcseZh3Hq2GIC4ShX/2M9ext7XpbK6Q9hjdXt60DMaosz2zM24Fbcdl7JilbTfYuthLZiqSRPqBY7v0G5f6Vbwy7GyGIb30WVzFjP/uQSKCLNSua015RmceIYVuX1G6UQzY3dJ1CYo8r7x2TvedaooVD5MWHy7u90jDZeVy63x/N1xMTBuXynV2sqtii6wyXZkNJo0dZfSFnYPfXUUxx//PFdjpEkKestv7KNiLET9BRPMMxPtG9zi2459a6eWzME7XG3tSgBRVILtc7sFSnWhmIZoh3HCxU4rNSiuLhaajJf18rXUodGkpGRYObvW8Xdaz9Hu/M//OmSoxhX5qDeHWT+0+tw+ntWBsXpa+20IRs7sFIWtgq77RkQdmHVFevXJpdBWZbbpkhx8+7kOz+ows6rV65VuhmxMaxGXbxnrDPZzFi1REjYWtqjudGb8akZp96m7rOxrWqdQHMGigU7BihJIznBGoh0/INKr5Yf0Voy0yf2QHRaDXnDJhGRW4WcR/v9bgOasrC77rrr+NOf/tRu++OPP84NN9yQiTX1CUSMnaCneAMh/k/3HL/QvU6wMXstpr7P+LxOdFJrFn6R1JzVeEadWtOtM2Gn1Ug0aJQbpisLbcXCTsUK6NPnKnFZM++H8bOV+lnLr8Ba9wVPXXkMxXYj22rc/OK5jUllSnaG0x9qTU7pyP0cE3aaSnZkQNhF1Hgsf5IZlMV2I1UUEpUlpJA3Lti6RXXFurW5QPo17BIOmaOInGiSPWMNHjVW0pF+ceL43AblPRfqpq2YP9Ta3suSgRi0soFDCMg6tETB2bHVzqgWENfbsiPsAI4dO4Tv1FqCkJ0M3EOJlIXdyy+/zIknnthu+wknnMBLL72UkUUJBP0Bv8eJXlLcsRGnqPOYDQJtO0EARWRX2Bkiyk1Rb+ncIuDUKxYkX0Pm44xltyLsgkbV2qLRwI8Ww4gfQMgDz86mLLiXJVdOxqzXsnpbHZ/tbEh7PqevNXmCDi12igtsgNTIvuqeu54jPrUAtD65rgh6rYYcm40a1UqadAKFKgBbJLXUSQ8tdgA6tROHtSW53rmWgHK+9GomdU8ImpX3XNTV9feM0+2Nt8SzOXou7MoL7eyXleMEOyl5YokoMX3GDLYTO5CTRxW21rMDgvrcrM11KJCysGtoaCAnp/2XmsPhoL4+yV9LAsH3gKC3Nchddmc5qP97SsidmCSQTVdsJCpjUmvm6a2dCzufGjMVaek87ihdJNXSFDG3uSnrDHDhP2HA0eBrhGfOZ7zNzeRhioWksiX98+H0h7uMscOcR0hdS7g2uaSBLlGFXTiF0hht3bFJCzu1T2xzBmrYxcgZMgGA3FANBLq3XuaGle8Eaw+KE8eIqgkUmm561bqaW+/RUgZi3vKtBqo0yvu9YV/74ryyLGOVlXNhyYCQ7IzB+RYqzWPjj8NZStQ4VEhZ2I0cOZKVK1e22/72228zfHjfqOEiEPQFQt5Wa5LkS99qIuicsO8Ai53UTF2WLHZuf2u8mbELYRdSY6akLtospYsx1l3AdkCNLqMN5rwIBaOgZS88cz6DTYoga+hB4o7TF4pnHcuGjq1okuqOzfVW4OphTJ9GrbEWNaYg7Bxp1LKLFydWrmNPkycAygcPok5W3xf1XYvcUDhCkawkiuSU9FzYae3Ke87g71rYxdt7YelRiZUYkiThMik15Dw17WvZuQJhclCEnS03u+VHLMOOif8/asqe2/dQIGVhd9NNN3Hrrbdy11138eGHH/Lhhx9y5513snDhQm688cZsrFEgOCSJ+Fotdga/EHbZIKqeY5daMiIfF/VOT1bmahtvpjd3LuwkNWbK6M28+90UUMSAtqN2SdZCuPwVsA+Aum/5+f7/w0SAxh6UIXH6gq2u2E5KvOiK1dZimkp21PXs3OuCyvWUTckHv5fmmNgXK3mSoiu2NqK8pp7WsYPEzFjXvq4TKBoaajBLynXJK+l5oqEhVxF2lmDX3zM+Z+bbe4XsSsmTSGP7c9/iCZKjxvQZM5CF2xUjJxxHWFYkjZSlRI1DhZSF3bx583jkkUd46qmnmD59OtOnT+eZZ55h8eLFXH311dlYY68gsmIFPaWtsLNFmvGHMl/X7PtOrO+l0zyIqKRFI8n4W2qyMpc7EO6yplsMfZ5yc7cGMlvuJBCO4IgormdjbieZlLlD4LKXwZTDYM9mFun/RIMr/a4nPq+3NTmloxg7SMyM7WFrMZ1aPFeTgptwQK6JfWqcV9IWO9WlXR1ShV0GXLEWg44qo2J96y4ztqlKWWczdjQGc8/nVrtP5Eeb8HRR7idWTsanTS6GMRk0eeUA6F372j3X0tLY+v4x96wgcndMGT2QbSjnX3OgRft7RlqdJ6655hr27dtHTU0NTqeTnTt3csUVV2R6bb2KyIoV9BTZ33qTK5BaRC27bKAWDA4bHIRMikUg6syOsHP5w9hjGaJduAotBcrNJTdcl3z5jSRo8oTixYmNOV3UPisZB5f+i7DGxKnajRQ1rk97zgRXt6HjNmoUKcJupLSf7+p6lhlrDCvXU2vJTXqf0rZFipO22CmWq8qQBchM8gSA16Ekk0Rru86M9dQpGdNNusz0MzXltbYV6yoUIaKWkwlksL2XvVQJwXL425da8bQo5zmIHvQ9F7BdYTHoWDnwOp4Kz8JxxKysztXX6VGv2KKiImy2zJl0BYL+hBxsFXaFOKl3fw9q2UWj3Y/JIBr1HMsGe7wCv+TJlrDrugtDDIcaM2UiEO99mgkaPAEKJeV4Gls3RW2HHIezcCIAuh6cj7CvTW9cqZPbhWqxK5eq2VXTs9driijCUJeCsCvLMVERVc9HUwXs/7zrHYJeJYMY2BNQ7l+ZiLED0KqdOCzdZMYGGhXrltuQoUb16vuhSGrpMis8Em/vlbk6bwWDFDGbH62HcOJ3nK9FiflzazJnIeyKn86dy8m/fJIJ5en33u0PpCzsampquPzyyxkwYAA6nQ6tVpvwJxAIFKRAW4udk/r+3i+2ZR/8vxHwzh0HbUptTNgZHWjVnpm2UENW3N6ubtprxSjOz6VJVgRDqLm9eypd2lrsYt0GukKyKK4vqQfiUlazO+XOrHUAjkFEtCYMUqTDAPqkCQcwyMpnxJBCzbOyHBPVFPB29FiQo/DST7oW1KobFq2B6oASW5cJVyyAQ+0ZmxeshFDn2ciyU7FuBSw9LE4cQxV2+biob+nCaqr2sY0YMyfsBg0aik82oEHGU59oMQ2qrt9k6xL2FJtRx+iSgyMi+zIpR4xeeeWV7Nmzh9/85jeUlZV9r9t2CARdoQm2fsEWSE6+6u+u2MqNSrmNb9+CmfcdlCn1qutOY85BZ1B+WCo9MwMMzrdkdC6v19Um3qzzm0e+xcA2OY88yY2zdg8FZeMzMn+D289kVMFi617Y6VVhpw02pz9pwAWSYhHt9JteoyGaPxJt3VeYnTvwhyKY9Gn8yPcrojUqS5hsuUnvVuIwIUlwW/AqZhRXoW2qgDevh9lLoaP7k5o4IVsK8dYp1zMTyRMAQ4cMo1m2kit5oOE7KO342ms9Ssa0bMuQZclSQAQNWimKs6EK6DjTVlKzjslgOZAci4GdUjHD2Uftnm0MKx0Vfy6sCrvA97xg8MEm5Xfz//73P9asWcPEiROzsByBoP8Qaz8FkI+T+iQadB/SxCyULfsVl6ymR5EeSWEIqy2+zDlIFsWqpHSf8Gdc2MWKIUeR0HRhwdJoJBp1RRDdi6duD5nKBXS1NGKU1MD4JCx2BjUL0Rp14w2GsRhS+7qPRGU0ITcYQOpCyALoSsZC3VcMp5Jd9R4OK0vjRq4mwrgxYzcbk95Nr9VQZDNS64KK6X9mxJsXwNevwrBT4JiftN9BrWEXMbdaBe0ZstiNKLbztTyQY6RtOPd9haMTYWf2Ke5xbe7ADp9PGY0Gjz4fR6ieQBdtxfSqsItZczNFs6EUgvtwViW6oCNeRdhl0vUr6J6Uv3kHDx6MnMGAYIGgv6ILt5Z+MEphXM6mLkb3A2LCLhoC98HptGFUO0HoLDlt4oyaqXVm3joa9CgWpaDW2rElqA2x2KlgU+ZcscEW5Zz6NZakAtH1VuXmnYOHhjTiO5W6fcqPEa2pa2EXq2U3Uqrku3Rbi6nu0xbZij3FFl9lOSYAdhjGwql3KRtXLoSar9sPVi12IaMi7OxGHVpNZjxPZoOWKr1SvqSli8xYe0iJPTMV9LzrRAy/US0U3UXykCGsvId11swKO59VeR3B+oqE7ZKafBP9nhcMPtikLOwee+wxFi5cSEVFRRaW03cQ5U4EPUUfTrzBBbJUhqPPEGgt70JL5gRNV5jVThAGS268aG93AeTpEmt3FdR2EW+mEjArsVNyS/dN2ZMl1ifWb0gu/izWWSBH8tCQRi07pz+EVc0C7qw3bhy1tdgITU+EXbMyL5aUhV2pKuyqnX44/hcwagaE/fDilRA8oLaeWpw4dh4zlREbw5szAoBIzbedjimMKOLSXtTzGnYxwmpbMdydf8+Y1NAFgzXDdd5y1TIjLYn9kTUB9cdsBrpcCJInZWF30UUX8cEHHzBixAjsdjv5+fkJf/0FUe5E0FNi1qQYUVdm65r1Odq2UWre0/m4DBGKRLHISlcEo63VYleM4orNNLG6hGF995UAIjZF2GkzablUBUnQlGQF/7bCLo34zpbu+sS2pU0tu+9q06tlF/E2A+CUrSm7RstyFAtmVYtf7Z/7F7CXKR0gVvwqcbCaPOFV+4mmKiK7Q1IzY83N7VtsAbhcLeRIyndDwYDyzE2svv913s67T1iiyrUxZbhvq6lIKXli9SW20YsVnNZk2PUr6JqU39GPPfZYFpYhEPQ/TFFv4oZu+jge8rTJAqZlb9an8wTC8XZXJlte1l2xsmqRjCYh7LS5g2AvmPyZE3Ya9Yad0Ce2K1T3lwMPu9O02MXr9nXSdSJOwUhkJHIlD/U1+4FJKc8XcDdhIT2LXcwVW9WsrtdaABf8HZ4+GzY9C8OmwpEXK8+pNexcmlwg8xY7x+DD4VsoCOyDSAi0icdvrKrADngxYrFnzhiiy1F+TJiCHfdsj0RlbLIbJLBkWNjlDlCslPmhRGuhMaRYufWZthAKuiRlYTd37txsrEMg6HeYVGtSRGNAGw2i9zf28oqyTIKwy74r1h1oLRiss+TGhZ1VCtDS0pzx+WLla+RuEgkADGr3CXswc2I+1pZOSiIjFjjAYpeGsPOFscbLu3Qj7PQmwo4h6J270TVuJxyJotOm5hAKuhuxAB7Jij7FfWOu2KqWNpba8pNg2u3w/n3w1k0wcJLiMlYtdi2azPWJbcugoaPxyEasUgC5cSdS0ZiE5521yo+eBk0hlgxWlTCrRYodoQYiUbld3KDTF4q397LmZqh+nkrpUOU1FtNIi9NFjkP5jJgiymcm2+3EBImklba2Y8cOfv3rX3PJJZdQW6u4l1auXMnXX3cQqCoQfA+RZRmLrNwUwzlKHI0l1EgwfHAL+B5U2gq75uxb7NxtLHYYHWC0EdYpmbBhZ+aTN+Lla5IQdrZi5Zrbo04Ipd/Sqy3mkPLDQOfopjhxDNVil4OHRk/qFkynv60rtvvXHOsZO1Tez96m1F9z2NMMgE+Xeh2yAbmKK7baeYAL/uSbFWtdyKPE24X88eSJJpS4wUzVsIsxssTODlkR9s697RMofA3KZ8Opz6y4irUVK5Saaejgeje7PVglZbs+w8kT1txivCjiunKP4oKORmVsquvX7EjSyizICCkLuw8//JAJEybw2Wef8corr+B2K192X375JXfddVfGFygQHIoEwlGsqujQ5A8DIF9ydfiF229oU7fvYFjsPG5Pa/kPNbg/YlFulhpP5oWdNqTWzOsukQAoKCjCJxuUB66qHs8dicrYwmqf2K7aibVF7c1plQI0ubzdDG6P0xfCFnfFdi+2pKK2cXapJ1BE1OK5wTSEXasr1p/440mjhfOfBEsh1HwF794RD4mol1Vhl6EadjFMei2VBkXYN+9uL+zCzUocms+U5HVMkliB7lgdxwPxtLRx0ZoyXH5EkqjXKfM37VeEncsfjscSWnKFxe5gkrKwW7hwIffeey+rVq3CYDDEt0+fPp1PPvkko4sTCA5VvMEIdtWNpS1U4k8KpRbqXf24rVhCVmz2LXY+1cIDxGPANHbl5mL01xOOZNY6qlfL12jN3Qu7EoeZKlmJKwo27e9mdPe0+ELxdmLmvCSL2ra5eftdDSnP6fSH23TaSKJ1ZCwzNk1hJ6tZx+E0+pgOyDGTZ9ETjET5qvKArhP2Ujj/r8r/1/09nthTG1FeU6YtdgAeh/KZD9d80+45jUvJlA7bMtR1IkY8K7y5w6xwryrs3JJVEbwZxmNWLIae2l0ANHmDcdevwSYsdgeTlIXd5s2bOe+889ptLyoqoqEh9S8PgaA/4vH6MEkhADQFypd8AU7q+3H3iVjWKKCIPF9zVucLqQ3NvZIlfqPSqVaLQloy2ps3GpUxqKVV9JburR25Fj21KMLOVZtkY/ouaPQEKECtQWYvTm4njTaewRvypB7f6fSFsEqqa7O75AmAQsUVm66wi3VFiBhSF3YajcTkcuV8r9vVwWsdeRqcdKP6QKnDWh1WhV2GY+yAeFydubl9z1iDV0kwkBwDMjtpmxjTpqb2NTP9LuW8eDXZ6e8edgxW/tOkvN+bPd7W5BtRx+6gkrKwy83NpaqqvWth48aNDByYoSraAsEhjt/TxmqQp7hiCyRnhy6S/kK4rbCDrLtjg27lHPs1rXXlJHubzNgMljxxB1utVwZr98JOkiSa1RgqX0PPz0ODOxi32MUsM8kQNeYq/3pSL47t9IfiBYqTibGLlTwZKNWzp7bjzMyu0KrCTk7C1d0RU4Ypwm5tR8IOYPodMPhY5f8aPTUBpbtFppMnAOyDlI4Thf7dEE3sW2wLKnHphrzMFScGwGjDr1FiDb0N7a3EQVesb2t2eqnqC5XvOaNbeb97mrPo+hV0ScrC7tJLL+W2226juroaSZKIRqN89NFH3HLLLVxxxRXZWKNAcMgRUN2EAQyKKwhV2PVji502pFi0mmTVIpBld2xEtQgG2hYMjrmjaMloyRNXG7ekzpzcTcpnVNYSzkD3iWanC0fM+mFNwa2lWkpkf3PKHYOcvjC2eHJKEmLAWkDElIdGkonWfZfyfNqgmnWcpnUnJuzWVTQSjXYwt1YPFzwFeeUwZhbOgBKf6chwHTuAAcPGEpD1GAgiH1DTMTesCB5bUcf9XHuC16DEsgVb2htfwqq4D+qzI7IcpUotu5xgFbIs42tRPHgeyQrazJ9jQeekLOzuu+8+hgwZwsCBA3G73YwbN46pU6dywgkn8Otf/zoba+wVROcJQU+ItZ/ySpZ4X898XDSkEcR+SBAJoYsq1p1tsmqJyHJmbLwTRFsLRKxIsdSU0e4TrjY13brrmxojaFVjqDKQPOFpUpJBwuhScmtprcpYS8SNJxjpevABKJ0nUnDF0tpabEBoT/sM1W4whJTPjDbNLgXjyhxYDVqc/jBbazopkpw7GH65ES76Jy0+JVQiG67YESU57JSVWMi2rcXCoSAFsiKw8krLMz5v0KR810Q7aCsmq8kp6bi6k6FgoBJjOUCuocETJKDGdfq02bEQCjonZWGn1+t59tln2bZtG//617945pln+Pbbb/nnP/+JVpv5gMzeQnSeEPSEkCo6fBoLWJRf0RpJxteSuovqkKBNqZNvo2qsTbYTKPwddIJQA9KVtmKZc8W6EhIJkrtRyXYlhsrg7bmwC6nt6Dy6vG771LZFq1b8T6f7hNPX1hWbnLDTqCVPRmgq2dOQwo8YWcYYUeLyNJbcVJYZR6fVcPRQ5fV26o4FpTMFikUSspM80TYztmn35vj2xpp9aCWZkKwlrzjzoUtRq2IlljwdtBVTW7bF3POZxlikuGKLJCd7qusJupVrEEgjGUbQM9KqYwcwYsQIZs+ezYUXXsioUaMyuSaB4JAnlkgQ0FhAqyNgyAVa+332O1Rh55f17JZVS1W2hZ06Z6StNalNZmBNRl2xoZSFnS5XsVxa/D2/5mGXcqP2GVOs4N+mll2q/WJ9Ph9GNQEo2dfctrVYSrXswn50sjKXoQc11o7tLs6uDU6/Ml+OJfPCDsBtV1yTbTNjW2qUxIJ6KS8rhhCNarE2+Nv/gIzFMGatb6spV/FQALV7txNVLYRhg4ivO9ik7PieN29el88vWbIk7cUIBP2FmLCLNYyPmAoh2IzcX9uKqTXs3JjZL6s1q7LsipXUPpRyW9eSemMrwEm909PRbmnh8odba7olKXIshYqws4cblAD6HpSYkNT3TciUYj2wHnSfiPhdEDMOGuzQUdzagajCbqRUycrGFCx2arxkWNZgTiI5pTNimbFrKxqRZRmpE+umPxSJ17vLRowdgFQ0FprA2CYz1lOvxNu16IpIsmhNShjU7hPWDtqK6YOKsMta31ZJosU4AIv/O9w1O8Crun6zZCEUdE7K7+gD06hDoRBfffUVzc3N/OAHP8jYwgSCQxlZdROGdGpgv7UQnN+h9fVvV6xbNrNfVoL75ZZ9JO80TJ14sL2xjbCzFiIjoZOi+J2ZO9cufzhelzBZYecoGEhY1qCTouCuBUf6t3Kt2gYrakmxW0Ga3SciUVkR60aQdSYkrQ6ioe53VGvZDZOq2NfYSZxbR/gV0eHEgr0HMW9HDs7FoNVQ5wpQ0eBlWKG1w3Exa51GAqshO8LONmgcbINC3y6QZZAkgmoijduYfGZzKsS6T+RFm/AEwliNra/NEFbL5WSxb2vANgj83xFu2I0h2AyknwwjSJ+U39Gvvvpqu23RaJRrr72W4cOHZ2RRAsGhjhxQLFix+C+tvRiqwBRsIhSJptwLs88TE3a0CjvJXQ3hAOiMWZlSH1JjstpmqWr1hE356P0NyBlsK+bxpu6WLMm1UEsuA2gEZ2WPhJ0hoPaJtaYo7NpY7L5LwWLn9rfpE5tk4gQAuUOJaAyYokECdbuBScntFxN2shV7D2LeTHotEwfnsraikXW7GjsXdmrihN2kR6PJzs+PsuHjCf9HgwUfsnM/Us4g5BalOHHIktmuEzGMua0xpnWuQIKwM4djfVuzJ+w0eUOhHnTOvegkJVknaxZCQadk5O6i0Wi48cYb+cMf/pCJwwkEhzySak2KqsJO71B+oRdILTSmGOt0SKB2nXBjpgl7azstZ8+7LnRGLNhee2D5EdUdq/fVdVz2Ig0CbesSJtFeC6DYYaI63n2iZyVPLKn2iY0Rs9hJnpTed237xCabBQyARksgR/mBb2renvx+bSx2th66RicPU4TEZ13E2bWoiRPZqGEXY3hpHhUoQqt5j9JHXae2uotmujiximRrrePYtrSSLMtYZeXzYnZkr72XpUS59nb//niWsy7DfWkF3ZMxs8GOHTsIh8OZOpxAcEgTaxgfE3Yam2JpKaSfFilWLZRu2cSgPEvcapfNOLuYsNMd0OIr1jMzX26m0ZsZER3yKsIjpDElXZPLYdJRJynCzl23p5vRnSPLMna1T6wpL8U2VDGLHallxbb4Wsu7JJsRG0NTMg6AQv9OAuEkS6zELXYW7D0UdlOGKcJlbUXnnZBirthM94lti1GnpUqv1KprrPgSAItfSYLR52apmL+9tfNKbUtrjKM3GMGhtvey5mSvvVfugJEAlMp16FVXrNEu+sQebFJ+V990000Jj2VZpqqqin//+9/MnTs3YwsTCA5ltKqbMO62U4vKFkj9tK1YG1fslPJ8Kr8qYCSVWc2MNUWVG5fBlpuwXdO2+4QzQKGt567gWM28kM5GsjYeSZJw6oshDIHG9C127kCYPJT5rcn2iY1hUqwlDim1rFjFYhdzxaZWh8xYNg6+fYVR0j4qm/2dukPbIvubkQAnVkb3UNhNGpqHRoK9jT6qWnyU5ZjbjYm5YrNR6qQtLtsIaP6UULWSGesIKUkw5oLMFycGwFJIFAmtJONqqAaUBJ5mX4g8VdgZbdlzxerylRIvg6VaQqq8EMLu4JPyJ2jjxo0JjzUaDUVFRTzyyCPdZswKBN8XdGrD+FZhpxYplpzszmAP0z5Dm+SJycPy2bdZtQpkqa2YLMtYZA9IYDzQ1RMveaLUshtHz+toRf3K60uomZcEfnMJuCDSkr5LutHT2k4sFkOVNAkWuxSEnS/cWpw4FVcsIBUfBsBoaR97GztPYGhL0N2IEWiRrT0WWzajjsMH5LB5fwtrdzVy7sT21rGDJezkojHQDMam70CWKYw2gAT24iwJO60Ory4XW7gJX1NlfHOLy8VASflBKWUz5i1XeV35khs/SsKPNTfFuFBBj0lZ2L3//vvZWIdA0K8whGOB/aqoUIVdAU429GuLnYUfDM3jDdQ+qXUVtLeXZGC6cBS72u7KeIDFLlakOJPdJ2JZztEUhV3YWgYu0LjTL1Lc4PIxCDXDNNXkCTXGzib5aXEnX37EmVC3L8Wm8UVjAaXkySuNbqD7NYc8TRhR2k8ZdT2PEJoyLL9rYefPfowdgHXQ4bAdCny78DTXYVUTcArLhmZtTr+xEFu4iXCb7hMetb1XFAmNMYt15Uw5+LR2zBEXAyQ1LjSLWbiCjulnqXmd4/V6GTp0KLfccktvL0XwPcAQUSx2WlOixa5QclLfD2PsonGLnYkim5GgVQkO99dXZGW+tuVHzAdm+bWx2GUqnlGKl1ZJzXqlyVHOg8mbfoauu6kWraQmgVhSjI9q03w97G1Kun+r0nUijaxYgPxhhCQDZimIs2pnUrtEvM0A+HW2TmvPpUKsnt26io4TKOIWuyzG2AGUDZ9AVJZwyE5adq0HoEF2YLN2b8VMl5BZFdLuVmEX69vqlazxzhvZwmsZlLhBlDs56KT8rj7qqKOS/uB9/vnnKS8oW9x3330ce+yxvb0MwfcEoxr/FW8Yr8bYOSQvTc4U6nsdIoS9LRhQYuysRh2moqGwDzTO7LhiPV4fRaprSdNJVmwRzdSm2K+0M2LJMClliAKGfOUmZwvWxWuZpYpX7RPr0jiwp9pMXatDNtiQgm4sUTeuQDgp96PTn74rFo0Wp7WcAvc2qPsGOLXbXaJqgeKQLjPtpyaXK+7GbTVuGj1B8q2GhOfjyRNZdsWWlxWylyKGUotryyoAGrWFZDXqzFailBzxtnY8CajtvXxaOynK9JSJ5g4BV2u3DcwiK/Zgk7J0/+EPf8iOHTswGo1MmzaNadOmYTKZ2LFjBzNmzODcc8+N//UVtm/fzrfffssZZ5zR20sRfE8wq8JOH6uib8olKik35ZCz/3WfiKquSr/GgkGnIb9MyY6z+qshGs34fD53c+uDA4VHvORDS8ZcsboDXetJYitU+uYa5AD4mroZ3TGhFkXYeXTpubQk9caaSpxdgsUuVWEHBPKVDhTW5m1JjZfV5JRwiokanVFgMzKqWJEwHVnt4n1is+yKNeq0VOmUuDPr3jUAuAzZjTnT5SihCKZAa4HuWN9Wvy4z57crTIXl8f+H0YIhe9ZJQcekLOzq6uq47rrr+OSTT3j00Ud59NFH+fjjj7nhhhuora3lrrvuiv8lw+rVqzn77LMZMGAAkiTx2muvtRvzxBNPMGzYMEwmE5MmTWLNmjUprfmWW27h/vvvT2kfgaAnWNT4L4NFFXaSRMik3Jij7v7XLzaqljuJqDFog4cOJyJLSv9PT+Zfr8/VDIAfA2gPuDmrrliH5KXZ6ezxXLIsow93XFqlOwpyHTTI6s3UlV6cXcSt/BDwp9onNkZCLbvkhK7TH2ptoZaqKxbQqQkU+b5dSY3XqH1MIxmM/5qs9o1d10E9u5aD5IoFcNpHADAooNT185mzU5w4hknNnLaFGpQOIkDEq5yDkD4zFtGusJaMiP/fq3WkZaUW9IyUhd2LL77IFVdc0W77ZZddxssvv5zyAjweD0ceeSSPP/54h88vX76cG264gTvuuIONGzdy8sknM2vWLPbsaa0LNWnSJMaPH9/ur7Kyktdff53Ro0czevTolNcmEKSFLGOVlZui0dZ6o4qq8VGStx+2FVMLFEdVETB2UAHVqMV5G3dnfjqPYv3yajqwBphyiGqVEieRDHSf8AQjceuV3pKa8ChxmKhRixTjrOx6cGeoPwRCpjTrj7XJjK1P2mIXjhcoTjl5ArAPmQDA0Mge3IHu65tq1b6/mYzHOnZYa9/YA4m5YrOdPAEQLRyT+NiajS6xrVjUfrFFUgsNMSGvurozKZw7Q5PXmhgSOAgWQkF7Uv65Yjab+d///seoUaMStv/vf//DZDKlvIBZs2Yxa9asTp9/9NFHmT9/PldddRUAjz32GO+88w6LFy+OW+E2bNjQ6f6ffvopL7zwAi+++CJut5tQKITD4eDOO+/scHwgECAQaP1V61R/8YdCIUKhJHolpkHsuNk6viB5MnEtQn4PFklxP+qNtvixJEshNIA+0Ig/EESbpVZGvYEUaO20EQqFKLJo+UoqYiANVO7aysABR6d13M6uRyxmyK+xdnytLMVoXHuR3LUEg8EeBeQ3uvxxYSe1uZ7JkG/W8rmczzh2462rQF+e+vtK51MsdhFzQVrvS63RgQalll2d05fUMVp8QWyqsAtrzchtvv+S2V9XooiZkVIl39U0M7asa0GhU4WdbLBn7HvwqEGKderrSidNbh+2Nu21WrzKHBadlPXvXUvZWPiu9bFsL8vInJ1dD0mN5y2imaomD3kmbVzYRY052b/P2AbEaz0G9Qdhvj7AwbiHp3LslIXdDTfcwDXXXMOGDRs47rjjAEU8LVmypFOxlC7BYJANGzawcOHChO0zZszg448/TuoY999/f1wALlu2jK+++qrLdd5///3cfffd7ba/++67WCyWFFafOqtWrcrq8QXJ05NrEfW1cB4QlSU+XPMROrUv7FHuMEOAfJy8+MbbOAxdHuaQYobaOaDFH2XFihUA2DT5EIVvPv+IL1p6Zik48Ho0VmxmGuCKGlmnzteWkyIGCoDcaBMvv/k2lh543Kq8MFR1S369fQ8Vze3n6wq3Giq/dcOH7KlLvfm7zqVYHaucIbZ28Fq7Y2Kdi6EoFrtPPt+MvfbLbvfZX6vFqr7mdV9upXZ367xJfTbkKD9Ej1kKsnrlq+ws6eJ1yzJnqX1MKxvd8fdPJsg3amkMwN9eXsXY3NaM4HqnFpDYtO4Tar7O2HQdUufyMbXN4z31Xpoy+BoPvB42fxWnohTofvG/H1GRJxNuUTJka51Bvs3g3B2hjfg5S/1/IKLJ6PXs62TzHu71Jl+uKOWvu4ULFzJ8+HD++Mc/8txzzwFw2GGHsWzZMi688MJUD9cl9fX1RCIRSkoSYxJKSkqors5cg++23H777QndNZxOJ4MHD2bGjBk4HNmJTwiFQqxatYrTTz8dvT77rgFB52TiWtTt3gLfggcT55x9Vny7ZtXHsPZjCiUnE487mbGl/cRNIctoNylW7tziwZxxxukAfLJ/FdR+TL4hxNFpJi51dj0+eHEXNAHm3A6TorSeF2DbDoqkZiYeN5WRxennAn6+p5nglocAOHzScYwbn9pr+dvXH0AQBlnDjE/jPKz94o8gw9CxExl8aur7a977FD5bTY7koXDQMM44Y2y3+zy4ZTW2sCLsJp84DXnI8Sl/Nqq3DGFwcAcjCrSc2tXrDrrRblIs3OWjx3PGzAnJvbAk+MC3mVc3VUHxSM44TfEyybLMTZ+9B8j8//bePEyO6jz7vqt6X2eVZrSMRhJCAiEQIAkbxexIIBm8ZDH25wVe48REIgGTxAGTvARiLBwneAnCDi/Bym7jGGzsCIMwezC2EBLISAiElpHQjGbtfa863x/nVHX3zPRe1d3qeX7XpUsz3TV1qru6u+5+lvv58PrL0euvPNNUCamMiqH77kCv8HU7d+0VWLZiVc37LXg+EiFg/1/CL8WxeOnp2LhmCZ7/7TZABeYsXIYP1qGJkB28A1JsDAtPW4q+GdC0WI9reKiCeuGqvsd+4hOfMFzEFWNyGoUxVlVq5YYbbii5jcPhgMMxdQSRzWYzXXTVYw2iPGo5F0qCe9jFJBd8ufsQo666EEQgobTOuU7HAcbrqGxuv/64PLMXAsOANfx+zY918vmQ0jzCo9h80+9bzIudJQUxEa/tuY5lGDpE9Mrq7gAq3FfU2w+MA7bg4aqOw68EAAnwdM2r7nEIg9g2RPHbWKasfYQTGXiF3YnV3Z73mMt9b4R8pwFj78E6dqD49jH+fkkxC7xev6Hviw8s7sbjewaxayCo7zeazOhNBd0+N2w2i2HrTYfNBuy3LkCvwoXdrHmnGfoYp5wPayfSkh02lkIiMAyb7Uw4FS4KHP7u+nzutPcDsTHI7k7IrfI5VwZmXsMr2W9VToWBQAAPP/wwvvKVr2B8nL9YX3/9dbz/fvVjc6aju7sbFotlSnRueHh4ShTPaLZu3Yrly5djzZo1pq5DtB4pMTA+Jk1K3bfqvNhk1pfP7s5GITvnccsTb3KwbGPcctFq+hRbgainmD4xCxM4Ga7Nyy6cyNRk/ZFuWwgAcEcGuJddBSTSSnZObOfcitcGMKkrtnTzhKIyhJOZ7KzYKponACDVyevsPMGDxTfU0vjwwOcytj7hAtFAsedYAIm0AiDbOGGzSHDa6uPRH/ItBgCEmQtdXSbPTpUkRO18jVSAd2K7Fd7V7Zxs5m0WYrQYedg1hopf1W+++SaWLl2Kr3/96/jGN76BQCAAAHj88cdxxx13GHpwdrsdq1atmpK33rFjB9auXWvoWpPZvHkz9u3bh507d5q6DtF6pGNZT7c8tLFiUgij4RaaFytEVpi54HZmL8xz+3nqq5eN4GTIWCEraV24jgLlEbnzYmtcO5xr/VGFsPP2nAaFSbArMb3DtVwmokl0a8Kuo8ovszk+duV8oYgkMpChwiMMoFHoOS6BrZdbnnTHS0yfEMIuxDzwOo21H1nU7UG314FURsWbx8U6moed02bIlItyULu4yB2VOvWaWzNJOflnjRo+iVRGhZdxYefyV9lZXSnnfxaYtwo489r6rEfkUfEr7LbbbsMNN9yAd999N68LdsOGDXjxxRcrPoBIJII9e/Zgz549AIDDhw9jz549up3JbbfdhocffhiPPPII9u/fjy996UsYGBjATTfdVPFaxMxGS7+Yvo4wW01MtuLIFXYtGLGLwAVfTueho4vbHrRJMbwzYGw03yJSsXAWEnaaSXGgZpPiWiN2fbM7cIKJC+p4eSO2NCYmxuEU80Ulb+WNFwAqjtiFEums1QlQlY8dAPiF5UmfcgxMKWJ5IoRdGG74DBZ2kiThgkVc2GpGxVkPu/qlCG1Lr8AE82K329yAhIbi5q8VOXoSwXgabRJPd7v9JkcLNZZcCfzhs0DvivqsR+RRsbDbuXMnvvjFL065fd68eVU1NLz22ms477zzcN555wHgQu68887TO1evu+46fOtb38I999yDc889Fy+++CK2b9+O/n7zhigDlIptNf76J7/FmnufwXCNablyUMQUhpSlQCoWIYzU4TjqhhB2UebMs5SAw4uozIXX+0ffNXRJW1pMgnAW6LY1cPpEJJ7Q682qiV71d3lwmPHUMMbfq2ztcZ5Ki8NZvYO/8LHzC2FXKi0ejKezaVjZClin1hyXw+wFZyDBbHBKaQQGi6Rj9YidG36DhR0AXCDmxv5aGBWHGiDs1q5ehX+76Fks+uQ36rKeLOp5bfERBOMptIELO9ndXpf1icZSsbBzOp3TdmccOHAAs2ZVPirl0ksvBWNsyr9t27bp22zatAlHjhxBMpnErl27cPHFFxfeoUFQKra1ePXQGMajKewZCJi+lprgQidjnRTpEBE7l5RCOFz7RITCB6ACP/5D4F8/Brz2CBAdM28tIC9i5500ezPm5nVhgROVCZpS2EXNkKVgxE6kYhHAcDBe01rJWM5s3yoidgu73DjK+IU2M1LZ85AQc2JDlhpqlbSIHaLIqExPRRYilEjnz4mtMl3pdNhxVOazcieO7C28ofBYC8EDnwmzWy9YxKNUrx+dQEZRc+bEmj91QsNmkfGnVy7DeQvqU3Nma+cmyO7UGIKhCFySiNRSzduMoGJh99GPfhT33HNP1nRVkjAwMIDbb78dv/d7v2f4ARKEkYzUIwUqhE56srCze6BYXAAAJWTiWLGxg8DeR4FDzwE//xLw96cD//Zx4PV/0+eVqirDi++M4LZH9+A/fz1QYoclyKmxy4vYAZDa+IU9NWbs9AmHwiMQlkKTIISws0sKkuHaJn1kRM2kItmqil51euwYtHKBGz9Z3uxUjZR4ncRsNVyQRcTOJ8VhgYLREmPFQvGc1HONs1uHHTyzkjxR2CyOJQIAgCDzGJ6KBYBlvT74nFZEkhnsHww3JGJXb9ydXNi1qxMYHeMediqkqusliVOLioXd3//932NkZASzZ89GPB7HJZdcgiVLlsDn8+Hee+814xgbAqViW5MRg4bCF0OfwjBNbZLiErVW0RHzDkCktmD3Ab3nAEwB3nsWeOJmsG+cjqP/+GF87b67sPmR5/HY6+/j/h2ViY0ppHIidpOEnXv2QgCAI3pC70o0ApfKhZ3d2z79BlYHFAe/j9U4m1eJi+kzk4V6mUiShIRvET+WscoidizML8oJRw21UTnpaj9K19nlzYmtsiNW35ePd0ZbRt8uuI0SC/Bt4TYlYmeRJazR07FjCCWyzROtikNE7GZJAQwO8nR+TPYCcn26gInGUvHXI7/fj5dffhnPPvssXn/9daiqivPPPx9XXnmlGcfXMDZv3ozNmzcjFAqhrc38+XpEfaiHsJNF/Zdqm+ai6OkCIsdgTYxCVRlkM8aKacKuYyFw00tgo+/i+Mv/BXnf45iXOoT+sZfxV3gZf+Gw4QX1HPxz4lowdkX1HYJajR1cmDMp4uLq5hGbudIoDgyFsbKvvcoHlY9HjQIS4PAUiWT5eoBkAJ70GGKpDNz26qJBqng+lSqbCABA7loMhAFn+Ci3PCn3uRZzhTPOGoSdxcYbIFIRtElRjJWIWofiaX2cWLWNExqZrmXACOAJFRa06egErOCpWLdJnnJrFnbi2beHsfPIOOZ38NrXesyJbRSSj9d0zpYCGBnhX2wSFh9qO5vEqUJFn3SZTAZOpxN79uzB5Zdfjssvv9ys4yIIU6i1kL4cLELYTRftsPpnAyeBDoQQiKfR6TFhrphIbWXsPvzb/x7Gf/z6BA4OrwKwCqdJ7+ML7bvxYflX8EcPY71lF85QBxBP/0nVwqdoKlb4Wc2TRrF/MGSIsFNVBg/4eB2nr7Cwk309wOgBXmcXSmJhd22PT7VV2bwAwNt7GpTDOZYnvvKsS2xiTizzVF6/nIeznQs7RDFWMmKX62FXWyrWNucs4G1gVvIIoCqAPFW4aV3kSavXnC86yPrZ/ebwuB4V9LvqV2NXd0TzUDeCCI5zYZe0tsikG6IkFb2yrVYr+vv7oSjGpVQIop7UI2JnzfA04XRWHLKH1351IYjRSNIUYccSIUgAnh9I4e539wEA3HYLPnruXHz6Ax/Cinl/BDAGdvhFSP/6EfRIAUzE0lULO229CFxTa6Ta+wAAc6UxPDFoTMNINJkdUO/2tRfcTvJmp08Mh5NY2F2dMJNEqrkWkdM3i1ue9Ekj3PKkTGHnSPJOTouvSqsTDVc7EDouInYlhF08t3mithhP57zTeWcsUsDEEaDrtCnbMNE8kbGZV/919rw2OG0yJmJpvD7A60xbORWrNWrZJYVHia1Axk6Zp5lCxQn3v/qrv8Idd9yhT5xoVajGrjWph7CzZ7SOzWmEgD59IoxRk44lMMGjPEHVhTN6ffjbj63Ar79yBbb87jlYMU98uEsSpHnnAwAcUhrhSLjQ7kqiJPjjjUwTsUMbF3Y9mMA7J4z5zIiGA5Albtnh8LQX3lA0UMyWAjXZ3GgRWKlQB24ZLOx2V2V54snw58wmRqRVTU5nbDk1dj6Dmif6un14j/HGEWV4/7TbSHqq2zzhYbfKOK+PR3cPjfAvXq3cPAGrAzELf70uBveQVBwk7GYKFX9F/853voODBw9i7ty56O/vh8eT/y349ddfN+zgGgnV2LUmI5Fk1bOGy8Wu8DThtFYcuklx0LQO3WQkAABQ7H48ectFhR+r3YsMZFihIhYcBeZXFxXKxIOwAojCCbd9UqrNMwuqxQFZSWLi5FFDnvtEOAAASMMCm81VeMNck+Iqp08wxmBNRwFrEWuVMljY5cF21gNgLzIj75X9wetXeHTJVe3UCQ3RGdsmlZ4+EYpncJpBzRO9fidew3ychaOIHnsL/jOvmbKNnOTCruAUEYO4YFEnfnUoa/1TT7uTRhB3dMMdC+E0+QQAgJHVyYyh4lf2xz72MRMOgyDqQyqjIpTImFo47VCFsHNN84VAE3YI4Z0SKbFq0boMFbuvuIiSJEQkH9pZEPFQ9ZYgqjBkzli9U9eTJEhtfcD4QbSnhnB8Io6+Tvc0eymfeISLnQg86Cj2+DRhhwAOVBkdjacVeBg/n9ZC1ipl0OmxY9CStTwpJw6mqAwdLAhINcyJ1RAROz+iOFRGxE6fPFFjjZ1FlnDSsQhI/y9Sg9Nbnli1VHchs2mD+MCi/Dmprdw8AQBp1ywgdginS8cBALIQ90TrU7awe+SRR/DpT38ad911l5nHQxCmMxJOmPqh7hLCzu6ZLmLHU7HdUgj/a1IqVuviVO2lIyAxiw/tmSBSkRpMjEVzgTJdFzAAqX0+MH4Q88AbKGoVdikRkYzLHhSNQeTOi60yFRtOZPR6M6ur+oiSJElI+BcC4fItTyZCYXRL/LXk75pT9doA8iJ25dTY+bSIXY1dsQAQ8S8BxgDL2IGpd6oqbBn++jFbeJy3oANWWUJGjBZs6VQswL/YjAGdEi8lsBbrICdairJr7P7wD/8QwWBQ/33u3Lk4cuSIGcdEEKZiamesqsAt6pPs00V49Bo7E+fFCmHHykgdJqx8m0x0ovr1ksLepVA9VpvWQDGK/YPV1/JppKMBAFzYFcWnNU8Eqq6tDCfS8IoOXKnGVKEkGgd0y5MShMa4/1gaFlg9nSW2LkFOjV2prthwIpMTsatd2KndywAAvshh3hmbSyoMCfy5sJg87splt+Ds+Tmefq3cPAHAOqku0+6t8TVEnDKULewmzxcMh8NQVdXwA2oWqHmidTG1gSIV0X90egunYjsRwli4tlFXhbAkeWpUFhfzYqREJ6Iarb6xQda7RguIACHsNMuTWskIe4yUpYSwE6nYTimC8WCk+LYFCCUy2ehVjWlJX+9pUFiO5UkJIkLYBaW22o1lcyJ2E7EUVLWwsAzlzoqtsXkCANw9pyHBbLCqSd4Zm4voiI0zO9zu6u1kykWbGwu0uN0JAEdHfpTX6a/BC5E4pSAb6gLQrNjWxVRhJ6JXKWaB2zXNhcrNI3ZWSUU8bE5nuTXNhZalWMeoIGMX28Srj9hlffsKiID2HGE3VLuw02r6UqUmQTjbwWQelVHEBIdKCSdyxmvVKOzmC8sTANzypATJID/msKW9pnUB5EXsFJUhKMZqTUZRGcLJDLySMTV2ADC/y4eDbB7/ZWTSBAoRXeZTJ8wXWpqfndMmw2E1xwy5WXBPqst0kbCbMZQt7CRJyiuMnvw7QZwqmCnsFCE6InDBPdn6AwCsdmRE7RuLmDNWTLNbsZeR2lLEBV8WpsYVo6qwFesCBnJSsWM4OhZDJFl8CH0pmIjYZWwlRIcsQ9X8vBKjSGYq998M543Xqk3kLKrQ8iQlhF3UZsAFWUTsOiz8XI0VmBcbEeO2smK29lRsX4cL72rCbrLliSbsTJoTO5kPLu7C6bO9WLe8RvuYUwDL5FRsrel84pSholTs0qVL0dnZic7OTkQiEZx33nn679o/gmhGvCyC5dIRAOYKu6So/4oyFzzTCTtkpwhY4qNTShyMwKFwYefwlfF+FMLOkgoW364QOalnS6Hmgrb5AIB58hgAhgO1Ru1EqlkpI00o+7KWJ9WcdyMjdv1dHhxl/HgyI6WFnTbjNuUw4HNVnOcO0YxRqIEilOCRPD1iZ0DzRF+nG++q/DWgnCwg7OoUsfM4rHj6SxfjHz91nulrNRzvJIsc6oqdMZT9Tvr+979v5nEQ9YYxIHAUaFswIwZD/1Xs77DKsQdXJL+BkUi3aeukIkG4wSN2863TP68W7yxg4j20qUEE42m0uw2cPqFk4GJciLiKjNvSkN18G3u1wk50xKaZBU5XgW5X/zwAEpxIoRsh7BsMY1V/9WLFImr6WBlCa/L0CW1OaLmEEzkdojU2T3RVaHkiR3lEN+My4PUqLuo+cCFeqIFCS9F6DXrMAH/cR2Q+Wi5zcj/yEqB6xM4Nn6M+zQwzJtM0RdhRV+xMoWxhd/3115t5HE3H1q1bsXXr1tYdn7b734Enbgauvg/44B83+mhMp4txO4+l0nEcDp9h2jrJGI8mxSR3wQuI7NVMinlnrKHCLpmNhnnKEHYWD0/zOTNVRtGEsOPjxApcmK12wDcHCJ8QnbG1Rewsaa1ZowzfM236BKozKTYyYsctT/qF5UnpGjtrQljQ1DonFtAjdh4Wgwy1oLDjETsGt4FdsZIkIdq2BIgA1omD+TNjRQlAEB7MaXHD4Lrj6kAGVliRgQoZsgGNMMSpQeuHaqqk5ZsntBqfwy829jjqzCwpYKrdSTrGIxAJuUhkyK152QUxEjbYpFhEQOLMjjZf6YuyXaRrXUqVNiSasGMueItdmLV0rDSKt2sUdjbRrCFPN7JtMt7cVGzlXnbheNowYQcAUtcSAIAzfKSk5YkrxYWdpcy5ssV31q7/6EcUYwWsdkLxDFxIwgLheGBAKhYAbF0LkWA2WJRJnbF5NXatbT9SdyQJUTv/4haTvTMiM0Nw6EzPdIb3NfoI6sosKYjxaAppxRyrHiXORUvSUkTY5UyfMNrLTo1na5ba3aUvlE4f/+D3siqFXUqL2DnhLZZKa8962b09FC5qt1EKm6ghlMuZBJFnUlz5cx2PR2CVxGvFAGHn7Snf8sSd5p3KtrbqRr3lYbEBNt6l3SYVnhcbSqTh1aJ1kAC7MRYkeZ2xuQ0Uda6xm2kwD3/tZMowKydaBxJ2M52JI0Aq2uijqBvd4BeSUu771aIkyvBY07zsTDApjke4hUqIecqaruFq49FDP4tUJ7ZyUrEeRxH7CNEZu8AyhlhKwcB4rPK1BE6Fv15t5RSD55gUV5OKzYjUOjNI5PTNLt/ypE3lws7dUePUCQ3Nyw6Fp0+E4ml4cruADapHm9/hwjuMR20xkhV2qvCxCzESdmbQNouL6bZOA9L5xCkDCTtiqrdUCzNLCgAwrzOWJXg0KV3MYy1nrJjhwi7EhV1UcsNpK+3T5WnjH/h+KYZIoopjyUnFFr0wi1TsMkcAAGqqs3OpQtiV4dOXOy+2mrFi2ni2jNVriMhZ2FWe5QlTFbQz/hx5ax0npqF52UnRgnYnodypEwalYYH8zliMZEeLZcRc4xA88BboIieqRxKvf6kMs3KidSBhR0z1lmphuiV+oR6JVDc7tCSieSFjLR2x60IIowbX2CX0OarlXZSdOZYo4YnRyhcUhswRuEqkYnlXZJ+F143VIuzcjAs7h7e99MZa84QUwHCoGmEn5uAaJHIWdpdneRKeGNFTwO3d9Y3YZbuAjRN2CzrdOV522S+SaoxHJRMWH6wWuhwZjtYZS1YnM4qKvyIpioJt27bhl7/8JYaHh6eMFXv22WcNOziiTswgYdcjhF01ablykFJlCAFPfleskaQi/EKZLDWVQcNi46IMccRCo8C8+ZUtmBOxO61YxEVE7LoUXle2r9qZsYzBy2KABLjKEXaixsgppRENVz5dQxKPjxnUUdjlseNEGZYn4dET8AMIMg/anC5D1s6N2BWrscuOEzM2YqelYtnoO5C0zlgtIkodm+aw9Gpgz38CZ1zT6CMh6kjFwu6WW27Btm3b8OEPfxgrVqyYOZ5ArcwMaqDokoIAmGmpWFkY9jJbaWHXIUUwETa2vjEtUlvpUlMZcohIPnhZHIngWOULJrOTNoqnYnmNnTMdhBuJqiN2qXgEdhHJcpdjwGx3Q7X7IKfCsMSGoagMFrn8zyy51Li0CpEkCckyLE8iE3xO7ITcjjJaRMojJ2I3HktN+1yE4jmpWIMeMwB4HVZEnHMRV+1wKUlg/DDQvQSSEHYKFfebw/xVwG1vNfooiDpTsbD7wQ9+gEcffRQbN24043iahpb3sctlBkXs7MjAjyhGDI6UaVjSXKixYsaurg4wSYbEVGQiVaQ/i6AKYadWEAGJWbxAZhjJSOXCTk2EIINH7ApN2gAAOP2Asw1IBDFXGsXBgBPBeLqsBo+8Yw1PwA5AYRI8vvIkj+TrBcbC6GZBjEWSmO13lvV3jDFuhmwFpELj0qqhawkQzrE8mebLcTIwBACIWAw0lRURO78U5f7ksRS6vI68TUKJNHoMGqE2mfldXhwcnouzpSO8rrd7CSwp0ZxCNWAEYRgVFzXY7XYsWbLEjGNpKlrexy6X8CAQM2cgfTMySwqaFrGzijmtUjGPNVmG6hTRpqjBY8VEBE0tx7xXkLBy0ZKJVv4ayMTL7IoF9KjdOR5+jNX42cVFOjUCN6xlDnGXcrzsKrE8SWZUuJiYg1toXFoV+MqwPMmE+O0xu4FjGkXEbraVC7fpTIpDcXNSsQDQ1+HGu7mdsUoG1gz/IiQ5DYtLEsSMp2Jh92d/9mf49re/bcqMS6KBzKjO2Oo8zcrBLi5UcolohySmT/jUAMLJjGHra6mtSiJMSSu/qKpVCDslwcVZQnbDUUpoCWF3rp+LwWrSsVpzSFSqYDRYbgNFBZ2x3NONixyrgcJu/qz2kpYn2TmxBo6/E1GxLmvhebHhRAYeybipE7nM73RlO2OH386bkmJ1txu6FkHMZCpOxb788st47rnn8OSTT+Kss86CzZafSnnssccMOziijgzvA/rXNvoo6sIsBPCGWcJOWHGUivDI3lnAyH7RGZuE3yDXfW3cllxBF1zG0Q5EAMQrby5gQtgp5TRrCJPipS4uPvdX0UCRivJjjMsVeMrlRuwqaJoJJzJ6h6hkwMxUjUXdHhxmvejDCLc86b9wyjZyjKfoFbeBwk68JjplLWI39bkIxXMMig1OxfZ1uPGc1hk78rY+TizCnPC4ykuPEwRRmoqFXXt7Oz7+8Y+bcSxEIxmeWRG7kXASjDHDm3+cCo+GWEtNRRANFNzLLoXFBvmH2oWws3nKr83S0rZSMlj5gskK7EBEZ+x8iYuWgyORipdLR8sY2TaZnOkTJyoQ9EbOic2lv8uDX7AeAHuRGX1v2g9hW4I/R8yIObEaWlcs+PM+uTNWURnCyQy8VpNSsZ05qdjRd/XyD5o6QRDGUvG76fvf/74Zx0E0ClcnEB+fUQ0U3VIQ8bSCaEox1hQ1k4QNaQCAzVWesOuSgoZanjjEuC2Ht3xhpxWuW5OBiteTtC7gcpo1RCq2PcUbA4aClfvKZcTItKSlAtGhTZ9AAHsqSMWGE2l4TWgk6PbmWJ4MTW954kpx0WP1GTBOTN9pOwDACx5VHp2Uio0keEmAx6Tmib4OF46xWYgz0Rk7uAcAn5JS1AORIIiKIEfImU7PWfz/4X0lh5K3CnMswqTY6HRsMhuBcnhKXBTF9Amj58W6VH4MTl/5wk728AJ9W6ryiJ1mB8LKqcfSLE9iJwAAw+FExbW62izclLUC0ZE7L7bCVKwZETtJkpDw9QMobHnizfCUs6Ot17B1tYidW+FR1vFJqdhQgn8p8cvidoMjdvM6XIAk4yDjohYDv+brUsSOIAylqnfTf//3f+PRRx/FwMAAUqn8b32vv/66IQdG1Inu04Gj/8ujdpFhwNfT6CMynbnWMJAEhkMJLOo2Zsg5AL0YPMYccDtL1AyJ2qkuKYS9RglMxuARUxnK8ngTWEXa1pmpvJnBKoSdpZxmDVFjZ4kOwYoM0ooV49GplhvF0Gr6MsV8Aiej19hNVNQ0E06k0W/CFAYAkLqWAJHClid+NQAAcHUaKOxExM6pRCBDndI8EYxzYdcmJwAGwyN2DqsFPT4n3onPx9k4Ahx7FQDNiSUIo6k4Yved73wH/+f//B/Mnj0bu3fvxgUXXICuri4cOnQIGzZsMOMYCTOxuoDOxfznGWJUPEsfK2ZwxC6VHa/ltpfoEM2psRspMN6pYtIxWMHNe33tXWX/mc3Dt3UpFTYzZFKwqPzY5XK6Rj2zAYsdElOxzM0F6FCFY74kIZ6VSiYVCGHXhTDGguUbQudH7Iw10PX2Li5seZKMwAX+2vR3zTVu0RyvOB9iU+xOtIidGSPFNPo6XTiodcYGBgAAQXjgM6h5iCCIKoTdgw8+iIceeggPPPAA7HY7vvzlL2PHjh340z/9UwSDVRRfE41n9pn8/xlSZ9fJAgDMSMVyYRQuZdYL6MKu08BUbCLM67IyTEZbW3vZf+f08+ihV61Q2KWyqWebqwyhJcuAn3dFnuWubrSbnKpixJe7C0yyQJYYlOhI2enfkEmpWACYP6ujoOVJMshrEOPMjo72duMWtdoBG49Qt0lRjE163YXiosZO64o1YcxXX4cb72idsdq6zAM/RewIwjAqFnYDAwNYu5bbYrhcLoTD/IP2s5/9LP7rv/7L2KNrIFu3bsXy5cuxZs2aRh+K+cxezv+fIRE7vxKABNVwYaelCaNwwlMyYpdNxRol7MJBLuzCcJcWljm42nja1s/CldVZiuhZnNnhLteuQqRjT3fyGrJKI3ZW0fVb0SQI2ZId46YGMBFLl/VnZjVPAFnLEwDc8iSH0CgfJzaKNniNjmTljhUrELFzmyRmAWB+zsxYfV244SVhRxCGUbGw6+3txdgYHz3U39+PV1/ldRKHDx9uKdPiGTV5YoZF7GQo6EDEcJPidFzMTWUuuMuM2HmlBMLh6uamTiYW4u/LqOSpyMbF08aPxSYpSCcqsCBJZqdOeO1lXpjbFgAAFli4CK20M1ar6at0xJekN1CUb1IcTmTgM0nk9He5cZTxFHFmNF/Yxca5sAvK7cbP4tYsT6QoJmJpZBRVvyskauxczMRUbIcLx9ksJKVsXSWvsaNULEEYRcXC7vLLL8fPfvYzAMCNN96IL33pS1i3bh2uu+468rc7VZklhN3I24CqFt+2RZglBQyP2KWEx1oELrhsJSJ2Dh+YhV/cmEHzYuMhLpYqMu8F4PO1IcX48UYDFRxLTuq57IiL8LKbixEAwMkKI3Z2YediqXRSQRUmxbF4DA5JRPcMFnazvA6ckLOWJ7kkRCo2YjVwTqyGiNi1S7zWMDd6GUpkYEcaViZuM7grFuBedgwyjkjZdGwIHmqeIAgDqfjd9NBDD0EVF/+bbroJnZ2dePnll3HttdfipptuMvwAiTrQdRog23jNVPAY0NHf6CMynW4T5sVmRMQuLrlhkUtEWiQJzN0FKXwCnswEoslMRenT6UiKcVuJSjzeAFitFozCi24EEQ2OoH3OonIXBCAiduUeu0jFdiu8YaBSYedUuCCxVjBZA0BW2KH8ebGZWE7NocH1ZpIkIe7rB6JTLU+U8EkAQMJm4JxYDRGxm2NPAHE+fWKWj3/ByJsTC5gi7BZ0cmPpfZl5WCbzxx1iJOwIwkgqfjfJsgxZzgb6PvGJT+ATn/iEoQdF1IHctLnFBnQvBYbf4unYFhR2msxSJStklsEsBPGOwV2xWfPc8qYiyN5ZQPiEXmdXq7BLx3jdWtpWuQiJyF50syDiIp1bFqLGLspc5V+YRcTOr5kUV9g84RIj2+ze9or+TrPx4XOCyxOTqjYuzeKCxWK88JC7lwDRqZYnLMKjmUmngePENIQg7hHCbjynIzuUSGfnxNrcgAmPucfvhM0i4YAyT88XxSze0nOGCYIom6oMil966SV85jOfwYUXXoj3338fAPBv//ZvePnllw09OKKO6HV2rd1AkXbxi+UsKYCxSBKKalxdqCIidqlyI2a65Ykx0yeUGBeWmSqEXVTmNWupcCXCLqfGrtzJAcKk2BU9AYBVHLFzC58+h6fEZI/J5KRiTwTiJTbm6HNwTYhcAYCnJ8fyJDqi324Rc2JVd/mWNWUjInazxNiw0ZwGilA8p6bQpMdskSXMa3flNVAoVbxeCYIoTMXC7sc//jGuuuoquFwu7N69G8kkvyCFw2F87WtfM/wAiToxQxoossIuCJVNPwi9WpgQOmlrmTVuuuVJGCPh2r3sWIILO7UKz7WElf9NOlK5sAuj8ho7WUmgE2GMR1NIZpTy/jaThEOMbHNWMDINQN70iefeHoFajqCvxlqlAvpm51iejGUbKOxJcQ48Bo4T0xARuy4Ln2k8nvOFIpTIScWa0DihwWfGZmvstFnFBEEYQ8XC7qtf/Sq+973v4f/9v/8Hmy37LX3t2rU0deJURrc8aXFh5+Riar6NX7QNrbNL8H2WPRXBaMsTEWFCFRfKpI0LO1Wkc8tCM2RmLngdZabSrA49etZv5c0e5TYzaBE0APD4KxV2fM1eKYD3A3G8PlD6cVrE45NMsP0AgIVdHhwRnbG5lif6nFi/CVNgRMSuXeLCbiwvYpeGV0vFmhSxA4D5HW4cZ7NwUu7BBPMi5TJBwBLEDKZiYXfgwAFcfPHFU273+/0IBAJGHBPRCLSI3egBQMk09lhMRIvY9VqMF3aSiPAotsoidl0GpWKtKS585Eo7RgFk7FwMsnig7L/J9e2raIi7SMcud/O1yk3Has0hEeaE11X+GDIAurDrEXOCn3jjRNHNE2kFTpWLn7KmalTBwm43jggvu1zLE21OrLPdBGEnInZ+cNGaK+zCiUzWnNjgSRu59HW6wCBjQ2oLrkj+PZyu8mpSCYIoj4qF3Zw5c3Dw4MEpt7/88stYvHixIQdFNID2fl4wraSmOOG3Bjz1lnZpdW0BAMYKO1lEeMpO3Wk1dgZNn7BmuLC0uSuP2CmOdgCAnCg/YqcIYVeR3Qmgd8YusQcAlG9SHItkDZhLjmybjBB2TjUONxLYvncwz8NtMuFERjcnLmsObhVMa3mipOFj/Dy6O+YYv6iI2HmYEHa5qdh4riGzianYDi7kxjNOjMNPHbEEYTAVC7svfvGLuOWWW/DrX/8akiThxIkT+I//+A/8+Z//OTZt2mTGMRL1QJaBWWfwn1u4gSLl1CYQcAFjpEmxJcML+1m5qTt3NhVrhMB0aMLOU7n/GXPxv7EkA2X/jdYsEoUL7lK+fbmIOjstFVuuSXEizM9ZDK7KjXsdXn2c1mmuKEYjKbzyXuF6wkgyAy94xK5SM+Ry0S1PkGN5EuWNExkmw99pXo2dS7xWtOkTisoQTuaMUDMxFdvXmR+hqyjaSxBESSr+qvTlL38ZwWAQl112GRKJBC6++GI4HA78+Z//OW6++WYzjrFmrFYrVqxYAQBYvXo1Hn744QYfUZMyezlw4nVuVNyiaKlYrxKEBYqhETt9KkK5wi6vxq725gmnMO91+ioXdrK44NvT5U/BUBJas4gXcinfvlzE9AnNpLhccZ0WBtCxCg2YdbyzgYnDuOY0C/b+lqdjL146a9pNzRwnlovUfVqe5Uk6dBI2AOPwo9Nb5pi2ShARO3uGn+cx8bqLJCbNiTU1YufK+50idgRhLFW9o+69917ceeed2LdvH1RVxfLly+H1mvdBUCvt7e3Ys2dPow+j+ZkBlicZRwcgyZCYik6EMGKgl51dmOfKzspSsV0IYrRMb7ViuNUoIAFuX+XGtlYPt9ZwVCDsWKLCmkKNdi7sujPCy67MiF0qFgAAJKoWdj3AxGFcNk/Flt8CT/12CF/92Ao4p4k2mjlOLP+QToNyJGt5Eh0/gXYAY8yPpW678QsKAW9JhSFB1WvsgmKcWJtFvB9MfMydHjvcdgtiKd4N7SdhRxCGUpWPHQC43W6sXr0aF1xwQVOLOqICZoLliWTRU6CzjZw+wRi/OAOwuMqscRMRO7ukICEaA6olrajwidShx1+5sLN5+d+4lfKFnSQMitVK7UA6eS1ue3wAACu7xk7z6Utaq/y8EZYnS1wx9PqdCCczeP7AyLSbhhPpbFrSRJHTNyvf8iQ+wadOBOT20tNLqkFE7CQw+BBDMJ5GWlERSnBh12nVumJNjFJKkl5nB4DmxBKEwZT9Venzn/98Wds98sgjFR3Aiy++iG984xvYtWsXBgcH8fjjj+NjH/tY3jYPPvggvvGNb2BwcBBnnXUWvvWtb+Giiy4qe41QKIRVq1bB5XLh3nvvxSWXXFLRMc4YNMuTsfeAdAKwmZAKaga8PUB0GN1SEANGCbt0DDJ4Mb613C5KmwvM7oWUisCVHkc8pcBVaVOAIBiJoVvij8XbXvnEAocQgx41XGLLLFKlzSIaHf2AJMOqxDELQQyHyovAqWKyR7paYefjHahydBjXnHMRHn75MH72xglcvaJ3yqahRAa+OqRiF3Zzy5M+jADj7yEZHAQARG0mzIkFAKudN0mlY+iQoggxLyaiKYS0iJ2cBFSYmooFeGfsgZP8tUapWIIwlrIjdtu2bcNzzz2HQCCAiYmJgv8qJRqNYuXKlXjggQemvf+HP/whbr31Vtx5553YvXs3LrroImzYsAEDAwP6NqtWrcKKFSum/DtxglsaHDlyBLt27cL3vvc9fO5zn0MoVH5UYkbh6+Xf6JkCjL3b6KMxD92sNmBcxE6Y9SpMgsNVwUVRT8fW1hkbCozrP5cdMczB7efH4UEcUNIlthbriGYRqdzUs4bVoTdQLJSGMBRKgLHShsFMRAirmawBQF8Tw/vxkXN5N+ov3z6JSHKqvU84kdNIYKL1x8KurOWJMvoelDCPIMZtJkyd0BBRuz43T8OORlJ6xM4vazV25k6DmJ8Tsauoo5ogiJKU/Y666aab8IMf/ACHDh3C5z//eXzmM59BZ2ftQ6o3bNiADRs2FLz//vvvx4033ogvfOELAIBvfetbeOqpp/Dd734XW7ZsAQDs2rWr6Bpz5/IP8RUrVmD58uV45513sHr16mm3TSaT+jQNALoITKfTSKfLu+BVirZfs/Y/HbKqwgJAURWoOetaZp0B+diryAzuBes6o27HYzaablBUBaq7GzKAbgQRSWYQjMbhtvO3QtXnIjoOG3iHqMMqlf33Fnc3pInD6JZCGAxE0eurLi0VmhgGAMTghE1lgFrZ8Ts8WfGSCo1A8k7fVKDDWLZZxO6p+PmydCyGHBjAQnkIO9NnYCwcR5tr6mPPOx9iskbG6q3uvbLgQ7ABYIeexxkfldHf6cbR8Rh+sfcEProy31okGE3qEbuMxQVm0nuz3SnjhMzXjg4eAOPZdKScnaZ9HlidbZDCJzDXkQCiwHAopk+g0GbFTveYjfycmtuW9SF0V/B+IbI04rpBTE89zkUl+y5b2D344IP45je/icceewyPPPII7rjjDnz4wx/GjTfeiPXr11duP1AGqVQKu3btwu233553+/r16/HKK6+UtY+JiQm43W44HA4cP34c+/btK+q3t2XLFtx9991Tbn/66afhdptrpLljxw5T95/L8vffw+kADh8+jLe2b9dvPyfuxiIAh179H+wfqLJIvQk5h/E06b59++DMRHA6gB45ACjAf//8aXRPyjpXei7aY4dwCfh4rf1v7oHl+O6y/u6CSAZzwDtjn37hVxjsrG527ejQEZwLPrf1Vznns1wSGaCDueGXYvjlk08g5SnuoWZRkrhGpJ4HxyPYXuGa54RlLAJwujwEKMCP/mcH5hZ5e+3YsQNzx3macjSarng9AABjWG/rgCs9gV0//jbOcJ2Lo5DxyDNvwPZ+/vl684iMj4iI3au792HsYGHPu1oZs8wGGBA7/lskGRfYI3G5usdYBr8TU9ENwJcaBXA6fvnybxBOA4AFjjQXzzvf2I/hI9M3bxjxOTU8LgHgZQd7X/8Nwi2cIDCbel43iOKYeS5isVjZ21YUA3c4HPjUpz6FT33qUzh69Ci2bduGTZs2IZ1OY9++fYY3UYyOjkJRFPT05Duw9/T0YGhoqKx97N+/H1/84hchyzIkScK3v/3topHGO+64A7fddpv+eygUQl9fH9avXw+/35yUTDqdxo4dO7Bu3bq8MW1mIj/7GjAMLFq0CP3rNmZvf20QeOpZLPGlsGjjxiJ7OLV4f89XAAYsX74ci+ECnvkf9NkjQBpYsXotzl/QDqD6cyEdeRE4wMdrXbz2A/jg4vKi2Zb/eRrYsxtdCGL2srOxcc380n80Df+74zFgEEhZfdhYxXljjOHkmx74EcOqlWeiY+na4n8QOQm8CahMwqLTlmHjxjMrWk/+9VHgmWexzDEGpIFl516Ai5ZMrQ3MPR/HDz4ApIDOOQuxvsrXpiw9C+z+F1zQPoH2yz+Ep/7xFbwTsuDCSy9HR04X6kuPvwXvOBd2H7jkSqD3nKrWK4fXRxXgGNCpjEC12YE00L3gjKrOYzlYIv8JvHsAizusQARYsHQ5ArE0cPQQ/FYFSAFrPnQZWN8H8/7OyM+p04bCePjArwAA6y+7GEtmUwNepTTiukFMTz3ORSUlZFUXN0iSBEmSwBiDqpr3bVZbKxfGWNkRwrVr12Lv3r1lr+VwOOBwTB1XZLPZTH/z1GMNHZmXV1osVlhy15zD/f7k0bcht9CHhSQBYIBFtsDi5dGoHlmYtMYyU573is9FhouACFzwux3l/62Pf2npkkKYiE89jnLRPOVSVn/V+whLXgAjSEcnSu9D4Sm7CFxo81TweDVmnQ6A19gBwGi0+GO32WywZ3jq1+pur/59csZGYPe/wHJwB8685h9w5hw/9g+G8MzbY/j/PrBA3yyaUvQaO5unAzDxveCbswTKALc86VaPAgDs7b3mfRa4+ZeO2TZ+DgPxDKIp/hnuErlgq7vwYzbic2rR7OyX5E6fi4RJDdT1ukEUxcxzUcl+K7I7SSaT+K//+i+sW7cOy5Ytw969e/HAAw9gYGDAFMuT7u5uWCyWKdG54eHhKVE8o9m6dSuWL1+ONWvWmLpO0zFLRF4CA3pDQMshmie6EQAAY7zsxHMVZU54HBV0turzYmtrnsgIj7eMrfr3YVTmF9tUpPBEBh3RyBCBE15HFd8PO08DAPQoJwAwnCzDy04TdnIVzSE6iy4GrE4gOAAM78NHVvL62yfeeD9vs0g8Ca9k/txUAFiQY3liZ7yhwdE+tVPXMISXXYfMRdx4TlesZtljdlesx2HFH128GH+waj5m+yqc+0sQRFHKFnabNm3CnDlz8PWvfx3XXHMNjh8/jh/96EfYuHEjZLlqO7yi2O12rFq1akreeseOHVi7tkSqqEY2b96Mffv2YefOnaau03R4uvS5mhg50NhjMQsh7NrEWDFDOmOF9UcYLr0RoywM6opl8QAAQKlBhCSsvBMyHRkvsSV0IRupdE6shrA8carc8qQcLzunMIC2eWoQdnY3sPhS/vM7v8C1omni14fH84ySM/GcLzUmd4j2d7lxhOV/UfV0mPjFVXTFtoE/n1pXrAUKbKr5PnYaX9l4Jr7xBytNqc8miJlM2Z/I3/ve97BgwQIsWrQIL7zwAl544YVpt3vssccqOoBIJIKDBw/qvx8+fBh79uxBZ2cnFixYgNtuuw2f/exnsXr1alx44YV46KGHMDAwgJtuuqmidYgKmH0mr6Ea3gfMn757+JRGCFe3EoIdaUOEXSYehBVAhLnhqUjY5YwVC9cwViwh6i9qEHYpmx9IAUq0DNsiIWQjcFUXsdMsTwID6JeGcDK0rOSfuFQuROzu9srXy2XpVcA7vwDeeQrzL/ozrOrvwK6jE/j5myfwhYt4Y5Uqnk9VtkO2mhtRWtTtwVOsFxfhtwCAIHOjw2+isBIROy/j53A8moJFkrLjxADTI3YEQZhH2Z/In/vc50z5ZvXaa6/hsssu03/XGheuv/56bNu2Dddddx3GxsZwzz33YHBwECtWrMD27dvR399v+LHksnXrVmzduhWKopi6TlMyezlw6HlTJlD8dM/7eODZg/jeZ1fhtFkNung42wHZBqhpdCFU9qzSYmRiIS7s4KrMZNigVKyUEsLOWX00K21vB6IA4mUIu9yIXTXCDuATKAIDWCifxIFSETs1A5cQHg5ve3XraZx+Ff//2G+A6Cg+snIudh2dwM/eyAo7Jh6favdWP56nTGb5HLrlCQCMsjZ0eUwUkyJi5xZm1GORJJw2S1bYWexceBMEcUpS9ifytm3bTDmASy+9tKQ56aZNm7Bp0yZT1i/E5s2bsXnzZoRCIbS11ZD6ORUxcWbsk3uH8O5wBM/uH26csJNlLqjCJ9Bt0FgxRUxFiEsu2K0VSAEx3qwTYYxF4lWvb03xi7TVXf3EAsXRDgCQEhUIO7jQVbWwOw049DwWSkN4oZSwS0b0H53eGqcytM3jXa5DbwLv7sDGs38Xd//sLbxxPIgjo1Es7PZASob5p6Pd/NeoJEmI+xZCZEYxijbM95hYDC8ido6MEHbRFHwOK7zapI06PGaCIMzD7C+jxKmINlrMhIgdAxfx2vDxhmHw9IlsV2qF3n9uPmFAlhisiQkk0tVFiO3iIm3ztFf19wCyA+KTwdLbas0T1dbYAfrM2IXSSYxGkkgrRbrrxXoJZoPXY4Cf5NKr+f/v/AKzfA78jrBa+dkbJ5DKqHCoVU7VqJau0/Qfg1IbHNbqRsuVhYjY2dL8OQ0nMhiLpnImbZCwI4hTGRJ2BZixXbEAMEvUO0VOArEyCumrYDxq0CivahF1drOkIEYjSahqdcbAGkwIu7SlwouixQrm4vYTXVIIw6HqnhenwiNajhqiWZKL/609VYZfUjJbY+dzVBldEmJmkTQExlA0Fa3NiQ3DXX3qNxdN2B38JZBJ4Vq9O/YEwom0LnJkp7kdsRqentOgMF7qErXVPtGnKELAy4kALDJfM5lR9akT9WicIAjCPEjYFaD1u2KLCBmHD2gXnl4GR+1sagKrpbcxHi7dBWkKWp2oGJnVjSAyKkMgXuMoGH2OaeXTOqScOrsTwcrTsarK4Fa50HL5qxd2Fg8XFI5M6YidqqdinbVH7OSTAFheV+pkkqKhI1xLhDCXuecBntlAKgwMvIKrzuqF3SLj3eEIdh6Z0NOSkslWJxoLZrfrlicJu8nCTkTspEQQXe7sc+nRI3Yk7AjiVIaEHTE9ejrW2Dq73x/7J/y34x6cPvG8ofutGBGx67NzgTJco9CURJeoUo2PXI7lyWAVwi6cyMAH7j/m9lcvCuxe/rcupXTELhPj4i/CXJX59uXSsRCABA/i6EYIJ4vU2SUjYj24jUlTyjKwdD3/+Z2n0Oay4dJl/Dz8x6+PwldnkdPf5cFhxr3rUq7Z5i4mInYAQ587o9/caRURU0rFEsQpDQk7Ynr0BgpjI3ZtCk/tLoi9Zeh+K8bDL55zrVzE1FpnJ6e5sGPVFJ7nWJ4MlmHUO5lAPAW/xIWdvYbmCYeP1/t51AhQoqFJifPnLS57qhdaVgfQ1gcAwvKk8DlIRQL6eoahpWMPPAkwpqdjX3p3NNshWidht6jbg29lfg//nrkC73VfYe5iVgdgdQEA+tzZWtcum/iZmicI4pSGhF0BZnSNHWBqAwUAzEkfM2W/ZSOaJ2ZLPBJUq7Cz6MKuitSdnooNYjBQubALxlJ6xK4WuxNnGxeYVii6T10htFRsxlqjCOji6dhF8lBRk+K01nVsMVDYLb6MW3tMHAZG38WVZ/bALaxq9A7ROgm72T4H9lvPxF9lboSrferMXMMRUbs5juzrPhuxo1QsQZzKkLArQOvX2JUg1/KkRPSmGvrZ+4inGugRKFKxndpYsRqFnVUIu6q6KEXErrvaVGwoAIskzlENws7v9SHJRCNEKS873eetRqEl6uz6pZNFx4ppdjKpSptTiuHwAgsv4j+/8wu47BasW85fF15NKNepxk6SJPR38W7fTrfd/AVFnV2vLfuct1tExI6EHUGc0pCwI6an63RAsgCJABAeKrl5pSyQhjEWLMNWwyxExM6vcAFTk0mxktFHMck1CLtqU7HRMH8MGVgAm6vy9QV+lx1BcKGmxooLO0kIO2arUQQIYbdIKh6xU0XqN12pnUwpcmxPAOizY311jtgBwHkL2gEAp/fUIRUqInazrDH9pjaZfOwIohUgYUdMj82pX3TNMCq2SAzRwXcN32/ZCGHnVKJwIllbxC4nbWmtSthlu2KrEXbJ0BgAIC57s12/VeB32RBgXDglQqNFt5XFY2a1+rx1cssTXmNX5LHrXccGC62lYgrFwKtAbBwXnT4LbS5bjqdb/YTdX1+zHD/d/Du4bJnJzROAHrHrsmYjxD6ZmicIohUgYVeAGV9jB5jWQKGRPvm2KfstC4cfsDoBoPbpEyJ6lWQ2OFxVmOcKYdeJEMajqYpNipOisSBZY72b02ZBSOL7iIfHim5ryXBhJ9cqfHJMiosJO21kmlJNDWMxOvp5PSlTgPeehd0q45pz5tS9xg4A3HYrVva1mzK6cQoiYtcuRfWbvFJ9G0YIgjAHEnYFmPE1doDpDRQYbWDETpL0zthZCGKkhjmtmrALwwVPJXNiNYSw65a4eCnm5zYd6ViA/29ANCsm831oUcBpURXYFC58LLUa+HYsBIMEnxSHIzmOSDIz7WZ6hNAM0aFF7Q48CQD4ysYzscQvahZbVeSIiJ0fWWGn+9iRQTFBnNKQsCMKY+LMWABwBA6ast+yMWqsmBB2UeaEu5qpCKLGzi/FYEe6YpNiNR4AAChGCDsLF2qZaJGJI+LxAoDFXaOwszkhtc0HACyUhgqKWmtarGlGM4M+hWIHoGTgcVjRYdHSkvVpnqg7ImLnUbPn0qVqDSOUiiWIUxkSdkRhtIjdyNuAWmSOZ5V4w4cM32dlB8CFXbcUQjCernpOK1LaFIYqI3bOdsDiAADMlgIVW56wBG9CYY7qO2I1kjYuZJQyhF2SWeGuJvU8mZx07HCBdKxNdB3LNXT9FmT+GsDVCSSCwLFf89tETV+rR+ycSlbYOTRhR80TBHFKQ8KOKEznYu7zlY4BgaPG7z5+1BTBWIy86iUh7HplLoyKzSotSjJH2FUTsZMkwMenDvRgvGLLE1kTIa7aRU/GzvfBinXFprJzYg2Z26qPFivcGesQs3AtbhOEnWwBTtemUHCzYj0q2arCTkTsbKkQbBb+rrArWsSuRR8zQcwQSNgRhbFYge5l/GcT6uwcLAGETxi+37KZNFas6nSsVmPHXPDYqxQ6fm6zMUcar7gzVhYRQ4sBwk5xtAMApGQRK5qc1HNVQnYyXbwzdmERyxOnwmvBrLWmfguh1dm98xSQjvNmCqB1RY4+LzaATg/3zbOJhpiWfcwEMUMgYVcA6orlHFDnAQD27zWpiWTkgDn7LQfRtNBrMUbYReGCu9q5qb45AICeKoSdXdSfWWsYJ6bBxAXfkgwU3khECCNww+s0MGInDU1vUsxUOFntI9OKsuQKQLYCo+8Ag2+IGyWgVgPmZkWbFxsPYN3yHsz22mHJUCqWIFoBEnYFoK5Y4Nh4DL8Z4t2BgWDAnEUa2RkrInazpACAGkyKkyI1aUDErleawIlA+alYxpiepnR426tbOwfJ3QkAsKdKR+zCcMFnSCpW87KbfvqEVU1CBn8dOnwmCTtnG9C/lv+890f8f4e/Jl/ApkYIeCQC+OrHzsarf/5BSNA6gUnYEcSpDAk7oiB/99QBKCaMEwOgG+Gy0XdM2X9ZaGPFGK8nqz5ixyNYYbj0WaMVowu78aITGCYTTyvwMp6mdBogeqwevg9HupiwywpZQyJ2wvLEL8URD56cekzCWiXNLHC7TRQdWnfsW4/z/1s5JalF7BJBQFUhp4XtiSQDNgMaYgiCaBgk7IhpeX1gAj97w7z6t4OMp3jVRqZivTwV68tMAGBVe9mpWvMEq7J5AshJxU4gEEuXPUc3EEvDL4k0pae9urVzsHq6AACunG7JKeipZ6cxzRM2J9IeLmwdoSNT7xZF/RG44HXaal+vEJqwi4uO4FYWdlrEjqm8q1trFrH7WjdKSRAzBBJ2xBQYY/jqz83xrtM4qPILOUYaGLETBsU2NQEPElVH7LQB9VE4a47YzZF49LBcL7tALA2fGFgvaRfrGnD6ubBzsjiQSU2/UY6Q9RkRsQPAOhcBAPzxY1DV/CixFrELGxUhLETXaXxGskYrCzubU5+8gnggpwuY0rAEcapDwo6Ywva9Q3h9IACXzYL+TnPSMocYj1BZYsP8wtIIHF7AxlPC3VKw6ho7NcFTk1HJDYe1yrdUTsQOYGVPnwjEU3rEDgZ4vLly07mJwLTbsJzUsyFdsQBss5YAABZgEKPR/PMgi6L+MNzGRAiLoXXHAq0t7IC8Ojt93nGrP2aCmAGQsCPySGYU3PcLbm3yxUsWVx+BKoBWsheBG0NMiIixBk6g0KZPIIDRqoUdFzppq6f6OZ9C2NmRRgfCZTdQBHMidqh1vBcAv9uJIBNivoDgVhLZiJ1RQkvWLU9O4mRw0nlIi4hdXYTd1dmfW13k5HTGanWT1BFLEKc+JOwKMFPtTv7llSM4Nh5Hj9+BP7p4salrvaelY5uggWKWFMRIOAlWTbOISGMp1houilY74OajxSrxsgtFY3BJImVqQMSuzWVDgInHEZ/epDgT40I2ihq6gCdTzMsuw4VdVPLAIptc/7Xgg9nnsdWFXW7EjlKxBNEykLArwEy0OxmPpvCPz/Lo2Z+vXwa3URftAhxk9RN2upXDZEQDRbcUREpREUpMP4S+6L6FQbBSa7TDn03Hljt9IhbKEV8GzDX1u2wIgD8OJTo27Ta5EUrZKKGVM1bs5KTHrqViE3IdPOUsNmDJlfxnM8aXNRO5EbuU1jxBwo4gTnVMzmsQTcs0kalvP/MOwokMls/x4/fOn2/q8k6rjPcy9W+gYJNTpSJiN98WApTqLE/kFLeKYLYaIzy+ucDQXvRWELFLCvGVlN1wyLWnzX1OK4LCiiYRGsN0UoqJ6E7GZqAI6FgEFRL8Ugyh8SEAC/W7ZBGxS1vrZBZ82Z3c9mP15+uzXqPIjdhpnwcGfDkgCKKxUMSOAAAcHI7g3389AAD4qw+faVwkZhJa5MzjsOK9OkbsCqILO15jVLHlCWOwigH1rNY0lojY9UoTGAyUJ+zSEd6Rm7Iakza0WWREZP44kpHx6TfSUs9GCjubE1EHr3dUR9/Lu0vrik3VkuquhK7TgN97WE8Ptyx5ETuteYIidgRxqkPCjgAA3Pfkfigqw5VnzsbaJd2mr+dxWLI1dhOHASVt+prTH4gYKybz9OJIuIDFRyEySciMH7tUa/OCn3v79WC8bLsTJRbgh2E3rh4sbuGPIx2ZPhWrpZ6NrkGLefsBANbg4bzbrcLHLmOQeCUE09XYUSqWIE55SNgReOXgKJ7ZPwyLLOH2DWfWZU2Pw4ohdCIGF6BmgPHDpf+oFgr1RIiIXbcYKzZaacQumTXytdQa7fBlI3bhRAaRZOl6P5bgETvVblwKLWXj+1Ki00fsLCJCKRks7NQOXmfniw7k3W5VuchVWr2Zod5M1xVLETuCOOUhYTfDURnw1f/h9iaf+cACLJldnw923k0p4YikpWMbNIFC2J20qQEAwEikwohdSrP+cMLtsNd2LCIVO9fCGyKGyonaCU85Iwv9U3a+L1bA7sQqxk/JBtir5O23m6c+O5PH8253iFQsM1C8EpjkY6dFYek5JohTHRJ2M5x3h8PYNxiCz2nFLVcurdu6Xicv9H9X4WKmYXV2Qtj50mMAWOXNE9oUBrjgdtTYvODT5sWK6RNl1NlZkzxiJ7uME3aqox0AICWmsTvJJGERqWeLy1gR4J6zDAAwVx1EIp0dqeZQjfPpI3LIi9hRKpYgWgUSdjOcnUf4xftPLl+CTk+NEacKcNv5zM93FC1i927d1s5DjBWzsDT8iFbePJE7J7ZWexgRsfOzMBxIlTV9wpLhKTSr20Bh5+TG0ZZkYOqdOalnm8vY1Ki7l0+fWCQNYSjHoNnJ+M+ywUJyxpNXY0epWIJoFUjYzXBiKQV9nS5cv3ZhXde1WyS47Za6etlNi80JOLgomiUFMVpp84QesXPWHrFztgNWFwCgVyrdQJHMKHAp/IJs83TWtnYOsojk2FLBaRYV5sTMAY/TYdiaACAJLzu/FMPYyKB+u5PxiJ3FwKgkgem7YiliRxCnPCTsCtDqkyeiqWyq6y+vPgMOq7Gjw8qhy2vPWp6MvDOtt15d0MaKScEqInb8gmhIxE6SspYnKG15Eoyn9Tmxdk97bWvnIAuR6EiHp96pPV644HUabINpc2FU5h3Z0SEh9BmDW0TsbO52Y9eb6egRu2DO5AmKihLEqQ4JuwK0+uSJA0P8g3y2z4EPnz2nIcfQ6XFggPVAlSy8eDs81JDj0DtjEcRELA1FreBvRQQrArcxc3VFnV2PNI7ByaO1JhGMpeEXc2KNrLGzebmwcykhQJ30ZOSkns2Y2zru7AMAKCNifnAmASv4lxASdgajReyYkn3vUSqWIE55SNjNUNJCvfT4HdUPrq+Rbo8dKdgQcYkpFw1roOBedj0yTz2GK7HUE0InCic8Rggd3aR4HIOB4qnYQDwNn2R8Y4HD1wUAkKFmuyU1cppFfEZH7ABEPQsAAJaAsL8RwlllEpxeiiYZis0FWEQ6XRUvekrFEsQpDwk7omFozRpjTn4xb5yw4xG7BXaeZgxVIezCzGVMxM6f7YwtNVYskBOxM9LuxOv1IcF4cwsmW56YHLFLty0CALgiR8V6WkTUCZ+zfs09MwYtaqdBETuCOOUhYUcUZcpsVQPp8vJowQlbg4WdmD4x18ZFSyhdwWNOZWvODInY5aRiI8kMQonCKjMQS8EvcU85vV7KAPxOGwIQF/j4JMuTVDZiZ8jjnYQsxnh1xI8BAKSEEM5wG1/TR0x93VDEjiBOeUjYEQ2jS0Tsjkj1S8VOq1NFxE5PxVbSGJsTwTImYsdTsfMsAQAoankSjKfh0yJ2Bha9+11WBFgBYZeTijUjYufsPR0A0JN5nzfTiIhdmLlNEZIzntyInc0DyPVvoiIIwlhI2BENo8vLhV3WpLhBXnZC2HUxLmIqS8VqqUIDumIBPWI3RzcpLlxnF4wl4YO438BUbJvLhiA8/JdCwo45Tamxa5/LhZ0PMbDYmD4yLQwXfCTsjCc3YkdpWIJoCUjYEeVhghWJlor9bZILK4TezzPArRuiecIvxoqFUuWnYlluxK5WHztAj9h1sXFIUIvW2cUjIciSOC8GNk/4XTYEGRd26Wi+sFMTWSFrRsRudlcnTjDelRs68Q7ScSHsGKViTSE3YkezeAmiJSBhRxTFzH5ZLRU7ELfrEyAwdtDEFQsgInaeNBdTlUTsWCKn5syIiJ23B4AEKxR0I1RU2KWE6FIkG2B11r62dgh2q15jlwyP5t2XiWtC1hyhZbfKOCFxcRs58Q5SUS7sInDDZaM0oeHkRuyovo4gWgISdjMcZqp0K76uloodj6bAunkKDiMNaKAQzRMyU9COCMIVNE+oOcLOEOFhsemGyT0lLE/S0QD/3+YrUDxYHbIsIWHh0Zt0ZDzvPkVE0GKyyzRT6xE7r7lMDb+LjBB2CYu7YbY8LQ1F7Aii5SBhRzQMze4kozKk2vmc0IZ0xlpsgIun/2ZJQYSqaJ7IWD2QZYOEh0/zsituecISAQCAajfe3y1h4/tUpqRi+eNVrB7D19QIu7lJsTRxSBeSSQtFk0yBInYE0XKQsCMahsNq0Qviwz7uX9ZoL7tZUgChNMDKrCmUhP2HYjPwoqh72Y1jsNi8WCF6mIH1dRppWzvfdzw/YsdEs4ihj3cSSeFl5wwd0Zsn0lYSHaZAETuCaDlmhLA7fPgwLrvsMixfvhxnn302otFoow+JEGjp2DHnQn5DwzpjeTq2G0GkVQmRpFLiDwCoKuQ0fy0xIy+Kfs3LjkfsColMKcVFlmRgR6xGxsH3KU0yKJaEb58ZUUKdzsUAAH/8GJiYTZu2kegwBeqKJYiWY0YIuxtuuAH33HMP9u3bhxdeeAEOh6PRh0QItHTsoI2n3zB2EFAyhq8joUQETkTs5tm4WBoJJ0vvNB3V98vsBgoPLRWLccRSCkLxqc+HojLY0jxaaDFjhqq44FuSgbybZSHszIzuOGZzk2K3EoYjwo2KVRJ25pAbsaNULEG0BC0v7N566y3YbDZcdNFFAIDOzk5YrWSbgFJCp05olifH1S7A6uIzKwNH638gQtgtdHDhMjARK/03Wn0dk2Gzu4w7FhGxm28NAAAGQ1PTsaF4Wvews5kg7CR3BwDAmg7l3W5J8+dHdpontGZ1dGBQWJ60hXkE19CIKJElL2JHzzFBtAINF3Yvvvgirr32WsydOxeSJOEnP/nJlG0efPBBLFq0CE6nE6tWrcJLL71U9v7fffddeL1efOQjH8H555+Pr33tawYePVE5XFBqDY6a5cl4LAN016OBokCDg+iM7Xfw1OqBoUjpXSWz48TcTpshRwdAj9jNlQMAgMHA1AaKQDytjxOTXcanYmU3F1aOdDB7I2OwZfiakonCrsfvxFHGhbaFiWilgZM1iByoxo4gWo6GC7toNIqVK1figQcemPb+H/7wh7j11ltx5513Yvfu3bjooouwYcMGDAwM6NusWrUKK1asmPLvxIkTSKfTeOmll7B161b86le/wo4dO7Bjx456PTyiBHqNXSQJdC/lN5og7ErGJ0XEbo6FR6gOnCzDKDmZ62FnoPWHiNjNwhgA4MQ0DRSBWCo7TsyEGju7hws7u5oAMiItncqmnq0miEmNHr8Dh9XevNskExpECFBXLEG0IA3PSW7YsAEbNmwoeP/999+PG2+8EV/4whcAAN/61rfw1FNP4bvf/S62bNkCANi1a1fBv58/fz7WrFmDvj5ew7Vx40bs2bMH69atm3b7ZDKJZDJbXxUK8Qt9Op1GOl3JrKny0fZr1v6nQyvIZ4wVXVdlKGu78hcW+1VVpNNptAmT29FwEsrcJbAAUE++DcWk50JRlGkfh+TqghVAu8rtPQ4MhUs+Xik2ASuAMHPBaZONO3+uWbABcKtRuJHA8fHolH2PRxLwS1zYKTYvVIOfL6vbD5VJkCWGdHiEC9/oOGwAFCbBZnea9nr12SUclyYJO4e3ru+PmYMVVosdkpJCxuoCK/IcN+JziigMnY/moR7nopJ9N1zYFSOVSmHXrl24/fbb825fv349XnnllbL2sWbNGpw8eRITExNoa2vDiy++iC9+8YsFt9+yZQvuvvvuKbc//fTTcLvdlT2ACqlnJNERCAAAgsEgtm/fXnA7lxC2ExMTRbcrlx4hmt9//32c3L4dx0YkABa8M3ACr2ciWAMg8N5v8JIBa+VyrhCyv/3tb/HeyNQ3iD82gMsA2KJDAID3RiJ44ufbYS0S054T2IkLAEThwuiJ49i+faDwxhXyYdkJq5pArzSOXfvew/ZUfrfwayMS1omI3Z4DR3B8xNjn69CIhCA86EAELz31BMKuefAmTuAK8Ajl4LHD2L79kKFr5nJS7sn7/ejgqCGvP2IqV0kuOJHCzjf2Y/hw6ZICyng0F3Q+mgczz0UsVkbdt6Cphd3o6CgURUFPT/6HfE9PD4aGhsrah9Vqxde+9jVcfPHFYIxh/fr1uOaaawpuf8cdd+C2227Tfw+FQujr68P69evh95uTDkqn09ixYwfWrVsHm83AWq0i7Br8BTAEtLW14YKNGwtu99qxJ4BRoKOjA2uKbFcuv337e0AcmDdvHlZt3Ii298bwbwd3AU4fzr3yE8DDW9GhjGLjhg2GTlMY3HM7wIAVK1Zg2ZpporWRYeDAX8GphOGxZBBVrFi2+iIs6y1cdyS9GQIO8zmxZy5djI3rlxp2vJaB+cDYQfRIE4BvBTZuXJ13/9irA/AP8Df6ygs+hHOWFo56V4PrwAiCAx50SBFcfMHZYH0fhPT+68B+IAw3zjt7OTZe2G/omrm8cDQM5FjonbHifGxce5pp681kLKMPAsdexep1vwd0nV5wu0Z8ThGFofPRPNTjXGjZw3JoamGnMXmUEGOsovFCpdK9uTgcDjgcDmzduhVbt26FonA/M5vNZvqbpx5raGjPnyRJRdfUhimU2q78hcV+ZRk2mw2z/TwKOhFLw9azDIAEKRGALRXUveWMxGKxTP842noBSYbEVCxzhfF6pAPvjcWxoq+z8M4yXFhF4ILPaTf23PnnAmMH0Ytx7Aklp+w7nFT1Gjurpwsw+HXT6XUiAD5dwpoK8/2rvNYvwlxoczlMfa2yjkW6sIsxB3xeN128zOIPtgGBAdh6l5e1eT0/p4jS0PloHsw8F5Xst+HNE8Xo7u6GxWKZEp0bHh6eEsUzms2bN2Pfvn3YuXOnqevMdLpz5sWqFifQIaJAowfqeyCyBXB3AwDOcAYAAPsHSzRQiCkMYeaC22HwdySfNn1iAoPB+BST4kA8pdfYwYTGAr/LhiATxfRidJnWLBKFE16nud8JO9o7MMS45UoYLniNfn6JLP45wIIPNPooCIIwiKYWdna7HatWrZqSt96xYwfWrl3boKMijKRD2J2ojFt4mNkZWxLRGbvIHgAAHBgqEfrOETqGdsUCOdMnxpFIqwjE8usCg7E0/CZ2xba5bAiACzsWE6EzrQuYmS+0etscOMJ4A0WYuUnYEQRBlEnDhV0kEsGePXuwZ88eAHz81549e3Q7k9tuuw0PP/wwHnnkEezfvx9f+tKXMDAwgJtuusnU49q6dSuWL1+ONWvWmLrOTMdmkdHm4iHmfMuTBowWE6nfPksAAPD2UKmIXY6PndHCQwi7BTZ+LJMtTyLRKBySEHsmeLz5nTYEGU/FpiOasOOPNwyX6RG7Hr8TR9QesR4JO4IgiHJp+Kfla6+9hssuu0z/XWtcuP7667Ft2zZcd911GBsbwz333IPBwUGsWLEC27dvR3+/eYXbAE/Fbt68GaFQCG1t5nl2NTuskKGvgXR57AjG0xiLpnB6tyjebmDErtfCTXkHgwkEY2m0uQvUNogIVpgZ7GMH6CbF84TIHAomcNbc7OswHeO2LAwSJBOEndMmIyzxiF06Og47oKeeI8wFn8lCq8fvxPN6xM6FeQ6Dn1+CIIgWpeERu0svvRSMsSn/tm3bpm+zadMmHDlyBMlkErt27cLFF1/cuAMmDCdrUpwCupfxG0caIexm8/+UIOa2OQEAbxdLx+oGxW647UZH7Liwm800k+L86RNKnItPxeYBZOPfxpIkIWnjgjET5RE7pqeezY/Y9fqdeEY9H++zLvxCvYAidgRBEGXScGFHEF0ePi92LJqTig0OAKnyfXsMwcOFnTMTxLJeHq0qOoFCEzrMCY/RESXRPNGmTMACBYOBSdMnhLBjDvOiyWkb3zcT0UElLiJ2cMFjeo2dEwfZfPxO8h/xn8oVpq9HEATRKpCwKwDV2NWPztyInacLcAmLkbGD9T0QkYp1pINY1sP964rW2aWyI8UMj9h5ZwOSBTJUdCOIoZyIHWNMT4ua0TihkXG0AwCkBBd2mUS2C9hj9OOdhNNm0WsvLRKDo5hTNEEQBKFDn5YFILuT+tEtOmPHomKUW6M6Y0XzhCMTwrIeHrF7e7B0KjbMXMZH7GRLtuZPGs9rnogkM/CyKN/MxJmtqpgjakny6KAqInZpiwcW2fzayx4/j+Q6qbyOIAiibEjYEQ2n05P1sgMA6A0Ude6MFULKmQ7oEbt3TkagqmzazVnSxIgdoNfZcS+7bMQuGE/rHnZmCjvJxX3krCkh7BL88So2j2lr5tLj53WOJOwIgiDKh4Qd0XC6vDwyMxoRwm6WaKAwwaS4aJxJCDu7EsXCDitsFgmRZAbvT65v08hpJnAb3RUL6J2xPdI4BoMJ3aQ4EEvrUyckE1OxFg8Xdo50CFBV/fEqdq9pa+bSS8KOIAiiYkjYFYBq7CYzfdTKCLqmROwa5GXnbAeTeeTNlhjHktlF6uyUNKQMj6IlLR7YLCa8lfzzAPCIXSqj6s9PbsTOzBo7q4fXOkrgNX1SivvYMXvh+blG0ttGwo4gCKJSSNgVgGrs6ocWsRuLaDV2IhU7dhBQlfodiCzrUTLpyIs4o1cIu+nq7JI5Ys+s1KRIxfbbsr56QH7EzgxzYg2vx4sY4+cG8QlY0uIxO+oj7Oa1uwAAbqt5XyoIgiBaDRJ2Mx7zi+BLrav52E3E0sgoKtDeD1gcQCYBBI/V9ajU864HAFh++TdY2aUCAN6ezvJECLs4s8PucJpzMMLyZL41AAA4IVLC+XNizYvY+V1WBCFEa3wC1rRo2HDWR9h9+Jw5uPF3+nHVfLUu6xEEQbQCJOyIhtPhtkMSOm8iluYdoV1L+A11NipWP7gJIec8SLFRrB/8HgDgwHSpWL1xwgQPOw0RsesBNykeCk0TsTNT2DltCIixYoiOwqry9S0mrpmLz2nD7VcvQ199SvoIgiBaAhJ2BaAau/phkSV0uCdZngjrESQCdT4YO97ouwEAMPe9H2K19DYOj0aRSE9KCYt6swgzqSMW0CN2ncoYAIYTAS6s8mvszEvFtrlsCEKoqpzIqd1t3poEQRBEbZCwK0DL19ix5qpb0i1PtM5Yg1PElext3LsM6spPAwDuc3wfsprGweFI/kY5VidmR+zsahw+xDEovOwCsRT84GlRc1OxNgSYEHaBAQBAgtngcrlMW5MgCIKoDRJ2RFOgdcaOap2xplGexFMuvwtwd2EJjuELlu1TO2PF5IcIM2FOrIbdA4iRYZrlCSBSsZKwYDFxpJjfaUVQS8UKYVePcWIEQRBE9ZCwI5qCbtEZO651xjYadyew/l4AwJ9aH8PQ0bfz78+tsTPDw04jz6RYa55I1yVi1+ayISCaJ1iAp2IjzAWvk4QdQRBEs0LCjmgKOvWxYmZH7Cpg5SdxsusCuKQULnrnvvz0dVLU2MEFt5kRLJ8m7MYxFExAVRlC0SS8EJMoTKyx8zltCIpULMuJ2PkoYkcQBNG0kLAjmgLN8kSfPtEMSBLGLt2CJLNiZWInsO8n2fu0iB1zmRyxy5oUpxWGsWgKmXgIsiREpok+dnarjJiFW5vI0ZMAuLDzkrAjCIJoWkjYFYC6YutLdvpEk6RiBf1Lz8V3lY8AANTtfwkkuFkwzJ4TqyFSsQvtmklxHKroFFYtTsBmkoeeIG3LT/VGmJNSsQRBEE0MCbsCtHxXbJORnT7RRBE7AB6HFf/j/yQOqb08avXLv+V36M0TJnbFAnoqts/Khd3h0SgcGdGha2IaViMzqYaPInYEQRDNDQk7oinonDwvtolY1NuFOzM38l92Pgwc35X1sTM9Yse97OZI3KR4/2BYNyeW6mAUrDo68n6PMBJ2BEEQzQwJO6IE9Rk51q3X2DVXKhYAzpjjx6/Us7CrbT0ABvz8FiA+AaB+ETtuUgy8PRTSzYmlOkTsJFd73u8RuCkVSxAE0cSQsCOagi4PT8WGEhmkMs01G/SMXt5A8G3L9YCzHRjaCxx+EUD9InaezASsyGD/YKguVicaFk9n3u8R5qSIHUEQRBNDwo4oTn0Cdmhz2WCR+WITseZKxy4Twu43IxaoV97Nb2RcfEZgstBxdwOyDRIYZiOAk6Fkjjmx+RE7u7sNGZb9mIjLbjis9LFBEATRrNAnNNEUyLnzYpusgWJhlwcOq4xEWsXR/t8D+j6o38dnxZqYipVlwNcLgHvZAahrxM7vtiMEt/572uqBJNVJ7RMEQRAVQ8KuAGR3Un+6dJPi5qqzs8gSlvbwqN3bQxHgmm8CMo/SjaHN/BFbos6uR+J1fXrErg7Cri13XiyAjM1bZGuCIAii0ZCwKwDZndQfzaS42SJ2QLbO7u2hMNCzHPjUD3Cn8kc4zmaZG7EDdC+7uTIXdtmInfmpWL/ThiCyYo6RsCMIgmhqqAqaaBqacqyYQKuzOzDEjYkzi6/Af6T5cXrMbJ4A9OkTixwhIAP4RFcsnO3mrgvA77IiwDzZGxw+09ckCIIgqocidkTT0K2bFDdXKhYAzpzDo2NvD3Fj4lha0e9zm2l3Auip2AXWAADAL3zs6tE84XfZEERW2ElOEnYEQRDNDAk7omnoamKTYi1id3Q8hlgqg1iSCzurLMFuMfltpJkUy6J5Qo/Y1aF5wplfYyfXYU2CIAiiekjYEU1Dp25S3HzCrtvrQLfXAcaAd05GEE1lAABuu8X8LlERsetSuUmxNnmiHjV2ba78GjuriyJ2BEEQzQwJO6Jp0EyKx83siq1Bg52h19mF9Iid6R2xgN484U+PAmD1jdi5bAjm1NjZSNgRBEE0NSTsiPJgzPQl9K7YJkzFAtl07P7BcF7EznRExM6qJtGGaDZiV4caO5/DqtfYhZkLXqfD9DUJgiCI6iFhN9NplNnsNOvqPnZNmIoFciN2YcSEsKtLxM7mAlwdAIB+6SQcEl+7HhE7WZaQsPF1InDRnFiCIIgmh4QdUWcKR/66RFdsJJmBokUI6xApLJczerOdsRGRiq1LxA4AfLyBYql8HADAJBmw18dTbtzRBwAYYLPhozmxBEEQTQ0JuwK0/uSJ5hFMGn6nFTYLj+TluIkYgmTA4z29xwtZAiZiaRwd5SbBpnvYaYg6u88v5dFMyeHj48bqQNC9EB9N3oPNqVvqE6EkCIIgqoaEXQFo8oQ5SJP+z7tPknST4rSimnwEleO0WbCwm9ebvT7Ap0C46yV0hOXJcuv74mDqZzvid1nxBluCUbRRKpYgCKLJIWFHNBWdojPWPGFXG2eKdOzrAwEAgKfOqVgM7+f/O+oo7Jw2/WcvRewIgiCaGhJ2RFPR7TU7YlcbWmdsMJ4GALjrnIpF8Bj/v44RuzZXVtj5KGJHEATR1JCwI5qKbCq2+WoAgayw0/CYPU5MQ4vYadTBnFjD76KIHUEQxKkCCTuiqdBMilNNGrHTUrEadY/YadSzxi43FUsRO4IgiKaGhB1Rgvr63GkmxelMcwq7+R2uPIuTxkXs6pmKzYq5unUBEwRBEFVBwo5oKrpM74qtDVmW8tKxdYvYuTsBS87UhzpMndDQUrFuuwUWuUGG1gRBEERZkLAjmgrNpLhZa+yA7AQKoI5dsZIE+HqzvzcgFUv1dQRBEM0PCTuiqdCaJ5q1xg7ITqAA6uhjB+hedgDq2jwxr8MFAJjT5qzbmgRBEER10FdwoqnIsztp0qzfskZE7IBJwq5+Ebsz5/jxT59dhdNn12eEGUEQBFE9LR+xO3DgAM4991z9n8vlwk9+8pNGHxZRAC1ip6qnRiq2bjV2AODL6YytY40dAFx1Vi8WzyJhRxAE0ey0fMRu2bJl2LNnDwAgEolg4cKFWLduXWMPiiiI12GF3drc3zfa3XasmOfHoZGonqasCw2K2BEEQRCnDi0v7HJ54okncMUVV8Dj8TT6UIgCSJKEbo8diDX6SIrzoy+uRSyVyZvKYDq5ETsSdgRBEMQ0NDw08uKLL+Laa6/F3LlzIUnStGnSBx98EIsWLYLT6cSqVavw0ksvVbXWo48+iuuuu67GIybMplPU2TUzLrtF7+CtGxSxIwiCIErQcGEXjUaxcuVKPPDAA9Pe/8Mf/hC33nor7rzzTuzevRsXXXQRNmzYgIGBAX2bVatWYcWKFVP+nThxQt8mFArhf//3f7Fx40bTHxNRG9r0CWISucKuzjV2BEEQxKlBw1OxGzZswIYNGwref//99+PGG2/EF77wBQDAt771LTz11FP47ne/iy1btgAAdu3aVXKdn/70p7jqqqvgdBa3bEgmk0gmk/rvoVAIAJBOp5FOp0uuUw3afs3a/3QwxrQfiq6rbcdKbFfpuqqqFtxfR86kg4ySATPweVGUwusCjTkXZePuhWXlp8Fc7VCZBDTjMRpMU5+PGQadi+aCzkfzUI9zUcm+Gy7sipFKpbBr1y7cfvvtebevX78er7zySkX7evTRR/FHf/RHJbfbsmUL7r777im3P/3003C73RWtWSk7duwwdf+5OINBAEAgEMT27dsLbyeE7URgouh25dIrRPPx4+9jqMD+gsPZQPIbb7yB48dq78Y8T+X2Kb/97V4cHE6U3L6e56Ii5KuAJAADzsWpRNOejxkInYvmgs5H82DmuYjFyi88b2phNzo6CkVR0NPTk3d7T08PhoaGyt5PMBjEb37zG/z4xz8uue0dd9yB2267Tf89FAqhr68P69evh99vTvornU5jx44dWLduHWy2+hTj7zqxHUgA7e1tWFMkPf3a8Z8DI0BHe0fR7crlt28/CMSB+fPn4fwC+zv24mHgef7zypUrcc7Zta97cs9fAgxYseJsLF19ecHtGnEuiMLQ+Wge6Fw0F3Q+mod6nAste1gOTS3sNCQp36mWMTbltmK0tbXh5MmTZW3rcDjgcEyt8bLZbKa/eeqxhob+/ElS0TW17aQS21W6rizLBfc3uy1rIWK1WAEDnxOLpfC6udTzXBClofPRPNC5aC7ofDQPZp6LSvbb8OaJYnR3d8NisUyJzg0PD0+J4hnN1q1bsXz5cqxZs8bUdYipdJ8CXbEEQRAE0Yw0tbCz2+1YtWrVlLz1jh07sHbtWlPX3rx5M/bt24edO3eaus6pQ/0mQXSa2hXbpHPKCIIgCMIAGp6KjUQiOHjwoP774cOHsWfPHnR2dmLBggW47bbb8NnPfharV6/GhRdeiIceeggDAwO46aabGnjUpz6S1hVbb6GjNeMWWbfLY4dWTcCYatARNu+IMoIgCIIwioYLu9deew2XXXaZ/rvWuHD99ddj27ZtuO666zA2NoZ77rkHg4ODWLFiBbZv347+/n5Tj2vr1q3YunUrFEUxdZ2ZBxdYxcRal9eOQ2KLZEZFcYOaCqGAHUEQBNHCNFzYXXrppVlPtQJs2rQJmzZtqtMRcTZv3ozNmzcjFAqhrY1c/uuJ226FLHMFFklmjBV2BEEQBNHCNHWNHVEHmjSCZbfwl2YkmWnwkRAEQRDEqQMJuwJQV6ygAlsZI7FZRMQuQcKOIAiCIMqFhF0BqCu2sVhl/tJMZtQGHwlBEARBnDqQsCOakyZNERMEQRBEM0PCjiAIgiAIokUgYVcAqrEjCIIgCOJUg4RdAajGjiAIgiCIUw0SdgRBEARBEC0CCTuCIAiCIIgWgYQdQRAEQRBEi0DCrgDUPEEQBEEQxKkGCbsCUPMEQRAEQRCnGiTsCIIgCIIgWgQSdgRBEARBEC0CCTuCIAiCIIgWgYRdAah5giAIgiCIUw0SdgWg5gmCIAiCIE41SNgRBEEQBEG0CCTsCIIgCIIgWgRrow+g2WGMAQBCoZBpa6TTacRiMYRCIdhsNtPWySUSTyGUZIjGU0UfW1TbLlZ8u7LXTWQQSjJEYomi+9O2C0ejhqwbTigIgSESKb6/RpwLojB0PpoHOhfNBZ2P5qEe50K7bmmapBgSK2erGczx48fR19fX6MMgCIIgCGKGc+zYMcyfP7/oNiTsSqCqKk6cOAGfzwdJkkxZIxQKoa+vD8eOHYPf7zdlDaI86Fw0F3Q+mgc6F80FnY/moR7ngjGGcDiMuXPnQpaLV9FRKrYEsiyXVMdG4ff76Q3aJNC5aC7ofDQPdC6aCzofzYPZ56Ktra2s7ah5giAIgiAIokUgYUcQBEEQBNEikLBrAhwOB+666y44HI5GH8qMh85Fc0Hno3mgc9Fc0PloHprtXFDzBEEQBEEQRItAETuCIAiCIIgWgYQdQRAEQRBEi0DCjiAIgiAIokUgYddgHnzwQSxatAhOpxOrVq3CSy+91OhDmhG8+OKLuPbaazF37lxIkoSf/OQnefczxvA3f/M3mDt3LlwuFy699FK89dZbjTnYFmfLli1Ys2YNfD4fZs+ejY997GM4cOBA3jZ0PurDd7/7XZxzzjm6H9eFF16IJ598Ur+fzkPj2LJlCyRJwq233qrfRuejfvzN3/wNJEnK+9fb26vf30zngoRdA/nhD3+IW2+9FXfeeSd2796Niy66CBs2bMDAwECjD63liUajWLlyJR544IFp7/+7v/s73H///XjggQewc+dO9Pb2YgzazswAAAq8SURBVN26dQiHw3U+0tbnhRdewObNm/Hqq69ix44dyGQyWL9+PaLRqL4NnY/6MH/+fNx333147bXX8Nprr+Hyyy/HRz/6Uf0CReehMezcuRMPPfQQzjnnnLzb6XzUl7POOguDg4P6v7179+r3NdW5YETDuOCCC9hNN92Ud9sZZ5zBbr/99gYd0cwEAHv88cf131VVZb29vey+++7Tb0skEqytrY1973vfa8ARziyGh4cZAPbCCy8wxuh8NJqOjg728MMP03loEOFwmJ1++ulsx44d7JJLLmG33HILY4zeF/XmrrvuYitXrpz2vmY7FxSxaxCpVAq7du3C+vXr825fv349XnnllQYdFQEAhw8fxtDQUN65cTgcuOSSS+jc1IFgMAgA6OzsBEDno1EoioIf/OAHiEajuPDCC+k8NIjNmzfjwx/+MK688sq82+l81J93330Xc+fOxaJFi/DJT34Shw4dAtB854JmxTaI0dFRKIqCnp6evNt7enowNDTUoKMiAOjP/3Tn5ujRo404pBkDYwy33XYbPvShD2HFihUA6HzUm7179+LCCy9EIpGA1+vF448/juXLl+sXKDoP9eMHP/gBXn/9dezcuXPKffS+qC8f+MAH8K//+q9YunQpTp48ia9+9atYu3Yt3nrrraY7FyTsGowkSXm/M8am3EY0Bjo39efmm2/Gm2++iZdffnnKfXQ+6sOyZcuwZ88eBAIB/PjHP8b111+PF154Qb+fzkN9OHbsGG655RY8/fTTcDqdBbej81EfNmzYoP989tln48ILL8Rpp52Gf/mXf8EHP/hBAM1zLigV2yC6u7thsVimROeGh4enqH6ivmidTnRu6suf/Mmf4IknnsBzzz2H+fPn67fT+agvdrsdS5YswerVq7FlyxasXLkS3/72t+k81Jldu3ZheHgYq1atgtVqhdVqxQsvvIDvfOc7sFqt+nNO56MxeDwenH322Xj33Xeb7r1Bwq5B2O12rFq1Cjt27Mi7fceOHVi7dm2DjooAgEWLFqG3tzfv3KRSKbzwwgt0bkyAMYabb74Zjz32GJ599lksWrQo7346H42FMYZkMknnoc5cccUV2Lt3L/bs2aP/W716NT796U9jz549WLx4MZ2PBpJMJrF//37MmTOn+d4bdW/XIHR+8IMfMJvNxv75n/+Z7du3j916663M4/GwI0eONPrQWp5wOMx2797Ndu/ezQCw+++/n+3evZsdPXqUMcbYfffdx9ra2thjjz3G9u7dyz71qU+xOXPmsFAo1OAjbz3++I//mLW1tbHnn3+eDQ4O6v9isZi+DZ2P+nDHHXewF198kR0+fJi9+eab7Ctf+QqTZZk9/fTTjDE6D40mtyuWMTof9eTP/uzP2PPPP88OHTrEXn31VXbNNdcwn8+nX6+b6VyQsGswW7duZf39/cxut7Pzzz9ft3ggzOW5555jAKb8u/766xljvH39rrvuYr29vczhcLCLL76Y7d27t7EH3aJMdx4AsO9///v6NnQ+6sPnP/95/fNo1qxZ7IorrtBFHWN0HhrNZGFH56N+XHfddWzOnDnMZrOxuXPnst/93d9lb731ln5/M50LiTHG6h8nJAiCIAiCIIyGauwIgiAIgiBaBBJ2BEEQBEEQLQIJO4IgCIIgiBaBhB1BEARBEESLQMKOIAiCIAiiRSBhRxAEQRAE0SKQsCMIgiAIgmgRSNgRBEEQBEG0CCTsCIIgpkGSJPzkJz9p9GEQBEFUBAk7giBmFDfccAMkSYIkSbDZbOjp6cG6devwyCOPQFVVfbvBwUFs2LChrH2SCCQIolkgYUcQxIzj6quvxuDgII4cOYInn3wSl112GW655RZcc801yGQyAIDe3l44HI4GHylBEERlkLAjCGLG4XA40Nvbi3nz5uH888/HV77yFfz0pz/Fk08+iW3btgHIj8KlUincfPPNmDNnDpxOJxYuXIgtW7YAABYuXAgA+PjHPw5JkvTf33vvPXz0ox9FT08PvF4v1qxZg2eeeSbvOBYuXIivfe1r+PznPw+fz4cFCxbgoYceytvm+PHj+OQnP4nOzk54PB6sXr0av/71r/X7f/azn2HVqlVwOp1YvHgx7r77bl2cEgQx8yBhRxAEAeDyyy/HypUr8dhjj0257zvf+Q6eeOIJPProozhw4AD+/d//XRdwO3fuBAB8//vfx+DgoP57JBLBxo0b8cwzz2D37t246qqrcO2112JgYCBv3//wD/+A1atXY/fu3di0aRP++I//GG+//ba+j0suuQQnTpzAE088gTfeeANf/vKX9ZTxU089hc985jP40z/9U+zbtw//9E//hG3btuHee+8162kiCKLJsTb6AAiCIJqFM844A2+++eaU2wcGBnD66afjQx/6ECRJQn9/v37frFmzAADt7e3o7e3Vb1+5ciVWrlyp//7Vr34Vjz/+OJ544gncfPPN+u0bN27Epk2bAAB/+Zd/iW9+85t4/vnnccYZZ+A///M/MTIygp07d6KzsxMAsGTJEv1v7733Xtx+++24/vrrAQCLFy/G3/7t3+LLX/4y7rrrLiOeEoIgTjFI2BEEQQgYY5AkacrtN9xwA9atW4dly5bh6quvxjXXXIP169cX3Vc0GsXdd9+Nn//85zhx4gQymQzi8fiUiN0555yj/yxJEnp7ezE8PAwA2LNnD8477zxd1E1m165d2LlzZ16ETlEUJBIJxGIxuN3ush87QRCtAQk7giAIwf79+7Fo0aIpt59//vk4fPgwnnzySTzzzDP4xCc+gSuvvBL//d//XXBff/EXf4GnnnoKf//3f48lS5bA5XLh93//95FKpfK2s9lseb9LkqSnWl0uV9HjVVUVd999N373d393yn1Op7Po3xIE0ZqQsCMIggDw7LPPYu/evfjSl7407f1+vx/XXXcdrrvuOvz+7/8+rr76aoyPj6OzsxM2mw2KouRt/9JLL+GGG27Axz/+cQC8Xu7IkSMVHdM555yDhx9+WF9nMueffz4OHDiQl54lCGJmQ8KOIIgZRzKZxNDQEBRFwcmTJ/GLX/wCW7ZswTXXXIPPfe5zU7b/5je/iTlz5uDcc8+FLMv40Y9+hN7eXrS3twPg3a2//OUv8Tu/8ztwOBzo6OjAkiVL8Nhjj+Haa6+FJEn467/+6zyfvHL41Kc+ha997Wv42Mc+hi1btmDOnDnYvXs35s6diwsvvBD/9//+X1xzzTXo6+vDH/zBH0CWZbz55pvYu3cvvvrVrxrxVBEEcYpBXbEEQcw4fvGLX2DOnDlYuHAhrr76ajz33HP4zne+g5/+9KewWCxTtvd6vfj617+O1atXY82aNThy5Ai2b98OWeYfof/wD/+AHTt2oK+vD+eddx4ALgY7Ojqwdu1aXHvttbjqqqtw/vnnV3ScdrsdTz/9NGbPno2NGzfi7LPPxn333acf41VXXYWf//zn2LFjB9asWYMPfvCDuP/++/OaOwiCmFlIjDHW6IMgCIIgCIIgaocidgRBEARBEC0CCTuCIAiCIIgWgYQdQRAEQRBEi0DCjiAIgiAIokUgYUcQBEEQBNEikLAjCIIgCIJoEUjYEQRBEARBtAgk7AiCIAiCIFoEEnYEQRAEQRAtAgk7giAIgiCIFoGEHUEQBEEQRItAwo4gCIIgCKJF+P8BGAyWLUqCS/8AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "for status, group in avg_long.groupby('cmv_status'):\n",
    "    plt.plot(group['distance'], group['mean_freq'], label=f'CMV: {status}')\n",
    "\n",
    "plt.xlabel('Distance')\n",
    "plt.ylabel('Mean Frequency')\n",
    "plt.title('TCR Frequency by Distance and CMV Status')\n",
    "plt.legend()\n",
    "plt.grid(True)\n",
    "# y log scale\n",
    "plt.yscale('log')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (tcr_env)",
   "language": "python",
   "name": "tcr_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
